// @unbrained/pm-cli/dist/cli/extension-command-options.js
const UNSAFE_LOOSE_OPTION_KEYS = new Set(["__proto__", "prototype", "constructor"]);
function toLooseOptionKey(rawKey) {
    const key = rawKey.trim().toLowerCase();
    return key.replaceAll(/-([a-z0-9])/g, (_match, group) => group.toUpperCase());
}
function setLooseOptionValue(options, key, value) {
    const existing = options[key];
    if (existing === undefined) {
        options[key] = value;
        return;
    }
    if (Array.isArray(existing)) {
        existing.push(value);
        options[key] = existing;
        return;
    }
    options[key] = [existing, value];
}
function parseLooseOptionToken(args, index) {
    const token = args[index];
    if (!token.startsWith("--") || token === "--") {
        return null;
    }
    const equalsIndex = token.indexOf("=");
    if (equalsIndex >= 0) {
        return {
            consumed: 1,
            key: token.slice(2, equalsIndex).trim(),
            value: token.slice(equalsIndex + 1),
        };
    }
    if (token.startsWith("--no-")) {
        return {
            consumed: 1,
            key: token.slice(5).trim(),
            value: false,
        };
    }
    const key = token.slice(2).trim();
    const next = args[index + 1];
    if (typeof next === "string" && !next.startsWith("-")) {
        return {
            consumed: 2,
            key,
            value: next,
        };
    }
    return {
        consumed: 1,
        key,
        value: true,
    };
}
function isUnsafeLooseOptionKey(key) {
    return UNSAFE_LOOSE_OPTION_KEYS.has(key);
}
function resolveLooseOptionCoercionKind(definition) {
    const explicitType = typeof definition.type === "string" ? definition.type.trim().toLowerCase() : undefined;
    const valueType = typeof definition.value_type === "string" ? definition.value_type.trim().toLowerCase() : undefined;
    const kind = explicitType ?? valueType;
    if (kind === "string") {
        return "string";
    }
    if (kind === "number" || kind === "int" || kind === "integer" || kind === "float") {
        return "number";
    }
    if (kind === "boolean" || kind === "bool") {
        return "boolean";
    }
    return null;
}
function collectLooseOptionKeys(definition) {
    const keys = [];
    const long = typeof definition.long === "string" ? definition.long.trim() : "";
    const short = typeof definition.short === "string" ? definition.short.trim() : "";
    if (long.startsWith("--")) {
        const normalized = toLooseOptionKey(long.slice(2));
        if (normalized.length > 0 && !isUnsafeLooseOptionKey(normalized)) {
            keys.push(normalized);
        }
    }
    if (short.startsWith("-") && !short.startsWith("--")) {
        const normalized = toLooseOptionKey(short.slice(1));
        if (normalized.length > 0 && !isUnsafeLooseOptionKey(normalized)) {
            keys.push(normalized);
        }
    }
    return [...new Set(keys)];
}
function coerceLooseOptionValue(value, kind) {
    if (Array.isArray(value)) {
        return value.map((entry) => coerceLooseOptionValue(entry, kind));
    }
    if (kind === "string") {
        if (typeof value === "string") {
            return value;
        }
        if (value === null || value === undefined) {
            return value;
        }
        return String(value);
    }
    if (kind === "number") {
        if (typeof value === "number") {
            return Number.isFinite(value) ? value : value;
        }
        if (typeof value === "string" && value.trim().length > 0) {
            const parsed = Number(value);
            return Number.isFinite(parsed) ? parsed : value;
        }
        return value;
    }
    if (typeof value === "boolean") {
        return value;
    }
    if (typeof value === "string") {
        const normalized = value.trim().toLowerCase();
        if (normalized === "true" || normalized === "1") {
            return true;
        }
        if (normalized === "false" || normalized === "0") {
            return false;
        }
    }
    return value;
}
export function coerceLooseCommandOptionsWithFlagDefinitions(options, definitions) {
    if (definitions.length === 0) {
        return options;
    }
    const coerced = Object.create(null);
    for (const [key, value] of Object.entries(options)) {
        coerced[key] = value;
    }
    for (const definition of definitions) {
        const kind = resolveLooseOptionCoercionKind(definition);
        if (!kind) {
            continue;
        }
        for (const key of collectLooseOptionKeys(definition)) {
            if (!Object.hasOwn(coerced, key)) {
                continue;
            }
            coerced[key] = coerceLooseOptionValue(coerced[key], kind);
        }
    }
    return coerced;
}
export function parseLooseCommandOptions(args) {
    const options = Object.create(null);
    let index = 0;
    while (index < args.length) {
        const parsed = parseLooseOptionToken(args, index);
        if (!parsed) {
            index += 1;
            continue;
        }
        const normalizedKey = toLooseOptionKey(parsed.key);
        if (normalizedKey.length === 0 || isUnsafeLooseOptionKey(normalizedKey)) {
            index += parsed.consumed;
            continue;
        }
        setLooseOptionValue(options, normalizedKey, parsed.value);
        index += parsed.consumed;
    }
    return options;
}
