// @unbrained/pm-cli/dist/cli/extension-command-help.js
export function normalizeExtensionCommandPath(commandPath) {
    return commandPath
        .trim()
        .toLowerCase()
        .split(/\s+/)
        .filter((part) => part.length > 0)
        .join(" ");
}
function toNonEmptyFlagString(value) {
    if (typeof value !== "string") {
        return null;
    }
    const trimmed = value.trim();
    return trimmed.length > 0 ? trimmed : null;
}
function toOptionalBoolean(value) {
    return typeof value === "boolean" ? value : undefined;
}
function formatDynamicExtensionFlagHelpLine(definition) {
    const visible = toOptionalBoolean(definition.visible);
    if (visible === false) {
        return null;
    }
    const longName = toNonEmptyFlagString(definition.long);
    if (!longName || !longName.startsWith("--") || longName.length < 3) {
        return null;
    }
    const shortName = toNonEmptyFlagString(definition.short);
    const shortPrefix = shortName && shortName.startsWith("-") && !shortName.startsWith("--") ? `${shortName}, ` : "";
    const valueName = toNonEmptyFlagString(definition.value_name);
    const valueSuffix = valueName ? ` <${valueName}>` : "";
    const description = toNonEmptyFlagString(definition.description) ?? "Extension-provided option.";
    const markers = [];
    if (toOptionalBoolean(definition.required) === true) {
        markers.push("required");
    }
    if (toOptionalBoolean(definition.enabled) === false) {
        markers.push("disabled");
    }
    const markerSuffix = markers.length > 0 ? ` [${markers.join(", ")}]` : "";
    return `${shortPrefix}${longName}${valueSuffix}  ${description}${markerSuffix}`;
}
function buildDynamicExtensionFlagHelp(definitions) {
    const lines = [
        ...new Set(definitions
            .map(formatDynamicExtensionFlagHelpLine)
            .filter((line) => line !== null)),
    ].sort((left, right) => left.localeCompare(right));
    if (lines.length === 0) {
        return null;
    }
    return `\nExtension-provided flags:\n  ${lines.join("\n  ")}`;
}
export function collectDynamicExtensionFlagHelpByCommand(registrations) {
    const grouped = new Map();
    for (const registration of registrations) {
        const commandPath = normalizeExtensionCommandPath(registration.target_command);
        if (commandPath.length === 0) {
            continue;
        }
        const existing = grouped.get(commandPath) ?? [];
        existing.push(...registration.flags);
        grouped.set(commandPath, existing);
    }
    const entries = [...grouped.entries()].sort(([left], [right]) => left.localeCompare(right));
    const helpByCommand = new Map();
    for (const [commandPath, definitions] of entries) {
        const helpText = buildDynamicExtensionFlagHelp(definitions);
        if (!helpText) {
            continue;
        }
        helpByCommand.set(commandPath, helpText);
    }
    return helpByCommand;
}
function normalizeExtensionCommandAction(commandPath, action) {
    if (typeof action !== "string" || action.trim().length === 0) {
        return commandPath.replace(/\s+/g, "-");
    }
    return action.trim().toLowerCase();
}
function normalizeExtensionCommandStringList(values) {
    if (!Array.isArray(values)) {
        return [];
    }
    const normalized = [];
    const seen = new Set();
    for (const value of values) {
        if (typeof value !== "string") {
            continue;
        }
        const trimmed = value.trim();
        if (trimmed.length === 0 || seen.has(trimmed)) {
            continue;
        }
        seen.add(trimmed);
        normalized.push(trimmed);
    }
    return normalized;
}
function normalizeExtensionCommandArguments(values) {
    if (!Array.isArray(values)) {
        return [];
    }
    return values
        .map((value) => {
        const name = typeof value.name === "string" ? value.name.trim() : "";
        if (name.length === 0) {
            return null;
        }
        const normalized = {
            name,
            required: value.required === true,
            variadic: value.variadic === true,
        };
        if (typeof value.description === "string" && value.description.trim().length > 0) {
            normalized.description = value.description.trim();
        }
        return normalized;
    })
        .filter((entry) => entry !== null);
}
export function collectExtensionCommandHelpDescriptors(commandHandlers, commandDefinitions, flagRegistrations) {
    const definitionsByCommand = new Map();
    for (const definition of commandDefinitions) {
        const commandPath = normalizeExtensionCommandPath(definition.command);
        if (commandPath.length === 0) {
            continue;
        }
        const description = typeof definition.description === "string" && definition.description.trim().length > 0
            ? definition.description.trim()
            : undefined;
        const intent = typeof definition.intent === "string" && definition.intent.trim().length > 0
            ? definition.intent.trim()
            : undefined;
        definitionsByCommand.set(commandPath, {
            command: commandPath,
            action: normalizeExtensionCommandAction(commandPath, definition.action),
            description,
            intent,
            examples: normalizeExtensionCommandStringList(definition.examples),
            failure_hints: normalizeExtensionCommandStringList(definition.failure_hints),
            arguments: normalizeExtensionCommandArguments(definition.arguments),
            flags: [],
            source: {
                layer: definition.layer,
                name: definition.name,
            },
        });
    }
    const flagsByCommand = new Map();
    for (const registration of flagRegistrations) {
        const commandPath = normalizeExtensionCommandPath(registration.target_command);
        if (commandPath.length === 0) {
            continue;
        }
        const existing = flagsByCommand.get(commandPath) ?? [];
        existing.push(...registration.flags);
        flagsByCommand.set(commandPath, existing);
    }
    const commandSet = new Set();
    for (const commandPath of commandHandlers) {
        const normalized = normalizeExtensionCommandPath(commandPath);
        if (normalized.length > 0) {
            commandSet.add(normalized);
        }
    }
    for (const commandPath of definitionsByCommand.keys()) {
        commandSet.add(commandPath);
    }
    for (const commandPath of flagsByCommand.keys()) {
        commandSet.add(commandPath);
    }
    const descriptors = new Map();
    const sortedCommands = [...commandSet].sort((left, right) => left.localeCompare(right));
    for (const commandPath of sortedCommands) {
        const definition = definitionsByCommand.get(commandPath);
        const flags = flagsByCommand.get(commandPath) ?? [];
        if (definition) {
            descriptors.set(commandPath, {
                ...definition,
                flags,
            });
            continue;
        }
        descriptors.set(commandPath, {
            command: commandPath,
            action: normalizeExtensionCommandAction(commandPath, undefined),
            examples: [],
            failure_hints: [],
            arguments: [],
            flags,
        });
    }
    return descriptors;
}
function buildExtensionArgumentToken(argument) {
    const variadicSuffix = argument.variadic ? "..." : "";
    if (argument.required) {
        return `<${argument.name}${variadicSuffix}>`;
    }
    return `[${argument.name}${variadicSuffix}]`;
}
export function applyDynamicExtensionArguments(command, descriptor) {
    for (const argument of descriptor.arguments) {
        command.argument(buildExtensionArgumentToken(argument), argument.description ?? "Extension argument.");
    }
}
function formatDynamicExtensionOptionFlags(definition) {
    const visible = toOptionalBoolean(definition.visible);
    if (visible === false) {
        return null;
    }
    const longName = toNonEmptyFlagString(definition.long);
    const shortName = toNonEmptyFlagString(definition.short);
    const normalizedShort = shortName && shortName.startsWith("-") && !shortName.startsWith("--") ? shortName : null;
    const normalizedLong = longName && longName.startsWith("--") && longName.length > 2 ? longName : null;
    if (!normalizedLong && !normalizedShort) {
        return null;
    }
    const optionValueName = toNonEmptyFlagString(definition.value_name);
    const optionValueSuffix = optionValueName ? ` <${optionValueName}>` : "";
    const optionNames = [normalizedShort, normalizedLong].filter((entry) => entry !== null);
    return `${optionNames.join(", ")}${optionValueSuffix}`;
}
function formatDynamicExtensionOptionDescription(definition) {
    const description = toNonEmptyFlagString(definition.description) ?? "Extension-provided option.";
    const markers = [];
    if (toOptionalBoolean(definition.required) === true) {
        markers.push("required");
    }
    if (toOptionalBoolean(definition.enabled) === false) {
        markers.push("disabled");
    }
    const markerSuffix = markers.length > 0 ? ` [${markers.join(", ")}]` : "";
    return `${description}${markerSuffix}`;
}
export function applyDynamicExtensionOptions(command, descriptor) {
    const seen = new Set();
    for (const definition of descriptor.flags) {
        const optionFlags = formatDynamicExtensionOptionFlags(definition);
        if (!optionFlags || seen.has(optionFlags)) {
            continue;
        }
        seen.add(optionFlags);
        command.option(optionFlags, formatDynamicExtensionOptionDescription(definition));
    }
}
function buildDynamicExtensionHelpOptionSummary(definition) {
    const flags = formatDynamicExtensionOptionFlags(definition);
    if (!flags) {
        return null;
    }
    const longName = toNonEmptyFlagString(definition.long);
    const shortName = toNonEmptyFlagString(definition.short);
    const normalizedLong = longName && longName.startsWith("--") && longName.length > 2 ? longName : null;
    const normalizedShort = shortName && shortName.startsWith("-") && !shortName.startsWith("--") ? shortName : null;
    const valueName = toNonEmptyFlagString(definition.value_name);
    const required = toOptionalBoolean(definition.required) === true;
    return {
        flags,
        long: normalizedLong,
        short: normalizedShort,
        description: formatDynamicExtensionOptionDescription(definition),
        takes_value: valueName !== null,
        value_required: valueName !== null,
        value_name: valueName,
        variadic: false,
        required,
        aliases: [],
        alias_for: null,
    };
}
export function buildDynamicExtensionHelpOptionSummaries(descriptor) {
    if (!descriptor) {
        return [];
    }
    const summaries = [];
    const seen = new Set();
    for (const definition of descriptor.flags) {
        const summary = buildDynamicExtensionHelpOptionSummary(definition);
        if (!summary || seen.has(summary.flags)) {
            continue;
        }
        seen.add(summary.flags);
        summaries.push(summary);
    }
    return summaries;
}
export function mergeHelpOptionSummaries(base, extension) {
    if (extension.length === 0) {
        return base;
    }
    const merged = [...base];
    const seen = new Set(base.map((entry) => entry.flags));
    for (const entry of extension) {
        if (seen.has(entry.flags)) {
            continue;
        }
        seen.add(entry.flags);
        merged.push(entry);
    }
    return merged;
}
export function buildDynamicExtensionCommandMetadataHelp(descriptor) {
    const lines = [];
    if (descriptor.intent) {
        lines.push(`Intent: ${descriptor.intent}`);
    }
    if (descriptor.action) {
        lines.push(`Action contract: ${descriptor.action}`);
    }
    if (descriptor.examples.length > 0) {
        lines.push("Examples:");
        for (const example of descriptor.examples) {
            lines.push(`  - ${example}`);
        }
    }
    if (descriptor.failure_hints.length > 0) {
        lines.push("Common failure hints:");
        for (const hint of descriptor.failure_hints) {
            lines.push(`  - ${hint}`);
        }
    }
    if (lines.length === 0) {
        return null;
    }
    return `\nExtension command metadata:\n  ${lines.join("\n  ")}`;
}
export function commandAliases(command) {
    const commandRecord = command;
    if (typeof commandRecord.aliases === "function") {
        return commandRecord.aliases().map((value) => value.trim().toLowerCase()).filter((value) => value.length > 0);
    }
    if (typeof commandRecord.alias === "function") {
        const alias = commandRecord.alias();
        if (typeof alias === "string" && alias.trim().length > 0) {
            return [alias.trim().toLowerCase()];
        }
    }
    if (Array.isArray(commandRecord._aliases)) {
        return commandRecord._aliases.map((value) => value.trim().toLowerCase()).filter((value) => value.length > 0);
    }
    return [];
}
export function findDirectChildCommand(parent, name) {
    const normalizedTarget = name.trim().toLowerCase();
    return (parent.commands.find((entry) => {
        if (entry.name().trim().toLowerCase() === normalizedTarget) {
            return true;
        }
        return commandAliases(entry).includes(normalizedTarget);
    }) ?? null);
}
export function findCommandByPath(root, pathParts) {
    let current = root;
    for (const part of pathParts) {
        const next = findDirectChildCommand(current, part);
        if (!next) {
            return null;
        }
        current = next;
    }
    return current;
}
export function ensureCommandPath(root, pathParts) {
    if (pathParts.length === 0) {
        return null;
    }
    let current = root;
    for (let index = 0; index < pathParts.length; index += 1) {
        const part = pathParts[index];
        const existing = findDirectChildCommand(current, part);
        if (existing) {
            current = existing;
            continue;
        }
        const created = current.command(part);
        if (index < pathParts.length - 1) {
            created.description("Extension-provided command group.");
        }
        else {
            created.description("Extension-provided command path.");
        }
        current = created;
    }
    return current;
}
