// @unbrained/pm-cli/dist/core/fs/fs-utils.js
import fs from "node:fs/promises";
import path from "node:path";
import crypto from "node:crypto";
export async function ensureDir(dirPath) {
    await fs.mkdir(dirPath, { recursive: true });
}
export async function pathExists(targetPath) {
    try {
        await fs.access(targetPath);
        return true;
    }
    catch {
        return false;
    }
}
export async function readFileIfExists(targetPath) {
    try {
        return await fs.readFile(targetPath, "utf8");
    }
    catch (error) {
        if (isErrno(error, "ENOENT")) {
            return null;
        }
        throw error;
    }
}
export async function writeFileAtomic(targetPath, contents) {
    const dirPath = path.dirname(targetPath);
    await ensureDir(dirPath);
    const tempPath = path.join(dirPath, `.${path.basename(targetPath)}.${process.pid}.${Date.now()}.${crypto.randomBytes(4).toString("hex")}.tmp`);
    await fs.writeFile(tempPath, contents, "utf8");
    await fs.rename(tempPath, targetPath);
}
export async function appendLineAtomic(targetPath, line) {
    const dirPath = path.dirname(targetPath);
    await ensureDir(dirPath);
    const handle = await fs.open(targetPath, "a");
    try {
        await handle.writeFile(`${line}\n`, "utf8");
    }
    finally {
        await handle.close();
    }
}
export async function removeFileIfExists(targetPath) {
    try {
        await fs.unlink(targetPath);
    }
    catch (error) {
        if (!isErrno(error, "ENOENT")) {
            throw error;
        }
    }
}
function isErrno(error, code) {
    return typeof error === "object" && error !== null && "code" in error && error.code === code;
}
