// @unbrained/pm-cli/dist/pi/native.js
import { PmCliError } from "../core/shared/errors.js";
import { readSettings } from "../core/store/settings.js";
import { resolvePmRoot } from "../core/store/paths.js";
import { normalizeStatusInput } from "../core/item/status.js";
import { resolveRuntimeStatusRegistry } from "../core/schema/runtime-schema.js";
import { runActivity, runAggregate, runAppend, runCalendar, runClaim, runClose, runComments, runCommentsAudit, runCompletion, runConfig, runContext, runContracts, runCreate, runDedupeAudit, runDelete, runDeps, runDocs, runExtension, runFiles, runFilesDiscover, runGc, runGet, runGuide, runHealth, runHistory, runInit, runLearnings, runList, runNormalize, runNotes, runReindex, runRelease, runRestore, runSearch, runStats, runTemplatesList, runTemplatesSave, runTemplatesShow, runTest, runTestAll, runTestRunsList, runTestRunsLogs, runTestRunsResume, runTestRunsStatus, runTestRunsStop, runUpdate, runUpdateMany, runValidate, } from "../cli/commands/index.js";
const GLOBAL_KEYS = new Set([
    "action",
    "cwd",
    "path",
    "json",
    "quiet",
    "profile",
    "noExtensions",
    "noPager",
    "pmExecutable",
    "timeoutMs",
    "options",
]);
function asRecord(value) {
    return typeof value === "object" && value !== null && !Array.isArray(value) ? { ...value } : {};
}
function readString(args, key) {
    const value = args[key];
    return typeof value === "string" && value.trim().length > 0 ? value : undefined;
}
function mutationOptions(options) {
    return {
        author: readString(options, "author"),
        message: typeof options.message === "string" ? options.message : undefined,
    };
}
function requiredString(args, key, action) {
    const value = readString(args, key);
    if (!value) {
        throw new PmCliError(`Action "${action}" requires "${key}".`, 64);
    }
    return value;
}
export function nativeGlobalOptions(args) {
    return {
        json: args.json !== false,
        quiet: args.quiet === true,
        profile: args.profile === true,
        noExtensions: args.noExtensions === true,
        noPager: args.noPager !== false,
        path: readString(args, "path"),
    };
}
export function nativeCommandOptions(args) {
    const options = asRecord(args.options);
    for (const [key, value] of Object.entries(args)) {
        if (GLOBAL_KEYS.has(key) || value === undefined)
            continue;
        if (options[key] === undefined)
            options[key] = value;
    }
    const author = readString(args, "author");
    if (author && options.author === undefined)
        options.author = author;
    return options;
}
async function withCwd(cwd, run) {
    if (typeof cwd !== "string" || cwd.trim().length === 0)
        return run();
    const previous = process.cwd();
    process.chdir(cwd);
    try {
        return await run();
    }
    finally {
        process.chdir(previous);
    }
}
async function runStartTask(id, options, global) {
    const pmRoot = resolvePmRoot(process.cwd(), global.path);
    const settings = await readSettings(pmRoot);
    const statusRegistry = resolveRuntimeStatusRegistry(settings.schema);
    const status = normalizeStatusInput("in_progress", statusRegistry) ?? statusRegistry.open_status;
    const force = options.force === true;
    const baseOptions = mutationOptions(options);
    const claim = await runClaim(id, force, global, baseOptions);
    const update = await runUpdate(id, { ...baseOptions, status, force }, global);
    return { id, action: "start_task", claim, update };
}
async function runPauseTask(id, options, global) {
    const pmRoot = resolvePmRoot(process.cwd(), global.path);
    const settings = await readSettings(pmRoot);
    const statusRegistry = resolveRuntimeStatusRegistry(settings.schema);
    const force = options.force === true;
    const baseOptions = mutationOptions(options);
    const update = await runUpdate(id, { ...baseOptions, status: statusRegistry.open_status, force }, global);
    const release = await runRelease(id, force, global, baseOptions);
    return { id, action: "pause_task", update, release };
}
async function runCloseTask(id, reason, options, global) {
    const force = options.force === true;
    const baseOptions = mutationOptions(options);
    const close = await runClose(id, reason, { ...baseOptions, validateClose: readString(options, "validateClose"), force }, global);
    const release = await runRelease(id, force, global, baseOptions);
    return { id, action: "close_task", close, release };
}
export async function runNativePmAction(args) {
    return withCwd(args.cwd, async () => {
        const action = requiredString(args, "action", "pm").trim().toLowerCase();
        const global = nativeGlobalOptions(args);
        const options = nativeCommandOptions(args);
        const id = readString(args, "id") ?? readString(options, "id");
        const force = args.force === true || options.force === true;
        switch (action) {
            case "init": return runInit(readString(args, "prefix") ?? readString(options, "prefix"), global, options);
            case "config": return runConfig(readString(args, "scope") ?? readString(options, "scope") ?? "project", readString(args, "configAction") ?? readString(options, "configAction") ?? requiredString(options, "action", action), readString(args, "key") ?? readString(options, "key"), options, global);
            case "extension-init": return runExtension(requiredString(args, "target", action), { ...options, init: true }, global);
            case "extension-install": return runExtension(readString(args, "target") ?? readString(options, "target"), { ...options, install: true }, global);
            case "extension-uninstall": return runExtension(requiredString(args, "target", action), { ...options, uninstall: true }, global);
            case "extension-explore": return runExtension(undefined, { ...options, explore: true }, global);
            case "extension-manage": return runExtension(undefined, { ...options, manage: true }, global);
            case "extension-doctor": return runExtension(undefined, { ...options, doctor: true }, global);
            case "extension-adopt": return runExtension(requiredString(args, "target", action), { ...options, adopt: true }, global);
            case "extension-adopt-all": return runExtension(undefined, { ...options, adoptAll: true }, global);
            case "extension-activate": return runExtension(requiredString(args, "target", action), { ...options, activate: true }, global);
            case "extension-deactivate": return runExtension(requiredString(args, "target", action), { ...options, deactivate: true }, global);
            case "extension": return runExtension(readString(args, "target") ?? readString(options, "target"), options, global);
            case "create": return runCreate(options, global);
            case "list":
            case "list-all":
            case "list-draft":
            case "list-open":
            case "list-in-progress":
            case "list-blocked":
            case "list-closed":
            case "list-canceled": return runList(action === "list" ? readString(args, "status") : action === "list-all" ? undefined : action.replace("list-", "").replaceAll("-", "_"), options, global);
            case "aggregate": return runAggregate(options, global);
            case "dedupe-audit": return runDedupeAudit(options, global);
            case "calendar": return runCalendar(options, global);
            case "context": return runContext(options, global);
            case "get": return runGet(id ?? requiredString(options, "id", action), global);
            case "search": return runSearch(readString(args, "query") ?? readString(args, "keywords") ?? requiredString(options, "query", action), options, global);
            case "reindex": return runReindex(options, global);
            case "history": return runHistory(id ?? requiredString(options, "id", action), options, global);
            case "activity": return runActivity(options, global);
            case "restore": return runRestore(id ?? requiredString(options, "id", action), readString(args, "target") ?? requiredString(options, "target", action), options, global);
            case "update": return runUpdate(id ?? requiredString(options, "id", action), options, global);
            case "update-many": return runUpdateMany(options, global);
            case "normalize": return runNormalize(options, global);
            case "close": return runClose(id ?? requiredString(options, "id", action), readString(args, "text") ?? readString(args, "reason") ?? requiredString(options, "text", action), options, global);
            case "delete": return runDelete(id ?? requiredString(options, "id", action), options, global);
            case "append": return runAppend(id ?? requiredString(options, "id", action), options, global);
            case "comments": return runComments(id ?? requiredString(options, "id", action), options, global);
            case "comments-audit": return runCommentsAudit(options, global);
            case "notes": return runNotes(id ?? requiredString(options, "id", action), options, global);
            case "learnings": return runLearnings(id ?? requiredString(options, "id", action), options, global);
            case "files": return runFiles(id ?? requiredString(options, "id", action), options, global);
            case "files-discover": return runFilesDiscover(id ?? requiredString(options, "id", action), options, global);
            case "docs": return runDocs(id ?? requiredString(options, "id", action), options, global);
            case "deps": return runDeps(id ?? requiredString(options, "id", action), options, global);
            case "test": return runTest(id ?? requiredString(options, "id", action), options, global);
            case "test-all": return runTestAll(options, global);
            case "test-runs-list": return runTestRunsList(options, global);
            case "test-runs-status": return runTestRunsStatus(requiredString(args, "runId", action), global);
            case "test-runs-logs": return runTestRunsLogs(requiredString(args, "runId", action), options, global);
            case "test-runs-stop": return runTestRunsStop(requiredString(args, "runId", action), options, global);
            case "test-runs-resume": return runTestRunsResume(requiredString(args, "runId", action), options, global);
            case "stats": return runStats(global);
            case "health": return runHealth(global, options);
            case "validate": return runValidate(options, global);
            case "gc": return runGc(global, options);
            case "contracts": return runContracts(options, global);
            case "completion": return runCompletion(requiredString(args, "shell", action), undefined, undefined, options.eagerTags === true, options.runtime);
            case "templates-save": return runTemplatesSave(readString(args, "template") ?? requiredString(options, "name", action), options, global);
            case "templates-list": return runTemplatesList(global);
            case "templates-show": return runTemplatesShow(readString(args, "template") ?? requiredString(options, "name", action), global);
            case "claim": return runClaim(id ?? requiredString(options, "id", action), force, global, options);
            case "release": return runRelease(id ?? requiredString(options, "id", action), force, global, options);
            case "start-task": return runStartTask(id ?? requiredString(options, "id", action), options, global);
            case "pause-task": return runPauseTask(id ?? requiredString(options, "id", action), options, global);
            case "close-task": return runCloseTask(id ?? requiredString(options, "id", action), readString(args, "text") ?? readString(args, "reason") ?? requiredString(options, "text", action), options, global);
            case "guide": return runGuide(options, global);
            default: throw new PmCliError(`Unsupported native pm action: ${action}`, 64);
        }
    });
}
