// @unbrained/pm-cli/dist/core/history/history-stream-policy.js
import fs from "node:fs/promises";
import path from "node:path";
import { runActiveOnWriteHooks } from "../extensions/index.js";
import { ensureDir, pathExists } from "../fs/fs-utils.js";
import { EXIT_CODE } from "../shared/constants.js";
import { PmCliError } from "../shared/errors.js";
import { getHistoryPath } from "../store/paths.js";
function toSortedUniqueItemIds(itemIds) {
    return [...new Set(itemIds.filter((value) => value.trim().length > 0))].sort((left, right) => left.localeCompare(right));
}
function strictMissingStreamError(itemId, commandLabel) {
    return new PmCliError(`Missing history stream for ${itemId}. ${commandLabel} requires history streams when settings.history.missing_stream is strict_error.`, EXIT_CODE.NOT_FOUND);
}
async function createStream(historyPath, commandLabel) {
    await ensureDir(path.dirname(historyPath));
    const handle = await fs.open(historyPath, "a");
    await handle.close();
    return runActiveOnWriteHooks({
        path: historyPath,
        scope: "project",
        op: `${commandLabel}:history:auto_create`,
    });
}
export async function enforceHistoryStreamPolicyForItem(params) {
    return enforceHistoryStreamPolicyForItems({
        pmRoot: params.pmRoot,
        settings: params.settings,
        itemIds: [params.itemId],
        commandLabel: params.commandLabel,
    });
}
export async function enforceHistoryStreamPolicyForItems(params) {
    const ids = toSortedUniqueItemIds(params.itemIds);
    const autoCreated = [];
    const warnings = [];
    for (const itemId of ids) {
        const historyPath = getHistoryPath(params.pmRoot, itemId);
        if (await pathExists(historyPath)) {
            continue;
        }
        if (params.settings.history.missing_stream === "strict_error") {
            throw strictMissingStreamError(itemId, params.commandLabel);
        }
        warnings.push(...(await createStream(historyPath, params.commandLabel)));
        warnings.push(`history_stream_auto_created:${itemId}`);
        autoCreated.push(itemId);
    }
    return {
        auto_created_ids: autoCreated,
        warnings,
    };
}
