// @unbrained/pm-cli/dist/core/history/history.js
import jsonPatch from "fast-json-patch";
import { FRONT_MATTER_KEY_ORDER } from "../shared/constants.js";
import { runActiveServiceOverride } from "../extensions/index.js";
import { appendLineAtomic } from "../fs/fs-utils.js";
import { canonicalDocument } from "../item/item-format.js";
import { orderObject, sha256Hex, stableStringify } from "../shared/serialization.js";
const EMPTY_LEGACY_HASH_DOCUMENT = {
    front_matter: {},
    body: "",
};
function decodeJsonPointer(path) {
    if (!path || path === "/") {
        return [];
    }
    if (!path.startsWith("/")) {
        return [];
    }
    return path
        .slice(1)
        .split("/")
        .map((segment) => segment.replaceAll("~1", "/").replaceAll("~0", "~"));
}
function isDefinedPointerPath(document, path) {
    const segments = decodeJsonPointer(path);
    if (segments.length === 0) {
        return true;
    }
    let cursor = document;
    for (const segment of segments) {
        if (Array.isArray(cursor)) {
            if (segment === "-" || !/^(0|[1-9]\d*)$/.test(segment)) {
                return false;
            }
            const index = Number(segment);
            if (!Number.isInteger(index) || index < 0 || index >= cursor.length) {
                return false;
            }
            const next = cursor[index];
            if (next === undefined) {
                return false;
            }
            cursor = next;
            continue;
        }
        if (typeof cursor !== "object" || cursor === null) {
            return false;
        }
        if (!Object.prototype.hasOwnProperty.call(cursor, segment)) {
            return false;
        }
        const next = cursor[segment];
        if (next === undefined) {
            return false;
        }
        cursor = next;
    }
    return true;
}
function normalizeHistoryPatchOps(beforeDocument, patch) {
    const normalized = [];
    let replayCursor = structuredClone(beforeDocument);
    for (const operation of patch) {
        const normalizedOperation = operation.op === "replace" && !isDefinedPointerPath(replayCursor, operation.path)
            ? { ...operation, op: "add" }
            : operation;
        normalized.push(normalizedOperation);
        replayCursor = jsonPatch.applyPatch(replayCursor, [normalizedOperation], true, true).newDocument;
    }
    return normalized;
}
function canonicalHashDocument(document) {
    const hasMetadata = document.metadata && Object.keys(document.metadata).length > 0;
    if (!hasMetadata) {
        return {
            front_matter: {},
            body: document.body ?? "",
        };
    }
    const canonical = canonicalDocument(document);
    const orderedFrontMatter = orderObject(canonical.metadata, FRONT_MATTER_KEY_ORDER);
    return {
        front_matter: orderedFrontMatter,
        body: canonical.body,
    };
}
function canonicalPatchDocument(document) {
    const hasMetadata = document.metadata && Object.keys(document.metadata).length > 0;
    if (!hasMetadata) {
        return {
            metadata: {},
            body: document.body ?? "",
        };
    }
    const canonical = canonicalDocument(document);
    const orderedMetadata = orderObject(canonical.metadata, FRONT_MATTER_KEY_ORDER);
    return {
        metadata: orderedMetadata,
        body: canonical.body,
    };
}
export function hashDocument(document) {
    return sha256Hex(stableStringify(canonicalHashDocument(document)));
}
export function hashEmptyDocument() {
    return sha256Hex(stableStringify(EMPTY_LEGACY_HASH_DOCUMENT));
}
export function createHistoryEntry(params) {
    const beforeHashCanonical = canonicalHashDocument(params.before);
    const afterHashCanonical = canonicalHashDocument(params.after);
    const beforePatchCanonical = canonicalPatchDocument(params.before);
    const afterPatchCanonical = canonicalPatchDocument(params.after);
    const rawPatch = jsonPatch.compare(beforePatchCanonical, afterPatchCanonical);
    const patch = normalizeHistoryPatchOps(beforePatchCanonical, rawPatch);
    return {
        ts: params.nowIso,
        author: params.author,
        op: params.op,
        patch,
        before_hash: sha256Hex(stableStringify(beforeHashCanonical)),
        after_hash: sha256Hex(stableStringify(afterHashCanonical)),
        message: params.message === undefined ? undefined : params.message,
    };
}
export async function appendHistoryEntry(historyPath, entry) {
    const override = await runActiveServiceOverride("history_append", {
        history_path: historyPath,
        entry,
    });
    if (override.handled) {
        if (override.result === false) {
            return;
        }
        if (typeof override.result === "string") {
            await appendLineAtomic(historyPath, override.result);
            return;
        }
        if (typeof override.result === "object" && override.result !== null) {
            const record = override.result;
            if (record.skip === true) {
                return;
            }
            const nextHistoryPath = typeof record.history_path === "string" ? record.history_path : historyPath;
            if (typeof record.line === "string") {
                await appendLineAtomic(nextHistoryPath, record.line);
                return;
            }
            const nextEntry = (record.entry ?? entry);
            await appendLineAtomic(nextHistoryPath, JSON.stringify(nextEntry));
            return;
        }
    }
    await appendLineAtomic(historyPath, JSON.stringify(entry));
}
