// @unbrained/pm-cli/dist/core/item/id.js
import crypto from "node:crypto";
import path from "node:path";
import { TYPE_TO_FOLDER } from "../shared/constants.js";
import { pathExists } from "../fs/fs-utils.js";
export function normalizePrefix(input) {
    const normalized = (input ?? "").trim().toLowerCase();
    if (!normalized)
        return "pm-";
    return normalized.endsWith("-") ? normalized : `${normalized}-`;
}
export function normalizeRawItemId(input) {
    let normalized = input.trim().toLowerCase();
    if (normalized.startsWith("#")) {
        normalized = normalized.slice(1);
    }
    return normalized;
}
export function normalizeItemId(input, prefix) {
    const canonicalPrefix = normalizePrefix(prefix);
    const normalized = normalizeRawItemId(input);
    if (normalized.startsWith(canonicalPrefix)) {
        return normalized;
    }
    return `${canonicalPrefix}${normalized}`;
}
function randomToken(length) {
    const bytes = crypto.randomBytes(length);
    let token = "";
    for (const byte of bytes) {
        token += (byte % 36).toString(36);
    }
    return token;
}
async function idExists(pmRoot, id) {
    const checks = Object.values(TYPE_TO_FOLDER).flatMap((folder) => [
        path.join(pmRoot, folder, `${id}.md`),
        path.join(pmRoot, folder, `${id}.toon`),
    ]);
    for (const target of checks) {
        if (await pathExists(target)) {
            return true;
        }
    }
    return false;
}
export async function generateItemId(pmRoot, prefix) {
    let tokenLength = 4;
    let attempts = 0;
    while (tokenLength <= 10) {
        for (let i = 0; i < 32; i += 1) {
            const id = `${normalizePrefix(prefix)}${randomToken(tokenLength)}`;
            if (!(await idExists(pmRoot, id))) {
                return id;
            }
            attempts += 1;
        }
        tokenLength += 1;
    }
    throw new Error(`Unable to generate unique id after ${attempts} attempts`);
}
