// @unbrained/pm-cli/dist/core/extensions/index.js
import { runCommandHandler, runCommandOverride, runOnIndexHooks, runParserOverride, runPreflightOverride, runOnReadHooks, runRendererOverride, runServiceOverride, runServiceOverrideSync, runOnWriteHooks, } from "./loader.js";
let activeExtensionHooks = null;
let activeExtensionCommands = null;
let activeExtensionParsers = null;
let activeExtensionPreflight = null;
let activeExtensionServices = null;
let activeExtensionRenderers = null;
let activeExtensionRegistrations = null;
let activeCommandContext = null;
let activeCommandResult = undefined;
export function setActiveExtensionHooks(hooks) {
    activeExtensionHooks = hooks;
}
export function setActiveExtensionCommands(commands) {
    activeExtensionCommands = commands;
}
export function setActiveExtensionParsers(parsers) {
    activeExtensionParsers = parsers;
}
export function setActiveExtensionPreflight(preflight) {
    activeExtensionPreflight = preflight;
}
export function setActiveExtensionServices(services) {
    activeExtensionServices = services;
}
export function setActiveExtensionRenderers(renderers) {
    activeExtensionRenderers = renderers;
}
export function setActiveExtensionRegistrations(registrations) {
    activeExtensionRegistrations = registrations;
}
export function getActiveExtensionRegistrations() {
    return activeExtensionRegistrations;
}
export function setActiveCommandContext(context) {
    activeCommandContext = context;
}
export function getActiveCommandContext() {
    if (!activeCommandContext) {
        return null;
    }
    return {
        command: activeCommandContext.command,
        args: [...activeCommandContext.args],
        options: activeCommandContext.options ? { ...activeCommandContext.options } : {},
        global: activeCommandContext.global ? { ...activeCommandContext.global } : undefined,
        pm_root: activeCommandContext.pm_root,
    };
}
export function setActiveCommandResult(result) {
    activeCommandResult = result;
}
export function getActiveCommandResult() {
    return activeCommandResult;
}
export function clearActiveExtensionHooks() {
    activeExtensionHooks = null;
    activeExtensionCommands = null;
    activeExtensionParsers = null;
    activeExtensionPreflight = null;
    activeExtensionServices = null;
    activeExtensionRenderers = null;
    activeExtensionRegistrations = null;
    activeCommandContext = null;
    activeCommandResult = undefined;
}
export async function runActiveOnWriteHooks(context) {
    if (!activeExtensionHooks) {
        return [];
    }
    return runOnWriteHooks(activeExtensionHooks, context);
}
export async function runActiveOnReadHooks(context) {
    if (!activeExtensionHooks) {
        return [];
    }
    return runOnReadHooks(activeExtensionHooks, context);
}
export async function runActiveOnIndexHooks(context) {
    if (!activeExtensionHooks) {
        return [];
    }
    return runOnIndexHooks(activeExtensionHooks, context);
}
export function runActiveCommandOverride(result) {
    if (!activeExtensionCommands || !activeCommandContext) {
        return {
            overridden: false,
            result,
            warnings: [],
        };
    }
    return runCommandOverride(activeExtensionCommands, {
        command: activeCommandContext.command,
        args: [...activeCommandContext.args],
        options: activeCommandContext.options ? { ...activeCommandContext.options } : {},
        global: activeCommandContext.global ? { ...activeCommandContext.global } : undefined,
        pm_root: activeCommandContext.pm_root,
        result,
    });
}
export async function runActiveCommandHandler(context) {
    if (!activeExtensionCommands) {
        return {
            handled: false,
            result: null,
            warnings: [],
        };
    }
    return runCommandHandler(activeExtensionCommands, context);
}
export async function runActiveParserOverride(context) {
    if (!activeExtensionParsers) {
        return {
            overridden: false,
            context: {
                command: context.command,
                args: [...context.args],
                options: { ...context.options },
                global: { ...context.global },
                pm_root: context.pm_root,
            },
            warnings: [],
        };
    }
    return runParserOverride(activeExtensionParsers, context);
}
export async function runActivePreflightOverride(context) {
    if (!activeExtensionPreflight) {
        return {
            overridden: false,
            context: {
                command: context.command,
                args: [...context.args],
                options: { ...context.options },
                global: { ...context.global },
                pm_root: context.pm_root,
            },
            decision: { ...context.decision },
            warnings: [],
        };
    }
    return runPreflightOverride(activeExtensionPreflight, context);
}
export function runActiveRendererOverride(format, result) {
    if (!activeExtensionRenderers) {
        return {
            overridden: false,
            rendered: null,
            warnings: [],
        };
    }
    return runRendererOverride(activeExtensionRenderers, {
        format,
        command: activeCommandContext?.command,
        args: activeCommandContext ? [...activeCommandContext.args] : [],
        options: activeCommandContext?.options ? { ...activeCommandContext.options } : {},
        global: activeCommandContext?.global ? { ...activeCommandContext.global } : undefined,
        pm_root: activeCommandContext?.pm_root,
        result,
    });
}
function buildServiceContext(service, payload) {
    return {
        service,
        command: activeCommandContext?.command,
        args: activeCommandContext ? [...activeCommandContext.args] : [],
        options: activeCommandContext?.options ? { ...activeCommandContext.options } : {},
        global: activeCommandContext?.global ? { ...activeCommandContext.global } : undefined,
        pm_root: activeCommandContext?.pm_root,
        payload,
    };
}
export async function runActiveServiceOverride(service, payload) {
    if (!activeExtensionServices) {
        return {
            handled: false,
            result: payload,
            warnings: [],
        };
    }
    return runServiceOverride(activeExtensionServices, buildServiceContext(service, payload));
}
export function runActiveServiceOverrideSync(service, payload) {
    if (!activeExtensionServices) {
        return {
            handled: false,
            result: payload,
            warnings: [],
        };
    }
    return runServiceOverrideSync(activeExtensionServices, buildServiceContext(service, payload));
}
export * from "./loader.js";
