// @unbrained/pm-cli/dist/core/extensions/loader.js
import fs from "node:fs/promises";
import path from "node:path";
import { pathToFileURL } from "node:url";
import { pathExists } from "../fs/fs-utils.js";
import { resolveGlobalPmRoot } from "../store/paths.js";
import { KNOWN_EXTENSION_CAPABILITIES, EXTENSION_CAPABILITY_CONTRACT_VERSION, EXTENSION_CAPABILITY_LEGACY_ALIASES, } from "./extension-types.js";
export * from "./extension-types.js";
const DEFAULT_EXTENSION_PRIORITY = 100;
/* Types now in extension-types.ts - re-exported via `export * from "./extension-types.js"` above */
function normalizeNames(values) {
    return [...new Set(values.map((value) => value.trim()).filter((value) => value.length > 0))].sort((a, b) => a.localeCompare(b));
}
function isKnownExtensionCapability(value) {
    return KNOWN_EXTENSION_CAPABILITIES.includes(value);
}
function collectUnknownExtensionCapabilities(capabilities) {
    return capabilities.filter((capability) => !isKnownExtensionCapability(capability));
}
function resolveLegacyExtensionCapabilityAlias(capability) {
    const normalized = capability.trim().toLowerCase();
    if (normalized.length === 0) {
        return null;
    }
    return EXTENSION_CAPABILITY_LEGACY_ALIASES[normalized] ?? null;
}
function normalizeManifestCapabilities(rawCapabilities) {
    const normalizedCapabilities = normalizeNames([...rawCapabilities].map((value) => value.toLowerCase()));
    const remappedCapabilities = [];
    const legacyAliases = [];
    for (const capability of normalizedCapabilities) {
        const legacyAliasTarget = resolveLegacyExtensionCapabilityAlias(capability);
        if (legacyAliasTarget) {
            remappedCapabilities.push(legacyAliasTarget);
            legacyAliases.push({
                alias: capability,
                target: legacyAliasTarget,
            });
            continue;
        }
        remappedCapabilities.push(capability);
    }
    const dedupedLegacyAliases = [...new Map(legacyAliases.map((entry) => [`${entry.alias}>${entry.target}`, entry])).values()].sort((left, right) => left.alias.localeCompare(right.alias));
    return {
        capabilities: normalizeNames(remappedCapabilities),
        legacy_aliases: dedupedLegacyAliases,
    };
}
function levenshteinDistance(left, right) {
    if (left === right) {
        return 0;
    }
    if (left.length === 0) {
        return right.length;
    }
    if (right.length === 0) {
        return left.length;
    }
    const previous = new Array(right.length + 1);
    const current = new Array(right.length + 1);
    for (let j = 0; j <= right.length; j += 1) {
        previous[j] = j;
    }
    for (let i = 1; i <= left.length; i += 1) {
        current[0] = i;
        for (let j = 1; j <= right.length; j += 1) {
            const substitutionCost = left[i - 1] === right[j - 1] ? 0 : 1;
            current[j] = Math.min(previous[j] + 1, current[j - 1] + 1, previous[j - 1] + substitutionCost);
        }
        for (let j = 0; j <= right.length; j += 1) {
            previous[j] = current[j];
        }
    }
    return previous[right.length] ?? left.length;
}
function suggestKnownExtensionCapability(capability) {
    const normalized = capability.trim().toLowerCase();
    if (normalized.length === 0) {
        return null;
    }
    const legacyAlias = resolveLegacyExtensionCapabilityAlias(normalized);
    if (legacyAlias) {
        return legacyAlias;
    }
    let bestMatch = null;
    let bestDistance = Number.POSITIVE_INFINITY;
    for (const candidate of KNOWN_EXTENSION_CAPABILITIES) {
        const distance = levenshteinDistance(normalized, candidate);
        if (distance < bestDistance) {
            bestDistance = distance;
            bestMatch = candidate;
        }
    }
    const maxDistance = Math.max(1, Math.floor(normalized.length * 0.34));
    return bestMatch !== null && bestDistance <= maxDistance ? bestMatch : null;
}
function formatUnknownExtensionCapabilityWarning(layer, name, capability) {
    const allowed = KNOWN_EXTENSION_CAPABILITIES.join(",");
    const suggested = suggestKnownExtensionCapability(capability) ?? "none";
    return `extension_capability_unknown:${layer}:${name}:${capability}:allowed=${allowed}:suggested=${suggested}`;
}
function formatLegacyExtensionCapabilityAliasWarning(layer, name, aliases) {
    const aliasesToken = aliases.map((entry) => `${entry.alias}>${entry.target}`).join(",");
    return `extension_capability_legacy_alias:${layer}:${name}:aliases=${aliasesToken}`;
}
export function parseUnknownExtensionCapabilityWarning(warning) {
    const match = /^extension_capability_unknown:(global|project):([^:]+):([^:]+):allowed=([^:]+):suggested=([^:]+)$/.exec(warning.trim());
    if (!match) {
        return null;
    }
    const [, layerRaw, name, capability, allowedRaw, suggestedRaw] = match;
    const layer = layerRaw;
    const allowed_capabilities = allowedRaw
        .split(",")
        .map((value) => value.trim())
        .filter((value) => value.length > 0);
    const legacyAlias = resolveLegacyExtensionCapabilityAlias(capability);
    const suggestedFromWarning = suggestedRaw === "none" ? undefined : suggestedRaw;
    const suggested_capability = suggestedFromWarning ?? legacyAlias ?? undefined;
    const suggestion_source = suggested_capability
        ? legacyAlias === suggested_capability
            ? "legacy_alias"
            : "nearest_match"
        : undefined;
    return {
        layer,
        name,
        capability,
        allowed_capabilities,
        capability_contract_version: EXTENSION_CAPABILITY_CONTRACT_VERSION,
        suggested_capability,
        suggestion_source,
        legacy_alias_target: legacyAlias ?? undefined,
    };
}
export function parseLegacyExtensionCapabilityAliasWarning(warning) {
    const match = /^extension_capability_legacy_alias:(global|project):([^:]+):aliases=(.+)$/.exec(warning.trim());
    if (!match) {
        return [];
    }
    const [, layerRaw, name, aliasesRaw] = match;
    const layer = layerRaw;
    const allowedCapabilities = [...KNOWN_EXTENSION_CAPABILITIES];
    const parsed = [];
    for (const token of aliasesRaw.split(",")) {
        const [rawAlias, rawTarget] = token.split(">");
        const alias = rawAlias?.trim();
        const target = rawTarget?.trim().toLowerCase();
        if (!alias || !target || !isKnownExtensionCapability(target)) {
            continue;
        }
        parsed.push({
            layer,
            name,
            capability: alias,
            allowed_capabilities: allowedCapabilities,
            capability_contract_version: EXTENSION_CAPABILITY_CONTRACT_VERSION,
            suggested_capability: target,
            suggestion_source: "legacy_alias",
            legacy_alias_target: target,
        });
    }
    return parsed;
}
function parseManifest(raw) {
    if (typeof raw !== "object" || raw === null) {
        return null;
    }
    const candidate = raw;
    if (typeof candidate.name !== "string" || candidate.name.trim().length === 0) {
        return null;
    }
    if (typeof candidate.version !== "string" || candidate.version.trim().length === 0) {
        return null;
    }
    if (typeof candidate.entry !== "string" || candidate.entry.trim().length === 0) {
        return null;
    }
    let priority = DEFAULT_EXTENSION_PRIORITY;
    if ("priority" in candidate && candidate.priority !== undefined && candidate.priority !== null) {
        if (typeof candidate.priority !== "number" || !Number.isInteger(candidate.priority)) {
            return null;
        }
        priority = candidate.priority;
    }
    let capabilities = [];
    let legacyCapabilityAliases = [];
    if ("capabilities" in candidate && candidate.capabilities !== undefined && candidate.capabilities !== null) {
        if (!Array.isArray(candidate.capabilities) || candidate.capabilities.some((value) => typeof value !== "string")) {
            return null;
        }
        const normalizedCapabilities = normalizeManifestCapabilities(candidate.capabilities);
        capabilities = normalizedCapabilities.capabilities;
        legacyCapabilityAliases = normalizedCapabilities.legacy_aliases;
    }
    return {
        name: candidate.name.trim(),
        version: candidate.version.trim(),
        entry: candidate.entry.trim(),
        priority,
        capabilities,
        legacy_capability_aliases: legacyCapabilityAliases.length > 0 ? legacyCapabilityAliases : undefined,
    };
}
function asRecord(value) {
    if (typeof value !== "object" || value === null) {
        return null;
    }
    return value;
}
function shouldEnable(name, enabled, disabled) {
    if (disabled.has(name)) {
        return false;
    }
    if (enabled.size === 0) {
        return true;
    }
    return enabled.has(name);
}
function isPathWithinDirectory(directory, targetPath) {
    const relative = path.relative(directory, targetPath);
    if (relative.length === 0) {
        return true;
    }
    return !relative.startsWith("..") && !path.isAbsolute(relative);
}
async function isCanonicalPathWithinDirectory(directory, targetPath) {
    const [resolvedDirectory, resolvedTargetPath] = await Promise.all([fs.realpath(directory), fs.realpath(targetPath)]);
    return isPathWithinDirectory(resolvedDirectory, resolvedTargetPath);
}
export function resolveExtensionRoots(pmRoot, cwd = process.cwd()) {
    return {
        global: path.join(resolveGlobalPmRoot(cwd), "extensions"),
        project: path.join(pmRoot, "extensions"),
    };
}
export function createEmptyExtensionHookRegistry() {
    return {
        beforeCommand: [],
        afterCommand: [],
        onWrite: [],
        onRead: [],
        onIndex: [],
    };
}
export function createEmptyExtensionCommandRegistry() {
    return {
        overrides: [],
        handlers: [],
    };
}
export function createEmptyExtensionParserRegistry() {
    return {
        overrides: [],
    };
}
export function createEmptyExtensionPreflightRegistry() {
    return {
        overrides: [],
    };
}
export function createEmptyExtensionServiceRegistry() {
    return {
        overrides: [],
    };
}
export function createEmptyExtensionRendererRegistry() {
    return {
        overrides: [],
    };
}
export function createEmptyExtensionRegistrationRegistry() {
    return {
        commands: [],
        flags: [],
        item_fields: [],
        item_types: [],
        migrations: [],
        importers: [],
        exporters: [],
        search_providers: [],
        vector_store_adapters: [],
    };
}
async function listExtensionDirectories(extensionsRoot) {
    try {
        const entries = await fs.readdir(extensionsRoot, { withFileTypes: true });
        return entries
            .filter((entry) => entry.isDirectory())
            .map((entry) => entry.name)
            .sort((a, b) => a.localeCompare(b));
    }
    catch {
        return [];
    }
}
function summarizeCandidate(candidate) {
    return {
        layer: candidate.layer,
        directory: candidate.directory,
        manifest_path: candidate.manifest_path,
        name: candidate.manifest.name,
        version: candidate.manifest.version,
        entry: candidate.manifest.entry,
        priority: candidate.manifest.priority,
        entry_path: candidate.entry_path,
        capabilities: [...candidate.manifest.capabilities],
    };
}
function sortCandidates(candidates) {
    return [...candidates].sort((left, right) => {
        if (left.manifest.priority !== right.manifest.priority) {
            return left.manifest.priority - right.manifest.priority;
        }
        const byName = left.manifest.name.localeCompare(right.manifest.name);
        if (byName !== 0) {
            return byName;
        }
        return left.directory.localeCompare(right.directory);
    });
}
function buildEffectiveExtensions(globalCandidates, projectCandidates) {
    const ordered = [...sortCandidates(globalCandidates), ...sortCandidates(projectCandidates)];
    const effective = [];
    for (const candidate of ordered) {
        const existingIndex = effective.findIndex((entry) => entry.manifest.name === candidate.manifest.name);
        if (existingIndex >= 0) {
            effective.splice(existingIndex, 1);
        }
        effective.push(candidate);
    }
    return effective;
}
async function scanExtensionLayer(layer, extensionsRoot, enabled, disabled) {
    const diagnostics = [];
    const warnings = [];
    const candidates = [];
    const directories = await listExtensionDirectories(extensionsRoot);
    for (const directory of directories) {
        const scanned = await scanExtensionDirectory(layer, extensionsRoot, directory, enabled, disabled);
        diagnostics.push(scanned.diagnostic);
        warnings.push(...scanned.warnings);
        if (scanned.candidate) {
            candidates.push(scanned.candidate);
        }
    }
    return { diagnostics, warnings, candidates };
}
async function scanExtensionDirectory(layer, extensionsRoot, directory, enabled, disabled) {
    const extensionDir = path.join(extensionsRoot, directory);
    const manifestPath = path.join(extensionDir, "manifest.json");
    if (!(await pathExists(manifestPath))) {
        return {
            diagnostic: {
                layer,
                directory,
                manifest_path: manifestPath,
                name: null,
                version: null,
                entry: null,
                priority: null,
                entry_path: null,
                enabled: null,
                status: "warn",
            },
            warnings: [`extension_manifest_missing:${layer}:${directory}`],
            candidate: null,
        };
    }
    let manifest = null;
    try {
        const parsed = JSON.parse(await fs.readFile(manifestPath, "utf8"));
        manifest = parseManifest(parsed);
    }
    catch {
        manifest = null;
    }
    if (!manifest) {
        return {
            diagnostic: {
                layer,
                directory,
                manifest_path: manifestPath,
                name: null,
                version: null,
                entry: null,
                priority: null,
                entry_path: null,
                enabled: null,
                status: "warn",
            },
            warnings: [`extension_manifest_invalid:${layer}:${directory}`],
            candidate: null,
        };
    }
    const entryPath = path.resolve(extensionDir, manifest.entry);
    const entryWithinDirectoryByPath = isPathWithinDirectory(extensionDir, entryPath);
    const entryExists = entryWithinDirectoryByPath ? await pathExists(entryPath) : false;
    const entryWithinDirectory = entryWithinDirectoryByPath && entryExists
        ? await isCanonicalPathWithinDirectory(extensionDir, entryPath)
        : entryWithinDirectoryByPath;
    const enabledForLoad = shouldEnable(manifest.name, enabled, disabled);
    const extensionWarnings = [];
    if (Array.isArray(manifest.legacy_capability_aliases) && manifest.legacy_capability_aliases.length > 0) {
        extensionWarnings.push(formatLegacyExtensionCapabilityAliasWarning(layer, manifest.name, manifest.legacy_capability_aliases));
    }
    for (const capability of collectUnknownExtensionCapabilities(manifest.capabilities)) {
        extensionWarnings.push(formatUnknownExtensionCapabilityWarning(layer, manifest.name, capability));
    }
    if (!entryWithinDirectory) {
        extensionWarnings.push(`extension_entry_outside_extension:${layer}:${manifest.name}`);
    }
    else if (!entryExists) {
        extensionWarnings.push(`extension_entry_missing:${layer}:${manifest.name}`);
    }
    return {
        diagnostic: {
            layer,
            directory,
            manifest_path: manifestPath,
            name: manifest.name,
            version: manifest.version,
            entry: manifest.entry,
            priority: manifest.priority,
            entry_path: entryPath,
            enabled: enabledForLoad,
            status: entryWithinDirectory && entryExists ? "ok" : "warn",
        },
        warnings: extensionWarnings,
        candidate: entryWithinDirectory && entryExists && enabledForLoad
            ? {
                layer,
                directory,
                manifest_path: manifestPath,
                entry_path: entryPath,
                manifest,
            }
            : null,
    };
}
export async function discoverExtensions(options) {
    const roots = resolveExtensionRoots(options.pmRoot, options.cwd ?? process.cwd());
    const configured_enabled = normalizeNames(options.settings.extensions.enabled);
    const configured_disabled = normalizeNames(options.settings.extensions.disabled);
    if (options.noExtensions) {
        return {
            disabled_by_flag: true,
            roots,
            configured_enabled,
            configured_disabled,
            discovered: [],
            effective: [],
            warnings: [],
        };
    }
    const enabled = new Set(configured_enabled);
    const disabled = new Set(configured_disabled);
    const globalScan = await scanExtensionLayer("global", roots.global, enabled, disabled);
    const projectScan = await scanExtensionLayer("project", roots.project, enabled, disabled);
    const effective = buildEffectiveExtensions(globalScan.candidates, projectScan.candidates).map(summarizeCandidate);
    return {
        disabled_by_flag: false,
        roots,
        configured_enabled,
        configured_disabled,
        discovered: [...globalScan.diagnostics, ...projectScan.diagnostics],
        effective,
        warnings: [...globalScan.warnings, ...projectScan.warnings],
    };
}
function formatUnknownError(error) {
    if (error instanceof Error && error.message.trim().length > 0) {
        return error.message;
    }
    return String(error);
}
export async function loadExtensions(options) {
    const discovery = await discoverExtensions(options);
    const loaded = [];
    const failed = [];
    const warnings = [...discovery.warnings];
    if (discovery.disabled_by_flag) {
        return {
            ...discovery,
            warnings,
            loaded,
            failed,
        };
    }
    for (const extension of discovery.effective) {
        try {
            const module = await import(pathToFileURL(extension.entry_path).href);
            loaded.push({
                ...extension,
                module,
            });
        }
        catch (error) {
            warnings.push(`extension_load_failed:${extension.layer}:${extension.name}`);
            failed.push({
                layer: extension.layer,
                name: extension.name,
                entry_path: extension.entry_path,
                error: formatUnknownError(error),
            });
        }
    }
    return {
        ...discovery,
        warnings,
        loaded,
        failed,
    };
}
function resolveActivatableExtension(module) {
    const moduleRecord = asRecord(module);
    if (!moduleRecord) {
        return null;
    }
    if (typeof moduleRecord.activate === "function") {
        return {
            activate: moduleRecord.activate,
        };
    }
    const defaultExport = asRecord(moduleRecord.default);
    if (defaultExport && typeof defaultExport.activate === "function") {
        return {
            activate: defaultExport.activate,
        };
    }
    return null;
}
function normalizeCommandName(command) {
    return command
        .trim()
        .toLowerCase()
        .split(/\s+/)
        .filter((part) => part.length > 0)
        .join(" ");
}
function defaultGlobalOptions() {
    return {
        json: false,
        quiet: false,
        noExtensions: false,
        profile: false,
    };
}
function cloneCommandOptionsSnapshot(options) {
    return options ? cloneContextSnapshot(options) : {};
}
function cloneGlobalOptionsSnapshot(options) {
    return options ? cloneContextSnapshot(options) : defaultGlobalOptions();
}
function cloneContextSnapshot(value) {
    return structuredClone(value);
}
function isOutputRendererFormat(value) {
    return value === "toon" || value === "json";
}
const EXTENSION_SERVICE_NAMES = [
    "output_format",
    "error_format",
    "help_format",
    "lock_acquire",
    "lock_release",
    "history_append",
    "item_store_write",
    "item_store_delete",
];
function isExtensionServiceName(value) {
    return EXTENSION_SERVICE_NAMES.includes(value);
}
function assertHookHandler(hookName, hook) {
    if (typeof hook !== "function") {
        throw new TypeError(`api.hooks.${hookName} requires a function handler`);
    }
}
function assertNonEmptyString(name, value) {
    if (typeof value !== "string" || value.trim().length === 0) {
        throw new TypeError(`${name} requires a non-empty string`);
    }
    return value.trim();
}
function assertFunctionHandler(name, value) {
    if (typeof value !== "function") {
        throw new TypeError(`${name} requires a function handler`);
    }
}
function normalizeRegistrationName(name) {
    return name
        .trim()
        .toLowerCase()
        .split(/\s+/)
        .filter((part) => part.length > 0)
        .join(" ");
}
function toRegistrationCommandPath(name, action) {
    return normalizeCommandName(`${name} ${action}`);
}
function sanitizeRegistrationValue(value) {
    if (value === null) {
        return null;
    }
    if (Array.isArray(value)) {
        return value.map((entry) => sanitizeRegistrationValue(entry));
    }
    if (typeof value === "function") {
        return "[Function]";
    }
    if (typeof value === "bigint") {
        return value.toString();
    }
    if (typeof value === "symbol") {
        return value.toString();
    }
    if (typeof value === "object") {
        const record = value;
        const normalized = {};
        const keys = Object.keys(record).sort((left, right) => left.localeCompare(right));
        for (const key of keys) {
            normalized[key] = sanitizeRegistrationValue(record[key]);
        }
        return normalized;
    }
    return value;
}
function cloneRuntimeRegistrationValue(value) {
    if (Array.isArray(value)) {
        return value.map((entry) => cloneRuntimeRegistrationValue(entry));
    }
    if (typeof value === "object" && value !== null) {
        const record = value;
        const cloned = {};
        for (const key of Object.keys(record).sort((left, right) => left.localeCompare(right))) {
            cloned[key] = cloneRuntimeRegistrationValue(record[key]);
        }
        return cloned;
    }
    return value;
}
const EXTENSION_REGISTRATION_TRACE_SYMBOL = Symbol("extension_registration_trace");
function createRegistrationValidationError(message, trace) {
    const error = new TypeError(message);
    Object.defineProperty(error, EXTENSION_REGISTRATION_TRACE_SYMBOL, {
        value: trace,
        enumerable: false,
        configurable: false,
        writable: false,
    });
    return error;
}
function extractRegistrationValidationTrace(error) {
    if (!(error instanceof Error)) {
        return undefined;
    }
    return error[EXTENSION_REGISTRATION_TRACE_SYMBOL];
}
function normalizeRegistrationRecord(name, value) {
    if (typeof value !== "object" || value === null || Array.isArray(value)) {
        throw new TypeError(`${name} requires an object definition`);
    }
    return sanitizeRegistrationValue(value);
}
function normalizeRuntimeRegistrationRecord(name, value) {
    if (typeof value !== "object" || value === null || Array.isArray(value)) {
        throw new TypeError(`${name} requires an object definition`);
    }
    return cloneRuntimeRegistrationValue(value);
}
function normalizeRegistrationRecordList(name, value) {
    if (!Array.isArray(value)) {
        throw new TypeError(`${name} requires an array of object definitions`);
    }
    return value.map((entry) => normalizeRegistrationRecord(name, entry));
}
function asRegistrationRecord(name, value) {
    if (typeof value !== "object" || value === null || Array.isArray(value)) {
        throw new TypeError(`${name} requires an object definition`);
    }
    return value;
}
function assertOptionalBooleanField(name, value) {
    if (value !== undefined && typeof value !== "boolean") {
        throw new TypeError(`${name} must be a boolean when provided`);
    }
}
function assertOptionalStringField(name, value) {
    if (value === undefined) {
        return;
    }
    if (typeof value !== "string" || value.trim().length === 0) {
        throw new TypeError(`${name} must be a non-empty string when provided`);
    }
}
function assertOptionalStringArrayField(name, value) {
    if (value === undefined) {
        return;
    }
    if (!Array.isArray(value)) {
        throw new TypeError(`${name} must be an array of non-empty strings when provided`);
    }
    for (const [index, entry] of value.entries()) {
        if (typeof entry !== "string" || entry.trim().length === 0) {
            throw new TypeError(`${name}[${index}] must be a non-empty string`);
        }
    }
}
function normalizeOptionalStringArrayField(name, value) {
    assertOptionalStringArrayField(name, value);
    if (!Array.isArray(value)) {
        return [];
    }
    const normalized = [];
    const seen = new Set();
    for (const entry of value) {
        const trimmed = entry.trim();
        if (trimmed.length === 0 || seen.has(trimmed)) {
            continue;
        }
        seen.add(trimmed);
        normalized.push(trimmed);
    }
    return normalized;
}
function normalizeCommandActionName(value) {
    return value
        .trim()
        .toLowerCase()
        .replace(/[\s_]+/g, "-")
        .replace(/[^a-z0-9-]/g, "-")
        .replace(/-+/g, "-")
        .replace(/^-+|-+$/g, "");
}
function resolveCommandDefinitionAction(commandPath, action) {
    if (action === undefined) {
        return commandPath.replace(/\s+/g, "-");
    }
    if (typeof action !== "string" || action.trim().length === 0) {
        throw new TypeError("registerCommand definition.action must be a non-empty string when provided");
    }
    const normalized = normalizeCommandActionName(action);
    if (normalized.length === 0) {
        throw new TypeError("registerCommand definition.action must contain alphanumeric characters");
    }
    return normalized;
}
function normalizeCommandDefinitionArguments(value) {
    if (value === undefined) {
        return [];
    }
    if (!Array.isArray(value)) {
        throw new TypeError("registerCommand definition.arguments must be an array when provided");
    }
    const normalized = [];
    for (const [index, entry] of value.entries()) {
        const record = asRegistrationRecord(`registerCommand definition.arguments[${index}]`, entry);
        const name = assertNonEmptyString(`registerCommand definition.arguments[${index}].name`, record.name);
        assertOptionalBooleanField(`registerCommand definition.arguments[${index}].required`, record.required);
        assertOptionalBooleanField(`registerCommand definition.arguments[${index}].variadic`, record.variadic);
        assertOptionalStringField(`registerCommand definition.arguments[${index}].description`, record.description);
        if (name.includes(" ")) {
            throw new TypeError(`registerCommand definition.arguments[${index}].name must not contain spaces`);
        }
        const definition = {
            name,
        };
        if (record.required === true) {
            definition.required = true;
        }
        if (record.variadic === true) {
            definition.variadic = true;
        }
        if (typeof record.description === "string") {
            const trimmedDescription = record.description.trim();
            if (trimmedDescription.length > 0) {
                definition.description = trimmedDescription;
            }
        }
        normalized.push(definition);
    }
    let variadicCount = 0;
    for (const [index, argument] of normalized.entries()) {
        if (!argument.variadic) {
            continue;
        }
        variadicCount += 1;
        if (variadicCount > 1) {
            throw new TypeError("registerCommand definition.arguments supports at most one variadic argument");
        }
        if (index !== normalized.length - 1) {
            throw new TypeError("registerCommand definition.arguments variadic argument must be the final argument");
        }
    }
    return normalized;
}
function validateFlagDefinitions(flags) {
    if (!Array.isArray(flags)) {
        throw new TypeError("registerFlags flags requires an array of object definitions");
    }
    for (const [index, raw] of flags.entries()) {
        const record = asRegistrationRecord(`registerFlags flags[${index}]`, raw);
        const long = record.long;
        const short = record.short;
        if (long === undefined && short === undefined) {
            throw new TypeError(`registerFlags flags[${index}] requires at least one of long or short`);
        }
        assertOptionalStringField(`registerFlags flags[${index}].long`, long);
        assertOptionalStringField(`registerFlags flags[${index}].short`, short);
        assertOptionalStringField(`registerFlags flags[${index}].value_name`, record.value_name);
        assertOptionalStringField(`registerFlags flags[${index}].description`, record.description);
        assertOptionalBooleanField(`registerFlags flags[${index}].required`, record.required);
        assertOptionalBooleanField(`registerFlags flags[${index}].enabled`, record.enabled);
        assertOptionalBooleanField(`registerFlags flags[${index}].visible`, record.visible);
    }
}
function validateItemFieldDefinitions(fields) {
    if (!Array.isArray(fields)) {
        throw new TypeError("registerItemFields fields requires an array of object definitions");
    }
    for (const [index, raw] of fields.entries()) {
        const record = asRegistrationRecord(`registerItemFields fields[${index}]`, raw);
        assertNonEmptyString(`registerItemFields fields[${index}].name`, record.name);
        assertNonEmptyString(`registerItemFields fields[${index}].type`, record.type);
        assertOptionalBooleanField(`registerItemFields fields[${index}].optional`, record.optional);
    }
}
function validateItemTypeDefinitions(types) {
    if (!Array.isArray(types)) {
        throw new TypeError("registerItemTypes types requires an array of object definitions");
    }
    for (const [typeIndex, raw] of types.entries()) {
        const record = asRegistrationRecord(`registerItemTypes types[${typeIndex}]`, raw);
        assertNonEmptyString(`registerItemTypes types[${typeIndex}].name`, record.name);
        assertOptionalStringField(`registerItemTypes types[${typeIndex}].folder`, record.folder);
        assertOptionalStringArrayField(`registerItemTypes types[${typeIndex}].aliases`, record.aliases);
        assertOptionalStringArrayField(`registerItemTypes types[${typeIndex}].required_create_fields`, record.required_create_fields);
        assertOptionalStringArrayField(`registerItemTypes types[${typeIndex}].required_create_repeatables`, record.required_create_repeatables);
        if (record.command_option_policies !== undefined) {
            if (!Array.isArray(record.command_option_policies)) {
                throw new TypeError(`registerItemTypes types[${typeIndex}].command_option_policies must be an array when provided`);
            }
            for (const [policyIndex, rawPolicy] of record.command_option_policies.entries()) {
                const policy = asRegistrationRecord(`registerItemTypes types[${typeIndex}].command_option_policies[${policyIndex}]`, rawPolicy);
                assertNonEmptyString(`registerItemTypes types[${typeIndex}].command_option_policies[${policyIndex}].command`, policy.command);
                assertNonEmptyString(`registerItemTypes types[${typeIndex}].command_option_policies[${policyIndex}].option`, policy.option);
                assertOptionalBooleanField(`registerItemTypes types[${typeIndex}].command_option_policies[${policyIndex}].enabled`, policy.enabled);
                assertOptionalBooleanField(`registerItemTypes types[${typeIndex}].command_option_policies[${policyIndex}].required`, policy.required);
                assertOptionalBooleanField(`registerItemTypes types[${typeIndex}].command_option_policies[${policyIndex}].visible`, policy.visible);
            }
        }
        if (record.options !== undefined) {
            if (!Array.isArray(record.options)) {
                throw new TypeError(`registerItemTypes types[${typeIndex}].options must be an array when provided`);
            }
            for (const [optionIndex, rawOption] of record.options.entries()) {
                const option = asRegistrationRecord(`registerItemTypes types[${typeIndex}].options[${optionIndex}]`, rawOption);
                assertNonEmptyString(`registerItemTypes types[${typeIndex}].options[${optionIndex}].key`, option.key);
                assertOptionalStringArrayField(`registerItemTypes types[${typeIndex}].options[${optionIndex}].values`, option.values);
                assertOptionalBooleanField(`registerItemTypes types[${typeIndex}].options[${optionIndex}].required`, option.required);
                assertOptionalStringArrayField(`registerItemTypes types[${typeIndex}].options[${optionIndex}].aliases`, option.aliases);
            }
        }
    }
}
function validateMigrationDefinition(definition) {
    const record = asRegistrationRecord("registerMigration definition", definition);
    if (record.id !== undefined && typeof record.id !== "string") {
        throw new TypeError("registerMigration definition.id must be a string when provided");
    }
    if (record.description !== undefined && typeof record.description !== "string") {
        throw new TypeError("registerMigration definition.description must be a string when provided");
    }
    if (record.status !== undefined && typeof record.status !== "string") {
        throw new TypeError("registerMigration definition.status must be a string when provided");
    }
    assertOptionalBooleanField("registerMigration definition.mandatory", record.mandatory);
    if (record.run !== undefined && typeof record.run !== "function") {
        throw new TypeError("registerMigration definition.run must be a function when provided");
    }
}
function attachRuntimeDefinition(entry, runtimeDefinition) {
    Object.defineProperty(entry, "runtime_definition", {
        value: runtimeDefinition,
        enumerable: false,
        writable: false,
        configurable: false,
    });
    return entry;
}
function getDeclaredExtensionCapabilities(extension) {
    if (!Array.isArray(extension.capabilities)) {
        return null;
    }
    const declared = new Set();
    for (const capability of extension.capabilities) {
        const normalized = capability.trim().toLowerCase();
        if (isKnownExtensionCapability(normalized)) {
            declared.add(normalized);
        }
    }
    return declared;
}
function assertExtensionCapability(extension, capability, method) {
    const declared = getDeclaredExtensionCapabilities(extension);
    // Keep direct unit tests that construct LoadedExtension fixtures without
    // capability metadata backwards-compatible while enforcing manifest-declared
    // capabilities for runtime-loaded extensions.
    if (declared === null) {
        return;
    }
    if (!declared.has(capability)) {
        throw new TypeError(`${method} requires capability '${capability}' in extension manifest capabilities`);
    }
}
function createExtensionApi(extension, hooks, commands, parsers, preflight, services, renderers, registrations, activationWarnings) {
    const registerCommandTrace = (mode, command, expectedSchema, received, hint) => ({
        method: "registerCommand",
        registration_index: mode === "override" ? commands.overrides.length : commands.handlers.length,
        command,
        expected_schema: expectedSchema,
        received: sanitizeRegistrationValue(received),
        hint,
    });
    const registerCommand = (commandOrDefinition, override) => {
        assertExtensionCapability(extension, "commands", "registerCommand");
        if (typeof commandOrDefinition === "string") {
            const normalizedCommand = normalizeCommandName(commandOrDefinition);
            if (normalizedCommand.length === 0) {
                throw createRegistrationValidationError("registerCommand requires a non-empty command name", registerCommandTrace("override", commandOrDefinition, 'registerCommand("<command>", (context) => unknown)', commandOrDefinition, "Provide a non-empty command path as the first argument."));
            }
            if (typeof override !== "function") {
                const trace = registerCommandTrace("override", normalizedCommand, 'registerCommand("<command>", (context) => unknown)', { command: commandOrDefinition, override }, "Provide a function as the second registerCommand argument.");
                throw createRegistrationValidationError(`registerCommand requires an override function when command name is provided (command="${normalizedCommand}", registration_index=${trace.registration_index})`, trace);
            }
            commands.overrides.push({
                layer: extension.layer,
                name: extension.name,
                command: normalizedCommand,
                run: override,
            });
            return;
        }
        if (typeof commandOrDefinition !== "object" || commandOrDefinition === null) {
            throw createRegistrationValidationError("registerCommand requires a command definition object", registerCommandTrace("definition", undefined, "{ name: string; run: (context) => unknown; }", commandOrDefinition, "Use registerCommand({ name: \"command path\", run: (context) => ... })."));
        }
        if (typeof commandOrDefinition.name !== "string") {
            throw createRegistrationValidationError("registerCommand requires a command definition name", registerCommandTrace("definition", undefined, "{ name: string; run: (context) => unknown; }", commandOrDefinition, "Set command definition.name to a non-empty string command path."));
        }
        const normalizedCommand = normalizeCommandName(commandOrDefinition.name);
        if (normalizedCommand.length === 0) {
            throw createRegistrationValidationError("registerCommand requires a non-empty command definition name", registerCommandTrace("definition", commandOrDefinition.name, "{ name: string; run: (context) => unknown; }", commandOrDefinition, "Ensure command definition.name contains a non-empty command path."));
        }
        const runHandler = typeof commandOrDefinition.run === "function" ? commandOrDefinition.run : undefined;
        const legacyHandler = typeof commandOrDefinition.handler === "function" ? commandOrDefinition.handler : undefined;
        if (!runHandler && legacyHandler) {
            activationWarnings.push(`extension_command_definition_legacy_handler_alias:${extension.layer}:${extension.name}:${normalizedCommand}`);
        }
        const resolvedHandler = runHandler ?? legacyHandler;
        if (typeof resolvedHandler !== "function") {
            const trace = registerCommandTrace("definition", normalizedCommand, "{ name: string; run: (context) => unknown; }", commandOrDefinition, "Define command definition.run as a function.");
            throw createRegistrationValidationError(`registerCommand requires a command definition run handler (command="${normalizedCommand}", registration_index=${trace.registration_index})`, trace);
        }
        try {
            assertOptionalStringField("registerCommand definition.action", commandOrDefinition.action);
            assertOptionalStringField("registerCommand definition.description", commandOrDefinition.description);
            assertOptionalStringField("registerCommand definition.intent", commandOrDefinition.intent);
            const action = resolveCommandDefinitionAction(normalizedCommand, commandOrDefinition.action);
            const description = commandOrDefinition.description?.trim();
            const intent = commandOrDefinition.intent?.trim();
            const examples = normalizeOptionalStringArrayField("registerCommand definition.examples", commandOrDefinition.examples);
            const failureHints = normalizeOptionalStringArrayField("registerCommand definition.failure_hints", commandOrDefinition.failure_hints);
            const argumentsDefinition = normalizeCommandDefinitionArguments(commandOrDefinition.arguments);
            if (commandOrDefinition.flags !== undefined) {
                assertExtensionCapability(extension, "schema", "registerCommand flags");
                validateFlagDefinitions(commandOrDefinition.flags);
                registrations.flags.push({
                    layer: extension.layer,
                    name: extension.name,
                    target_command: normalizedCommand,
                    flags: normalizeRegistrationRecordList("registerCommand definition.flags", commandOrDefinition.flags),
                });
            }
            const registration = {
                layer: extension.layer,
                name: extension.name,
                command: normalizedCommand,
                action,
                examples,
                failure_hints: failureHints,
                arguments: argumentsDefinition,
            };
            if (description) {
                registration.description = description;
            }
            if (intent) {
                registration.intent = intent;
            }
            registrations.commands.push(registration);
        }
        catch (error) {
            const reason = error instanceof Error ? error.message : "registerCommand definition validation failed";
            const trace = registerCommandTrace("definition", normalizedCommand, "{ name: string; run: (context) => unknown; action?: string; arguments?: object[]; flags?: object[]; }", commandOrDefinition, "Use schema-style metadata (action/arguments/flags/examples/intent) with valid values.");
            throw createRegistrationValidationError(`registerCommand definition metadata invalid (command="${normalizedCommand}", registration_index=${trace.registration_index}): ${reason}`, trace);
        }
        commands.handlers.push({
            layer: extension.layer,
            name: extension.name,
            command: normalizedCommand,
            run: resolvedHandler,
        });
    };
    const registerParser = (command, override) => {
        assertExtensionCapability(extension, "parser", "registerParser");
        const normalizedCommand = normalizeCommandName(assertNonEmptyString("registerParser command", command));
        assertFunctionHandler("registerParser override", override);
        parsers.overrides.push({
            layer: extension.layer,
            name: extension.name,
            command: normalizedCommand,
            run: override,
        });
    };
    const registerPreflight = (override) => {
        assertExtensionCapability(extension, "preflight", "registerPreflight");
        assertFunctionHandler("registerPreflight override", override);
        preflight.overrides.push({
            layer: extension.layer,
            name: extension.name,
            run: override,
        });
    };
    const registerService = (service, override) => {
        assertExtensionCapability(extension, "services", "registerService");
        const normalizedService = String(service).trim().toLowerCase();
        if (!isExtensionServiceName(normalizedService)) {
            throw new TypeError(`registerService service must be one of: ${EXTENSION_SERVICE_NAMES.join(", ")}`);
        }
        assertFunctionHandler("registerService override", override);
        services.overrides.push({
            layer: extension.layer,
            name: extension.name,
            service: normalizedService,
            run: override,
        });
    };
    const registerRenderer = (format, renderer) => {
        assertExtensionCapability(extension, "renderers", "registerRenderer");
        if (typeof renderer !== "function") {
            throw new TypeError("registerRenderer requires a renderer function");
        }
        const normalizedFormat = String(format).trim().toLowerCase();
        if (!isOutputRendererFormat(normalizedFormat)) {
            throw new Error(`registerRenderer format must be toon|json, received: ${String(format)}`);
        }
        renderers.overrides.push({
            layer: extension.layer,
            name: extension.name,
            format: normalizedFormat,
            run: renderer,
        });
    };
    const registerFlags = (targetCommand, flags) => {
        assertExtensionCapability(extension, "schema", "registerFlags");
        const normalizedTargetCommand = normalizeCommandName(assertNonEmptyString("registerFlags targetCommand", targetCommand));
        validateFlagDefinitions(flags);
        const normalizedFlags = normalizeRegistrationRecordList("registerFlags flags", flags);
        if (normalizedFlags.length === 0) {
            throw new TypeError("registerFlags requires at least one flag definition");
        }
        registrations.flags.push({
            layer: extension.layer,
            name: extension.name,
            target_command: normalizedTargetCommand,
            flags: normalizedFlags,
        });
    };
    const registerItemFields = (fields) => {
        assertExtensionCapability(extension, "schema", "registerItemFields");
        validateItemFieldDefinitions(fields);
        const normalizedFields = normalizeRegistrationRecordList("registerItemFields fields", fields);
        if (normalizedFields.length === 0) {
            throw new TypeError("registerItemFields requires at least one field definition");
        }
        registrations.item_fields.push({
            layer: extension.layer,
            name: extension.name,
            fields: normalizedFields,
        });
    };
    const registerItemTypes = (types) => {
        assertExtensionCapability(extension, "schema", "registerItemTypes");
        validateItemTypeDefinitions(types);
        const normalizedTypes = normalizeRegistrationRecordList("registerItemTypes types", types);
        if (normalizedTypes.length === 0) {
            throw new TypeError("registerItemTypes requires at least one type definition");
        }
        registrations.item_types.push({
            layer: extension.layer,
            name: extension.name,
            types: normalizedTypes,
        });
    };
    const registerMigration = (definition) => {
        assertExtensionCapability(extension, "schema", "registerMigration");
        validateMigrationDefinition(definition);
        const runtimeDefinition = normalizeRuntimeRegistrationRecord("registerMigration definition", definition);
        registrations.migrations.push(attachRuntimeDefinition({
            layer: extension.layer,
            name: extension.name,
            definition: normalizeRegistrationRecord("registerMigration definition", definition),
        }, runtimeDefinition));
    };
    const registerImporter = (name, importer) => {
        assertExtensionCapability(extension, "importers", "registerImporter");
        const normalizedName = normalizeRegistrationName(assertNonEmptyString("registerImporter name", name));
        assertFunctionHandler("registerImporter importer", importer);
        const commandPath = toRegistrationCommandPath(normalizedName, "import");
        registrations.importers.push({
            layer: extension.layer,
            name: extension.name,
            importer: normalizedName,
        });
        commands.handlers.push({
            layer: extension.layer,
            name: extension.name,
            command: commandPath,
            run: async (context) => importer({
                registration: normalizedName,
                action: "import",
                command: context.command,
                args: cloneContextSnapshot(context.args),
                options: cloneContextSnapshot(context.options),
                global: cloneContextSnapshot(context.global),
                pm_root: context.pm_root,
            }),
        });
    };
    const registerExporter = (name, exporter) => {
        assertExtensionCapability(extension, "importers", "registerExporter");
        const normalizedName = normalizeRegistrationName(assertNonEmptyString("registerExporter name", name));
        assertFunctionHandler("registerExporter exporter", exporter);
        const commandPath = toRegistrationCommandPath(normalizedName, "export");
        registrations.exporters.push({
            layer: extension.layer,
            name: extension.name,
            exporter: normalizedName,
        });
        commands.handlers.push({
            layer: extension.layer,
            name: extension.name,
            command: commandPath,
            run: async (context) => exporter({
                registration: normalizedName,
                action: "export",
                command: context.command,
                args: cloneContextSnapshot(context.args),
                options: cloneContextSnapshot(context.options),
                global: cloneContextSnapshot(context.global),
                pm_root: context.pm_root,
            }),
        });
    };
    const registerSearchProvider = (provider) => {
        assertExtensionCapability(extension, "search", "registerSearchProvider");
        const runtimeDefinition = normalizeRuntimeRegistrationRecord("registerSearchProvider provider", provider);
        registrations.search_providers.push(attachRuntimeDefinition({
            layer: extension.layer,
            name: extension.name,
            definition: normalizeRegistrationRecord("registerSearchProvider provider", provider),
        }, runtimeDefinition));
    };
    const registerVectorStoreAdapter = (adapter) => {
        assertExtensionCapability(extension, "search", "registerVectorStoreAdapter");
        const runtimeDefinition = normalizeRuntimeRegistrationRecord("registerVectorStoreAdapter adapter", adapter);
        registrations.vector_store_adapters.push(attachRuntimeDefinition({
            layer: extension.layer,
            name: extension.name,
            definition: normalizeRegistrationRecord("registerVectorStoreAdapter adapter", adapter),
        }, runtimeDefinition));
    };
    const registerBeforeCommand = (hook) => {
        assertExtensionCapability(extension, "hooks", "api.hooks.beforeCommand");
        assertHookHandler("beforeCommand", hook);
        hooks.beforeCommand.push({
            layer: extension.layer,
            name: extension.name,
            run: hook,
        });
    };
    const registerAfterCommand = (hook) => {
        assertExtensionCapability(extension, "hooks", "api.hooks.afterCommand");
        assertHookHandler("afterCommand", hook);
        hooks.afterCommand.push({
            layer: extension.layer,
            name: extension.name,
            run: hook,
        });
    };
    const registerOnWrite = (hook) => {
        assertExtensionCapability(extension, "hooks", "api.hooks.onWrite");
        assertHookHandler("onWrite", hook);
        hooks.onWrite.push({
            layer: extension.layer,
            name: extension.name,
            run: hook,
        });
    };
    const registerOnRead = (hook) => {
        assertExtensionCapability(extension, "hooks", "api.hooks.onRead");
        assertHookHandler("onRead", hook);
        hooks.onRead.push({
            layer: extension.layer,
            name: extension.name,
            run: hook,
        });
    };
    const registerOnIndex = (hook) => {
        assertExtensionCapability(extension, "hooks", "api.hooks.onIndex");
        assertHookHandler("onIndex", hook);
        hooks.onIndex.push({
            layer: extension.layer,
            name: extension.name,
            run: hook,
        });
    };
    return {
        registerCommand,
        registerParser,
        registerPreflight,
        registerService,
        registerFlags,
        registerItemFields,
        registerItemTypes,
        registerMigration,
        registerRenderer,
        registerImporter,
        registerExporter,
        registerSearchProvider,
        registerVectorStoreAdapter,
        hooks: {
            beforeCommand: registerBeforeCommand,
            afterCommand: registerAfterCommand,
            onWrite: registerOnWrite,
            onRead: registerOnRead,
            onIndex: registerOnIndex,
        },
    };
}
async function executeRegisteredHooks(entries, hookName, context) {
    const warnings = [];
    for (const entry of entries) {
        try {
            await entry.run(cloneContextSnapshot(context));
        }
        catch {
            warnings.push(`extension_hook_failed:${entry.layer}:${entry.name}:${hookName}`);
        }
    }
    return warnings;
}
function getRegistrationCounts(registrations) {
    const commandCount = registrations.commands.length;
    const flagCount = registrations.flags.reduce((total, entry) => total + entry.flags.length, 0);
    const itemFieldCount = registrations.item_fields.reduce((total, entry) => total + entry.fields.length, 0);
    const itemTypeCount = registrations.item_types.reduce((total, entry) => total + entry.types.length, 0);
    return {
        commands: commandCount,
        flags: flagCount,
        item_fields: itemFieldCount,
        item_types: itemTypeCount,
        migrations: registrations.migrations.length,
        importers: registrations.importers.length,
        exporters: registrations.exporters.length,
        search_providers: registrations.search_providers.length,
        vector_store_adapters: registrations.vector_store_adapters.length,
    };
}
function collectCommandCollisionWarnings(commands) {
    const warnings = [];
    const collectByCommand = (entries, codePrefix) => {
        const grouped = new Map();
        for (const entry of entries) {
            const bucket = grouped.get(entry.command) ?? [];
            bucket.push(entry);
            grouped.set(entry.command, bucket);
        }
        for (const command of [...grouped.keys()].sort((left, right) => left.localeCompare(right))) {
            const bucket = grouped.get(command) ?? [];
            if (bucket.length <= 1) {
                continue;
            }
            const winner = bucket[bucket.length - 1];
            for (const displaced of bucket.slice(0, -1)) {
                warnings.push(`${codePrefix}:${command}:${winner.layer}:${winner.name}:${displaced.layer}:${displaced.name}`);
            }
        }
    };
    collectByCommand(commands.handlers, "extension_command_handler_collision");
    collectByCommand(commands.overrides, "extension_command_override_collision");
    const handlerCommands = new Set(commands.handlers.map((entry) => entry.command));
    const overlapCommands = [...new Set(commands.overrides.map((entry) => entry.command))]
        .filter((command) => handlerCommands.has(command))
        .sort((left, right) => left.localeCompare(right));
    for (const command of overlapCommands) {
        warnings.push(`extension_command_override_handler_overlap:${command}`);
    }
    return warnings;
}
function collectRendererCollisionWarnings(renderers) {
    const grouped = new Map();
    for (const entry of renderers.overrides) {
        const bucket = grouped.get(entry.format) ?? [];
        bucket.push(entry);
        grouped.set(entry.format, bucket);
    }
    const warnings = [];
    for (const format of [...grouped.keys()].sort((left, right) => left.localeCompare(right))) {
        const bucket = grouped.get(format) ?? [];
        if (bucket.length <= 1) {
            continue;
        }
        const winner = bucket[bucket.length - 1];
        for (const displaced of bucket.slice(0, -1)) {
            warnings.push(`extension_renderer_collision:${format}:${winner.layer}:${winner.name}:${displaced.layer}:${displaced.name}`);
        }
    }
    return warnings;
}
function collectParserCollisionWarnings(parsers) {
    const warnings = [];
    const grouped = new Map();
    for (const entry of parsers.overrides) {
        const bucket = grouped.get(entry.command) ?? [];
        bucket.push(entry);
        grouped.set(entry.command, bucket);
    }
    for (const command of [...grouped.keys()].sort((left, right) => left.localeCompare(right))) {
        const bucket = grouped.get(command) ?? [];
        if (bucket.length <= 1) {
            continue;
        }
        const winner = bucket[bucket.length - 1];
        for (const displaced of bucket.slice(0, -1)) {
            warnings.push(`extension_parser_override_collision:${command}:${winner.layer}:${winner.name}:${displaced.layer}:${displaced.name}`);
        }
    }
    return warnings;
}
function collectPreflightCollisionWarnings(preflight) {
    if (preflight.overrides.length <= 1) {
        return [];
    }
    const winner = preflight.overrides[preflight.overrides.length - 1];
    return preflight.overrides.slice(0, -1).map((displaced) => `extension_preflight_override_collision:${winner.layer}:${winner.name}:${displaced.layer}:${displaced.name}`);
}
function collectServiceCollisionWarnings(services) {
    const warnings = [];
    const grouped = new Map();
    for (const entry of services.overrides) {
        const bucket = grouped.get(entry.service) ?? [];
        bucket.push(entry);
        grouped.set(entry.service, bucket);
    }
    for (const service of [...grouped.keys()].sort((left, right) => left.localeCompare(right))) {
        const bucket = grouped.get(service) ?? [];
        if (bucket.length <= 1) {
            continue;
        }
        const winner = bucket[bucket.length - 1];
        for (const displaced of bucket.slice(0, -1)) {
            warnings.push(`extension_service_override_collision:${service}:${winner.layer}:${winner.name}:${displaced.layer}:${displaced.name}`);
        }
    }
    return warnings;
}
export async function activateExtensions(loadResult) {
    const hooks = createEmptyExtensionHookRegistry();
    const commands = createEmptyExtensionCommandRegistry();
    const parsers = createEmptyExtensionParserRegistry();
    const preflight = createEmptyExtensionPreflightRegistry();
    const services = createEmptyExtensionServiceRegistry();
    const renderers = createEmptyExtensionRendererRegistry();
    const registrations = createEmptyExtensionRegistrationRegistry();
    const failed = [];
    const warnings = [];
    for (const extension of loadResult.loaded) {
        const activatable = resolveActivatableExtension(extension.module);
        if (!activatable) {
            continue;
        }
        try {
            await activatable.activate(createExtensionApi(extension, hooks, commands, parsers, preflight, services, renderers, registrations, warnings));
        }
        catch (error) {
            warnings.push(`extension_activate_failed:${extension.layer}:${extension.name}`);
            const trace = extractRegistrationValidationTrace(error);
            failed.push({
                layer: extension.layer,
                name: extension.name,
                entry_path: extension.entry_path,
                error: formatUnknownError(error),
                trace,
            });
        }
    }
    const collisionWarnings = [
        ...collectCommandCollisionWarnings(commands),
        ...collectParserCollisionWarnings(parsers),
        ...collectPreflightCollisionWarnings(preflight),
        ...collectServiceCollisionWarnings(services),
        ...collectRendererCollisionWarnings(renderers),
    ];
    const mergedWarnings = [...new Set([...warnings, ...collisionWarnings])];
    return {
        hooks,
        commands,
        parsers,
        preflight,
        services,
        renderers,
        registrations,
        failed,
        warnings: mergedWarnings,
        hook_counts: {
            before_command: hooks.beforeCommand.length,
            after_command: hooks.afterCommand.length,
            on_write: hooks.onWrite.length,
            on_read: hooks.onRead.length,
            on_index: hooks.onIndex.length,
        },
        command_override_count: commands.overrides.length,
        command_handler_count: commands.handlers.length,
        parser_override_count: parsers.overrides.length,
        preflight_override_count: preflight.overrides.length,
        service_override_count: services.overrides.length,
        renderer_override_count: renderers.overrides.length,
        registration_counts: getRegistrationCounts(registrations),
    };
}
export async function runBeforeCommandHooks(hooks, context) {
    return executeRegisteredHooks(hooks.beforeCommand, "beforeCommand", context);
}
export async function runAfterCommandHooks(hooks, context) {
    return executeRegisteredHooks(hooks.afterCommand, "afterCommand", context);
}
export async function runOnWriteHooks(hooks, context) {
    return executeRegisteredHooks(hooks.onWrite, "onWrite", context);
}
export async function runOnReadHooks(hooks, context) {
    return executeRegisteredHooks(hooks.onRead, "onRead", context);
}
export async function runOnIndexHooks(hooks, context) {
    return executeRegisteredHooks(hooks.onIndex, "onIndex", context);
}
export async function runCommandHandler(commands, context) {
    const command = normalizeCommandName(context.command);
    if (command.length === 0) {
        return {
            handled: false,
            result: null,
            warnings: [],
        };
    }
    const matched = [...commands.handlers].reverse().find((entry) => entry.command === command);
    if (!matched) {
        return {
            handled: false,
            result: null,
            warnings: [],
        };
    }
    try {
        const result = await matched.run({
            command,
            args: cloneContextSnapshot(context.args),
            options: cloneContextSnapshot(context.options),
            global: cloneContextSnapshot(context.global),
            pm_root: context.pm_root,
        });
        return {
            handled: true,
            result,
            warnings: [],
        };
    }
    catch {
        return {
            handled: false,
            result: null,
            warnings: [`extension_command_handler_failed:${matched.layer}:${matched.name}:${matched.command}`],
        };
    }
}
export async function runParserOverride(parsers, context) {
    const command = normalizeCommandName(context.command);
    if (command.length === 0) {
        return {
            overridden: false,
            context: {
                command,
                args: cloneContextSnapshot(context.args),
                options: cloneCommandOptionsSnapshot(context.options),
                global: cloneGlobalOptionsSnapshot(context.global),
                pm_root: context.pm_root,
            },
            warnings: [],
        };
    }
    const matched = [...parsers.overrides].reverse().find((entry) => entry.command === command);
    if (!matched) {
        return {
            overridden: false,
            context: {
                command,
                args: cloneContextSnapshot(context.args),
                options: cloneCommandOptionsSnapshot(context.options),
                global: cloneGlobalOptionsSnapshot(context.global),
                pm_root: context.pm_root,
            },
            warnings: [],
        };
    }
    try {
        const delta = (await Promise.resolve(matched.run({
            command,
            args: cloneContextSnapshot(context.args),
            options: cloneCommandOptionsSnapshot(context.options),
            global: cloneGlobalOptionsSnapshot(context.global),
            pm_root: context.pm_root,
        }))) ?? {};
        const nextArgs = Array.isArray(delta.args) ? cloneContextSnapshot(delta.args) : cloneContextSnapshot(context.args);
        const nextOptions = delta.options ? cloneCommandOptionsSnapshot(delta.options) : cloneCommandOptionsSnapshot(context.options);
        const nextGlobal = delta.global ? cloneGlobalOptionsSnapshot(delta.global) : cloneGlobalOptionsSnapshot(context.global);
        return {
            overridden: true,
            context: {
                command,
                args: nextArgs,
                options: nextOptions,
                global: nextGlobal,
                pm_root: context.pm_root,
            },
            warnings: [],
        };
    }
    catch {
        return {
            overridden: false,
            context: {
                command,
                args: cloneContextSnapshot(context.args),
                options: cloneCommandOptionsSnapshot(context.options),
                global: cloneGlobalOptionsSnapshot(context.global),
                pm_root: context.pm_root,
            },
            warnings: [`extension_parser_override_failed:${matched.layer}:${matched.name}:${matched.command}`],
        };
    }
}
export async function runPreflightOverride(preflight, context) {
    const matched = [...preflight.overrides].reverse()[0];
    const baseContext = {
        command: normalizeCommandName(context.command),
        args: cloneContextSnapshot(context.args),
        options: cloneCommandOptionsSnapshot(context.options),
        global: cloneGlobalOptionsSnapshot(context.global),
        pm_root: context.pm_root,
    };
    const baseDecision = cloneContextSnapshot(context.decision);
    if (!matched) {
        return {
            overridden: false,
            context: baseContext,
            decision: baseDecision,
            warnings: [],
        };
    }
    try {
        const delta = (await Promise.resolve(matched.run({
            command: baseContext.command,
            args: cloneContextSnapshot(baseContext.args),
            options: cloneCommandOptionsSnapshot(baseContext.options),
            global: cloneGlobalOptionsSnapshot(baseContext.global),
            pm_root: baseContext.pm_root,
            decision: cloneContextSnapshot(baseDecision),
        }))) ?? {};
        const nextContext = {
            command: baseContext.command,
            args: Array.isArray(delta.args) ? cloneContextSnapshot(delta.args) : baseContext.args,
            options: delta.options ? cloneCommandOptionsSnapshot(delta.options) : baseContext.options,
            global: delta.global ? cloneGlobalOptionsSnapshot(delta.global) : baseContext.global,
            pm_root: baseContext.pm_root,
        };
        const nextDecision = {
            enforce_item_format_gate: typeof delta.enforce_item_format_gate === "boolean"
                ? delta.enforce_item_format_gate
                : baseDecision.enforce_item_format_gate,
            run_preflight_item_format_sync: typeof delta.run_preflight_item_format_sync === "boolean"
                ? delta.run_preflight_item_format_sync
                : baseDecision.run_preflight_item_format_sync,
            run_extension_migrations: typeof delta.run_extension_migrations === "boolean"
                ? delta.run_extension_migrations
                : baseDecision.run_extension_migrations,
            enforce_mandatory_migration_gate: typeof delta.enforce_mandatory_migration_gate === "boolean"
                ? delta.enforce_mandatory_migration_gate
                : baseDecision.enforce_mandatory_migration_gate,
        };
        return {
            overridden: true,
            context: nextContext,
            decision: nextDecision,
            warnings: [],
        };
    }
    catch {
        return {
            overridden: false,
            context: baseContext,
            decision: baseDecision,
            warnings: [`extension_preflight_override_failed:${matched.layer}:${matched.name}`],
        };
    }
}
function resolveDefaultServiceResult(context) {
    return {
        handled: false,
        result: context.payload,
        warnings: [],
    };
}
export function runServiceOverrideSync(services, context) {
    const matched = [...services.overrides].reverse().find((entry) => entry.service === context.service);
    if (!matched) {
        return resolveDefaultServiceResult(context);
    }
    try {
        const result = matched.run({
            service: context.service,
            command: context.command ? normalizeCommandName(context.command) : undefined,
            args: context.args ? cloneContextSnapshot(context.args) : undefined,
            options: context.options ? cloneCommandOptionsSnapshot(context.options) : undefined,
            global: context.global ? cloneGlobalOptionsSnapshot(context.global) : undefined,
            pm_root: context.pm_root,
            payload: cloneContextSnapshot(context.payload),
        });
        if (result instanceof Promise) {
            return {
                handled: false,
                result: context.payload,
                warnings: [`extension_service_override_async_unsupported:${matched.layer}:${matched.name}:${matched.service}`],
            };
        }
        return {
            handled: true,
            result,
            warnings: [],
        };
    }
    catch {
        return {
            handled: false,
            result: context.payload,
            warnings: [`extension_service_override_failed:${matched.layer}:${matched.name}:${matched.service}`],
        };
    }
}
export async function runServiceOverride(services, context) {
    const matched = [...services.overrides].reverse().find((entry) => entry.service === context.service);
    if (!matched) {
        return resolveDefaultServiceResult(context);
    }
    try {
        const result = await Promise.resolve(matched.run({
            service: context.service,
            command: context.command ? normalizeCommandName(context.command) : undefined,
            args: context.args ? cloneContextSnapshot(context.args) : undefined,
            options: context.options ? cloneCommandOptionsSnapshot(context.options) : undefined,
            global: context.global ? cloneGlobalOptionsSnapshot(context.global) : undefined,
            pm_root: context.pm_root,
            payload: cloneContextSnapshot(context.payload),
        }));
        return {
            handled: true,
            result,
            warnings: [],
        };
    }
    catch {
        return {
            handled: false,
            result: context.payload,
            warnings: [`extension_service_override_failed:${matched.layer}:${matched.name}:${matched.service}`],
        };
    }
}
export function runCommandOverride(commands, context) {
    const command = normalizeCommandName(context.command);
    if (command.length === 0) {
        return {
            overridden: false,
            result: context.result,
            warnings: [],
        };
    }
    const matched = [...commands.overrides].reverse().find((entry) => entry.command === command);
    if (!matched) {
        return {
            overridden: false,
            result: context.result,
            warnings: [],
        };
    }
    try {
        const overrideOptions = cloneCommandOptionsSnapshot(context.options);
        const overrideGlobal = cloneGlobalOptionsSnapshot(context.global);
        const overrideResult = matched.run({
            command,
            args: cloneContextSnapshot(context.args),
            options: overrideOptions,
            global: overrideGlobal,
            pm_root: context.pm_root,
            result: cloneContextSnapshot(context.result),
        });
        if (overrideResult instanceof Promise) {
            return {
                overridden: false,
                result: context.result,
                warnings: [`extension_command_override_async_unsupported:${matched.layer}:${matched.name}:${matched.command}`],
            };
        }
        return {
            overridden: true,
            result: overrideResult,
            warnings: [],
        };
    }
    catch {
        return {
            overridden: false,
            result: context.result,
            warnings: [`extension_command_override_failed:${matched.layer}:${matched.name}:${matched.command}`],
        };
    }
}
export function runRendererOverride(renderers, context) {
    const matched = [...renderers.overrides].reverse().find((entry) => entry.format === context.format);
    if (!matched) {
        return {
            overridden: false,
            rendered: null,
            warnings: [],
        };
    }
    try {
        const rendererCommand = typeof context.command === "string" ? normalizeCommandName(context.command) : "";
        const rendererArgs = Array.isArray(context.args) ? cloneContextSnapshot(context.args) : [];
        const rendererOptions = cloneCommandOptionsSnapshot(context.options);
        const rendererGlobal = cloneGlobalOptionsSnapshot(context.global);
        const rendererPmRoot = typeof context.pm_root === "string" ? context.pm_root : "";
        const rendered = matched.run({
            format: context.format,
            command: rendererCommand,
            args: rendererArgs,
            options: rendererOptions,
            global: rendererGlobal,
            pm_root: rendererPmRoot,
            result: cloneContextSnapshot(context.result),
        });
        if (typeof rendered !== "string") {
            return {
                overridden: false,
                rendered: null,
                warnings: [`extension_renderer_invalid_result:${matched.layer}:${matched.name}:${matched.format}`],
            };
        }
        return {
            overridden: true,
            rendered,
            warnings: [],
        };
    }
    catch {
        return {
            overridden: false,
            rendered: null,
            warnings: [`extension_renderer_failed:${matched.layer}:${matched.name}:${matched.format}`],
        };
    }
}
