// @unbrained/pm-cli/dist/cli/commands/aggregate.js
import { EXIT_CODE } from "../../core/shared/constants.js";
import { PmCliError } from "../../core/shared/errors.js";
import { nowIso } from "../../core/shared/time.js";
import { normalizeStatusInput } from "../../core/item/status.js";
import { resolveRuntimeStatusRegistry } from "../../core/schema/runtime-schema.js";
import { resolvePmRoot } from "../../core/store/paths.js";
import { readSettings } from "../../core/store/settings.js";
import { runList } from "./list.js";
const AGGREGATE_GROUP_FIELDS = [
    "parent",
    "type",
    "priority",
    "status",
    "assignee",
    "tags",
    "sprint",
    "release",
];
function parseStatus(raw, statusRegistry) {
    if (raw === undefined) {
        return undefined;
    }
    const normalized = normalizeStatusInput(raw, statusRegistry);
    if (!normalized) {
        throw new PmCliError(`Status filter must be one of ${statusRegistry.definitions.map((definition) => definition.id).join("|")}`, EXIT_CODE.USAGE);
    }
    return normalized;
}
function parseGroupBy(raw) {
    const value = raw?.trim() ?? "parent,type";
    if (value.length === 0) {
        throw new PmCliError("--group-by requires at least one field name", EXIT_CODE.USAGE);
    }
    const requested = [...new Set(value.split(",").map((entry) => entry.trim()).filter((entry) => entry.length > 0))];
    if (requested.length === 0) {
        throw new PmCliError("--group-by requires a comma-separated list of fields", EXIT_CODE.USAGE);
    }
    for (const field of requested) {
        if (!AGGREGATE_GROUP_FIELDS.includes(field)) {
            throw new PmCliError(`Aggregate group fields must be one of ${AGGREGATE_GROUP_FIELDS.join("|")}; received "${field}"`, EXIT_CODE.USAGE);
        }
    }
    return requested;
}
function normalizeTagGroupValue(tags) {
    const normalized = [...new Set(tags.map((tag) => tag.trim().toLowerCase()).filter((tag) => tag.length > 0))].sort((left, right) => left.localeCompare(right));
    if (normalized.length === 0) {
        return null;
    }
    return normalized.join(",");
}
function resolveGroupValue(field, item) {
    switch (field) {
        case "parent":
            return item.parent ?? null;
        case "type":
            return item.type;
        case "priority":
            return item.priority;
        case "status":
            return item.status;
        case "assignee":
            return item.assignee ?? null;
        case "tags":
            return normalizeTagGroupValue(item.tags);
        case "sprint":
            return item.sprint ?? null;
        case "release":
            return item.release ?? null;
        default:
            return null;
    }
}
function compareNullableGroupValue(left, right) {
    const leftValue = left ?? null;
    const rightValue = right ?? null;
    if (leftValue === rightValue) {
        return 0;
    }
    if (leftValue === null) {
        return 1;
    }
    if (rightValue === null) {
        return -1;
    }
    if (typeof leftValue === "number" && typeof rightValue === "number") {
        return leftValue - rightValue;
    }
    return String(leftValue).localeCompare(String(rightValue));
}
function buildGroupKey(groupBy, group) {
    return groupBy.map((field) => `${field}:${JSON.stringify(group[field] ?? null)}`).join("|");
}
function compareAggregateRows(left, right, groupBy) {
    for (const field of groupBy) {
        const byField = compareNullableGroupValue(left.group[field], right.group[field]);
        if (byField !== 0) {
            return byField;
        }
    }
    return 0;
}
export async function runAggregate(options, global) {
    if (options.count === false) {
        throw new PmCliError("Aggregate currently supports grouped counts only. Pass --count.", EXIT_CODE.USAGE);
    }
    const pmRoot = resolvePmRoot(process.cwd(), global.path);
    const settings = await readSettings(pmRoot);
    const statusRegistry = resolveRuntimeStatusRegistry(settings.schema);
    const groupBy = parseGroupBy(options.groupBy);
    const status = parseStatus(options.status, statusRegistry);
    const includeUnparented = options.includeUnparented === true;
    const listed = await runList(status, {
        type: options.type,
        tag: options.tag,
        priority: options.priority,
        deadlineBefore: options.deadlineBefore,
        deadlineAfter: options.deadlineAfter,
        assignee: options.assignee,
        assigneeFilter: options.assigneeFilter,
        parent: options.parent,
        sprint: options.sprint,
        release: options.release,
    }, global);
    const grouped = new Map();
    let skippedUnparented = 0;
    let groupedItemCount = 0;
    for (const listedItem of listed.items) {
        const item = listedItem;
        const group = {};
        for (const field of groupBy) {
            group[field] = resolveGroupValue(field, item);
        }
        if (groupBy.includes("parent") && group.parent === null && !includeUnparented) {
            skippedUnparented += 1;
            continue;
        }
        const key = buildGroupKey(groupBy, group);
        const existing = grouped.get(key);
        if (existing) {
            existing.count += 1;
        }
        else {
            grouped.set(key, {
                group,
                count: 1,
            });
        }
        groupedItemCount += 1;
    }
    const groups = [...grouped.values()].sort((left, right) => compareAggregateRows(left, right, groupBy));
    const warnings = listed.warnings && listed.warnings.length > 0 ? listed.warnings : undefined;
    return {
        groups,
        count: groups.length,
        totals: {
            items_considered: listed.items.length,
            items_grouped: groupedItemCount,
            items_skipped_unparented: skippedUnparented,
        },
        filters: {
            group_by: groupBy,
            count: true,
            include_unparented: includeUnparented,
            status: status ?? null,
            type: options.type ?? null,
            tag: options.tag ?? null,
            priority: options.priority ?? null,
            deadline_before: options.deadlineBefore ?? null,
            deadline_after: options.deadlineAfter ?? null,
            assignee: options.assignee ?? null,
            assignee_filter: options.assigneeFilter ?? null,
            parent: options.parent ?? null,
            sprint: options.sprint ?? null,
            release: options.release ?? null,
        },
        now: nowIso(),
        ...(warnings ? { warnings } : {}),
    };
}
