// @unbrained/pm-cli/dist/cli/commands/append.js
import { pathExists } from "../../core/fs/fs-utils.js";
import { EXIT_CODE } from "../../core/shared/constants.js";
import { PmCliError } from "../../core/shared/errors.js";
import { createStdinTokenResolver } from "../../core/item/parse.js";
import { mutateItem } from "../../core/store/item-store.js";
import { getSettingsPath, resolvePmRoot } from "../../core/store/paths.js";
import { readSettings } from "../../core/store/settings.js";
function resolveAuthor(candidate, defaultAuthor) {
    const resolved = candidate ?? process.env.PM_AUTHOR ?? defaultAuthor;
    const trimmed = resolved.trim();
    return trimmed || "unknown";
}
export async function runAppend(id, options, global) {
    if (options.body === undefined) {
        throw new PmCliError("Missing required --body text", EXIT_CODE.USAGE);
    }
    const stdinResolver = createStdinTokenResolver();
    const pmRoot = resolvePmRoot(process.cwd(), global.path);
    if (!(await pathExists(getSettingsPath(pmRoot)))) {
        throw new PmCliError(`Tracker is not initialized at ${pmRoot}. Run pm init first.`, EXIT_CODE.NOT_FOUND);
    }
    const settings = await readSettings(pmRoot);
    const author = resolveAuthor(options.author, settings.author_default);
    const bodyInput = await stdinResolver.resolveValue(options.body, "--body");
    const appended = (bodyInput ?? "").trim();
    const result = await mutateItem({
        pmRoot,
        settings,
        id,
        op: "append",
        author,
        message: options.message,
        force: options.force,
        mutate(document) {
            if (appended.length === 0) {
                return { changedFields: [] };
            }
            const spacer = document.body.trim().length > 0 ? "\n\n" : "";
            document.body = `${document.body.replace(/\s+$/, "")}${spacer}${appended}\n`;
            return { changedFields: ["body"] };
        },
    });
    return {
        item: result.item,
        appended: appended.length > 0 ? appended : "",
        changed_fields: result.changedFields,
    };
}
