// @unbrained/pm-cli/dist/cli/commands/close.js
import { pathExists } from "../../core/fs/fs-utils.js";
import { normalizeStatusInput } from "../../core/item/status.js";
import { resolveRuntimeStatusRegistry } from "../../core/schema/runtime-schema.js";
import { EXIT_CODE } from "../../core/shared/constants.js";
import { PmCliError } from "../../core/shared/errors.js";
import { mutateItem } from "../../core/store/item-store.js";
import { getSettingsPath, resolvePmRoot } from "../../core/store/paths.js";
import { readSettings } from "../../core/store/settings.js";
function toAuthor(candidate, defaultAuthor) {
    const resolved = candidate ?? process.env.PM_AUTHOR ?? defaultAuthor;
    const trimmed = resolved.trim();
    return trimmed || "unknown";
}
function ensureCloseReason(reasonText) {
    const reason = reasonText.trim();
    if (reason.length === 0) {
        throw new PmCliError("Close reason text must not be empty", EXIT_CODE.USAGE);
    }
    return reason;
}
function isTerminal(status, statusRegistry) {
    const normalized = normalizeStatusInput(status, statusRegistry) ?? status;
    return statusRegistry.terminal_statuses.has(normalized);
}
const CLOSE_VALIDATION_FIELDS = [
    { key: "resolution", label: "resolution" },
    { key: "expected_result", label: "expected_result" },
    { key: "actual_result", label: "actual_result" },
];
function parseValidateCloseMode(raw) {
    if (raw === undefined) {
        return undefined;
    }
    const normalized = raw.trim().toLowerCase();
    if (normalized.length === 0 || normalized === "warn") {
        return "warn";
    }
    if (normalized === "off" || normalized === "none" || normalized === "disabled") {
        return "off";
    }
    if (normalized === "strict") {
        return "strict";
    }
    throw new PmCliError(`Invalid --validate-close mode "${raw}" (expected "off", "warn", or "strict")`, EXIT_CODE.USAGE);
}
function findMissingCloseValidationFields(frontMatter) {
    const missing = [];
    for (const field of CLOSE_VALIDATION_FIELDS) {
        const rawValue = frontMatter[field.key];
        if (typeof rawValue !== "string" || rawValue.trim().length === 0) {
            missing.push(field.label);
        }
    }
    return missing;
}
export async function runClose(id, closeReasonText, options, global) {
    const pmRoot = resolvePmRoot(process.cwd(), global.path);
    if (!(await pathExists(getSettingsPath(pmRoot)))) {
        throw new PmCliError(`Tracker is not initialized at ${pmRoot}. Run pm init first.`, EXIT_CODE.NOT_FOUND);
    }
    const settings = await readSettings(pmRoot);
    const statusRegistry = resolveRuntimeStatusRegistry(settings.schema);
    const author = toAuthor(options.author, settings.author_default);
    const closeReason = ensureCloseReason(closeReasonText);
    const validateCloseMode = parseValidateCloseMode(options.validateClose) ?? settings.governance.close_validation_default;
    const result = await mutateItem({
        pmRoot,
        settings,
        id,
        op: "close",
        author,
        message: options.message,
        force: options.force,
        mutate(document) {
            if (isTerminal(document.metadata.status, statusRegistry) && !options.force) {
                throw new PmCliError(`Item ${document.metadata.id} is already terminal; use --force to close again.`, EXIT_CODE.CONFLICT);
            }
            const mutationWarnings = [];
            if (validateCloseMode !== "off") {
                const missingFields = findMissingCloseValidationFields(document.metadata);
                if (missingFields.length > 0) {
                    if (validateCloseMode === "strict") {
                        throw new PmCliError(`Cannot close item ${document.metadata.id}: missing ${missingFields.join(", ")}. Populate fields or use --validate-close warn.`, EXIT_CODE.USAGE);
                    }
                    mutationWarnings.push(`close_validation_missing_fields:${document.metadata.id}:${missingFields.join(",")}`);
                }
            }
            document.metadata.status = statusRegistry.close_status;
            document.metadata.close_reason = closeReason;
            const changedFields = ["status", "close_reason"];
            if (document.metadata.assignee !== undefined) {
                delete document.metadata.assignee;
                changedFields.push("assignee");
            }
            return {
                changedFields,
                ...(mutationWarnings.length > 0 ? { warnings: mutationWarnings } : {}),
            };
        },
    });
    return {
        item: result.item,
        changed_fields: result.changedFields,
        warnings: result.warnings,
    };
}
