// @unbrained/pm-cli/dist/cli/commands/completion.js
import { EXIT_CODE } from "../../core/shared/constants.js";
import { PmCliError } from "../../core/shared/errors.js";
import { ACTIVITY_FLAG_CONTRACTS, APPEND_FLAG_CONTRACTS, CALENDAR_FLAG_CONTRACTS, COMPLETION_FLAG_CONTRACTS, CONTRACTS_FLAG_CONTRACTS, CONTEXT_FLAG_CONTRACTS, CREATE_FLAG_CONTRACTS, DEPS_FLAG_CONTRACTS, GUIDE_FLAG_CONTRACTS, GLOBAL_FLAG_CONTRACTS, HEALTH_FLAG_CONTRACTS, INIT_FLAG_CONTRACTS, LIST_FILTER_FLAG_CONTRACTS, NORMALIZE_FLAG_CONTRACTS, PM_CORE_COMMAND_NAMES, SEARCH_FLAG_CONTRACTS, UPDATE_FLAG_CONTRACTS, UPDATE_MANY_FLAG_CONTRACTS, toCompletionFlagString, } from "../../sdk/cli-contracts.js";
import { BUILTIN_ITEM_TYPE_VALUES } from "../../types/index.js";
import { listGuideTopicIds } from "../guide-topics.js";
const VALID_SHELLS = ["bash", "zsh", "fish"];
const DEFAULT_ITEM_TYPES = [...BUILTIN_ITEM_TYPE_VALUES];
const DEFAULT_STATUS_VALUES = ["draft", "open", "in_progress", "blocked", "closed", "canceled"];
const ALL_COMMANDS = [...PM_CORE_COMMAND_NAMES];
const LIST_FLAGS = toCompletionFlagString(LIST_FILTER_FLAG_CONTRACTS);
const APPEND_FLAGS = toCompletionFlagString(APPEND_FLAG_CONTRACTS);
const CREATE_FLAGS = toCompletionFlagString(CREATE_FLAG_CONTRACTS);
const UPDATE_FLAGS = toCompletionFlagString(UPDATE_FLAG_CONTRACTS);
const UPDATE_MANY_FLAGS = toCompletionFlagString(UPDATE_MANY_FLAG_CONTRACTS);
const NORMALIZE_FLAGS = toCompletionFlagString(NORMALIZE_FLAG_CONTRACTS);
const ACTIVITY_FLAGS = toCompletionFlagString(ACTIVITY_FLAG_CONTRACTS);
const CALENDAR_FLAGS = toCompletionFlagString(CALENDAR_FLAG_CONTRACTS);
const CONTEXT_FLAGS = toCompletionFlagString(CONTEXT_FLAG_CONTRACTS);
const DEPS_FLAGS = toCompletionFlagString(DEPS_FLAG_CONTRACTS);
const GUIDE_FLAGS = toCompletionFlagString(GUIDE_FLAG_CONTRACTS);
const SEARCH_FLAGS = toCompletionFlagString(SEARCH_FLAG_CONTRACTS);
const HEALTH_FLAGS = toCompletionFlagString(HEALTH_FLAG_CONTRACTS);
const INIT_FLAGS = toCompletionFlagString(INIT_FLAG_CONTRACTS);
const CONTRACTS_FLAGS = toCompletionFlagString(CONTRACTS_FLAG_CONTRACTS);
const COMPLETION_FLAGS = toCompletionFlagString(COMPLETION_FLAG_CONTRACTS);
const COMPLETION_SHELL_CHOICES = `${COMPLETION_FLAGS} bash zsh fish`;
const GUIDE_TOPIC_CHOICES = joinCompletionValues(listGuideTopicIds());
const MUTATION_FLAGS = "--author --message --force --json --quiet --path --no-extensions --no-pager --profile --help";
const CLOSE_MUTATION_FLAGS = "--author --message --validate-close --force --json --quiet --path --no-extensions --no-pager --profile --help";
const RELEASE_MUTATION_FLAGS = "--allow-audit-release --author --message --force --json --quiet --path --no-extensions --no-pager --profile --help";
const GLOBAL_FLAGS = GLOBAL_FLAG_CONTRACTS.flatMap((entry) => [entry.short, entry.flag])
    .filter((value) => Boolean(value))
    .join(" ");
function joinCompletionValues(values) {
    return [...new Set(values.map((value) => value.trim()).filter((value) => value.length > 0))]
        .sort((left, right) => left.localeCompare(right))
        .join(" ");
}
function mergeFlagStrings(baseFlags, runtimeFlags) {
    const merged = [...baseFlags.split(/\s+/u).filter((value) => value.length > 0), ...(runtimeFlags ?? [])];
    return joinCompletionValues(merged);
}
export function generateBashScript(itemTypes = DEFAULT_ITEM_TYPES, tags = [], eagerTagExpansion = false, runtime = {}) {
    const cmds = ALL_COMMANDS.join(" ");
    const typeValues = itemTypes.join(" ");
    const statusValues = joinCompletionValues(runtime.statuses ?? DEFAULT_STATUS_VALUES);
    const tagValues = joinCompletionValues(tags);
    const listFlags = mergeFlagStrings(LIST_FLAGS, runtime.command_flags?.list);
    const createFlags = mergeFlagStrings(CREATE_FLAGS, runtime.command_flags?.create);
    const updateFlags = mergeFlagStrings(UPDATE_FLAGS, runtime.command_flags?.update);
    const updateManyFlags = mergeFlagStrings(UPDATE_MANY_FLAGS, runtime.command_flags?.["update-many"]);
    const normalizeFlags = NORMALIZE_FLAGS;
    const searchFlags = mergeFlagStrings(SEARCH_FLAGS, runtime.command_flags?.search);
    const calendarFlags = mergeFlagStrings(CALENDAR_FLAGS, runtime.command_flags?.calendar);
    const contextFlags = mergeFlagStrings(CONTEXT_FLAGS, runtime.command_flags?.context);
    const useEagerTagExpansion = eagerTagExpansion || tags.length > 0;
    // Note: "${...}" inside regular (non-template) strings are literal characters,
    // not JS interpolation. Only backtick template literals interpolate ${...}.
    const compgen = (flags) => `$(compgen -W "${flags}" -- "$cur")`;
    return [
        "# bash completion for pm",
        '# Source this file or add \'eval "$(pm completion bash)"\' to ~/.bashrc',
        "",
        "_pm_completion() {",
        "  local cur prev words cword",
        "  _init_completion 2>/dev/null || {",
        '    cur="${COMP_WORDS[COMP_CWORD]}"',
        '    prev="${COMP_WORDS[COMP_CWORD-1]}"',
        "    cword=$COMP_CWORD",
        "  }",
        "",
        "  if [[ $cword -eq 1 ]]; then",
        `    COMPREPLY=(${compgen(cmds)})`,
        "    return 0",
        "  fi",
        "",
        '  if [[ "$prev" == "--type" ]]; then',
        `    COMPREPLY=(${compgen(typeValues)})`,
        "    return 0",
        "  fi",
        "",
        '  if [[ "$prev" == "--status" ]]; then',
        `    COMPREPLY=(${compgen(statusValues)})`,
        "    return 0",
        "  fi",
        "",
        ...(useEagerTagExpansion
            ? [
                '  if [[ "$prev" == "--tag" ]]; then',
                `    COMPREPLY=(${compgen(tagValues)})`,
                "    return 0",
                "  fi",
            ]
            : [
                '  if [[ "$prev" == "--tag" ]]; then',
                '    local now ttl cache_ts tag_values',
                '    now="$(date +%s 2>/dev/null || echo 0)"',
                '    ttl="${PM_COMPLETION_TAG_TTL:-120}"',
                '    cache_ts="${PM_COMPLETION_TAG_CACHE_TS:-0}"',
                '    tag_values="${PM_COMPLETION_TAG_CACHE:-}"',
                '    if [[ -z "$tag_values" || "$now" -eq 0 || $((now - cache_ts)) -ge "$ttl" ]]; then',
                '      tag_values="$(pm completion-tags 2>/dev/null)"',
                '      PM_COMPLETION_TAG_CACHE="$tag_values"',
                '      PM_COMPLETION_TAG_CACHE_TS="$now"',
                "    fi",
                '    COMPREPLY=($(compgen -W "$tag_values" -- "$cur"))',
                "    return 0",
                "  fi",
            ]),
        "",
        '  local cmd="${COMP_WORDS[1]}"',
        "",
        '  case "$cmd" in',
        "    list|list-all|list-draft|list-open|list-in-progress|list-blocked|list-closed|list-canceled)",
        `      COMPREPLY=(${compgen(listFlags)})`,
        "      ;;",
        "    aggregate)",
        `      COMPREPLY=(${compgen("--group-by --count --include-unparented --status --type --tag --priority --deadline-before --deadline-after --assignee --assignee-filter --parent --sprint --release --json --quiet --path --no-extensions --no-pager --profile --help")})`,
        "      ;;",
        "    dedupe-audit)",
        `      COMPREPLY=(${compgen("--mode --limit --threshold --status --type --tag --priority --deadline-before --deadline-after --assignee --assignee-filter --parent --sprint --release --json --quiet --path --no-extensions --no-pager --profile --help")})`,
        "      ;;",
        "    create)",
        `      COMPREPLY=(${compgen(createFlags)})`,
        "      ;;",
        "    update)",
        `      COMPREPLY=(${compgen(updateFlags)})`,
        "      ;;",
        "    update-many)",
        `      COMPREPLY=(${compgen(updateManyFlags)})`,
        "      ;;",
        "    normalize)",
        `      COMPREPLY=(${compgen(normalizeFlags)})`,
        "      ;;",
        "    calendar|cal)",
        `      COMPREPLY=(${compgen(calendarFlags)})`,
        "      ;;",
        "    context|ctx)",
        `      COMPREPLY=(${compgen(contextFlags)})`,
        "      ;;",
        "    guide)",
        `      COMPREPLY=(${compgen(`${GUIDE_FLAGS} ${GUIDE_TOPIC_CHOICES}`)})`,
        "      ;;",
        "    search)",
        `      COMPREPLY=(${compgen(searchFlags)})`,
        "      ;;",
        "    reindex)",
        `      COMPREPLY=(${compgen("--mode --progress --json --quiet --path --no-extensions --no-pager --profile --help")})`,
        "      ;;",
        "    init)",
        `      COMPREPLY=(${compgen(`${INIT_FLAGS} --json --quiet --path --no-extensions --no-pager --profile --help`)})`,
        "      ;;",
        "    config)",
        `      COMPREPLY=(${compgen("--criterion --clear-criteria --format --policy --json --quiet --path --no-extensions --no-pager --profile --help")})`,
        "      ;;",
        "    extension)",
        `      COMPREPLY=(${compgen("init scaffold install uninstall explore manage doctor adopt adopt-all activate deactivate --init --scaffold --install --uninstall --explore --manage --doctor --adopt --adopt-all --activate --deactivate --project --local --global --gh --github --ref --detail --trace --runtime-probe --fix-managed-state --strict-exit --fail-on-warn --json --quiet --path --no-extensions --no-pager --profile --help")})`,
        "      ;;",
        "    comments)",
        `      COMPREPLY=(${compgen("--add --stdin --file --limit --author --message --allow-audit-comment --force --json --quiet --path --no-extensions --no-pager --profile --help")})`,
        "      ;;",
        "    comments-audit)",
        `      COMPREPLY=(${compgen("--status --type --tag --priority --parent --sprint --release --assignee --assignee-filter --limit-items --limit --full-history --latest --json --quiet --path --no-extensions --no-pager --profile --help")})`,
        "      ;;",
        "    notes)",
        `      COMPREPLY=(${compgen("--add --limit --author --message --allow-audit-note --allow-audit-comment --force --json --quiet --path --no-extensions --no-pager --profile --help")})`,
        "      ;;",
        "    learnings)",
        `      COMPREPLY=(${compgen("--add --limit --author --message --allow-audit-learning --allow-audit-comment --force --json --quiet --path --no-extensions --no-pager --profile --help")})`,
        "      ;;",
        "    files)",
        `      COMPREPLY=(${compgen("discover --add --add-glob --remove --migrate --list --apply --note --append-stable --validate-paths --audit --author --message --force --json --quiet --path --no-extensions --no-pager --profile --help")})`,
        "      ;;",
        "    docs)",
        `      COMPREPLY=(${compgen("--add --add-glob --remove --migrate --validate-paths --audit --author --message --force --json --quiet --path --no-extensions --no-pager --profile --help")})`,
        "      ;;",
        "    append)",
        `      COMPREPLY=(${compgen(APPEND_FLAGS)})`,
        "      ;;",
        "    deps)",
        `      COMPREPLY=(${compgen(DEPS_FLAGS)})`,
        "      ;;",
        "    test)",
        `      COMPREPLY=(${compgen("--add --remove --run --background --timeout --progress --env-set --env-clear --shared-host-safe --pm-context --override-linked-pm-context --fail-on-context-mismatch --fail-on-skipped --fail-on-empty-test-run --require-assertions-for-pm --check-context --auto-pm-context --author --message --force --json --quiet --path --no-extensions --no-pager --profile --help")})`,
        "      ;;",
        "    test-all)",
        `      COMPREPLY=(${compgen("--status --limit --offset --background --timeout --progress --env-set --env-clear --shared-host-safe --pm-context --override-linked-pm-context --fail-on-context-mismatch --fail-on-skipped --fail-on-empty-test-run --require-assertions-for-pm --check-context --auto-pm-context --json --quiet --path --no-extensions --no-pager --profile --help")})`,
        "      ;;",
        "    test-runs)",
        `      COMPREPLY=(${compgen("list status logs stop resume --status --limit --stream --tail --force --author --json --quiet --path --no-extensions --no-pager --profile --help")})`,
        "      ;;",
        "    validate)",
        `      COMPREPLY=(${compgen("--check-metadata --metadata-profile --check-resolution --check-lifecycle --check-stale-blockers --dependency-cycle-severity --check-files --scan-mode --include-pm-internals --verbose-file-lists --strict-exit --fail-on-warn --check-history-drift --check-command-references --json --quiet --path --no-extensions --no-pager --profile --help")})`,
        "      ;;",
        "    health)",
        `      COMPREPLY=(${compgen(HEALTH_FLAGS)})`,
        "      ;;",
        "    history)",
        `      COMPREPLY=(${compgen("--limit --diff --verify --json --quiet --path --no-extensions --no-pager --profile --help")})`,
        "      ;;",
        "    activity)",
        `      COMPREPLY=(${compgen(ACTIVITY_FLAGS)})`,
        "      ;;",
        "    contracts)",
        `      COMPREPLY=(${compgen(CONTRACTS_FLAGS)})`,
        "      ;;",
        "    gc)",
        `      COMPREPLY=(${compgen("--dry-run --scope --json --quiet --path --no-extensions --no-pager --profile --help")})`,
        "      ;;",
        "    close|close-task)",
        `      COMPREPLY=(${compgen(CLOSE_MUTATION_FLAGS)})`,
        "      ;;",
        "    release)",
        `      COMPREPLY=(${compgen(RELEASE_MUTATION_FLAGS)})`,
        "      ;;",
        "    claim|delete|restore|start-task|pause-task)",
        `      COMPREPLY=(${compgen(MUTATION_FLAGS)})`,
        "      ;;",
        "    completion)",
        `      COMPREPLY=(${compgen(COMPLETION_SHELL_CHOICES)})`,
        "      ;;",
        "    templates)",
        `      COMPREPLY=(${compgen("save list show")})`,
        "      ;;",
        "    *)",
        `      COMPREPLY=(${compgen(GLOBAL_FLAGS)})`,
        "      ;;",
        "  esac",
        "  return 0",
        "}",
        "",
        "complete -F _pm_completion pm",
    ].join("\n");
}
export function generateZshScript(itemTypes = DEFAULT_ITEM_TYPES, tags = [], eagerTagExpansion = false) {
    const cmds = ALL_COMMANDS.map((c) => `'${c}'`).join(" ");
    const typeChoices = itemTypes.join(" ");
    const guideTopicChoices = GUIDE_TOPIC_CHOICES;
    const tagChoices = joinCompletionValues(tags);
    const useEagerTagExpansion = eagerTagExpansion || tags.length > 0;
    const zshTagChoices = useEagerTagExpansion ? tagChoices : '${(f)"$(_pm_tag_choices)"}';
    const dynamicTagResolver = useEagerTagExpansion
        ? ""
        : `
_pm_tag_choices() {
  local now ttl cache_ts
  now=\${EPOCHSECONDS:-0}
  ttl=\${PM_COMPLETION_TAG_TTL:-120}
  cache_ts=\${PM_COMPLETION_TAG_CACHE_TS:-0}
  if [[ -n "\${PM_COMPLETION_TAG_CACHE:-}" && "$now" -ne 0 && $((now - cache_ts)) -lt "$ttl" ]]; then
    print -r -- "$PM_COMPLETION_TAG_CACHE"
    return
  fi
  PM_COMPLETION_TAG_CACHE="$(pm completion-tags 2>/dev/null)"
  PM_COMPLETION_TAG_CACHE_TS="$now"
  print -r -- "$PM_COMPLETION_TAG_CACHE"
}
`;
    return `#compdef pm
# zsh completion for pm
# Source this file or add 'eval "$(pm completion zsh)"' to ~/.zshrc

_pm_commands() {
  local -a commands
  commands=(
    'init:Initialize pm storage for the current workspace'
    'config:Read or update pm settings'
    'extension:Manage extension lifecycle operations'
    'create:Create a new project management item'
    'list:List active items with optional filters'
    'list-all:List all items with optional filters'
    'list-draft:List draft items with optional filters'
    'list-open:List open items with optional filters'
    'list-in-progress:List in-progress items with optional filters'
    'list-blocked:List blocked items with optional filters'
    'list-closed:List closed items with optional filters'
    'list-canceled:List canceled items with optional filters'
    'aggregate:Aggregate grouped item counts for governance queries'
    'dedupe-audit:Audit potential duplicate items and emit merge suggestions'
    'guide:Browse local progressive-disclosure guides'
    'calendar:Show calendar views for deadlines and reminders'
    'cal:Alias for calendar'
    'context:Show a token-efficient project context snapshot'
    'ctx:Alias for context'
    'get:Show item details by ID'
    'search:Search items with keyword, semantic, or hybrid modes'
    'reindex:Rebuild search artifacts'
    'history:Show item history entries'
    'activity:Show recent activity across items'
    'restore:Restore an item to an earlier state'
    'update:Update item fields and metadata'
    'update-many:Bulk-update matched items with dry-run and rollback checkpoints'
    'normalize:Normalize lifecycle metadata with dry-run planning or apply mode'
    'close:Close an item with a required reason'
    'delete:Delete an item and record the change'
    'append:Append text to an item body'
    'comments:List or add comments for an item'
    'comments-audit:Audit latest comments or full history across filtered items'
    'notes:List or add notes for an item'
    'learnings:List or add learnings for an item'
    'files:Manage linked files'
    'docs:Manage linked docs'
    'deps:Show dependency relationships for an item'
    'test:Manage linked tests and optionally run them'
    'test-all:Run linked tests across matching items'
    'test-runs:Manage background linked-test runs'
    'stats:Show project tracker statistics'
    'health:Show project tracker health checks'
    'validate:Run standalone validation checks'
    'gc:Clean optional cache artifacts'
    'contracts:Show machine-readable command and schema contracts'
    'claim:Claim an item for active work'
    'release:Release the active claim for an item'
    'start-task:Lifecycle alias to claim and set in_progress'
    'pause-task:Lifecycle alias to reopen and release claim'
    'close-task:Lifecycle alias to close and release claim'
    'templates:Manage reusable create templates'
    'completion:Generate shell completion'
    'help:Display help for a command'
  )
  _describe 'command' commands
}
${dynamicTagResolver}

_pm() {
  local context state line
  _arguments -C \\
    '--json[Output JSON instead of TOON]' \\
    '--quiet[Suppress stdout output]' \\
    '--path[Override PM path for this command]:path:_files -/' \\
    '--no-extensions[Disable extension loading]' \\
    '--no-pager[Disable pager integration for help and long output]' \\
    '--profile[Print deterministic timing diagnostics]' \\
    '(-V --version)--version[Output the version number]' \\
    '(-h --help)--help[Display help]' \\
    '1: :_pm_commands' \\
    '*:: :->args' && return 0

  case $state in
    args)
      case $line[1] in
        list|list-all|list-draft|list-open|list-in-progress|list-blocked|list-closed|list-canceled)
          _arguments \\
            '--type[Filter by item type]:(${typeChoices})' \\
            '--tag[Filter by tag]:(${zshTagChoices})' \\
            '--priority[Filter by priority]:(0 1 2 3 4)' \\
            '--deadline-before[Filter by deadline upper bound (ISO/date string or relative)]:date' \\
            '--deadline-after[Filter by deadline lower bound (ISO/date string or relative)]:date' \\
            '--assignee[Filter by assignee]:assignee' \\
            '--assignee-filter[Filter assignee presence]:(assigned unassigned)' \\
            '--sprint[Filter by sprint]:sprint' \\
            '--release[Filter by release]:release' \\
            '--limit[Limit returned item count]:number' \\
            '--offset[Skip the first n matching rows before limit]:number' \\
            '--include-body[Include item body in each returned list row]' \\
            '--compact[Render compact list projection fields]' \\
            '--fields[Render custom comma-separated list fields]:fields' \\
            '--sort[Sort field]:(priority deadline updated_at created_at title parent)' \\
            '--order[Sort order (requires --sort)]:(asc desc)' \\
            '--stream[Emit line-delimited JSON rows (requires --json)]' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]' \\
            '--path[Override PM path]:path:_files -/'
          ;;
        aggregate)
          _arguments \\
            '--group-by[Comma-separated group-by fields (supported: parent,type,priority,status,assignee,tags,sprint,release)]:fields' \\
            '--count[Return grouped counts]' \\
            '--include-unparented[Include unparented rows when grouping by parent]' \\
            '--status[Filter by status]:(draft open in_progress blocked closed canceled)' \\
            '--type[Filter by item type]:(${typeChoices})' \\
            '--tag[Filter by tag]:(${zshTagChoices})' \\
            '--priority[Filter by priority]:(0 1 2 3 4)' \\
            '--deadline-before[Filter by deadline upper bound (ISO/date string or relative)]:date' \\
            '--deadline-after[Filter by deadline lower bound (ISO/date string or relative)]:date' \\
            '--assignee[Filter by assignee]:assignee' \\
            '--assignee-filter[Filter assignee presence]:(assigned unassigned)' \\
            '--parent[Filter by parent item ID]:parent_id' \\
            '--sprint[Filter by sprint]:sprint' \\
            '--release[Filter by release]:release' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        dedupe-audit)
          _arguments \\
            '--mode[Dedupe mode]:(title_exact title_fuzzy parent_scope)' \\
            '--limit[Limit returned duplicate clusters]:number' \\
            '--threshold[Fuzzy mode token similarity threshold between 0 and 1]:number' \\
            '--status[Filter by status]:(draft open in_progress blocked closed canceled)' \\
            '--type[Filter by item type]:(${typeChoices})' \\
            '--tag[Filter by tag]:(${zshTagChoices})' \\
            '--priority[Filter by priority]:(0 1 2 3 4)' \\
            '--deadline-before[Filter by deadline upper bound (ISO/date string or relative)]:date' \\
            '--deadline-after[Filter by deadline lower bound (ISO/date string or relative)]:date' \\
            '--assignee[Filter by assignee]:assignee' \\
            '--assignee-filter[Filter assignee presence]:(assigned unassigned)' \\
            '--parent[Filter by parent item ID]:parent_id' \\
            '--sprint[Filter by sprint]:sprint' \\
            '--release[Filter by release]:release' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        create)
          _arguments \\
            '(-t --title)'{-t,--title}'[Item title]:title' \\
            '(-d --description)'{-d,--description}'[Item description]:description' \\
            '--type[Item type]:(${typeChoices})' \\
            '--create-mode[Create required-option policy mode]:(strict progressive)' \\
            '--schedule-preset[Scheduling preset for Reminder/Meeting/Event]:(lightweight)' \\
            '(-s --status)'{-s,--status}'[Item status]:(draft open in_progress blocked)' \\
            '(-p --priority)'{-p,--priority}'[Priority (0-4)]:(0 1 2 3 4)' \\
            '--tags[Comma-separated tags]:tags' \\
            '(-b --body)'{-b,--body}'[Item body]:body' \\
            '--deadline[Deadline (ISO/date string or relative +6h/+1d/+2w/+6m)]:deadline' \\
            '--estimate[Estimated minutes]:minutes' \\
            '--acceptance-criteria[Acceptance criteria]:criteria' \\
            '--reminder[Reminder entry at=<iso|relative>,text=<text>]:reminder' \\
            '--event[Event entry start=<iso|relative>,end=<iso|relative>,recur_*]:event' \\
            '--type-option[Type option key=value or key=<name>,value=<value>]:type_option' \\
            '--unset[Clear scalar metadata field by name]:field' \\
            '--replace-deps[Atomically replace dependencies with provided --dep values]' \\
            '--replace-tests[Atomically replace linked tests with provided --test values]' \\
            '--clear-deps[Clear dependency entries]' \\
            '--clear-comments[Clear comments]' \\
            '--clear-notes[Clear notes]' \\
            '--clear-learnings[Clear learnings]' \\
            '--clear-files[Clear linked files]' \\
            '--clear-tests[Clear linked tests]' \\
            '--clear-docs[Clear linked docs]' \\
            '--clear-reminders[Clear reminders]' \\
            '--clear-events[Clear events]' \\
            '--clear-type-options[Clear type options]' \\
            '--author[Mutation author]:author' \\
            '--message[History message]:message' \\
            '--assignee[Assignee]:assignee' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        update)
          _arguments \\
            '(-t --title)'{-t,--title}'[Item title]:title' \\
            '(-d --description)'{-d,--description}'[Item description]:description' \\
            '(-b --body)'{-b,--body}'[Item body]:body' \\
            '(-s --status)'{-s,--status}'[Item status]:(draft open in_progress blocked canceled)' \\
            '--close-reason[Set close reason]:close_reason' \\
            '(-p --priority)'{-p,--priority}'[Priority (0-4)]:(0 1 2 3 4)' \\
            '--type[Item type]:(${typeChoices})' \\
            '--tags[Comma-separated tags]:tags' \\
            '--comment[Comment seed author=<value>,created_at=<iso|now>,text=<value>]:comment' \\
            '--note[Note seed author=<value>,created_at=<iso|now>,text=<value>]:note' \\
            '--learning[Learning seed author=<value>,created_at=<iso|now>,text=<value>]:learning' \\
            '--file[Linked file path=<value>,scope=<project|global>,note=<text>]:file' \\
            '--test[Linked test command=<value>,path=<value>,scope=<project|global>]:test' \\
            '--doc[Linked doc path=<value>,scope=<project|global>,note=<text>]:doc' \\
            '--reminder[Reminder entry at=<iso|relative>,text=<text>]:reminder' \\
            '--event[Event entry start=<iso|relative>,end=<iso|relative>,recur_*]:event' \\
            '--type-option[Type option key=value or key=<name>,value=<value>]:type_option' \\
            '--unset[Clear scalar metadata field by name]:field' \\
            '--clear-deps[Clear dependency entries]' \\
            '--clear-comments[Clear comments]' \\
            '--clear-notes[Clear notes]' \\
            '--clear-learnings[Clear learnings]' \\
            '--clear-files[Clear linked files]' \\
            '--clear-tests[Clear linked tests]' \\
            '--clear-docs[Clear linked docs]' \\
            '--clear-reminders[Clear reminders]' \\
            '--clear-events[Clear events]' \\
            '--clear-type-options[Clear type options]' \\
            '--allow-audit-update[Allow non-owner metadata-only audit updates without requiring --force]' \\
            '--author[Mutation author]:author' \\
            '--message[History message]:message' \\
            '--force[Force override]' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        update-many)
          _arguments \\
            '--filter-status[Filter by status before applying updates]:(draft open in_progress blocked closed canceled)' \\
            '--filter-type[Filter by type before applying updates]:(${typeChoices})' \\
            '--filter-tag[Filter by tag before applying updates]:(${zshTagChoices})' \\
            '--filter-priority[Filter by priority before applying updates]:(0 1 2 3 4)' \\
            '--filter-deadline-before[Filter by deadline upper bound]:deadline' \\
            '--filter-deadline-after[Filter by deadline lower bound]:deadline' \\
            '--filter-assignee[Filter by assignee before applying updates]:assignee' \\
            '--filter-assignee-filter[Filter assignee presence]:(assigned unassigned)' \\
            '--filter-parent[Filter by parent item ID]:parent' \\
            '--filter-sprint[Filter by sprint]:sprint' \\
            '--filter-release[Filter by release]:release' \\
            '--limit[Limit matched item count]:number' \\
            '--offset[Skip first n matched rows]:number' \\
            '--dry-run[Preview updates without mutating]' \\
            '--rollback[Rollback checkpoint ID]:checkpoint_id' \\
            '--no-checkpoint[Disable checkpoint creation during apply mode]' \\
            '(-t --title)'{-t,--title}'[Item title]:title' \\
            '(-d --description)'{-d,--description}'[Item description]:description' \\
            '(-b --body)'{-b,--body}'[Item body]:body' \\
            '(-p --priority)'{-p,--priority}'[Priority (0-4)]:(0 1 2 3 4)' \\
            '--type[Item type]:(${typeChoices})' \\
            '--tags[Comma-separated tags]:tags' \\
            '--deadline[Deadline (ISO/date string or relative +6h/+1d/+2w/+6m)]:deadline' \\
            '--estimate[Estimated minutes]:minutes' \\
            '--acceptance-criteria[Acceptance criteria]:criteria' \\
            '--definition-of-ready[Definition of ready]:definition_of_ready' \\
            '--order[Planning order/rank]:order' \\
            '--goal[Goal identifier]:goal' \\
            '--objective[Objective identifier]:objective' \\
            '--value[Business value summary]:value' \\
            '--impact[Business impact summary]:impact' \\
            '--outcome[Expected outcome summary]:outcome' \\
            '--why-now[Why-now rationale]:why_now' \\
            '--reviewer[Reviewer]:reviewer' \\
            '--risk[Risk level]:risk' \\
            '--confidence[Confidence level]:confidence' \\
            '--sprint[Sprint identifier]:sprint' \\
            '--release[Release identifier]:release' \\
            '--reporter[Issue reporter]:reporter' \\
            '--severity[Issue severity]:severity' \\
            '--environment[Issue environment context]:environment' \\
            '--repro-steps[Issue reproduction steps]:repro_steps' \\
            '--resolution[Issue resolution summary]:resolution' \\
            '--expected-result[Issue expected behavior]:expected_result' \\
            '--actual-result[Issue observed behavior]:actual_result' \\
            '--affected-version[Affected version identifier]:affected_version' \\
            '--fixed-version[Fixed version identifier]:fixed_version' \\
            '--component[Issue component ownership]:component' \\
            '--regression[Regression marker true|false|1|0]:regression' \\
            '--customer-impact[Customer impact summary]:customer_impact' \\
            '--dep[Dependency seed id=<id>,kind=<kind>,author=<author>,created_at=<timestamp>]:dep' \\
            '--dep-remove[Dependency removal selector id=<id>,kind=<kind>,author=<author>,created_at=<timestamp>]:dep_remove' \\
            '--replace-deps[Atomically replace dependencies with provided --dep values]' \\
            '--replace-tests[Atomically replace linked tests with provided --test values]' \\
            '--comment[Comment seed author=<value>,created_at=<iso|now>,text=<value>]:comment' \\
            '--note[Note seed author=<value>,created_at=<iso|now>,text=<value>]:note' \\
            '--learning[Learning seed author=<value>,created_at=<iso|now>,text=<value>]:learning' \\
            '--file[Linked file path=<value>,scope=<project|global>,note=<text>]:file' \\
            '--test[Linked test command=<value>,path=<value>,scope=<project|global>]:test' \\
            '--doc[Linked doc path=<value>,scope=<project|global>,note=<text>]:doc' \\
            '--reminder[Reminder entry at=<iso|relative>,text=<text>]:reminder' \\
            '--event[Event entry start=<iso|relative>,end=<iso|relative>,recur_*]:event' \\
            '--type-option[Type option key=value or key=<name>,value=<value>]:type_option' \\
            '--unset[Clear scalar metadata field by name]:field' \\
            '--clear-deps[Clear dependency entries]' \\
            '--clear-comments[Clear comments]' \\
            '--clear-notes[Clear notes]' \\
            '--clear-learnings[Clear learnings]' \\
            '--clear-files[Clear linked files]' \\
            '--clear-tests[Clear linked tests]' \\
            '--clear-docs[Clear linked docs]' \\
            '--clear-reminders[Clear reminders]' \\
            '--clear-events[Clear events]' \\
            '--clear-type-options[Clear type options]' \\
            '--allow-audit-update[Allow non-owner metadata-only audit updates without requiring --force]' \\
            '--author[Mutation author]:author' \\
            '--message[History message]:message' \\
            '--force[Force override]' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        normalize)
          _arguments \\
            '--filter-status[Filter by status before planning or apply]:(draft open in_progress blocked closed canceled)' \\
            '--filter-type[Filter by type before planning or apply]:(${typeChoices})' \\
            '--filter-tag[Filter by tag before planning or apply]:(${zshTagChoices})' \\
            '--filter-priority[Filter by priority before planning or apply]:(0 1 2 3 4)' \\
            '--filter-deadline-before[Filter by deadline upper bound]:deadline' \\
            '--filter-deadline-after[Filter by deadline lower bound]:deadline' \\
            '--filter-assignee[Filter by assignee before planning or apply]:assignee' \\
            '--filter-assignee-filter[Filter assignee presence]:(assigned unassigned)' \\
            '--filter-parent[Filter by parent item ID]:parent' \\
            '--filter-sprint[Filter by sprint]:sprint' \\
            '--filter-release[Filter by release]:release' \\
            '--limit[Limit matched item count]:number' \\
            '--offset[Skip first n matched rows]:number' \\
            '--dry-run[Preview normalize findings without mutating]' \\
            '--apply[Apply normalize changes]' \\
            '--allow-audit-update[Allow non-owner metadata-only audit updates without requiring --force]' \\
            '--author[Mutation author]:author' \\
            '--message[History message]:message' \\
            '--force[Force override]' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        calendar|cal)
          _arguments \\
            '--view[Calendar view]:(agenda day week month)' \\
            '--date[Anchor date/time (ISO/date string or relative)]:date' \\
            '--from[Agenda lower bound (ISO/date string or relative)]:date' \\
            '--to[Agenda upper bound (ISO/date string or relative)]:date' \\
            '--past[Include past entries]' \\
            '--full-period[Include full anchored day/week/month period]' \\
            '--type[Filter by type]:(${typeChoices})' \\
            '--tag[Filter by tag]:(${zshTagChoices})' \\
            '--priority[Filter by priority]:(0 1 2 3 4)' \\
            '--status[Filter by status]:(draft open in_progress blocked closed canceled)' \\
            '--assignee[Filter by assignee]:assignee' \\
            '--assignee-filter[Filter assignee presence]:(assigned unassigned)' \\
            '--sprint[Filter by sprint]:sprint' \\
            '--release[Filter by release]:release' \\
            '--include[Include event sources]:(all deadlines reminders events)' \\
            '--recurrence-lookahead-days[Bound open-ended recurrence lookahead]:days' \\
            '--recurrence-lookback-days[Bound open-ended recurrence lookback]:days' \\
            '--occurrence-limit[Cap occurrences per recurring event]:number' \\
            '--limit[Limit returned events]:number' \\
            '--format[Output override]:(markdown toon json)' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        context|ctx)
          _arguments \\
            '--date[Anchor date/time (ISO/date string or relative)]:date' \\
            '--from[Agenda lower bound (ISO/date string or relative)]:date' \\
            '--to[Agenda upper bound (ISO/date string or relative)]:date' \\
            '--past[Include past entries in bounded windows]' \\
            '--type[Filter by type]:(${typeChoices})' \\
            '--tag[Filter by tag]:(${zshTagChoices})' \\
            '--priority[Filter by priority]:(0 1 2 3 4)' \\
            '--assignee[Filter by assignee]:assignee' \\
            '--assignee-filter[Filter assignee presence]:(assigned unassigned)' \\
            '--sprint[Filter by sprint]:sprint' \\
            '--release[Filter by release]:release' \\
            '--limit[Limit focus and agenda rows per section]:number' \\
            '--format[Output override]:(markdown toon json)' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        guide)
          _arguments \\
            '1:topic:(${guideTopicChoices})' \\
            '--list[Show guide topic index]' \\
            '--format[Output override]:(markdown toon json)' \\
            '--depth[Guide detail depth]:(brief standard deep)' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        search)
          _arguments \\
            '--mode[Search mode]:(keyword semantic hybrid)' \\
            '--include-linked[Include linked content in scoring]' \\
            '--limit[Max results]:number' \\
            '--type[Filter by type]:(${typeChoices})' \\
            '--tag[Filter by tag]:(${zshTagChoices})' \\
            '--priority[Filter by priority]:(0 1 2 3 4)' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        reindex)
          _arguments \\
            '--mode[Reindex mode]:(keyword semantic hybrid)' \\
            '--progress[Emit progress updates to stderr]' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        history)
          _arguments \\
            '--limit[Max entries]:number' \\
            '--diff[Include changed-field patch summary]' \\
            '--verify[Verify history hash chain and replay integrity]' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        activity)
          _arguments \\
            '--id[Filter by item ID]:id' \\
            '--op[Filter by history operation]:op' \\
            '--author[Filter by history author]:author' \\
            '--from[Lower timestamp bound (ISO/date string or relative)]:date' \\
            '--to[Upper timestamp bound (ISO/date string or relative)]:date' \\
            '--limit[Max entries]:number' \\
            '--stream[Emit line-delimited JSON rows]:mode' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        contracts)
          _arguments \\
            '--action[Filter schema by tool action]:action' \\
            '--command[Scope output to one command (narrow-by-default)]:command' \\
            '--schema-only[Return schema-only payload]' \\
            '--flags-only[Return command flag contracts only]' \\
            '--availability-only[Return action availability only]' \\
            '--runtime-only[Include only actions invocable in the current runtime]' \\
            '--active-only[Alias for --runtime-only]' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        gc)
          _arguments \\
            '--dry-run[Preview cleanup targets without deleting files]' \\
            '--scope[Limit cleanup to one or more scopes]:scope' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        comments)
          _arguments \\
            '--add[Add one entry (plain text, text=<value>, markdown pairs, or - for stdin)]:text' \\
            '--stdin[Read comment text from stdin (supports multiline markdown)]' \\
            '--file[Read comment text from file (supports multiline markdown)]:path' \\
            '--limit[Return only latest n entries]:number' \\
            '--author[Entry author (falls back to PM_AUTHOR/settings)]:author' \\
            '--message[History message]:message' \\
            '--allow-audit-comment[Allow non-owner append-only comment audits without requiring --force]' \\
            '--force[Force override]' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        notes)
          _arguments \\
            '--add[Add one entry (plain text, text=<value>, markdown pairs, or - for stdin)]:text' \\
            '--limit[Return only latest n entries]:number' \\
            '--author[Entry author (falls back to PM_AUTHOR/settings)]:author' \\
            '--message[History message]:message' \\
            '--allow-audit-note[Allow non-owner append-only note audits without requiring --force]' \\
            '--allow-audit-comment[Backward-compatible alias for --allow-audit-note]' \\
            '--force[Force override]' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        learnings)
          _arguments \\
            '--add[Add one entry (plain text, text=<value>, markdown pairs, or - for stdin)]:text' \\
            '--limit[Return only latest n entries]:number' \\
            '--author[Entry author (falls back to PM_AUTHOR/settings)]:author' \\
            '--message[History message]:message' \\
            '--allow-audit-learning[Allow non-owner append-only learning audits without requiring --force]' \\
            '--allow-audit-comment[Backward-compatible alias for --allow-audit-learning]' \\
            '--force[Force override]' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        deps)
          _arguments \\
            '--format[Output format]:(tree graph)' \\
            '--max-depth[Maximum traversal depth (0 keeps root only)]:depth' \\
            '--collapse[Collapse mode]:(none repeated)' \\
            '--summary[Return counts only without tree/graph payload]' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        test)
          _arguments \\
            '--add[Add linked test entry]:entry' \\
            '--remove[Remove linked test entry by command/path]:entry' \\
            '--run[Run linked tests]' \\
            '--background[Run linked tests in managed background mode]' \\
            '--timeout[Default timeout seconds]:seconds' \\
            '--progress[Emit linked-test progress to stderr]' \\
            '--env-set[Set linked-test runtime environment values]:entry' \\
            '--env-clear[Clear linked-test runtime environment values]:name' \\
            '--shared-host-safe[Apply shared-host-safe runtime defaults]' \\
            '--pm-context[PM linked-test context mode]:(schema tracker auto)' \\
            '--override-linked-pm-context[Force run-level --pm-context over per-linked-test pm_context_mode metadata]' \\
            '--fail-on-context-mismatch[Fail when context item counts mismatch]' \\
            '--fail-on-skipped[Treat skipped linked tests as dependency failures]' \\
            '--fail-on-empty-test-run[Treat empty linked-test selections as failures]' \\
            '--require-assertions-for-pm[Require assertions for linked PM command tests]' \\
            '--check-context[Preflight linked PM command context diagnostics before execution]' \\
            '--auto-pm-context[Auto-remediate tracker-read context mismatches using tracker context]' \\
            '--author[Mutation author]:author' \\
            '--message[History message]:message' \\
            '--force[Force override]' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        test-all)
          _arguments \\
            '--status[Filter by status]:(open in_progress)' \\
            '--limit[Limit matching items before running linked tests]:number' \\
            '--offset[Skip matching items before running linked tests]:number' \\
            '--background[Run linked tests in managed background mode]' \\
            '--timeout[Default timeout seconds]:seconds' \\
            '--progress[Emit linked-test progress to stderr]' \\
            '--env-set[Set linked-test runtime environment values]:entry' \\
            '--env-clear[Clear linked-test runtime environment values]:name' \\
            '--shared-host-safe[Apply shared-host-safe runtime defaults]' \\
            '--pm-context[PM linked-test context mode]:(schema tracker auto)' \\
            '--override-linked-pm-context[Force run-level --pm-context over per-linked-test pm_context_mode metadata]' \\
            '--fail-on-context-mismatch[Fail when context item counts mismatch]' \\
            '--fail-on-skipped[Treat skipped linked tests as dependency failures]' \\
            '--fail-on-empty-test-run[Treat empty linked-test selections as failures]' \\
            '--require-assertions-for-pm[Require assertions for linked PM command tests]' \\
            '--check-context[Preflight linked PM command context diagnostics before execution]' \\
            '--auto-pm-context[Auto-remediate tracker-read context mismatches using tracker context]' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        test-runs)
          _arguments \\
            '1:subcommand:(list status logs stop resume)' \\
            '--status[Filter by background run status]:status:(queued running passed failed stopped canceled)' \\
            '--limit[Limit returned runs]:number' \\
            '--stream[Background log stream]:stream:(stdout stderr both)' \\
            '--tail[Tail number of lines]:number' \\
            '--force[Force stop with SIGKILL]' \\
            '--author[Resume author]:author' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        init)
          _arguments \\
            '--preset[Governance preset for new setups]:preset:(minimal default strict)' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        init)
          _arguments \\
            '--preset[Governance preset for new setups]:preset:(minimal default strict)' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        config)
          _arguments \\
            '--criterion[Criteria value for definition-of-done metadata-required-fields or lifecycle pattern keys (repeatable for set)]:criterion' \\
            '--clear-criteria[Clear config criteria-list key values]' \\
            '--format[Item format for item-format key]:format:(toon)' \\
            '--policy[Policy value for supported policy keys]:policy' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        close)
          _arguments \\
            '--author[Mutation author]:author' \\
            '--message[History message]:message' \\
            '--validate-close[Validate closure metadata mode]:(off warn strict)' \\
            '--force[Force override]' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        claim)
          _arguments \\
            '--author[Mutation author]:author' \\
            '--message[History message]:message' \\
            '--force[Force override]' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        release)
          _arguments \\
            '--author[Mutation author]:author' \\
            '--message[History message]:message' \\
            '--allow-audit-release[Allow non-owner release handoffs without requiring --force]' \\
            '--force[Force override]' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        start-task|pause-task)
          _arguments \\
            '--author[Mutation author]:author' \\
            '--message[History message]:message' \\
            '--force[Force override]' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        close-task)
          _arguments \\
            '--author[Mutation author]:author' \\
            '--message[History message]:message' \\
            '--validate-close[Validate closure metadata mode]:(off warn strict)' \\
            '--force[Force override]' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        validate)
          _arguments \\
            '--check-metadata[Run metadata completeness checks]' \\
            '--metadata-profile[Select metadata validation profile for --check-metadata]:(core strict custom)' \\
            '--check-resolution[Run closed-item resolution metadata checks]' \\
            '--check-lifecycle[Run active-item lifecycle governance drift checks]' \\
            '--check-stale-blockers[Include stale blocker-pattern diagnostics in lifecycle checks]' \\
            '--dependency-cycle-severity[Set dependency-cycle warning policy for lifecycle checks]:(off warn error)' \\
            '--check-files[Run linked-file and orphaned-file checks]' \\
            '--scan-mode[Select file candidate scan mode for --check-files]:(default tracked-all tracked-all-strict)' \\
            '--include-pm-internals[Include PM storage internals in tracked-all candidate scans]' \\
            '--verbose-file-lists[Include full file-path lists for validate --check-files details]' \\
            '--strict-exit[Return non-zero exit when validation warnings are present]' \\
            '--fail-on-warn[Alias for --strict-exit]' \\
            '--check-history-drift[Run item/history hash drift checks]' \\
            '--check-command-references[Run linked-command PM-ID reference checks]' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        health)
          _arguments \\
            '--strict-directories[Treat optional item-type directories as required failures]' \\
            '--check-only[Run read-only health diagnostics without refreshing vectors]' \\
            '--no-refresh[Disable automatic vector refresh attempts during health checks]' \\
            '--refresh-vectors[Explicitly enable vector refresh attempts during health checks]' \\
            '--verbose-stale-items[Include full stale vectorization ID lists in health output]' \\
            '--strict-exit[Return non-zero exit when health warnings are present]' \\
            '--fail-on-warn[Alias for --strict-exit]' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        comments-audit)
          _arguments \\
            '--status[Filter by item status]:status:(draft open in_progress blocked closed canceled)' \\
            '--type[Filter by item type]:(${typeChoices})' \\
            '--tag[Filter by tag]:(${zshTagChoices})' \\
            '--priority[Filter by priority]:(0 1 2 3 4)' \\
            '--parent[Filter by parent item ID]:parent_id' \\
            '--sprint[Filter by sprint]:sprint' \\
            '--release[Filter by release]:release' \\
            '--assignee[Filter by assignee]:assignee' \\
            '--assignee-filter[Filter assignee presence]:(assigned unassigned)' \\
            '--limit-items[Limit returned item count]:number' \\
            '--limit[Alias for --limit-items]:number' \\
            '--full-history[Export full comment history rows (cannot be combined with --latest)]' \\
            '--latest[Return latest n comments per item (0 for summary-only rows)]:number' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]'
          ;;
        extension)
          _arguments \\
            '1:extension_action:(init scaffold install uninstall explore manage doctor adopt adopt-all activate deactivate)' \\
            '--init[Generate a starter extension scaffold at target path]' \\
            '--scaffold[Alias for --init]' \\
            '--install[Install extension from local path or GitHub source]' \\
            '--uninstall[Uninstall extension by name]' \\
            '--explore[List discovered extensions for selected scope]' \\
            '--manage[List managed extensions with update metadata]' \\
            '--doctor[Run consolidated extension diagnostics (summary/deep)]' \\
            '--adopt[Adopt an unmanaged extension into managed metadata]' \\
            '--adopt-all[Adopt all unmanaged extensions into managed metadata]' \\
            '--activate[Activate extension in selected scope settings]' \\
            '--deactivate[Deactivate extension in selected scope settings]' \\
            '--project[Use project extension scope (default)]' \\
            '--local[Alias for --project]' \\
            '--global[Use global extension scope]' \\
            '--gh[Install from GitHub shorthand owner/repo/path]:github_spec' \\
            '--github[Alias for --gh]:github_spec' \\
            '--ref[Git ref/branch/tag for GitHub source]:git_ref' \\
            '--detail[Detail mode for extension diagnostics]:detail_mode:(summary deep)' \\
            '--trace[Include registration traces in doctor deep diagnostics]' \\
            '--runtime-probe[Opt-in runtime activation probe for manage output]' \\
            '--fix-managed-state[Adopt unmanaged extensions before diagnostics/update checks]' \\
            '--strict-exit[Return non-zero exit when doctor warnings are present]' \\
            '--fail-on-warn[Alias for --strict-exit (doctor)]' \\
            '--json[Output JSON]' \\
            '--quiet[Suppress stdout]' \\
            '*:target_or_name:_files -/'
          ;;
        completion)
          _arguments \\
            '--eager-tags[Embed current tracker tags directly in script output]' \\
            '1:shell:(bash zsh fish)'
          ;;
        templates)
          local -a templates_cmds
          templates_cmds=('save:Save or update a create template' 'list:List saved create templates' 'show:Show saved template details')
          _describe 'templates command' templates_cmds
          ;;
      esac
      ;;
  esac
}

compdef _pm pm`;
}
export function generateFishScript(itemTypes = DEFAULT_ITEM_TYPES, tags = [], eagerTagExpansion = false) {
    const listCmds = ALL_COMMANDS.filter((command) => command === "list" || command.startsWith("list-")).join(" ");
    const noSubcommandList = ALL_COMMANDS.join(" ");
    const typeChoices = itemTypes.join(" ");
    const guideTopicChoices = GUIDE_TOPIC_CHOICES;
    const tagChoices = joinCompletionValues(tags);
    const useEagerTagExpansion = eagerTagExpansion || tags.length > 0;
    const fishTagChoices = useEagerTagExpansion ? `'${tagChoices}'` : "'(__pm_tag_choices)'";
    const dynamicTagResolver = useEagerTagExpansion
        ? ""
        : `
function __pm_tag_choices
  set -l now (date +%s ^/dev/null)
  if test -z "$now"
    set now 0
  end
  set -l ttl 120
  if set -q PM_COMPLETION_TAG_TTL
    set ttl $PM_COMPLETION_TAG_TTL
  end
  if set -q PM_COMPLETION_TAG_CACHE; and set -q PM_COMPLETION_TAG_CACHE_TS
    set -l age (math "$now - $PM_COMPLETION_TAG_CACHE_TS")
    if test $age -lt $ttl
      printf '%s\n' $PM_COMPLETION_TAG_CACHE
      return
    end
  end
  set -l resolved (pm completion-tags ^/dev/null)
  set -gx PM_COMPLETION_TAG_CACHE $resolved
  set -gx PM_COMPLETION_TAG_CACHE_TS $now
  printf '%s\n' $resolved
end
`;
    return `# Fish shell completion for pm
# Save to ~/.config/fish/completions/pm.fish
# or run: pm completion fish > ~/.config/fish/completions/pm.fish

# Disable file completion by default
complete -c pm -f

# Global flags (available for all subcommands)
complete -c pm -l json -d 'Output JSON instead of TOON'
complete -c pm -l quiet -d 'Suppress stdout output'
complete -c pm -l path -d 'Override PM path for this command' -r
complete -c pm -l no-extensions -d 'Disable extension loading'
complete -c pm -l profile -d 'Print deterministic timing diagnostics'
complete -c pm -s V -l version -d 'Output the version number'
complete -c pm -s h -l help -d 'Display help'

# Helper: true when no subcommand has been given yet
function __pm_no_subcommand
  not __fish_seen_subcommand_from ${noSubcommandList}
end
${dynamicTagResolver}

# Subcommands
complete -c pm -n __pm_no_subcommand -a init          -d 'Initialize pm storage for the current workspace'
complete -c pm -n __pm_no_subcommand -a config        -d 'Read or update pm settings'
complete -c pm -n __pm_no_subcommand -a extension     -d 'Manage extension lifecycle operations'
complete -c pm -n __pm_no_subcommand -a create        -d 'Create a new project management item'
complete -c pm -n __pm_no_subcommand -a list          -d 'List active items with optional filters'
complete -c pm -n __pm_no_subcommand -a list-all      -d 'List all items with optional filters'
complete -c pm -n __pm_no_subcommand -a list-draft    -d 'List draft items with optional filters'
complete -c pm -n __pm_no_subcommand -a list-open     -d 'List open items with optional filters'
complete -c pm -n __pm_no_subcommand -a list-in-progress -d 'List in-progress items with optional filters'
complete -c pm -n __pm_no_subcommand -a list-blocked  -d 'List blocked items with optional filters'
complete -c pm -n __pm_no_subcommand -a list-closed   -d 'List closed items with optional filters'
complete -c pm -n __pm_no_subcommand -a list-canceled -d 'List canceled items with optional filters'
complete -c pm -n __pm_no_subcommand -a aggregate     -d 'Aggregate grouped item counts for governance queries'
complete -c pm -n __pm_no_subcommand -a dedupe-audit  -d 'Audit potential duplicate items and emit merge suggestions'
complete -c pm -n __pm_no_subcommand -a guide         -d 'Browse local progressive-disclosure guides'
complete -c pm -n __pm_no_subcommand -a calendar      -d 'Show deadline/reminder calendar views'
complete -c pm -n __pm_no_subcommand -a cal           -d 'Alias for calendar'
complete -c pm -n __pm_no_subcommand -a context       -d 'Show a token-efficient project context snapshot'
complete -c pm -n __pm_no_subcommand -a ctx           -d 'Alias for context'
complete -c pm -n __pm_no_subcommand -a get           -d 'Show item details by ID'
complete -c pm -n __pm_no_subcommand -a search        -d 'Search items with keyword, semantic, or hybrid modes'
complete -c pm -n __pm_no_subcommand -a reindex       -d 'Rebuild search artifacts'
complete -c pm -n __pm_no_subcommand -a history       -d 'Show item history entries'
complete -c pm -n __pm_no_subcommand -a activity      -d 'Show recent activity across items'
complete -c pm -n __pm_no_subcommand -a restore       -d 'Restore an item to an earlier state'
complete -c pm -n __pm_no_subcommand -a update        -d 'Update item fields and metadata'
complete -c pm -n __pm_no_subcommand -a update-many   -d 'Bulk-update matched items with dry-run and rollback checkpoints'
complete -c pm -n __pm_no_subcommand -a normalize     -d 'Normalize lifecycle metadata with dry-run planning or apply mode'
complete -c pm -n __pm_no_subcommand -a close         -d 'Close an item with a required reason'
complete -c pm -n __pm_no_subcommand -a delete        -d 'Delete an item and record the change'
complete -c pm -n __pm_no_subcommand -a append        -d 'Append text to an item body'
complete -c pm -n __pm_no_subcommand -a comments      -d 'List or add comments for an item'
complete -c pm -n __pm_no_subcommand -a comments-audit -d 'Audit latest comments or full history across filtered items'
complete -c pm -n __pm_no_subcommand -a notes         -d 'List or add notes for an item'
complete -c pm -n __pm_no_subcommand -a learnings     -d 'List or add learnings for an item'
complete -c pm -n __pm_no_subcommand -a files         -d 'Manage linked files'
complete -c pm -n __pm_no_subcommand -a docs          -d 'Manage linked docs'
complete -c pm -n __pm_no_subcommand -a deps          -d 'Show dependency relationships for an item'
complete -c pm -n __pm_no_subcommand -a test          -d 'Manage linked tests and optionally run them'
complete -c pm -n __pm_no_subcommand -a test-all      -d 'Run linked tests across matching items'
complete -c pm -n __pm_no_subcommand -a test-runs     -d 'Manage background linked-test runs'
complete -c pm -n __pm_no_subcommand -a stats         -d 'Show project tracker statistics'
complete -c pm -n __pm_no_subcommand -a health        -d 'Show project tracker health checks'
complete -c pm -n __pm_no_subcommand -a validate      -d 'Run standalone validation checks'
complete -c pm -n __pm_no_subcommand -a gc            -d 'Clean optional cache artifacts'
complete -c pm -n __pm_no_subcommand -a contracts     -d 'Show machine-readable command and schema contracts'
complete -c pm -n __pm_no_subcommand -a claim         -d 'Claim an item for active work'
complete -c pm -n __pm_no_subcommand -a release       -d 'Release the active claim for an item'
complete -c pm -n __pm_no_subcommand -a start-task    -d 'Lifecycle alias to claim and set in-progress'
complete -c pm -n __pm_no_subcommand -a pause-task    -d 'Lifecycle alias to reopen and release claim'
complete -c pm -n __pm_no_subcommand -a close-task    -d 'Lifecycle alias to close and release claim'
complete -c pm -n __pm_no_subcommand -a templates     -d 'Manage reusable create templates'
complete -c pm -n __pm_no_subcommand -a completion    -d 'Generate shell completion'

# list* flags
for list_cmd in ${listCmds}
  complete -c pm -n "__fish_seen_subcommand_from $list_cmd" -l type     -d 'Filter by item type' -r -a '${typeChoices}'
  complete -c pm -n "__fish_seen_subcommand_from $list_cmd" -l tag      -d 'Filter by tag' -r -a ${fishTagChoices}
  complete -c pm -n "__fish_seen_subcommand_from $list_cmd" -l priority -d 'Filter by priority' -r -a '0 1 2 3 4'
  complete -c pm -n "__fish_seen_subcommand_from $list_cmd" -l assignee -d 'Filter by assignee' -r
  complete -c pm -n "__fish_seen_subcommand_from $list_cmd" -l assignee-filter -d 'Filter assignee presence' -r -a 'assigned unassigned'
  complete -c pm -n "__fish_seen_subcommand_from $list_cmd" -l sprint   -d 'Filter by sprint' -r
  complete -c pm -n "__fish_seen_subcommand_from $list_cmd" -l release  -d 'Filter by release' -r
  complete -c pm -n "__fish_seen_subcommand_from $list_cmd" -l limit    -d 'Limit returned item count' -r
  complete -c pm -n "__fish_seen_subcommand_from $list_cmd" -l offset   -d 'Skip the first n matching rows before limit' -r
  complete -c pm -n "__fish_seen_subcommand_from $list_cmd" -l include-body -d 'Include item body in each returned list row'
  complete -c pm -n "__fish_seen_subcommand_from $list_cmd" -l compact -d 'Render compact list projection fields'
  complete -c pm -n "__fish_seen_subcommand_from $list_cmd" -l fields -d 'Render custom comma-separated list fields' -r
  complete -c pm -n "__fish_seen_subcommand_from $list_cmd" -l sort -d 'Sort field' -r -a 'priority deadline updated_at created_at title parent'
  complete -c pm -n "__fish_seen_subcommand_from $list_cmd" -l order -d 'Sort order (requires --sort)' -r -a 'asc desc'
  complete -c pm -n "__fish_seen_subcommand_from $list_cmd" -l stream -d 'Emit line-delimited JSON rows (requires --json)'
  complete -c pm -n "__fish_seen_subcommand_from $list_cmd" -l deadline-before -d 'Filter by deadline upper bound (ISO/date string or relative)' -r
  complete -c pm -n "__fish_seen_subcommand_from $list_cmd" -l deadline-after  -d 'Filter by deadline lower bound (ISO/date string or relative)' -r
end

# aggregate flags
complete -c pm -n '__fish_seen_subcommand_from aggregate' -l group-by -d 'Comma-separated group-by fields (supported: parent,type,priority,status,assignee,tags,sprint,release)' -r
complete -c pm -n '__fish_seen_subcommand_from aggregate' -l count -d 'Return grouped counts'
complete -c pm -n '__fish_seen_subcommand_from aggregate' -l include-unparented -d 'Include unparented rows when grouping by parent'
complete -c pm -n '__fish_seen_subcommand_from aggregate' -l status -d 'Filter by status' -r -a 'draft open in_progress blocked closed canceled'
complete -c pm -n '__fish_seen_subcommand_from aggregate' -l type -d 'Filter by item type' -r -a '${typeChoices}'
complete -c pm -n '__fish_seen_subcommand_from aggregate' -l tag -d 'Filter by tag' -r -a ${fishTagChoices}
complete -c pm -n '__fish_seen_subcommand_from aggregate' -l priority -d 'Filter by priority' -r -a '0 1 2 3 4'
complete -c pm -n '__fish_seen_subcommand_from aggregate' -l deadline-before -d 'Filter by deadline upper bound (ISO/date string or relative)' -r
complete -c pm -n '__fish_seen_subcommand_from aggregate' -l deadline-after -d 'Filter by deadline lower bound (ISO/date string or relative)' -r
complete -c pm -n '__fish_seen_subcommand_from aggregate' -l assignee -d 'Filter by assignee' -r
complete -c pm -n '__fish_seen_subcommand_from aggregate' -l assignee-filter -d 'Filter assignee presence' -r -a 'assigned unassigned'
complete -c pm -n '__fish_seen_subcommand_from aggregate' -l parent -d 'Filter by parent item ID' -r
complete -c pm -n '__fish_seen_subcommand_from aggregate' -l sprint -d 'Filter by sprint' -r
complete -c pm -n '__fish_seen_subcommand_from aggregate' -l release -d 'Filter by release' -r

# dedupe-audit flags
complete -c pm -n '__fish_seen_subcommand_from dedupe-audit' -l mode -d 'Dedupe mode' -r -a 'title_exact title_fuzzy parent_scope'
complete -c pm -n '__fish_seen_subcommand_from dedupe-audit' -l limit -d 'Limit returned duplicate clusters' -r
complete -c pm -n '__fish_seen_subcommand_from dedupe-audit' -l threshold -d 'Fuzzy mode token similarity threshold between 0 and 1' -r
complete -c pm -n '__fish_seen_subcommand_from dedupe-audit' -l status -d 'Filter by status' -r -a 'draft open in_progress blocked closed canceled'
complete -c pm -n '__fish_seen_subcommand_from dedupe-audit' -l type -d 'Filter by item type' -r -a '${typeChoices}'
complete -c pm -n '__fish_seen_subcommand_from dedupe-audit' -l tag -d 'Filter by tag' -r -a ${fishTagChoices}
complete -c pm -n '__fish_seen_subcommand_from dedupe-audit' -l priority -d 'Filter by priority' -r -a '0 1 2 3 4'
complete -c pm -n '__fish_seen_subcommand_from dedupe-audit' -l deadline-before -d 'Filter by deadline upper bound (ISO/date string or relative)' -r
complete -c pm -n '__fish_seen_subcommand_from dedupe-audit' -l deadline-after -d 'Filter by deadline lower bound (ISO/date string or relative)' -r
complete -c pm -n '__fish_seen_subcommand_from dedupe-audit' -l assignee -d 'Filter by assignee' -r
complete -c pm -n '__fish_seen_subcommand_from dedupe-audit' -l assignee-filter -d 'Filter assignee presence' -r -a 'assigned unassigned'
complete -c pm -n '__fish_seen_subcommand_from dedupe-audit' -l parent -d 'Filter by parent item ID' -r
complete -c pm -n '__fish_seen_subcommand_from dedupe-audit' -l sprint -d 'Filter by sprint' -r
complete -c pm -n '__fish_seen_subcommand_from dedupe-audit' -l release -d 'Filter by release' -r

# create flags
complete -c pm -n '__fish_seen_subcommand_from create' -s t -l title              -d 'Item title' -r
complete -c pm -n '__fish_seen_subcommand_from create' -s d -l description        -d 'Item description' -r
complete -c pm -n '__fish_seen_subcommand_from create' -l type                    -d 'Item type' -r -a '${typeChoices}'
complete -c pm -n '__fish_seen_subcommand_from create' -l create-mode             -d 'Create required-option policy mode' -r -a 'strict progressive'
complete -c pm -n '__fish_seen_subcommand_from create' -l schedule-preset         -d 'Scheduling preset for Reminder/Meeting/Event' -r -a 'lightweight'
complete -c pm -n '__fish_seen_subcommand_from create' -s s -l status             -d 'Item status' -r -a 'draft open in_progress blocked'
complete -c pm -n '__fish_seen_subcommand_from create' -s p -l priority           -d 'Priority (0-4)' -r -a '0 1 2 3 4'
complete -c pm -n '__fish_seen_subcommand_from create' -l tags                    -d 'Comma-separated tags' -r
complete -c pm -n '__fish_seen_subcommand_from create' -s b -l body               -d 'Item body' -r
complete -c pm -n '__fish_seen_subcommand_from create' -l deadline                -d 'Deadline (ISO/date string or relative +6h/+1d/+2w/+6m)' -r
complete -c pm -n '__fish_seen_subcommand_from create' -l estimate                -d 'Estimated minutes' -r
complete -c pm -n '__fish_seen_subcommand_from create' -l acceptance-criteria     -d 'Acceptance criteria' -r
complete -c pm -n '__fish_seen_subcommand_from create' -l reminder                -d 'Reminder entry at=<iso|relative>,text=<text>' -r
complete -c pm -n '__fish_seen_subcommand_from create' -l event                   -d 'Event entry start=<iso|relative>,end=<iso|relative>,recur_*' -r
complete -c pm -n '__fish_seen_subcommand_from create' -l type-option             -d 'Type option key=value or key=<name>,value=<value>' -r
complete -c pm -n '__fish_seen_subcommand_from create' -l author                  -d 'Mutation author' -r
complete -c pm -n '__fish_seen_subcommand_from create' -l message                 -d 'History message' -r
complete -c pm -n '__fish_seen_subcommand_from create' -l assignee                -d 'Assignee' -r
complete -c pm -n '__fish_seen_subcommand_from create' -l unset                   -d 'Clear scalar metadata field by name' -r
complete -c pm -n '__fish_seen_subcommand_from create' -l clear-deps              -d 'Clear dependency entries'
complete -c pm -n '__fish_seen_subcommand_from create' -l clear-comments          -d 'Clear comments'
complete -c pm -n '__fish_seen_subcommand_from create' -l clear-notes             -d 'Clear notes'
complete -c pm -n '__fish_seen_subcommand_from create' -l clear-learnings         -d 'Clear learnings'
complete -c pm -n '__fish_seen_subcommand_from create' -l clear-files             -d 'Clear linked files'
complete -c pm -n '__fish_seen_subcommand_from create' -l clear-tests             -d 'Clear linked tests'
complete -c pm -n '__fish_seen_subcommand_from create' -l clear-docs              -d 'Clear linked docs'
complete -c pm -n '__fish_seen_subcommand_from create' -l clear-reminders         -d 'Clear reminders'
complete -c pm -n '__fish_seen_subcommand_from create' -l clear-events            -d 'Clear events'
complete -c pm -n '__fish_seen_subcommand_from create' -l clear-type-options      -d 'Clear type options'

# update flags
complete -c pm -n '__fish_seen_subcommand_from update' -s t -l title              -d 'Item title' -r
complete -c pm -n '__fish_seen_subcommand_from update' -s d -l description        -d 'Item description' -r
complete -c pm -n '__fish_seen_subcommand_from update' -s b -l body               -d 'Item body' -r
complete -c pm -n '__fish_seen_subcommand_from update' -s s -l status             -d 'Item status' -r -a 'draft open in_progress blocked canceled'
complete -c pm -n '__fish_seen_subcommand_from update' -l close-reason            -d 'Set close reason' -r
complete -c pm -n '__fish_seen_subcommand_from update' -s p -l priority           -d 'Priority (0-4)' -r -a '0 1 2 3 4'
complete -c pm -n '__fish_seen_subcommand_from update' -l type                    -d 'Item type' -r -a '${typeChoices}'
complete -c pm -n '__fish_seen_subcommand_from update' -l comment                 -d 'Comment seed author=<value>,created_at=<iso|now>,text=<value>' -r
complete -c pm -n '__fish_seen_subcommand_from update' -l note                    -d 'Note seed author=<value>,created_at=<iso|now>,text=<value>' -r
complete -c pm -n '__fish_seen_subcommand_from update' -l learning                -d 'Learning seed author=<value>,created_at=<iso|now>,text=<value>' -r
complete -c pm -n '__fish_seen_subcommand_from update' -l file                    -d 'Linked file path=<value>,scope=<project|global>,note=<text>' -r
complete -c pm -n '__fish_seen_subcommand_from update' -l test                    -d 'Linked test command=<value>,path=<value>,scope=<project|global>' -r
complete -c pm -n '__fish_seen_subcommand_from update' -l doc                     -d 'Linked doc path=<value>,scope=<project|global>,note=<text>' -r
complete -c pm -n '__fish_seen_subcommand_from update' -l reminder                -d 'Reminder entry at=<iso|relative>,text=<text>' -r
complete -c pm -n '__fish_seen_subcommand_from update' -l event                   -d 'Event entry start=<iso|relative>,end=<iso|relative>,recur_*' -r
complete -c pm -n '__fish_seen_subcommand_from update' -l type-option             -d 'Type option key=value or key=<name>,value=<value>' -r
complete -c pm -n '__fish_seen_subcommand_from update' -l unset                   -d 'Clear scalar metadata field by name' -r
complete -c pm -n '__fish_seen_subcommand_from update' -l replace-deps            -d 'Atomically replace dependencies with provided --dep values'
complete -c pm -n '__fish_seen_subcommand_from update' -l replace-tests           -d 'Atomically replace linked tests with provided --test values'
complete -c pm -n '__fish_seen_subcommand_from update' -l clear-deps              -d 'Clear dependency entries'
complete -c pm -n '__fish_seen_subcommand_from update' -l clear-comments          -d 'Clear comments'
complete -c pm -n '__fish_seen_subcommand_from update' -l clear-notes             -d 'Clear notes'
complete -c pm -n '__fish_seen_subcommand_from update' -l clear-learnings         -d 'Clear learnings'
complete -c pm -n '__fish_seen_subcommand_from update' -l clear-files             -d 'Clear linked files'
complete -c pm -n '__fish_seen_subcommand_from update' -l clear-tests             -d 'Clear linked tests'
complete -c pm -n '__fish_seen_subcommand_from update' -l clear-docs              -d 'Clear linked docs'
complete -c pm -n '__fish_seen_subcommand_from update' -l clear-reminders         -d 'Clear reminders'
complete -c pm -n '__fish_seen_subcommand_from update' -l clear-events            -d 'Clear events'
complete -c pm -n '__fish_seen_subcommand_from update' -l clear-type-options      -d 'Clear type options'
complete -c pm -n '__fish_seen_subcommand_from update' -l allow-audit-update      -d 'Allow non-owner metadata-only audit updates without requiring --force'
complete -c pm -n '__fish_seen_subcommand_from update' -l author                  -d 'Mutation author' -r
complete -c pm -n '__fish_seen_subcommand_from update' -l message                 -d 'History message' -r
complete -c pm -n '__fish_seen_subcommand_from update' -l force                   -d 'Force override'

# update-many flags
complete -c pm -n '__fish_seen_subcommand_from update-many' -l filter-status           -d 'Filter by status before applying updates' -r -a 'draft open in_progress blocked closed canceled'
complete -c pm -n '__fish_seen_subcommand_from update-many' -l filter-type             -d 'Filter by type before applying updates' -r -a '${typeChoices}'
complete -c pm -n '__fish_seen_subcommand_from update-many' -l filter-tag              -d 'Filter by tag before applying updates' -r -a ${fishTagChoices}
complete -c pm -n '__fish_seen_subcommand_from update-many' -l filter-priority         -d 'Filter by priority before applying updates' -r -a '0 1 2 3 4'
complete -c pm -n '__fish_seen_subcommand_from update-many' -l filter-deadline-before  -d 'Filter by deadline upper bound' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l filter-deadline-after   -d 'Filter by deadline lower bound' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l filter-assignee         -d 'Filter by assignee before applying updates' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l filter-assignee-filter  -d 'Filter assignee presence' -r -a 'assigned unassigned'
complete -c pm -n '__fish_seen_subcommand_from update-many' -l filter-parent           -d 'Filter by parent item ID' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l filter-sprint           -d 'Filter by sprint before applying updates' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l filter-release          -d 'Filter by release before applying updates' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l limit                   -d 'Limit matched item count' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l offset                  -d 'Skip first n matched rows' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l dry-run                 -d 'Preview updates without mutating'
complete -c pm -n '__fish_seen_subcommand_from update-many' -l rollback                -d 'Rollback checkpoint ID' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l no-checkpoint           -d 'Disable checkpoint creation during apply mode'
complete -c pm -n '__fish_seen_subcommand_from update-many' -s t -l title              -d 'Item title' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -s d -l description        -d 'Item description' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -s b -l body               -d 'Item body' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -s p -l priority           -d 'Priority (0-4)' -r -a '0 1 2 3 4'
complete -c pm -n '__fish_seen_subcommand_from update-many' -l type                    -d 'Item type' -r -a '${typeChoices}'
complete -c pm -n '__fish_seen_subcommand_from update-many' -l tags                    -d 'Comma-separated tags' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l deadline                -d 'Deadline (ISO/date string or relative)' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l estimate                -d 'Estimated minutes' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l acceptance-criteria     -d 'Acceptance criteria' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l definition-of-ready     -d 'Definition of ready' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l order                   -d 'Planning order/rank' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l goal                    -d 'Goal identifier' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l objective               -d 'Objective identifier' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l value                   -d 'Business value summary' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l impact                  -d 'Business impact summary' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l outcome                 -d 'Expected outcome summary' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l why-now                 -d 'Why-now rationale' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l reviewer                -d 'Reviewer' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l risk                    -d 'Risk level' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l confidence              -d 'Confidence level' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l sprint                  -d 'Sprint identifier' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l release                 -d 'Release identifier' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l reporter                -d 'Issue reporter' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l severity                -d 'Issue severity' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l environment             -d 'Issue environment context' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l repro-steps             -d 'Issue reproduction steps' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l resolution              -d 'Issue resolution summary' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l expected-result         -d 'Issue expected behavior' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l actual-result           -d 'Issue observed behavior' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l affected-version        -d 'Affected version identifier' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l fixed-version           -d 'Fixed version identifier' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l component               -d 'Issue component ownership' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l regression              -d 'Regression marker true|false|1|0' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l customer-impact         -d 'Customer impact summary' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l dep                     -d 'Dependency seed id=<id>,kind=<kind>,author=<author>,created_at=<timestamp>' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l dep-remove              -d 'Dependency removal selector id=<id>,kind=<kind>,author=<author>,created_at=<timestamp>' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l replace-deps            -d 'Atomically replace dependencies with provided --dep values'
complete -c pm -n '__fish_seen_subcommand_from update-many' -l replace-tests           -d 'Atomically replace linked tests with provided --test values'
complete -c pm -n '__fish_seen_subcommand_from update-many' -l comment                 -d 'Comment seed author=<value>,created_at=<iso|now>,text=<value>' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l note                    -d 'Note seed author=<value>,created_at=<iso|now>,text=<value>' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l learning                -d 'Learning seed author=<value>,created_at=<iso|now>,text=<value>' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l file                    -d 'Linked file path=<value>,scope=<project|global>,note=<text>' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l test                    -d 'Linked test command=<value>,path=<value>,scope=<project|global>' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l doc                     -d 'Linked doc path=<value>,scope=<project|global>,note=<text>' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l reminder                -d 'Reminder entry at=<iso|relative>,text=<text>' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l event                   -d 'Event entry start=<iso|relative>,end=<iso|relative>,recur_*' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l type-option             -d 'Type option key=value or key=<name>,value=<value>' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l unset                   -d 'Clear scalar metadata field by name' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l clear-deps              -d 'Clear dependency entries'
complete -c pm -n '__fish_seen_subcommand_from update-many' -l clear-comments          -d 'Clear comments'
complete -c pm -n '__fish_seen_subcommand_from update-many' -l clear-notes             -d 'Clear notes'
complete -c pm -n '__fish_seen_subcommand_from update-many' -l clear-learnings         -d 'Clear learnings'
complete -c pm -n '__fish_seen_subcommand_from update-many' -l clear-files             -d 'Clear linked files'
complete -c pm -n '__fish_seen_subcommand_from update-many' -l clear-tests             -d 'Clear linked tests'
complete -c pm -n '__fish_seen_subcommand_from update-many' -l clear-docs              -d 'Clear linked docs'
complete -c pm -n '__fish_seen_subcommand_from update-many' -l clear-reminders         -d 'Clear reminders'
complete -c pm -n '__fish_seen_subcommand_from update-many' -l clear-events            -d 'Clear events'
complete -c pm -n '__fish_seen_subcommand_from update-many' -l clear-type-options      -d 'Clear type options'
complete -c pm -n '__fish_seen_subcommand_from update-many' -l allow-audit-update      -d 'Allow non-owner metadata-only audit updates without requiring --force'
complete -c pm -n '__fish_seen_subcommand_from update-many' -l author                  -d 'Mutation author' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l message                 -d 'History message' -r
complete -c pm -n '__fish_seen_subcommand_from update-many' -l force                   -d 'Force override'

# normalize flags
complete -c pm -n '__fish_seen_subcommand_from normalize' -l filter-status           -d 'Filter by status before planning or apply' -r -a 'draft open in_progress blocked closed canceled'
complete -c pm -n '__fish_seen_subcommand_from normalize' -l filter-type             -d 'Filter by type before planning or apply' -r -a '${typeChoices}'
complete -c pm -n '__fish_seen_subcommand_from normalize' -l filter-tag              -d 'Filter by tag before planning or apply' -r -a ${fishTagChoices}
complete -c pm -n '__fish_seen_subcommand_from normalize' -l filter-priority         -d 'Filter by priority before planning or apply' -r -a '0 1 2 3 4'
complete -c pm -n '__fish_seen_subcommand_from normalize' -l filter-deadline-before  -d 'Filter by deadline upper bound' -r
complete -c pm -n '__fish_seen_subcommand_from normalize' -l filter-deadline-after   -d 'Filter by deadline lower bound' -r
complete -c pm -n '__fish_seen_subcommand_from normalize' -l filter-assignee         -d 'Filter by assignee before planning or apply' -r
complete -c pm -n '__fish_seen_subcommand_from normalize' -l filter-assignee-filter  -d 'Filter assignee presence' -r -a 'assigned unassigned'
complete -c pm -n '__fish_seen_subcommand_from normalize' -l filter-parent           -d 'Filter by parent item ID' -r
complete -c pm -n '__fish_seen_subcommand_from normalize' -l filter-sprint           -d 'Filter by sprint before planning or apply' -r
complete -c pm -n '__fish_seen_subcommand_from normalize' -l filter-release          -d 'Filter by release before planning or apply' -r
complete -c pm -n '__fish_seen_subcommand_from normalize' -l limit                   -d 'Limit matched item count' -r
complete -c pm -n '__fish_seen_subcommand_from normalize' -l offset                  -d 'Skip first n matched rows' -r
complete -c pm -n '__fish_seen_subcommand_from normalize' -l dry-run                 -d 'Preview normalize findings without mutating'
complete -c pm -n '__fish_seen_subcommand_from normalize' -l apply                   -d 'Apply normalize changes'
complete -c pm -n '__fish_seen_subcommand_from normalize' -l allow-audit-update      -d 'Allow non-owner metadata-only audit updates without requiring --force'
complete -c pm -n '__fish_seen_subcommand_from normalize' -l author                  -d 'Mutation author' -r
complete -c pm -n '__fish_seen_subcommand_from normalize' -l message                 -d 'History message' -r
complete -c pm -n '__fish_seen_subcommand_from normalize' -l force                   -d 'Force override'

# search flags
complete -c pm -n '__fish_seen_subcommand_from search' -l mode          -d 'Search mode' -r -a 'keyword semantic hybrid'
complete -c pm -n '__fish_seen_subcommand_from search' -l include-linked -d 'Include linked content in scoring'
complete -c pm -n '__fish_seen_subcommand_from search' -l limit          -d 'Max results' -r
complete -c pm -n '__fish_seen_subcommand_from search' -l type           -d 'Filter by type' -r -a '${typeChoices}'
complete -c pm -n '__fish_seen_subcommand_from search' -l tag            -d 'Filter by tag' -r -a ${fishTagChoices}
complete -c pm -n '__fish_seen_subcommand_from search' -l priority       -d 'Filter by priority' -r -a '0 1 2 3 4'

# calendar flags
complete -c pm -n '__fish_seen_subcommand_from calendar cal' -l view      -d 'Calendar view' -r -a 'agenda day week month'
complete -c pm -n '__fish_seen_subcommand_from calendar cal' -l date      -d 'Anchor date/time (ISO/date string or relative)' -r
complete -c pm -n '__fish_seen_subcommand_from calendar cal' -l from      -d 'Agenda lower bound (ISO/date string or relative)' -r
complete -c pm -n '__fish_seen_subcommand_from calendar cal' -l to        -d 'Agenda upper bound (ISO/date string or relative)' -r
complete -c pm -n '__fish_seen_subcommand_from calendar cal' -l past      -d 'Include past entries'
complete -c pm -n '__fish_seen_subcommand_from calendar cal' -l full-period -d 'Include full anchored day/week/month period'
complete -c pm -n '__fish_seen_subcommand_from calendar cal' -l type      -d 'Filter by type' -r -a '${typeChoices}'
complete -c pm -n '__fish_seen_subcommand_from calendar cal' -l tag       -d 'Filter by tag' -r -a ${fishTagChoices}
complete -c pm -n '__fish_seen_subcommand_from calendar cal' -l priority  -d 'Filter by priority' -r -a '0 1 2 3 4'
complete -c pm -n '__fish_seen_subcommand_from calendar cal' -l status    -d 'Filter by status' -r -a 'draft open in_progress blocked closed canceled'
complete -c pm -n '__fish_seen_subcommand_from calendar cal' -l assignee  -d 'Filter by assignee' -r
complete -c pm -n '__fish_seen_subcommand_from calendar cal' -l assignee-filter -d 'Filter assignee presence' -r -a 'assigned unassigned'
complete -c pm -n '__fish_seen_subcommand_from calendar cal' -l sprint    -d 'Filter by sprint' -r
complete -c pm -n '__fish_seen_subcommand_from calendar cal' -l release   -d 'Filter by release' -r
complete -c pm -n '__fish_seen_subcommand_from calendar cal' -l include   -d 'Include event sources' -r -a 'all deadlines reminders events'
complete -c pm -n '__fish_seen_subcommand_from calendar cal' -l recurrence-lookahead-days -d 'Bound open-ended recurrence lookahead' -r
complete -c pm -n '__fish_seen_subcommand_from calendar cal' -l recurrence-lookback-days -d 'Bound open-ended recurrence lookback' -r
complete -c pm -n '__fish_seen_subcommand_from calendar cal' -l occurrence-limit -d 'Cap occurrences per recurring event' -r
complete -c pm -n '__fish_seen_subcommand_from calendar cal' -l limit     -d 'Limit returned events' -r
complete -c pm -n '__fish_seen_subcommand_from calendar cal' -l format    -d 'Output override' -r -a 'markdown toon json'

# context flags
complete -c pm -n '__fish_seen_subcommand_from context ctx' -l date      -d 'Anchor date/time (ISO/date string or relative)' -r
complete -c pm -n '__fish_seen_subcommand_from context ctx' -l from      -d 'Agenda lower bound (ISO/date string or relative)' -r
complete -c pm -n '__fish_seen_subcommand_from context ctx' -l to        -d 'Agenda upper bound (ISO/date string or relative)' -r
complete -c pm -n '__fish_seen_subcommand_from context ctx' -l past      -d 'Include past entries in bounded windows'
complete -c pm -n '__fish_seen_subcommand_from context ctx' -l type      -d 'Filter by type' -r -a '${typeChoices}'
complete -c pm -n '__fish_seen_subcommand_from context ctx' -l tag       -d 'Filter by tag' -r -a ${fishTagChoices}
complete -c pm -n '__fish_seen_subcommand_from context ctx' -l priority  -d 'Filter by priority' -r -a '0 1 2 3 4'
complete -c pm -n '__fish_seen_subcommand_from context ctx' -l assignee  -d 'Filter by assignee' -r
complete -c pm -n '__fish_seen_subcommand_from context ctx' -l assignee-filter -d 'Filter assignee presence' -r -a 'assigned unassigned'
complete -c pm -n '__fish_seen_subcommand_from context ctx' -l sprint    -d 'Filter by sprint' -r
complete -c pm -n '__fish_seen_subcommand_from context ctx' -l release   -d 'Filter by release' -r
complete -c pm -n '__fish_seen_subcommand_from context ctx' -l limit     -d 'Limit focus and agenda rows per section' -r
complete -c pm -n '__fish_seen_subcommand_from context ctx' -l format    -d 'Output override' -r -a 'markdown toon json'

# guide flags
complete -c pm -n '__fish_seen_subcommand_from guide' -l list      -d 'Show guide topic index'
complete -c pm -n '__fish_seen_subcommand_from guide' -l format    -d 'Output override' -r -a 'markdown toon json'
complete -c pm -n '__fish_seen_subcommand_from guide' -l depth     -d 'Guide detail depth' -r -a 'brief standard deep'
complete -c pm -n '__fish_seen_subcommand_from guide' -a '${guideTopicChoices}' -d 'Guide topic'

# reindex flags
complete -c pm -n '__fish_seen_subcommand_from reindex' -l mode -d 'Reindex mode' -r -a 'keyword semantic hybrid'
complete -c pm -n '__fish_seen_subcommand_from reindex' -l progress -d 'Emit progress updates to stderr'

# history / activity flags
complete -c pm -n '__fish_seen_subcommand_from history'  -l limit -d 'Max history entries' -r
complete -c pm -n '__fish_seen_subcommand_from history'  -l diff -d 'Include changed-field patch summary'
complete -c pm -n '__fish_seen_subcommand_from history'  -l verify -d 'Verify history hash chain and replay integrity'
complete -c pm -n '__fish_seen_subcommand_from activity' -l id -d 'Filter by item ID' -r
complete -c pm -n '__fish_seen_subcommand_from activity' -l op -d 'Filter by history operation' -r
complete -c pm -n '__fish_seen_subcommand_from activity' -l author -d 'Filter by history author' -r
complete -c pm -n '__fish_seen_subcommand_from activity' -l from -d 'Lower timestamp bound (ISO/date string or relative)' -r
complete -c pm -n '__fish_seen_subcommand_from activity' -l to -d 'Upper timestamp bound (ISO/date string or relative)' -r
complete -c pm -n '__fish_seen_subcommand_from activity' -l limit -d 'Max activity entries' -r
complete -c pm -n '__fish_seen_subcommand_from activity' -l stream -d 'Emit line-delimited JSON rows'
complete -c pm -n '__fish_seen_subcommand_from contracts' -l action -d 'Filter schema by tool action' -r
complete -c pm -n '__fish_seen_subcommand_from contracts' -l command -d 'Scope output to one command (narrow-by-default)' -r
complete -c pm -n '__fish_seen_subcommand_from contracts' -l schema-only -d 'Return schema-only payload'
complete -c pm -n '__fish_seen_subcommand_from contracts' -l flags-only -d 'Return command flag contracts only'
complete -c pm -n '__fish_seen_subcommand_from contracts' -l availability-only -d 'Return action availability only'
complete -c pm -n '__fish_seen_subcommand_from contracts' -l runtime-only -d 'Include only actions invocable in the current runtime'
complete -c pm -n '__fish_seen_subcommand_from contracts' -l active-only -d 'Alias for --runtime-only'
complete -c pm -n '__fish_seen_subcommand_from deps' -l format -d 'Output format' -r -a 'tree graph'
complete -c pm -n '__fish_seen_subcommand_from deps' -l max-depth -d 'Maximum traversal depth (0 keeps root only)' -r
complete -c pm -n '__fish_seen_subcommand_from deps' -l collapse -d 'Collapse mode' -r -a 'none repeated'
complete -c pm -n '__fish_seen_subcommand_from deps' -l summary -d 'Return counts only without tree/graph payload'

# comments / notes / learnings flags
complete -c pm -n '__fish_seen_subcommand_from comments notes learnings' -l add -d 'Add one entry (text=<value> or plain text)' -r
complete -c pm -n '__fish_seen_subcommand_from comments' -l stdin -d 'Read comment text from stdin (supports multiline markdown)'
complete -c pm -n '__fish_seen_subcommand_from comments' -l file -d 'Read comment text from file (supports multiline markdown)' -r
complete -c pm -n '__fish_seen_subcommand_from comments notes learnings' -l limit -d 'Return only latest n entries' -r
complete -c pm -n '__fish_seen_subcommand_from comments notes learnings' -l author -d 'Entry author' -r
complete -c pm -n '__fish_seen_subcommand_from comments notes learnings' -l message -d 'History message' -r
complete -c pm -n '__fish_seen_subcommand_from comments notes learnings' -l allow-audit-comment -d 'Allow non-owner append-only comment audits (legacy alias for notes/learnings)'
complete -c pm -n '__fish_seen_subcommand_from notes' -l allow-audit-note -d 'Allow non-owner append-only note audits without requiring --force'
complete -c pm -n '__fish_seen_subcommand_from learnings' -l allow-audit-learning -d 'Allow non-owner append-only learning audits without requiring --force'
complete -c pm -n '__fish_seen_subcommand_from comments notes learnings' -l force -d 'Force override'

# test flags
complete -c pm -n '__fish_seen_subcommand_from test' -l add -d 'Add linked test entry' -r
complete -c pm -n '__fish_seen_subcommand_from test' -l remove -d 'Remove linked test entry' -r
complete -c pm -n '__fish_seen_subcommand_from test' -l run -d 'Run linked tests'
complete -c pm -n '__fish_seen_subcommand_from test' -l background -d 'Run linked tests in managed background mode'
complete -c pm -n '__fish_seen_subcommand_from test' -l timeout -d 'Default timeout seconds' -r
complete -c pm -n '__fish_seen_subcommand_from test' -l progress -d 'Emit linked-test progress to stderr'
complete -c pm -n '__fish_seen_subcommand_from test' -l env-set -d 'Set linked-test runtime environment values' -r
complete -c pm -n '__fish_seen_subcommand_from test' -l env-clear -d 'Clear linked-test runtime environment values' -r
complete -c pm -n '__fish_seen_subcommand_from test' -l shared-host-safe -d 'Apply shared-host-safe runtime defaults'
complete -c pm -n '__fish_seen_subcommand_from test' -l pm-context -d 'PM linked-test context mode' -r -a 'schema tracker auto'
complete -c pm -n '__fish_seen_subcommand_from test' -l override-linked-pm-context -d 'Force run-level --pm-context over per-linked-test metadata'
complete -c pm -n '__fish_seen_subcommand_from test' -l fail-on-context-mismatch -d 'Fail when context item counts mismatch'
complete -c pm -n '__fish_seen_subcommand_from test' -l fail-on-skipped -d 'Treat skipped linked tests as dependency failures'
complete -c pm -n '__fish_seen_subcommand_from test' -l fail-on-empty-test-run -d 'Treat empty linked-test selections as failures'
complete -c pm -n '__fish_seen_subcommand_from test' -l require-assertions-for-pm -d 'Require assertions for linked PM command tests'
complete -c pm -n '__fish_seen_subcommand_from test' -l check-context -d 'Preflight linked PM command context diagnostics before execution'
complete -c pm -n '__fish_seen_subcommand_from test' -l auto-pm-context -d 'Auto-remediate tracker-read context mismatches using tracker context'
complete -c pm -n '__fish_seen_subcommand_from test' -l author -d 'Mutation author' -r
complete -c pm -n '__fish_seen_subcommand_from test' -l message -d 'History message' -r
complete -c pm -n '__fish_seen_subcommand_from test' -l force -d 'Force override'

# test-all flags
complete -c pm -n '__fish_seen_subcommand_from test-all' -l status  -d 'Filter by status' -r -a 'open in_progress'
complete -c pm -n '__fish_seen_subcommand_from test-all' -l limit -d 'Limit matching items before running linked tests' -r
complete -c pm -n '__fish_seen_subcommand_from test-all' -l offset -d 'Skip matching items before running linked tests' -r
complete -c pm -n '__fish_seen_subcommand_from test-all' -l background -d 'Run linked tests in managed background mode'
complete -c pm -n '__fish_seen_subcommand_from test-all' -l timeout -d 'Default timeout seconds' -r
complete -c pm -n '__fish_seen_subcommand_from test-all' -l progress -d 'Emit linked-test progress to stderr'
complete -c pm -n '__fish_seen_subcommand_from test-all' -l env-set -d 'Set linked-test runtime environment values' -r
complete -c pm -n '__fish_seen_subcommand_from test-all' -l env-clear -d 'Clear linked-test runtime environment values' -r
complete -c pm -n '__fish_seen_subcommand_from test-all' -l shared-host-safe -d 'Apply shared-host-safe runtime defaults'
complete -c pm -n '__fish_seen_subcommand_from test-all' -l pm-context -d 'PM linked-test context mode' -r -a 'schema tracker auto'
complete -c pm -n '__fish_seen_subcommand_from test-all' -l override-linked-pm-context -d 'Force run-level --pm-context over per-linked-test metadata'
complete -c pm -n '__fish_seen_subcommand_from test-all' -l fail-on-context-mismatch -d 'Fail when context item counts mismatch'
complete -c pm -n '__fish_seen_subcommand_from test-all' -l fail-on-skipped -d 'Treat skipped linked tests as dependency failures'
complete -c pm -n '__fish_seen_subcommand_from test-all' -l fail-on-empty-test-run -d 'Treat empty linked-test selections as failures'
complete -c pm -n '__fish_seen_subcommand_from test-all' -l require-assertions-for-pm -d 'Require assertions for linked PM command tests'
complete -c pm -n '__fish_seen_subcommand_from test-all' -l check-context -d 'Preflight linked PM command context diagnostics before execution'
complete -c pm -n '__fish_seen_subcommand_from test-all' -l auto-pm-context -d 'Auto-remediate tracker-read context mismatches using tracker context'

# test-runs flags
complete -c pm -n '__fish_seen_subcommand_from test-runs' -a 'list status logs stop resume' -d 'test-runs subcommand'
complete -c pm -n '__fish_seen_subcommand_from test-runs' -l status -d 'Filter background runs by status' -r -a 'queued running passed failed stopped canceled'
complete -c pm -n '__fish_seen_subcommand_from test-runs' -l limit -d 'Limit returned runs' -r
complete -c pm -n '__fish_seen_subcommand_from test-runs' -l stream -d 'Background log stream selector' -r -a 'stdout stderr both'
complete -c pm -n '__fish_seen_subcommand_from test-runs' -l tail -d 'Tail number of lines from logs' -r
complete -c pm -n '__fish_seen_subcommand_from test-runs' -l force -d 'Force-stop run with SIGKILL'
complete -c pm -n '__fish_seen_subcommand_from test-runs' -l author -d 'Resume author' -r

# gc flags
complete -c pm -n '__fish_seen_subcommand_from gc' -l dry-run -d 'Preview cleanup targets without deleting files'
complete -c pm -n '__fish_seen_subcommand_from gc' -l scope -d 'Limit cleanup to index/embeddings/runtime scopes' -r

# append flags
complete -c pm -n '__fish_seen_subcommand_from append' -s b -l body -d 'Item body' -r
complete -c pm -n '__fish_seen_subcommand_from append' -l author -d 'Mutation author' -r
complete -c pm -n '__fish_seen_subcommand_from append' -l message -d 'History message' -r
complete -c pm -n '__fish_seen_subcommand_from append' -l force -d 'Force override'

# close flags
complete -c pm -n '__fish_seen_subcommand_from claim release start-task pause-task close close-task' -l author -d 'Mutation author' -r
complete -c pm -n '__fish_seen_subcommand_from claim release start-task pause-task close close-task' -l message -d 'History message' -r
complete -c pm -n '__fish_seen_subcommand_from claim release start-task pause-task close close-task' -l force -d 'Force override'
complete -c pm -n '__fish_seen_subcommand_from close close-task' -l validate-close -d 'Validate closure metadata mode' -r -a 'off warn strict'
complete -c pm -n '__fish_seen_subcommand_from release' -l allow-audit-release -d 'Allow non-owner release handoffs without requiring --force'

# validate flags
complete -c pm -n '__fish_seen_subcommand_from validate' -l check-metadata -d 'Run metadata completeness checks'
complete -c pm -n '__fish_seen_subcommand_from validate' -l metadata-profile -d 'Select metadata validation profile for --check-metadata' -r -a 'core strict custom'
complete -c pm -n '__fish_seen_subcommand_from validate' -l check-resolution -d 'Run closed-item resolution metadata checks'
complete -c pm -n '__fish_seen_subcommand_from validate' -l check-lifecycle -d 'Run active-item lifecycle governance drift checks'
complete -c pm -n '__fish_seen_subcommand_from validate' -l check-stale-blockers -d 'Include stale blocker-pattern diagnostics in lifecycle checks'
complete -c pm -n '__fish_seen_subcommand_from validate' -l dependency-cycle-severity -d 'Set dependency-cycle warning policy for lifecycle checks' -r -a 'off warn error'
complete -c pm -n '__fish_seen_subcommand_from validate' -l check-files -d 'Run linked-file and orphaned-file checks'
complete -c pm -n '__fish_seen_subcommand_from validate' -l scan-mode -d 'Select file candidate scan mode for --check-files' -r -a 'default tracked-all tracked-all-strict'
complete -c pm -n '__fish_seen_subcommand_from validate' -l include-pm-internals -d 'Include PM storage internals in tracked-all candidate scans'
complete -c pm -n '__fish_seen_subcommand_from validate' -l verbose-file-lists -d 'Include full file-path lists for validate --check-files details'
complete -c pm -n '__fish_seen_subcommand_from validate' -l strict-exit -d 'Return non-zero exit when validation warnings are present'
complete -c pm -n '__fish_seen_subcommand_from validate' -l fail-on-warn -d 'Alias for --strict-exit'
complete -c pm -n '__fish_seen_subcommand_from validate' -l check-history-drift -d 'Run item/history hash drift checks'
complete -c pm -n '__fish_seen_subcommand_from validate' -l check-command-references -d 'Run linked-command PM-ID reference checks'
complete -c pm -n '__fish_seen_subcommand_from init' -l preset -d 'Governance preset for new setups' -r -a 'minimal default strict'
complete -c pm -n '__fish_seen_subcommand_from config' -l criterion -d 'Criteria value for definition-of-done metadata-required-fields or lifecycle pattern keys (repeatable for set)' -r
complete -c pm -n '__fish_seen_subcommand_from config' -l clear-criteria -d 'Clear config criteria-list key values'
complete -c pm -n '__fish_seen_subcommand_from config' -l format -d 'Item format for item-format key' -r -a 'toon'
complete -c pm -n '__fish_seen_subcommand_from config' -l policy -d 'Policy value for supported policy keys' -r
complete -c pm -n '__fish_seen_subcommand_from health' -l strict-directories -d 'Treat optional item-type directories as required failures'
complete -c pm -n '__fish_seen_subcommand_from health' -l check-only -d 'Run read-only health diagnostics without refreshing vectors'
complete -c pm -n '__fish_seen_subcommand_from health' -l no-refresh -d 'Disable automatic vector refresh attempts during health checks'
complete -c pm -n '__fish_seen_subcommand_from health' -l refresh-vectors -d 'Explicitly enable vector refresh attempts during health checks'
complete -c pm -n '__fish_seen_subcommand_from health' -l verbose-stale-items -d 'Include full stale vectorization ID lists in health output'
complete -c pm -n '__fish_seen_subcommand_from health' -l strict-exit -d 'Return non-zero exit when health warnings are present'
complete -c pm -n '__fish_seen_subcommand_from health' -l fail-on-warn -d 'Alias for --strict-exit'
complete -c pm -n '__fish_seen_subcommand_from comments-audit' -l status -d 'Filter by item status' -r -a 'draft open in_progress blocked closed canceled'
complete -c pm -n '__fish_seen_subcommand_from comments-audit' -l type -d 'Filter by item type' -r -a '${typeChoices}'
complete -c pm -n '__fish_seen_subcommand_from comments-audit' -l tag -d 'Filter by tag' -r -a ${fishTagChoices}
complete -c pm -n '__fish_seen_subcommand_from comments-audit' -l priority -d 'Filter by priority' -r -a '0 1 2 3 4'
complete -c pm -n '__fish_seen_subcommand_from comments-audit' -l parent -d 'Filter by parent item ID' -r
complete -c pm -n '__fish_seen_subcommand_from comments-audit' -l sprint -d 'Filter by sprint' -r
complete -c pm -n '__fish_seen_subcommand_from comments-audit' -l release -d 'Filter by release' -r
complete -c pm -n '__fish_seen_subcommand_from comments-audit' -l assignee -d 'Filter by assignee' -r
complete -c pm -n '__fish_seen_subcommand_from comments-audit' -l assignee-filter -d 'Filter assignee presence' -r -a 'assigned unassigned'
complete -c pm -n '__fish_seen_subcommand_from comments-audit' -l limit-items -d 'Limit returned item count' -r
complete -c pm -n '__fish_seen_subcommand_from comments-audit' -l limit -d 'Alias for --limit-items' -r
complete -c pm -n '__fish_seen_subcommand_from comments-audit' -l full-history -d 'Export full comment history rows (cannot be combined with --latest)'
complete -c pm -n '__fish_seen_subcommand_from comments-audit' -l latest -d 'Return latest n comments per item (0 for summary-only rows)' -r

# completion shell argument
complete -c pm -n '__fish_seen_subcommand_from completion' -l eager-tags -d 'Embed current tracker tags directly in script output'
complete -c pm -n '__fish_seen_subcommand_from completion' -a 'bash zsh fish' -d 'Shell type'

# templates subcommands
complete -c pm -n '__fish_seen_subcommand_from templates' -a 'save list show' -d 'Templates command'

# extension lifecycle flags
complete -c pm -n '__fish_seen_subcommand_from extension' -a 'init scaffold install uninstall explore manage doctor adopt adopt-all activate deactivate' -d 'Extension action subcommand'
complete -c pm -n '__fish_seen_subcommand_from extension' -l init -d 'Generate starter extension scaffold'
complete -c pm -n '__fish_seen_subcommand_from extension' -l scaffold -d 'Alias for --init'
complete -c pm -n '__fish_seen_subcommand_from extension' -l install -d 'Install extension from local path or GitHub source'
complete -c pm -n '__fish_seen_subcommand_from extension' -l uninstall -d 'Uninstall extension by name'
complete -c pm -n '__fish_seen_subcommand_from extension' -l explore -d 'List discovered extensions for selected scope'
complete -c pm -n '__fish_seen_subcommand_from extension' -l manage -d 'List managed extensions with update metadata'
complete -c pm -n '__fish_seen_subcommand_from extension' -l doctor -d 'Run consolidated extension diagnostics'
complete -c pm -n '__fish_seen_subcommand_from extension' -l adopt -d 'Adopt an unmanaged extension into managed metadata'
complete -c pm -n '__fish_seen_subcommand_from extension' -l adopt-all -d 'Adopt all unmanaged extensions into managed metadata'
complete -c pm -n '__fish_seen_subcommand_from extension' -l activate -d 'Activate extension in selected scope settings'
complete -c pm -n '__fish_seen_subcommand_from extension' -l deactivate -d 'Deactivate extension in selected scope settings'
complete -c pm -n '__fish_seen_subcommand_from extension' -l project -d 'Use project extension scope'
complete -c pm -n '__fish_seen_subcommand_from extension' -l local -d 'Alias for --project'
complete -c pm -n '__fish_seen_subcommand_from extension' -l global -d 'Use global extension scope'
complete -c pm -n '__fish_seen_subcommand_from extension' -l gh -d 'GitHub shorthand owner/repo/path' -r
complete -c pm -n '__fish_seen_subcommand_from extension' -l github -d 'Alias for --gh' -r
complete -c pm -n '__fish_seen_subcommand_from extension' -l ref -d 'Git ref/branch/tag for GitHub source' -r
complete -c pm -n '__fish_seen_subcommand_from extension' -l detail -d 'Detail mode for extension diagnostics' -r -a 'summary deep'
complete -c pm -n '__fish_seen_subcommand_from extension' -l trace -d 'Include registration traces in doctor deep diagnostics'
complete -c pm -n '__fish_seen_subcommand_from extension' -l runtime-probe -d 'Opt-in runtime activation probe for manage output'
complete -c pm -n '__fish_seen_subcommand_from extension' -l fix-managed-state -d 'Adopt unmanaged extensions before diagnostics/update checks'
complete -c pm -n '__fish_seen_subcommand_from extension' -l strict-exit -d 'Return non-zero exit when doctor warnings are present'
complete -c pm -n '__fish_seen_subcommand_from extension' -l fail-on-warn -d 'Alias for --strict-exit (doctor)'`;
}
const SETUP_HINTS = {
    bash: 'Add to ~/.bashrc or ~/.bash_profile: eval "$(pm completion bash)"',
    zsh: 'Add to ~/.zshrc: eval "$(pm completion zsh)"',
    fish: "Run: pm completion fish > ~/.config/fish/completions/pm.fish",
};
export function runCompletion(shell, itemTypes = DEFAULT_ITEM_TYPES, tags = [], eagerTagExpansion = false, runtime = {}) {
    const normalized = shell.trim().toLowerCase();
    if (!VALID_SHELLS.includes(normalized)) {
        throw new PmCliError(`Unknown shell: "${shell}". Supported shells: ${VALID_SHELLS.join(", ")}.`, EXIT_CODE.USAGE);
    }
    const validShell = normalized;
    let script;
    if (validShell === "bash") {
        script = generateBashScript(itemTypes, tags, eagerTagExpansion, runtime);
    }
    else if (validShell === "zsh") {
        script = generateZshScript(itemTypes, tags, eagerTagExpansion);
    }
    else {
        script = generateFishScript(itemTypes, tags, eagerTagExpansion);
    }
    return {
        shell: validShell,
        script,
        setup_hint: SETUP_HINTS[validShell],
    };
}
