// @unbrained/pm-cli/dist/cli/commands/delete.js
import { pathExists } from "../../core/fs/fs-utils.js";
import { EXIT_CODE } from "../../core/shared/constants.js";
import { PmCliError } from "../../core/shared/errors.js";
import { deleteItem } from "../../core/store/item-store.js";
import { getSettingsPath, resolvePmRoot } from "../../core/store/paths.js";
import { readSettings } from "../../core/store/settings.js";
function toAuthor(candidate, defaultAuthor) {
    const resolved = candidate ?? process.env.PM_AUTHOR ?? defaultAuthor;
    const trimmed = resolved.trim();
    return trimmed || "unknown";
}
export async function runDelete(id, options, global) {
    const pmRoot = resolvePmRoot(process.cwd(), global.path);
    if (!(await pathExists(getSettingsPath(pmRoot)))) {
        throw new PmCliError(`Tracker is not initialized at ${pmRoot}. Run pm init first.`, EXIT_CODE.NOT_FOUND);
    }
    const settings = await readSettings(pmRoot);
    const author = toAuthor(options.author, settings.author_default);
    const result = await deleteItem({
        pmRoot,
        settings,
        id,
        author,
        message: options.message,
        force: options.force,
    });
    return {
        item: result.item,
        changed_fields: result.changedFields,
        warnings: result.warnings,
    };
}
