// @unbrained/pm-cli/dist/cli/commands/gc.js
import fs from "node:fs/promises";
import path from "node:path";
import { runActiveOnIndexHooks, runActiveOnReadHooks, runActiveOnWriteHooks } from "../../core/extensions/index.js";
import { pathExists } from "../../core/fs/fs-utils.js";
import { EXIT_CODE } from "../../core/shared/constants.js";
import { PmCliError } from "../../core/shared/errors.js";
import { nowIso } from "../../core/shared/time.js";
import { getSettingsPath, resolvePmRoot } from "../../core/store/paths.js";
import { readSettings } from "../../core/store/settings.js";
const GC_SCOPE_VALUES = ["index", "embeddings", "runtime"];
const GC_TARGETS = [
    {
        scope: "index",
        relativePath: "index/manifest.json",
        kind: "file",
    },
    {
        scope: "embeddings",
        relativePath: "search/embeddings.jsonl",
        kind: "file",
    },
    {
        scope: "runtime",
        relativePath: "runtime/test-runs",
        kind: "directory",
    },
];
function isErrno(error, code) {
    return typeof error === "object" && error !== null && "code" in error && error.code === code;
}
async function removeCacheFile(pmRoot, target, dryRun) {
    const relativePath = target.relativePath;
    const absolutePath = path.join(pmRoot, relativePath);
    const warnings = await runActiveOnReadHooks({
        path: absolutePath,
        scope: "project",
    });
    try {
        const stats = await fs.stat(absolutePath);
        if (target.kind === "file" && !stats.isFile()) {
            return {
                removed: false,
                warnings: [...warnings, `not_a_file:${relativePath}`],
            };
        }
        if (target.kind === "directory" && !stats.isDirectory()) {
            return {
                removed: false,
                warnings: [...warnings, `not_a_directory:${relativePath}`],
            };
        }
        if (dryRun) {
            return {
                removed: true,
                warnings,
            };
        }
        if (target.kind === "file") {
            await fs.unlink(absolutePath);
        }
        else {
            await fs.rm(absolutePath, { recursive: true, force: true });
        }
        const writeWarnings = await runActiveOnWriteHooks({
            path: absolutePath,
            scope: "project",
            op: "gc:remove",
        });
        return {
            removed: true,
            warnings: [...warnings, ...writeWarnings],
        };
    }
    catch (error) {
        if (isErrno(error, "ENOENT")) {
            return {
                removed: false,
                warnings,
            };
        }
        throw error;
    }
}
function parseScopes(raw) {
    if (!raw || raw.length === 0) {
        return [...GC_SCOPE_VALUES];
    }
    const tokens = raw
        .flatMap((entry) => entry.split(","))
        .map((entry) => entry.trim().toLowerCase())
        .filter((entry) => entry.length > 0);
    if (tokens.length === 0) {
        throw new PmCliError(`--scope requires at least one value (${GC_SCOPE_VALUES.join(",")})`, EXIT_CODE.USAGE);
    }
    const resolved = new Set();
    for (const token of tokens) {
        if (!GC_SCOPE_VALUES.includes(token)) {
            throw new PmCliError(`Invalid --scope value "${token}". Expected one or more of: ${GC_SCOPE_VALUES.join(", ")}`, EXIT_CODE.USAGE);
        }
        resolved.add(token);
    }
    return GC_SCOPE_VALUES.filter((scope) => resolved.has(scope));
}
function buildGcGuidance(params) {
    const guidance = [];
    if (params.dryRun) {
        guidance.push("Dry-run preview only: no cache artifacts were deleted.");
    }
    const searchScopeSelected = params.scopes.includes("index") || params.scopes.includes("embeddings");
    const searchArtifactsAffected = params.removed.some((entry) => entry === "index/manifest.json" || entry === "search/embeddings.jsonl");
    if (searchScopeSelected && searchArtifactsAffected) {
        guidance.push('Search artifacts were removed; run "pm reindex --mode keyword" (and "--mode semantic" when semantic search is enabled) before search-heavy workflows.');
    }
    return guidance;
}
export async function runGc(global, options = {}) {
    const pmRoot = resolvePmRoot(process.cwd(), global.path);
    if (!(await pathExists(getSettingsPath(pmRoot)))) {
        throw new PmCliError(`Tracker is not initialized at ${pmRoot}. Run pm init first.`, EXIT_CODE.NOT_FOUND);
    }
    await readSettings(pmRoot);
    const dryRun = options.dryRun === true;
    const scopes = parseScopes(options.scope);
    const selectedTargets = GC_TARGETS.filter((target) => scopes.includes(target.scope));
    const removed = [];
    const retained = [];
    const warnings = [];
    for (const target of selectedTargets) {
        const result = await removeCacheFile(pmRoot, target, dryRun);
        if (result.removed) {
            removed.push(target.relativePath);
        }
        else {
            retained.push(target.relativePath);
        }
        warnings.push(...result.warnings);
    }
    warnings.push(...(await runActiveOnIndexHooks({
        mode: dryRun ? "gc:dry-run" : "gc",
        total_items: selectedTargets.length,
    })));
    const guidance = buildGcGuidance({
        dryRun,
        scopes,
        removed,
    });
    return {
        ok: warnings.length === 0,
        dry_run: dryRun,
        scope: scopes,
        removed,
        retained,
        warnings,
        guidance,
        generated_at: nowIso(),
    };
}
