// @unbrained/pm-cli/dist/cli/commands/guide.js
import { readFile } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { EXIT_CODE } from "../../core/shared/constants.js";
import { PmCliError } from "../../core/shared/errors.js";
import { listGuideTopicIds, listGuideTopics, resolveGuideTopic, } from "../guide-topics.js";
export const GUIDE_OUTPUT_VALUES = ["markdown", "toon", "json"];
export const GUIDE_DEPTH_VALUES = ["brief", "standard", "deep"];
const PACKAGE_ROOT_FROM_SOURCE = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..", "..", "..");
function parseGuideOutputFormat(raw) {
    if (!raw) {
        return undefined;
    }
    const normalized = raw.trim().toLowerCase();
    if (!GUIDE_OUTPUT_VALUES.includes(normalized)) {
        throw new PmCliError("Guide format must be one of markdown|toon|json", EXIT_CODE.USAGE);
    }
    return normalized;
}
export function resolveGuideOutputFormat(options, global) {
    const commandFormat = parseGuideOutputFormat(options.format);
    if (global.json && commandFormat && commandFormat !== "json") {
        throw new PmCliError("Cannot combine --json with --format markdown|toon", EXIT_CODE.USAGE);
    }
    if (global.json) {
        return "json";
    }
    return commandFormat ?? "toon";
}
function parseGuideDepth(raw) {
    if (!raw) {
        return "brief";
    }
    const normalized = raw.trim().toLowerCase();
    if (!GUIDE_DEPTH_VALUES.includes(normalized)) {
        throw new PmCliError("Guide depth must be one of brief|standard|deep", EXIT_CODE.USAGE);
    }
    return normalized;
}
function resolvePackageRoot() {
    const envRoot = process.env.PM_CLI_PACKAGE_ROOT;
    if (typeof envRoot === "string" && envRoot.trim().length > 0) {
        return path.resolve(envRoot.trim());
    }
    return PACKAGE_ROOT_FROM_SOURCE;
}
function normalizeLineEndings(value) {
    return value.replaceAll("\r\n", "\n");
}
function toExcerpt(content, maxLines, maxCharacters) {
    const normalized = normalizeLineEndings(content);
    const lines = normalized.split("\n");
    const lineSlice = lines.slice(0, maxLines);
    let excerpt = lineSlice.join("\n");
    let truncated = lines.length > maxLines;
    if (excerpt.length > maxCharacters) {
        excerpt = excerpt.slice(0, maxCharacters);
        truncated = true;
    }
    return {
        excerpt: excerpt.trimEnd(),
        truncated,
        lineCount: lines.length,
    };
}
async function renderGuideDocs(topic, depth, packageRoot) {
    const includeContent = depth !== "brief";
    const contentMode = includeContent ? (depth === "deep" ? "full" : "excerpt") : "none";
    const docs = [];
    for (const doc of topic.docs) {
        const absolutePath = path.resolve(packageRoot, doc.path);
        if (!includeContent) {
            docs.push({
                path: doc.path,
                purpose: doc.purpose,
                optional: doc.optional === true,
                exists: true,
                line_count: null,
                content_mode: "none",
                content: null,
                truncated: false,
            });
            continue;
        }
        try {
            const rawContent = await readFile(absolutePath, "utf8");
            if (depth === "deep") {
                const normalized = normalizeLineEndings(rawContent).trimEnd();
                docs.push({
                    path: doc.path,
                    purpose: doc.purpose,
                    optional: doc.optional === true,
                    exists: true,
                    line_count: normalized.length === 0 ? 0 : normalized.split("\n").length,
                    content_mode: "full",
                    content: normalized,
                    truncated: false,
                });
            }
            else {
                const excerpt = toExcerpt(rawContent, 120, 12_000);
                docs.push({
                    path: doc.path,
                    purpose: doc.purpose,
                    optional: doc.optional === true,
                    exists: true,
                    line_count: excerpt.lineCount,
                    content_mode: contentMode,
                    content: excerpt.excerpt,
                    truncated: excerpt.truncated,
                });
            }
        }
        catch (error) {
            const code = error?.code;
            const missing = code === "ENOENT";
            docs.push({
                path: doc.path,
                purpose: doc.purpose,
                optional: doc.optional === true,
                exists: false,
                line_count: null,
                content_mode: "none",
                content: null,
                truncated: false,
            });
            if (!missing) {
                throw new PmCliError(`Failed to read guide document "${doc.path}".`, EXIT_CODE.GENERIC_FAILURE);
            }
        }
    }
    return docs;
}
function buildGuideIndex(depth) {
    const topics = listGuideTopics().map((topic) => ({
        id: topic.id,
        aliases: topic.aliases,
        title: topic.title,
        summary: topic.summary,
        intent: topic.intent,
        quick_commands: depth === "brief" ? topic.commands.slice(0, 3) : topic.commands.slice(0, 6),
        docs: topic.docs.map((doc) => ({ path: doc.path, purpose: doc.purpose })),
        related: topic.related,
    }));
    return {
        output_default: "toon",
        mode: "index",
        depth,
        topics,
        suggested_next_steps: [
            "pm guide <topic>",
            "pm guide <topic> --depth standard",
            "pm guide <topic> --depth deep --format markdown",
            "pm guide <topic> --json",
        ],
    };
}
function ensureGuideTopic(topic) {
    const resolved = resolveGuideTopic(topic);
    if (resolved) {
        return resolved;
    }
    const available = listGuideTopicIds().join(", ");
    throw new PmCliError(`Unknown guide topic "${topic ?? ""}". Available topics: ${available}. Use "pm guide --list" to browse.`, EXIT_CODE.USAGE);
}
export async function runGuide(options, global) {
    const depth = parseGuideDepth(typeof options.depth === "string" ? options.depth : undefined);
    const listRequested = options.list === true;
    const topicRaw = typeof options.topic === "string" ? options.topic : undefined;
    if (listRequested || !topicRaw) {
        return buildGuideIndex(depth);
    }
    const topic = ensureGuideTopic(topicRaw);
    const packageRoot = resolvePackageRoot();
    const docs = await renderGuideDocs(topic, depth, packageRoot);
    const warnings = docs
        .filter((doc) => !doc.exists && !doc.optional)
        .map((doc) => `Missing required guide document: ${doc.path}`);
    return {
        output_default: "toon",
        mode: "topic",
        depth,
        requested_topic: topicRaw,
        topic,
        docs,
        warnings,
    };
}
function markdownCodeFence(content) {
    return content.replaceAll("```", "``\\`");
}
export function renderGuideMarkdown(result) {
    if (result.mode === "index") {
        const lines = [
            "# pm guide",
            "",
            "Token-efficient local documentation index for agent and maintainer workflows.",
            "",
            "## Topics",
        ];
        for (const topic of result.topics) {
            lines.push(`- \`${topic.id}\` - ${topic.summary}`);
            if (result.depth !== "brief") {
                lines.push(`  - intent: ${topic.intent}`);
                lines.push(`  - docs: ${topic.docs.map((doc) => `\`${doc.path}\``).join(", ")}`);
            }
        }
        lines.push("", "## Next steps");
        for (const step of result.suggested_next_steps) {
            lines.push(`- \`${step}\``);
        }
        return lines.join("\n");
    }
    const lines = [
        `# pm guide ${result.topic.id}`,
        "",
        result.topic.summary,
        "",
        "## Intent",
        result.topic.intent,
        "",
        "## Key commands",
        ...result.topic.commands.map((command) => `- \`${command}\``),
        "",
        "## Workflow prompts",
    ];
    for (const workflow of result.topic.workflows) {
        lines.push(`### ${workflow.name}`);
        lines.push(`goal: ${workflow.goal}`);
        lines.push(`prompt: ${workflow.prompt}`);
        lines.push("commands:");
        lines.push(...workflow.commands.map((command) => `- \`${command}\``));
        lines.push("");
    }
    lines.push("## Documents");
    for (const doc of result.docs) {
        const status = doc.exists ? "available" : doc.optional ? "missing (optional)" : "missing (required)";
        lines.push(`- \`${doc.path}\` - ${doc.purpose} (${status})`);
        if (doc.exists && doc.content_mode !== "none" && doc.content) {
            lines.push("");
            lines.push(`### Excerpt: \`${doc.path}\``);
            if (doc.truncated) {
                lines.push("_truncated for context efficiency_");
            }
            lines.push("```markdown");
            lines.push(markdownCodeFence(doc.content));
            lines.push("```");
        }
    }
    if (result.topic.related.length > 0) {
        lines.push("", "## Related topics");
        lines.push(...result.topic.related.map((topic) => `- \`${topic}\``));
    }
    if (result.warnings.length > 0) {
        lines.push("", "## Warnings");
        lines.push(...result.warnings.map((warning) => `- ${warning}`));
    }
    return lines.join("\n");
}
