// @unbrained/pm-cli/dist/cli/commands/init.js
import fs from "node:fs/promises";
import path from "node:path";
import readline from "node:readline/promises";
import { stdin as input, stdout as output } from "node:process";
import { getActiveExtensionRegistrations, runActiveOnWriteHooks } from "../../core/extensions/index.js";
import { pathExists } from "../../core/fs/fs-utils.js";
import { normalizePrefix } from "../../core/item/id.js";
import { resolveItemTypeRegistry } from "../../core/item/type-registry.js";
import { ensureRuntimeSchemaFileScaffold } from "../../core/schema/runtime-schema.js";
import { EXIT_CODE, GOVERNANCE_PRESET_DEFAULTS, PM_REQUIRED_SUBDIRS, SETTINGS_DEFAULTS } from "../../core/shared/constants.js";
import { PmCliError } from "../../core/shared/errors.js";
import { resolvePmRoot } from "../../core/store/paths.js";
import { readSettings, writeSettings } from "../../core/store/settings.js";
function cloneDefaults() {
    return structuredClone(SETTINGS_DEFAULTS);
}
const BUILTIN_GOVERNANCE_PRESETS = ["minimal", "default", "strict"];
function normalizeInitGovernancePreset(rawValue) {
    if (rawValue === undefined) {
        return undefined;
    }
    const normalized = rawValue.trim().toLowerCase().replaceAll("-", "_");
    if (normalized.length === 0) {
        throw new PmCliError("--preset must not be empty", EXIT_CODE.USAGE);
    }
    if (normalized === "minimal" || normalized === "default" || normalized === "strict") {
        return normalized;
    }
    if (normalized === "lite" || normalized === "minimum") {
        return "minimal";
    }
    throw new PmCliError(`Invalid --preset value "${rawValue}". Allowed: ${BUILTIN_GOVERNANCE_PRESETS.join(", ")}`, EXIT_CODE.USAGE);
}
function parseYesNoChoice(answer, currentDefault) {
    const normalized = answer.trim().toLowerCase();
    if (normalized.length === 0) {
        return currentDefault;
    }
    if (normalized === "y" || normalized === "yes") {
        return true;
    }
    if (normalized === "n" || normalized === "no") {
        return false;
    }
    return currentDefault;
}
function applyGovernancePreset(settings, preset) {
    const knobs = GOVERNANCE_PRESET_DEFAULTS[preset];
    settings.governance = {
        preset,
        ...knobs,
    };
    settings.validation.parent_reference = knobs.parent_reference;
    settings.validation.metadata_profile = knobs.metadata_profile;
}
async function runInitWizard(initialPrefix, telemetryDefault) {
    const rl = readline.createInterface({ input, output });
    try {
        output.write("pm init setup wizard (agent-optimized)\n");
        output.write("This walkthrough is non-destructive and each choice can be changed later with pm config.\n\n");
        output.write("1/3 Item ID prefix\n");
        output.write("Prefix is prepended to generated IDs (for example pm-a1b2).\n");
        const prefixAnswer = await rl.question(`Item ID prefix [${initialPrefix}]: `);
        const resolvedPrefix = prefixAnswer.trim().length > 0 ? normalizePrefix(prefixAnswer) : initialPrefix;
        output.write("\n2/3 Governance preset\n");
        output.write("minimal: no ownership blocking, progressive create defaults, close validation off.\n");
        output.write("default: ownership conflict warnings, progressive create defaults, close validation warn.\n");
        output.write("strict: ownership blocking, strict create defaults, close validation strict.\n");
        const presetAnswer = await rl.question("Governance preset [minimal/default/strict] (default: minimal): ");
        const resolvedPreset = normalizeInitGovernancePreset(presetAnswer.trim().length > 0 ? presetAnswer : "minimal") ?? "minimal";
        output.write("\n3/3 Project telemetry\n");
        output.write("Telemetry helps improve reliability and can be disabled anytime via pm config.\n");
        const telemetryLabel = telemetryDefault ? "Y/n" : "y/N";
        const telemetryAnswer = await rl.question(`Enable telemetry for this project? [${telemetryLabel}] `);
        const telemetryEnabled = parseYesNoChoice(telemetryAnswer, telemetryDefault);
        output.write("\n");
        return {
            prefix: resolvedPrefix,
            preset: resolvedPreset,
            telemetry_enabled: telemetryEnabled,
        };
    }
    finally {
        rl.close();
    }
}
export async function runInit(prefixArg, global, options = {}) {
    const cwd = process.cwd();
    const pmRoot = resolvePmRoot(cwd, global.path);
    const createdDirs = [];
    const warnings = [];
    let wizardUsed = false;
    for (const subdir of PM_REQUIRED_SUBDIRS) {
        const target = subdir ? path.join(pmRoot, subdir) : pmRoot;
        const existed = await pathExists(target);
        await fs.mkdir(target, { recursive: true });
        if (existed) {
            warnings.push(`already_exists:${target}`);
        }
        else {
            createdDirs.push(target);
        }
        warnings.push(...(await runActiveOnWriteHooks({
            path: target,
            scope: "project",
            op: "init:ensure_dir",
        })));
    }
    const settingsPath = path.join(pmRoot, "settings.json");
    const settingsExists = await pathExists(settingsPath);
    let normalizedPrefix = normalizePrefix(prefixArg);
    const presetFromOption = normalizeInitGovernancePreset(options.preset);
    let chosenPreset = presetFromOption;
    let chosenTelemetryEnabled;
    let settings;
    if (settingsExists) {
        settings = await readSettings(pmRoot);
        warnings.push(`already_exists:${settingsPath}`);
        let changed = false;
        if (prefixArg !== undefined && settings.id_prefix !== normalizedPrefix) {
            settings.id_prefix = normalizedPrefix;
            warnings.push(`updated:id_prefix:${normalizedPrefix}`);
            changed = true;
        }
        if (presetFromOption !== undefined && settings.governance.preset !== presetFromOption) {
            applyGovernancePreset(settings, presetFromOption);
            warnings.push(`updated:governance_preset:${presetFromOption}`);
            changed = true;
        }
        if (changed) {
            await writeSettings(pmRoot, settings);
        }
    }
    else {
        if (presetFromOption === undefined && process.stdin.isTTY === true && process.stdout.isTTY === true) {
            const wizardChoices = await runInitWizard(normalizedPrefix, SETTINGS_DEFAULTS.telemetry.enabled);
            normalizedPrefix = wizardChoices.prefix;
            chosenPreset = wizardChoices.preset;
            chosenTelemetryEnabled = wizardChoices.telemetry_enabled;
            wizardUsed = true;
        }
        const effectivePreset = chosenPreset ?? "minimal";
        settings = cloneDefaults();
        settings.id_prefix = normalizedPrefix;
        applyGovernancePreset(settings, effectivePreset);
        if (chosenTelemetryEnabled !== undefined) {
            settings.telemetry.enabled = chosenTelemetryEnabled;
            settings.telemetry.first_run_prompt_completed = true;
        }
        await writeSettings(pmRoot, settings);
    }
    const runtimeSchemaScaffold = await ensureRuntimeSchemaFileScaffold(pmRoot, settings.schema);
    for (const createdPath of runtimeSchemaScaffold.created_paths) {
        createdDirs.push(createdPath);
        warnings.push(...(await runActiveOnWriteHooks({
            path: createdPath,
            scope: "project",
            op: "init:runtime_schema_file",
        })));
    }
    const typeRegistry = resolveItemTypeRegistry(settings, getActiveExtensionRegistrations());
    for (const typeFolder of typeRegistry.folders) {
        if (PM_REQUIRED_SUBDIRS.includes(typeFolder)) {
            continue;
        }
        const target = path.join(pmRoot, typeFolder);
        const existed = await pathExists(target);
        await fs.mkdir(target, { recursive: true });
        if (existed) {
            warnings.push(`already_exists:${target}`);
        }
        else {
            createdDirs.push(target);
        }
        warnings.push(...(await runActiveOnWriteHooks({
            path: target,
            scope: "project",
            op: "init:ensure_dir",
        })));
    }
    return {
        ok: true,
        path: pmRoot,
        settings,
        created_dirs: createdDirs,
        warnings,
        governance_preset: settings.governance.preset,
        wizard_used: wizardUsed,
    };
}
