// @unbrained/pm-cli/dist/cli/commands/list.js
import { pathExists } from "../../core/fs/fs-utils.js";
import { getActiveExtensionRegistrations } from "../../core/extensions/index.js";
import { normalizeStatusInput } from "../../core/item/status.js";
import { resolveItemTypeRegistry } from "../../core/item/type-registry.js";
import { parseIntegerLimit, parsePriority, parseType } from "../shared-parsers.js";
import { collectRuntimeFilterValues, matchesRuntimeFilters } from "../../core/schema/runtime-field-filters.js";
import { resolveRuntimeFieldRegistry, resolveRuntimeStatusRegistry, } from "../../core/schema/runtime-schema.js";
import { EXIT_CODE } from "../../core/shared/constants.js";
import { PmCliError } from "../../core/shared/errors.js";
import { compareTimestampStrings, nowIso, resolveIsoOrRelative } from "../../core/shared/time.js";
import { listAllFrontMatter, listAllFrontMatterWithBody } from "../../core/store/item-store.js";
import { getSettingsPath, resolvePmRoot } from "../../core/store/paths.js";
import { readSettings } from "../../core/store/settings.js";
export const LIST_SORT_FIELDS = ["priority", "deadline", "updated_at", "created_at", "title", "parent"];
export const LIST_SORT_ORDER_VALUES = ["asc", "desc"];
const DEFAULT_COMPACT_LIST_FIELDS = ["id", "title", "status", "type", "priority", "parent", "updated_at"];
const BRIEF_LIST_FIELDS = ["id", "status", "type", "title"];
function isTerminal(status, statusRegistry) {
    const normalized = normalizeStatusInput(status, statusRegistry) ?? status;
    return statusRegistry.terminal_statuses.has(normalized);
}
function compareDefaultSort(left, right, statusRegistry) {
    const leftTerminal = isTerminal(left.status, statusRegistry);
    const rightTerminal = isTerminal(right.status, statusRegistry);
    if (leftTerminal !== rightTerminal) {
        return leftTerminal ? 1 : -1;
    }
    const byPriority = left.priority - right.priority;
    if (byPriority !== 0) {
        return byPriority;
    }
    const byUpdated = compareTimestampStrings(right.updated_at, left.updated_at);
    if (byUpdated !== 0) {
        return byUpdated;
    }
    return left.id.localeCompare(right.id);
}
function sortItemsDefault(items, statusRegistry) {
    return [...items].sort((left, right) => compareDefaultSort(left, right, statusRegistry));
}
function parseDeadline(raw, fieldLabel) {
    if (raw === undefined)
        return undefined;
    return resolveIsoOrRelative(raw, new Date(), fieldLabel);
}
function parseOffset(raw) {
    if (raw === undefined)
        return undefined;
    const parsed = Number(raw);
    if (!Number.isInteger(parsed) || parsed < 0) {
        throw new PmCliError("Offset filter must be a non-negative integer", EXIT_CODE.USAGE);
    }
    return parsed;
}
function parseFieldSelectors(raw) {
    if (raw === undefined) {
        return undefined;
    }
    const selectors = raw
        .split(",")
        .map((entry) => entry.trim())
        .filter((entry) => entry.length > 0);
    if (selectors.length === 0) {
        throw new PmCliError("List --fields requires a comma-separated list of field names", EXIT_CODE.USAGE);
    }
    return [...new Set(selectors)];
}
function parseProjectionConfig(options) {
    const compactRequested = options.compact === true;
    const briefRequested = options.brief === true;
    const fieldSelectors = parseFieldSelectors(options.fields);
    const enabledModes = Number(compactRequested) + Number(briefRequested) + Number(fieldSelectors !== undefined);
    if (enabledModes > 1) {
        throw new PmCliError("List projection options are mutually exclusive. Use one of --compact, --brief, or --fields.", EXIT_CODE.USAGE);
    }
    if (briefRequested) {
        return {
            mode: "compact",
            fields: [...BRIEF_LIST_FIELDS],
        };
    }
    if (compactRequested) {
        return {
            mode: "compact",
            fields: [...DEFAULT_COMPACT_LIST_FIELDS],
        };
    }
    if (fieldSelectors) {
        return {
            mode: "fields",
            fields: fieldSelectors,
        };
    }
    return {
        mode: "full",
        fields: [],
    };
}
function parseSortField(raw) {
    if (raw === undefined) {
        return undefined;
    }
    const normalized = raw.trim().toLowerCase();
    if (!LIST_SORT_FIELDS.includes(normalized)) {
        throw new PmCliError(`Sort field must be one of ${LIST_SORT_FIELDS.join("|")}`, EXIT_CODE.USAGE);
    }
    return normalized;
}
function parseSortOrder(raw) {
    if (raw === undefined) {
        return undefined;
    }
    const normalized = raw.trim().toLowerCase();
    if (!LIST_SORT_ORDER_VALUES.includes(normalized)) {
        throw new PmCliError(`Sort order must be one of ${LIST_SORT_ORDER_VALUES.join("|")}`, EXIT_CODE.USAGE);
    }
    return normalized;
}
function parseAssigneeFilter(raw) {
    if (raw === undefined) {
        return undefined;
    }
    const normalized = raw.trim().toLowerCase();
    if (!normalized) {
        throw new PmCliError("Assignee filter must be one of assigned|unassigned", EXIT_CODE.USAGE);
    }
    if (normalized !== "assigned" && normalized !== "unassigned") {
        throw new PmCliError(`Invalid assignee filter "${raw}". Allowed: assigned|unassigned`, EXIT_CODE.USAGE);
    }
    return normalized;
}
function resolveStatusFilter(status, statusRegistry) {
    if (status === undefined) {
        return undefined;
    }
    const normalized = normalizeStatusInput(status, statusRegistry);
    const token = status.trim().toLowerCase();
    if (token === "open") {
        return statusRegistry.open_status;
    }
    if (token === "closed") {
        return statusRegistry.close_status;
    }
    if (token === "canceled" || token === "cancelled") {
        return statusRegistry.canceled_status;
    }
    return normalized ?? status;
}
function applyFilters(items, status, options, typeRegistry, statusRegistry, runtimeFieldFilters) {
    const typeFilter = parseType(options.type, typeRegistry);
    const tagFilter = options.tag?.trim().toLowerCase();
    const priorityFilter = parsePriority(options.priority);
    const deadlineBefore = parseDeadline(options.deadlineBefore, "deadline-before");
    const deadlineAfter = parseDeadline(options.deadlineAfter, "deadline-after");
    const assigneeFilter = options.assignee?.trim();
    const assigneeModeFilter = parseAssigneeFilter(options.assigneeFilter);
    const parentFilter = options.parent?.trim();
    const sprintFilter = options.sprint?.trim();
    const releaseFilter = options.release?.trim();
    if (assigneeFilter && (assigneeFilter.toLowerCase() === "none" || assigneeFilter.toLowerCase() === "null")) {
        throw new PmCliError('--assignee no longer accepts "none" or "null". Use --assignee-filter unassigned.', EXIT_CODE.USAGE);
    }
    if (assigneeFilter !== undefined && assigneeModeFilter === "unassigned") {
        throw new PmCliError("Cannot combine --assignee with --assignee-filter unassigned", EXIT_CODE.USAGE);
    }
    return items.filter((item) => {
        if (status && item.status !== status)
            return false;
        if (options.excludeTerminal && isTerminal(item.status, statusRegistry))
            return false;
        if (typeFilter && item.type !== typeFilter)
            return false;
        if (tagFilter && !item.tags.includes(tagFilter))
            return false;
        if (priorityFilter !== undefined && item.priority !== priorityFilter)
            return false;
        if (deadlineBefore && (!item.deadline || compareTimestampStrings(item.deadline, deadlineBefore) > 0))
            return false;
        if (deadlineAfter && (!item.deadline || compareTimestampStrings(item.deadline, deadlineAfter) < 0))
            return false;
        if (assigneeModeFilter === "assigned" && !item.assignee)
            return false;
        if (assigneeModeFilter === "unassigned" && item.assignee)
            return false;
        if (assigneeFilter !== undefined && item.assignee !== assigneeFilter) {
            return false;
        }
        if (parentFilter !== undefined && item.parent !== parentFilter)
            return false;
        if (sprintFilter !== undefined && item.sprint !== sprintFilter)
            return false;
        if (releaseFilter !== undefined && item.release !== releaseFilter)
            return false;
        if (!matchesRuntimeFilters(item, runtimeFieldFilters)) {
            return false;
        }
        return true;
    });
}
function compareNullableString(left, right) {
    if (left === right) {
        return 0;
    }
    if (left === null) {
        return 1;
    }
    if (right === null) {
        return -1;
    }
    return left.localeCompare(right);
}
function compareNullableTimestamp(left, right) {
    if (left === right) {
        return 0;
    }
    if (left === null) {
        return 1;
    }
    if (right === null) {
        return -1;
    }
    return compareTimestampStrings(left, right);
}
function compareBySortField(left, right, field) {
    switch (field) {
        case "priority":
            return left.priority - right.priority;
        case "deadline":
            return compareNullableTimestamp(left.deadline ?? null, right.deadline ?? null);
        case "updated_at":
            return compareTimestampStrings(left.updated_at, right.updated_at);
        case "created_at":
            return compareTimestampStrings(left.created_at, right.created_at);
        case "title":
            return left.title.localeCompare(right.title);
        case "parent":
            return compareNullableString(left.parent ?? null, right.parent ?? null);
        default:
            return 0;
    }
}
function sortItems(items, sortField, sortOrder, statusRegistry) {
    if (!sortField) {
        return sortItemsDefault(items, statusRegistry);
    }
    return [...items].sort((left, right) => {
        const byField = compareBySortField(left, right, sortField);
        if (byField !== 0) {
            return sortOrder === "desc" ? -byField : byField;
        }
        const fallback = compareDefaultSort(left, right, statusRegistry);
        return sortOrder === "desc" ? -fallback : fallback;
    });
}
function readListFieldValue(item, field) {
    const normalized = field.trim();
    if (normalized.length === 0) {
        return null;
    }
    if (normalized.startsWith("item.")) {
        const nestedKey = normalized.slice("item.".length);
        if (nestedKey.length === 0) {
            return null;
        }
        const itemRecord = item;
        return itemRecord[nestedKey] ?? null;
    }
    const itemRecord = item;
    if (Object.prototype.hasOwnProperty.call(itemRecord, normalized)) {
        return itemRecord[normalized] ?? null;
    }
    return null;
}
function projectListItems(items, projection) {
    if (projection.mode === "full") {
        return items;
    }
    return items.map((item) => {
        const projected = {};
        for (const field of projection.fields) {
            projected[field] = readListFieldValue(item, field);
        }
        return projected;
    });
}
export async function runList(status, options, global) {
    const pmRoot = resolvePmRoot(process.cwd(), global.path);
    if (!(await pathExists(getSettingsPath(pmRoot)))) {
        throw new PmCliError(`Tracker is not initialized at ${pmRoot}. Run pm init first.`, EXIT_CODE.NOT_FOUND);
    }
    const settings = await readSettings(pmRoot);
    const statusRegistry = resolveRuntimeStatusRegistry(settings.schema);
    const runtimeFieldRegistry = resolveRuntimeFieldRegistry(settings.schema);
    const runtimeFieldFilters = collectRuntimeFilterValues(options, runtimeFieldRegistry, "list");
    const typeRegistry = resolveItemTypeRegistry(settings, getActiveExtensionRegistrations());
    const listWarnings = [];
    const items = options.includeBody
        ? await listAllFrontMatterWithBody(pmRoot, settings.item_format, typeRegistry.type_to_folder, listWarnings, settings.schema)
        : await listAllFrontMatter(pmRoot, settings.item_format, typeRegistry.type_to_folder, listWarnings, settings.schema);
    const projection = parseProjectionConfig(options);
    const sortField = parseSortField(options.sort);
    const sortOrder = parseSortOrder(options.order) ?? "asc";
    if (!sortField && options.order !== undefined) {
        throw new PmCliError("List --order requires --sort", EXIT_CODE.USAGE);
    }
    const explicitStatus = resolveStatusFilter(options.status, statusRegistry);
    const resolvedStatus = explicitStatus ?? resolveStatusFilter(status, statusRegistry);
    const effectiveOptions = explicitStatus ? { ...options, excludeTerminal: false } : options;
    const filtered = applyFilters(items, resolvedStatus, effectiveOptions, typeRegistry, statusRegistry, runtimeFieldFilters);
    const sorted = sortItems(filtered, sortField, sortOrder, statusRegistry);
    const limit = parseIntegerLimit(options.limit);
    const offset = parseOffset(options.offset) ?? 0;
    const limited = limit === undefined ? sorted.slice(offset) : sorted.slice(offset, offset + limit);
    const projected = projectListItems(limited, projection);
    const now = nowIso();
    const warnings = [...new Set(listWarnings)].sort((left, right) => left.localeCompare(right));
    const projectionFields = projection.mode === "full" ? null : [...projection.fields];
    return {
        items: projected,
        count: projected.length,
        filters: {
            status: resolvedStatus ?? null,
            type: options.type ?? null,
            tag: options.tag ?? null,
            priority: options.priority ?? null,
            deadline_before: options.deadlineBefore ?? null,
            deadline_after: options.deadlineAfter ?? null,
            assignee: options.assignee ?? null,
            assignee_filter: options.assigneeFilter ?? null,
            parent: options.parent ?? null,
            sprint: options.sprint ?? null,
            release: options.release ?? null,
            limit: options.limit ?? null,
            offset: options.offset ?? null,
            include_body: options.includeBody ?? null,
            compact: options.compact ?? null,
            fields: options.fields ?? null,
            sort: sortField ?? null,
            order: sortField ? sortOrder : null,
            projection: projection.mode,
            runtime_filters: runtimeFieldFilters,
        },
        projection: {
            mode: projection.mode,
            fields: projectionFields,
        },
        sorting: {
            sort: sortField ?? "default",
            order: sortField ? sortOrder : "asc",
        },
        now,
        ...(warnings.length > 0 ? { warnings } : {}),
    };
}
