// @unbrained/pm-cli/dist/cli/commands/search.js
import fs from "node:fs/promises";
import path from "node:path";
import { toNonEmptyStringOrUndefined } from "../../core/shared/primitives.js";
import { getActiveExtensionRegistrations, runActiveOnReadHooks } from "../../core/extensions/index.js";
import { resolveRegisteredSearchProvider, resolveRegisteredVectorStoreAdapter, } from "../../core/extensions/runtime-registrations.js";
import { resolveItemTypeRegistry } from "../../core/item/type-registry.js";
import { parseLimit, parsePriority, parseType } from "../shared-parsers.js";
import { executeEmbeddingRequest, resolveEmbeddingProviders, } from "../../core/search/providers.js";
import { resolveSettingsWithSemanticRuntimeDefaults } from "../../core/search/semantic-defaults.js";
import { executeVectorQuery, resolveVectorStores, } from "../../core/search/vector-stores.js";
import { buildEventCorpus, buildReminderCorpus } from "../../core/search/corpus.js";
import { pathExists } from "../../core/fs/fs-utils.js";
import { parseItemDocument } from "../../core/item/item-format.js";
import { normalizeStatusInput } from "../../core/item/status.js";
import { collectRuntimeFilterValues, matchesRuntimeFilters } from "../../core/schema/runtime-field-filters.js";
import { resolveRuntimeFieldRegistry, resolveRuntimeStatusRegistry, } from "../../core/schema/runtime-schema.js";
import { EXIT_CODE } from "../../core/shared/constants.js";
import { PmCliError } from "../../core/shared/errors.js";
import { tokenizeAlphaNumeric } from "../../core/shared/text-normalization.js";
import { compareTimestampStrings, nowIso, resolveIsoOrRelative } from "../../core/shared/time.js";
import { listAllFrontMatter } from "../../core/store/item-store.js";
import { getSettingsPath, resolveGlobalPmRoot, resolvePmRoot, getItemPath } from "../../core/store/paths.js";
import { readSettings } from "../../core/store/settings.js";
const DEFAULT_COMPACT_SEARCH_FIELDS = [
    "id",
    "title",
    "status",
    "type",
    "priority",
    "updated_at",
    "score",
    "matched_fields",
];
const LONG_QUERY_TOKEN_THRESHOLD = 2;
const LONG_QUERY_TITLE_EXACT_BONUS = 120;
const LONG_QUERY_PHRASE_MULTIPLIER = 6;
const IMPLICIT_HYBRID_EMBEDDING_TIMEOUT_MS = 8_000;
const IMPLICIT_HYBRID_VECTOR_TIMEOUT_MS = 8_000;
function isTerminal(status, statusRegistry) {
    const normalized = normalizeStatusInput(status, statusRegistry) ?? status;
    return statusRegistry.terminal_statuses.has(normalized);
}
function classifyImplicitSemanticFallbackReason(error) {
    const message = (error instanceof Error ? error.message : String(error)).toLowerCase();
    if (message.includes("timed out") || message.includes("timeout")) {
        return "timeout";
    }
    if (message.includes("econnrefused") ||
        message.includes("connection refused") ||
        message.includes("connect ") ||
        message.includes("enotfound") ||
        message.includes("eai_again")) {
        return "connection";
    }
    return "error";
}
function buildImplicitSemanticFallbackWarning(error) {
    const reason = classifyImplicitSemanticFallbackReason(error);
    if (reason === "timeout") {
        return "search_implicit_semantic_fallback:timeout:using_keyword_mode";
    }
    if (reason === "connection") {
        return "search_implicit_semantic_fallback:connection:using_keyword_mode";
    }
    return "search_implicit_semantic_fallback:error:using_keyword_mode";
}
function parseMode(raw, context) {
    if (raw === undefined) {
        return context.hasProvider && context.hasVectorStore ? "hybrid" : "keyword";
    }
    const normalized = raw.trim().toLowerCase();
    if (normalized !== "keyword" && normalized !== "semantic" && normalized !== "hybrid") {
        throw new PmCliError("Search mode must be one of keyword|semantic|hybrid", EXIT_CODE.USAGE);
    }
    return normalized;
}
function parseIncludeLinked(raw) {
    return raw === true;
}
function parseTitleExact(raw) {
    return raw === true;
}
function parsePhraseExact(raw) {
    return raw === true;
}
function normalizeSearchPhrase(value) {
    return value
        .toLowerCase()
        .replace(/\s+/g, " ")
        .trim();
}
function parseDeadline(raw, fieldLabel) {
    if (raw === undefined)
        return undefined;
    return resolveIsoOrRelative(raw, new Date(), fieldLabel);
}
function parseFieldSelectors(raw) {
    if (raw === undefined) {
        return undefined;
    }
    const selectors = raw
        .split(",")
        .map((entry) => entry.trim())
        .filter((entry) => entry.length > 0);
    if (selectors.length === 0) {
        throw new PmCliError("Search --fields requires a comma-separated list of field names", EXIT_CODE.USAGE);
    }
    return [...new Set(selectors)];
}
function parseProjectionConfig(options) {
    const compactRequested = options.compact === true;
    const fullRequested = options.full === true;
    const fieldSelectors = parseFieldSelectors(options.fields);
    const enabledModes = Number(compactRequested) + Number(fullRequested) + Number(fieldSelectors !== undefined);
    if (enabledModes > 1) {
        throw new PmCliError("Search projection options are mutually exclusive. Use one of --compact, --full, or --fields.", EXIT_CODE.USAGE);
    }
    if (compactRequested) {
        return {
            mode: "compact",
            fields: [...DEFAULT_COMPACT_SEARCH_FIELDS],
        };
    }
    if (fullRequested) {
        return {
            mode: "full",
            fields: [],
        };
    }
    if (fieldSelectors) {
        return {
            mode: "fields",
            fields: fieldSelectors,
        };
    }
    return {
        mode: "full",
        fields: [],
    };
}
function parseTokens(query) {
    const normalized = normalizeSearchPhrase(query);
    if (!normalized) {
        throw new PmCliError("Search query must not be empty", EXIT_CODE.USAGE);
    }
    return normalized.split(/\s+/).filter(Boolean);
}
function collectExactPhraseFields(document) {
    const item = document.metadata;
    return [
        item.title,
        item.description,
        item.status,
        item.tags.join(" "),
        document.body,
        (item.comments ?? []).map((entry) => entry.text).join(" "),
        (item.notes ?? []).map((entry) => entry.text).join(" "),
        (item.learnings ?? []).map((entry) => entry.text).join(" "),
        buildReminderCorpus(item).join(" "),
        buildEventCorpus(item).join(" "),
        (item.dependencies ?? []).map((entry) => `${entry.id} ${entry.kind}`).join(" "),
    ];
}
function documentContainsExactPhrase(document, normalizedQuery) {
    return collectExactPhraseFields(document).some((fieldValue) => normalizeSearchPhrase(fieldValue).includes(normalizedQuery));
}
function applyExactQueryFilters(items, normalizedQuery, options) {
    if (!options.titleExact && !options.phraseExact) {
        return items;
    }
    return items.filter((document) => {
        if (options.titleExact && normalizeSearchPhrase(document.metadata.title) !== normalizedQuery) {
            return false;
        }
        if (options.phraseExact && !documentContainsExactPhrase(document, normalizedQuery)) {
            return false;
        }
        return true;
    });
}
function applyFilters(items, options, typeRegistry, runtimeFieldFilters) {
    const typeFilter = parseType(options.type, typeRegistry);
    const tagFilter = options.tag?.trim().toLowerCase();
    const priorityFilter = parsePriority(options.priority);
    const deadlineBefore = parseDeadline(options.deadlineBefore, "deadline-before");
    const deadlineAfter = parseDeadline(options.deadlineAfter, "deadline-after");
    return items.filter((document) => {
        const item = document.metadata;
        if (typeFilter && item.type !== typeFilter)
            return false;
        if (tagFilter && !item.tags.includes(tagFilter))
            return false;
        if (priorityFilter !== undefined && item.priority !== priorityFilter)
            return false;
        if (deadlineBefore && (!item.deadline || compareTimestampStrings(item.deadline, deadlineBefore) > 0))
            return false;
        if (deadlineAfter && (!item.deadline || compareTimestampStrings(item.deadline, deadlineAfter) < 0))
            return false;
        if (!matchesRuntimeFilters(item, runtimeFieldFilters))
            return false;
        return true;
    });
}
function countOccurrences(haystack, needle) {
    let count = 0;
    let index = 0;
    while (true) {
        const foundAt = haystack.indexOf(needle, index);
        if (foundAt < 0) {
            return count;
        }
        count += 1;
        index = foundAt + needle.length;
    }
}
function tokenizeForExactTokenMatch(value) {
    return tokenizeAlphaNumeric(value);
}
function collectLinkedPaths(item) {
    const fromFiles = (item.files ?? []).map((entry) => ({
        scope: entry.scope,
        path: entry.path.trim(),
    }));
    const fromDocs = (item.docs ?? []).map((entry) => ({
        scope: entry.scope,
        path: entry.path.trim(),
    }));
    const fromTests = (item.tests ?? [])
        .filter((entry) => typeof entry.path === "string" && entry.path.trim().length > 0)
        .map((entry) => ({
        scope: entry.scope,
        path: entry.path.trim(),
    }));
    const sorted = [...fromFiles, ...fromDocs, ...fromTests]
        .filter((entry) => entry.path.length > 0)
        .sort((a, b) => a.scope.localeCompare(b.scope) || a.path.localeCompare(b.path));
    const deduped = new Map();
    for (const entry of sorted) {
        deduped.set(`${entry.scope}:${entry.path}`, entry);
    }
    return [...deduped.values()];
}
function isPathWithinRoot(root, resolvedPath) {
    const relative = path.relative(root, resolvedPath);
    if (relative.length === 0) {
        return true;
    }
    return !relative.startsWith("..") && !path.isAbsolute(relative);
}
async function resolveContainmentRoot(root) {
    const resolved = path.resolve(root);
    try {
        const realpathRoot = await fs.realpath(resolved);
        return {
            resolved,
            realpath: realpathRoot,
        };
    }
    catch {
        return null;
    }
}
async function loadLinkedCorpus(document, projectRoot, globalRoot) {
    const linkedPaths = collectLinkedPaths(document.metadata);
    const chunks = [];
    const projectContainmentRoot = await resolveContainmentRoot(projectRoot);
    const globalContainmentRoot = await resolveContainmentRoot(globalRoot);
    for (const linkedPath of linkedPaths) {
        const containmentRoot = linkedPath.scope === "global" ? globalContainmentRoot : projectContainmentRoot;
        if (!containmentRoot) {
            continue;
        }
        const resolved = path.resolve(containmentRoot.resolved, linkedPath.path);
        if (!isPathWithinRoot(containmentRoot.resolved, resolved)) {
            continue;
        }
        let linkedRealpath;
        try {
            linkedRealpath = await fs.realpath(resolved);
        }
        catch {
            continue;
        }
        if (!isPathWithinRoot(containmentRoot.realpath, linkedRealpath)) {
            continue;
        }
        try {
            await runActiveOnReadHooks({
                path: resolved,
                scope: linkedPath.scope,
            });
            chunks.push(await fs.readFile(resolved, "utf8"));
        }
        catch {
            // Best-effort linked-content indexing: unreadable paths are ignored.
        }
    }
    return chunks.join("\n");
}
function scoreDocument(document, tokens, normalizedQuery, linkedCorpus, tuning) {
    const item = document.metadata;
    const titleTokenCounts = new Map();
    for (const token of tokenizeForExactTokenMatch(item.title)) {
        titleTokenCounts.set(token, (titleTokenCounts.get(token) ?? 0) + 1);
    }
    const searchableFields = [
        { name: "title", value: item.title, weight: tuning.title_weight },
        { name: "description", value: item.description, weight: tuning.description_weight },
        { name: "tags", value: item.tags.join(" "), weight: tuning.tags_weight },
        { name: "status", value: item.status, weight: tuning.status_weight },
        { name: "body", value: document.body, weight: tuning.body_weight },
        { name: "comments", value: (item.comments ?? []).map((entry) => entry.text).join(" "), weight: tuning.comments_weight },
        { name: "notes", value: (item.notes ?? []).map((entry) => entry.text).join(" "), weight: tuning.notes_weight },
        { name: "learnings", value: (item.learnings ?? []).map((entry) => entry.text).join(" "), weight: tuning.learnings_weight },
        { name: "reminders", value: buildReminderCorpus(item).join(" "), weight: tuning.reminders_weight },
        { name: "events", value: buildEventCorpus(item).join(" "), weight: tuning.events_weight },
        {
            name: "dependencies",
            value: (item.dependencies ?? []).map((entry) => `${entry.id} ${entry.kind}`).join(" "),
            weight: tuning.dependencies_weight,
        },
        { name: "linked_content", value: linkedCorpus, weight: tuning.linked_content_weight },
    ];
    let score = 0;
    const matched = new Set();
    for (const token of tokens) {
        const exactTitleMatches = titleTokenCounts.get(token) ?? 0;
        if (exactTitleMatches > 0) {
            score += exactTitleMatches * tuning.title_exact_bonus;
            matched.add("title");
        }
        for (const field of searchableFields) {
            const fieldValue = field.value.toLowerCase();
            const occurrences = countOccurrences(fieldValue, token);
            if (occurrences > 0) {
                score += occurrences * field.weight;
                matched.add(field.name);
            }
        }
    }
    const isLongPhraseQuery = tokens.length >= LONG_QUERY_TOKEN_THRESHOLD && normalizedQuery.includes(" ");
    if (isLongPhraseQuery) {
        const normalizedTitle = normalizeSearchPhrase(item.title);
        if (normalizedTitle === normalizedQuery) {
            score += LONG_QUERY_TITLE_EXACT_BONUS;
            matched.add("title");
        }
        for (const field of searchableFields) {
            const normalizedField = normalizeSearchPhrase(field.value);
            const phraseOccurrences = countOccurrences(normalizedField, normalizedQuery);
            if (phraseOccurrences > 0) {
                score += phraseOccurrences * field.weight * LONG_QUERY_PHRASE_MULTIPLIER;
                matched.add(field.name);
            }
        }
    }
    if (score <= 0) {
        return null;
    }
    return {
        item,
        score,
        matched_fields: [...matched].sort((a, b) => a.localeCompare(b)),
    };
}
function sortHits(items, statusRegistry) {
    return [...items].sort((a, b) => {
        const byScore = b.score - a.score;
        if (byScore !== 0)
            return byScore;
        const aTerminal = isTerminal(a.item.status, statusRegistry);
        const bTerminal = isTerminal(b.item.status, statusRegistry);
        if (aTerminal !== bTerminal) {
            return aTerminal ? 1 : -1;
        }
        const byPriority = a.item.priority - b.item.priority;
        if (byPriority !== 0)
            return byPriority;
        const byUpdated = compareTimestampStrings(b.item.updated_at, a.item.updated_at);
        if (byUpdated !== 0)
            return byUpdated;
        return a.item.id.localeCompare(b.item.id);
    });
}
function buildHybridLexicalScore(document, tokens, normalizedQuery, includeLinked, linkedCorpusById, tuning) {
    return scoreDocument(document, tokens, normalizedQuery, includeLinked ? linkedCorpusById.get(document.metadata.id) ?? "" : "", tuning);
}
function normalizeScoreMap(scoreById) {
    if (scoreById.size === 0) {
        return new Map();
    }
    const values = [...scoreById.values()];
    const minScore = Math.min(...values);
    const maxScore = Math.max(...values);
    if (maxScore === minScore) {
        return new Map([...scoreById.keys()].map((id) => [id, 1]));
    }
    const normalized = new Map();
    for (const [id, score] of scoreById) {
        normalized.set(id, (score - minScore) / (maxScore - minScore));
    }
    return normalized;
}
export function resolveSearchMaxResults(settings) {
    const candidate = settings.search?.max_results;
    if (typeof candidate === "number" && Number.isFinite(candidate) && candidate > 0) {
        return Math.floor(candidate);
    }
    return 50;
}
export function resolveSearchScoreThreshold(settings) {
    const candidate = settings.search?.score_threshold;
    if (typeof candidate === "number" && Number.isFinite(candidate)) {
        return candidate;
    }
    return 0;
}
export function resolveHybridSemanticWeight(settings) {
    const candidate = settings.search?.hybrid_semantic_weight;
    if (typeof candidate === "number" && Number.isFinite(candidate) && candidate >= 0 && candidate <= 1) {
        return candidate;
    }
    return 0.7;
}
export function resolveSearchTuning(settings) {
    const defaults = {
        title_exact_bonus: 10,
        title_weight: 8,
        description_weight: 5,
        tags_weight: 6,
        status_weight: 2,
        body_weight: 1,
        comments_weight: 1,
        notes_weight: 1,
        learnings_weight: 1,
        reminders_weight: 2,
        events_weight: 2,
        dependencies_weight: 3,
        linked_content_weight: 1,
    };
    const tuning = settings.search?.tuning;
    if (!tuning)
        return defaults;
    const resolveWeight = (candidate, fallback) => {
        if (typeof candidate === "number" && Number.isFinite(candidate) && candidate >= 0) {
            return candidate;
        }
        return fallback;
    };
    return {
        title_exact_bonus: resolveWeight(tuning.title_exact_bonus, defaults.title_exact_bonus),
        title_weight: resolveWeight(tuning.title_weight, defaults.title_weight),
        description_weight: resolveWeight(tuning.description_weight, defaults.description_weight),
        tags_weight: resolveWeight(tuning.tags_weight, defaults.tags_weight),
        status_weight: resolveWeight(tuning.status_weight, defaults.status_weight),
        body_weight: resolveWeight(tuning.body_weight, defaults.body_weight),
        comments_weight: resolveWeight(tuning.comments_weight, defaults.comments_weight),
        notes_weight: resolveWeight(tuning.notes_weight, defaults.notes_weight),
        learnings_weight: resolveWeight(tuning.learnings_weight, defaults.learnings_weight),
        reminders_weight: resolveWeight(tuning.reminders_weight, defaults.reminders_weight),
        events_weight: resolveWeight(tuning.events_weight, defaults.events_weight),
        dependencies_weight: resolveWeight(tuning.dependencies_weight, defaults.dependencies_weight),
        linked_content_weight: resolveWeight(tuning.linked_content_weight, defaults.linked_content_weight),
    };
}
function emptySearchResult(query, mode, options, includeLinked, scoreThreshold, hybridSemanticWeight, projection, warnings) {
    const projectionFields = projection.mode === "full" ? null : [...projection.fields];
    return {
        query: query.trim(),
        mode,
        items: [],
        count: 0,
        filters: {
            mode,
            type: options.type ?? null,
            tag: options.tag ?? null,
            priority: options.priority ?? null,
            deadline_before: options.deadlineBefore ?? null,
            deadline_after: options.deadlineAfter ?? null,
            include_linked: includeLinked,
            title_exact: options.titleExact === true,
            phrase_exact: options.phraseExact === true,
            score_threshold: scoreThreshold,
            hybrid_semantic_weight: mode === "hybrid" ? hybridSemanticWeight : null,
            limit: options.limit ?? null,
            projection: projection.mode,
            fields: projectionFields,
        },
        projection: {
            mode: projection.mode,
            fields: projectionFields,
        },
        now: nowIso(),
        ...(warnings.length > 0 ? { warnings } : {}),
    };
}
function requireSemanticDependencies(requestedMode, providerResolution, vectorResolution, hasExtensionVectorQuery) {
    if (!providerResolution.active) {
        throw new PmCliError(`Search mode '${requestedMode}' requires a configured embedding provider in settings.providers.openai or settings.providers.ollama`, EXIT_CODE.USAGE);
    }
    if (!vectorResolution.active && !hasExtensionVectorQuery) {
        throw new PmCliError(`Search mode '${requestedMode}' requires a configured vector store in settings.vector_store.qdrant/settings.vector_store.lancedb or an extension adapter selected by settings.vector_store.adapter`, EXIT_CODE.USAGE);
    }
    return {
        provider: providerResolution.active,
        vectorStore: vectorResolution.active ?? null,
    };
}
const toOptionalNonEmptyString = toNonEmptyStringOrUndefined;
function resolveExtensionSearchProvider(settings) {
    const registrations = getActiveExtensionRegistrations();
    const providerName = toOptionalNonEmptyString(settings.search?.provider);
    const resolved = resolveRegisteredSearchProvider(registrations, providerName);
    if (!resolved) {
        return null;
    }
    const runtimeDefinition = resolved.runtime_definition ?? resolved.definition;
    const query = runtimeDefinition.query;
    if (typeof query !== "function") {
        return null;
    }
    const registeredName = toOptionalNonEmptyString(runtimeDefinition.name) ??
        toOptionalNonEmptyString(resolved.definition.name) ??
        providerName;
    if (!registeredName) {
        return null;
    }
    return {
        providerName: registeredName,
        query: query,
    };
}
function resolveExtensionVectorAdapter(settings) {
    const registrations = getActiveExtensionRegistrations();
    const adapterName = toOptionalNonEmptyString(settings.vector_store?.adapter);
    const resolved = resolveRegisteredVectorStoreAdapter(registrations, adapterName);
    if (!resolved) {
        return null;
    }
    const runtimeDefinition = resolved.runtime_definition ?? resolved.definition;
    const query = runtimeDefinition.query;
    if (typeof query !== "function") {
        return null;
    }
    return {
        query: query,
    };
}
function normalizeExtensionProviderHits(providerName, raw, filteredById) {
    const rawHits = Array.isArray(raw)
        ? raw
        : raw?.hits;
    if (!Array.isArray(rawHits)) {
        throw new PmCliError(`Extension search provider "${providerName}" must return an array of hits or { hits: [...] }`, EXIT_CODE.GENERIC_FAILURE);
    }
    const seen = new Set();
    const hits = [];
    for (const rawHit of rawHits) {
        if (typeof rawHit !== "object" || rawHit === null) {
            continue;
        }
        const id = toOptionalNonEmptyString(rawHit.id);
        const score = rawHit.score;
        if (!id || typeof score !== "number" || !Number.isFinite(score) || seen.has(id)) {
            continue;
        }
        const document = filteredById.get(id);
        if (!document) {
            continue;
        }
        const matchedFieldsRaw = rawHit.matched_fields;
        const matchedFields = Array.isArray(matchedFieldsRaw) && matchedFieldsRaw.every((entry) => typeof entry === "string")
            ? [...new Set(matchedFieldsRaw.map((entry) => entry.trim()).filter((entry) => entry.length > 0))].sort((a, b) => a.localeCompare(b))
            : [`provider:${providerName}`];
        seen.add(id);
        hits.push({
            item: document.metadata,
            score,
            matched_fields: matchedFields,
        });
    }
    return hits;
}
function buildSemanticHits(vectorHits, filteredById) {
    const semanticHits = [];
    const semanticScores = new Map();
    for (const vectorHit of vectorHits) {
        if (semanticScores.has(vectorHit.id)) {
            continue;
        }
        const document = filteredById.get(vectorHit.id);
        if (!document) {
            continue;
        }
        semanticScores.set(vectorHit.id, vectorHit.score);
        semanticHits.push({
            item: document.metadata,
            score: vectorHit.score,
            matched_fields: ["semantic"],
        });
    }
    return {
        semanticHits,
        semanticScores,
    };
}
function combineHybridHits(filteredById, semanticScores, keywordHits, hybridSemanticWeight) {
    const keywordScores = new Map(keywordHits.map((entry) => [entry.item.id, entry.score]));
    const keywordMatches = new Map(keywordHits.map((entry) => [entry.item.id, entry.matched_fields]));
    const normalizedSemantic = normalizeScoreMap(semanticScores);
    const normalizedKeyword = normalizeScoreMap(keywordScores);
    const candidateIds = new Set([...semanticScores.keys(), ...keywordScores.keys()]);
    const keywordWeight = 1 - hybridSemanticWeight;
    return [...candidateIds]
        .map((id) => {
        const document = filteredById.get(id);
        const semanticScore = normalizedSemantic.get(id) ?? 0;
        const keywordScore = normalizedKeyword.get(id) ?? 0;
        const combinedScore = (semanticScore * hybridSemanticWeight) + (keywordScore * keywordWeight);
        if (combinedScore <= 0) {
            return null;
        }
        const matchedFields = new Set();
        if (semanticScores.has(id)) {
            matchedFields.add("semantic");
        }
        for (const field of keywordMatches.get(id) ?? []) {
            matchedFields.add(field);
        }
        return {
            item: document.metadata,
            score: combinedScore,
            matched_fields: [...matchedFields].sort((a, b) => a.localeCompare(b)),
        };
    })
        .filter((entry) => entry !== null);
}
async function computeSemanticOrHybridHits(context) {
    const semanticLimit = context.limit ?? context.maxResults;
    const embeddingOptions = context.embeddingTimeoutMs !== undefined ? { timeout_ms: context.embeddingTimeoutMs } : {};
    const vectorQueryOptions = context.vectorQueryTimeoutMs !== undefined ? { timeout_ms: context.vectorQueryTimeoutMs } : {};
    const queryVectors = await executeEmbeddingRequest(context.provider, context.query.trim(), embeddingOptions);
    const semanticVector = queryVectors[0];
    let vectorHits;
    if (context.extensionVectorAdapter?.query) {
        try {
            vectorHits = await Promise.resolve(context.extensionVectorAdapter.query({
                vector: semanticVector,
                limit: semanticLimit,
                settings: context.settings,
            }));
        }
        catch (error) {
            if (!context.vectorStore) {
                throw new PmCliError(`Extension vector adapter query failed and no built-in fallback store is configured (${error instanceof Error ? error.message : String(error)})`, EXIT_CODE.GENERIC_FAILURE);
            }
            vectorHits = await executeVectorQuery(context.vectorStore, semanticVector, semanticLimit, vectorQueryOptions);
        }
    }
    else if (context.vectorStore) {
        vectorHits = await executeVectorQuery(context.vectorStore, semanticVector, semanticLimit, vectorQueryOptions);
    }
    else {
        throw new PmCliError("Semantic search requires either a configured vector store or an extension vector adapter query handler", EXIT_CODE.USAGE);
    }
    const filteredById = new Map(context.filteredDocuments.map((document) => [document.metadata.id, document]));
    const { semanticHits, semanticScores } = buildSemanticHits(vectorHits, filteredById);
    if (context.requestedMode === "semantic") {
        return semanticHits;
    }
    return combineHybridHits(filteredById, semanticScores, context.keywordHits, context.hybridSemanticWeight);
}
function alternateFormat(itemFormat) {
    return itemFormat === "toon" ? "json_markdown" : "toon";
}
async function loadDocuments(pmRoot, itemFormat, typeToFolder, schema) {
    const listWarnings = [];
    const items = await listAllFrontMatter(pmRoot, itemFormat, typeToFolder, listWarnings, schema);
    const warnings = [...new Set(listWarnings)].sort((left, right) => left.localeCompare(right));
    const documents = [];
    for (const item of items) {
        const preferredPath = getItemPath(pmRoot, item.type, item.id, itemFormat, typeToFolder);
        try {
            const raw = await fs.readFile(preferredPath, "utf8");
            await runActiveOnReadHooks({
                path: preferredPath,
                scope: "project",
            });
            documents.push(parseItemDocument(raw, { format: itemFormat, schema, onWarning: (warning) => listWarnings.push(warning) }));
            continue;
        }
        catch {
            // Fallback to the alternate format when preferred format path is absent.
        }
        const fallbackFormat = alternateFormat(itemFormat);
        const fallbackPath = getItemPath(pmRoot, item.type, item.id, fallbackFormat, typeToFolder);
        const raw = await fs.readFile(fallbackPath, "utf8");
        await runActiveOnReadHooks({
            path: fallbackPath,
            scope: "project",
        });
        documents.push(parseItemDocument(raw, { format: fallbackFormat, schema, onWarning: (warning) => listWarnings.push(warning) }));
    }
    return {
        documents,
        warnings,
    };
}
function readSearchFieldValue(hit, field) {
    const normalized = field.trim();
    if (normalized.length === 0) {
        return null;
    }
    if (normalized === "score") {
        return hit.score;
    }
    if (normalized === "matched_fields") {
        return hit.matched_fields;
    }
    if (normalized.startsWith("item.")) {
        const itemKey = normalized.slice("item.".length);
        if (itemKey.length === 0) {
            return null;
        }
        const itemRecord = hit.item;
        return itemRecord[itemKey] ?? null;
    }
    const hitRecord = hit;
    const itemRecord = hit.item;
    if (Object.prototype.hasOwnProperty.call(itemRecord, normalized)) {
        return itemRecord[normalized] ?? null;
    }
    if (Object.prototype.hasOwnProperty.call(hitRecord, normalized)) {
        return hitRecord[normalized] ?? null;
    }
    return null;
}
function projectSearchHits(hits, projection) {
    if (projection.mode === "full") {
        return hits;
    }
    return hits.map((hit) => {
        const projected = {};
        for (const field of projection.fields) {
            projected[field] = readSearchFieldValue(hit, field);
        }
        return projected;
    });
}
export async function runSearch(query, options, global) {
    const includeLinked = parseIncludeLinked(options.includeLinked);
    const titleExact = parseTitleExact(options.titleExact);
    const phraseExact = parsePhraseExact(options.phraseExact);
    const tokens = parseTokens(query);
    const normalizedQuery = normalizeSearchPhrase(query);
    const limit = parseLimit(options.limit);
    const projection = parseProjectionConfig(options);
    const modeWasExplicit = typeof options.mode === "string" && options.mode.trim().length > 0;
    const pmRoot = resolvePmRoot(process.cwd(), global.path);
    if (!(await pathExists(getSettingsPath(pmRoot)))) {
        throw new PmCliError(`Tracker is not initialized at ${pmRoot}. Run pm init first.`, EXIT_CODE.NOT_FOUND);
    }
    const storedSettings = await readSettings(pmRoot);
    const runtimeDefaultsResolution = resolveSettingsWithSemanticRuntimeDefaults(storedSettings);
    const settings = runtimeDefaultsResolution.settings;
    const statusRegistry = resolveRuntimeStatusRegistry(settings.schema);
    const runtimeFieldRegistry = resolveRuntimeFieldRegistry(settings.schema);
    const runtimeFieldFilters = collectRuntimeFilterValues(options, runtimeFieldRegistry, "search");
    const typeRegistry = resolveItemTypeRegistry(settings, getActiveExtensionRegistrations());
    const maxResults = resolveSearchMaxResults(settings);
    const scoreThreshold = resolveSearchScoreThreshold(settings);
    const hybridSemanticWeight = resolveHybridSemanticWeight(settings);
    const tuning = resolveSearchTuning(settings);
    const providerResolution = resolveEmbeddingProviders(settings);
    const vectorResolution = resolveVectorStores(settings);
    const extensionSearchProvider = resolveExtensionSearchProvider(settings);
    const extensionVectorAdapter = resolveExtensionVectorAdapter(settings);
    let effectiveMode = parseMode(options.mode, {
        hasProvider: providerResolution.active !== null || extensionSearchProvider !== null,
        hasVectorStore: vectorResolution.active !== null || extensionVectorAdapter !== null,
    });
    const loadedDocuments = await loadDocuments(pmRoot, settings.item_format ?? "toon", typeRegistry.type_to_folder, settings.schema);
    const warnings = loadedDocuments.warnings;
    const allDocuments = loadedDocuments.documents;
    const metadataFilteredDocuments = applyFilters(allDocuments, options, typeRegistry, runtimeFieldFilters);
    const filteredDocuments = applyExactQueryFilters(metadataFilteredDocuments, normalizedQuery, {
        titleExact,
        phraseExact,
    });
    if (effectiveMode === "keyword" && (filteredDocuments.length === 0 || limit === 0)) {
        return emptySearchResult(query, effectiveMode, options, includeLinked, scoreThreshold, hybridSemanticWeight, projection, warnings);
    }
    const projectRoot = process.cwd();
    const globalRoot = resolveGlobalPmRoot(projectRoot);
    const linkedCorpusById = new Map();
    if (includeLinked && (effectiveMode === "keyword" || effectiveMode === "hybrid")) {
        for (const document of filteredDocuments) {
            linkedCorpusById.set(document.metadata.id, await loadLinkedCorpus(document, projectRoot, globalRoot));
        }
    }
    const keywordHits = filteredDocuments
        .map((document) => buildHybridLexicalScore(document, tokens, normalizedQuery, effectiveMode !== "semantic", linkedCorpusById, tuning))
        .filter((entry) => entry !== null);
    let hits = keywordHits;
    if (effectiveMode !== "keyword") {
        try {
            if (!extensionSearchProvider) {
                requireSemanticDependencies(effectiveMode, providerResolution, vectorResolution, extensionVectorAdapter !== null);
            }
            if (filteredDocuments.length === 0 || limit === 0) {
                return emptySearchResult(query, effectiveMode, options, includeLinked, scoreThreshold, hybridSemanticWeight, projection, warnings);
            }
            const filteredById = new Map(filteredDocuments.map((document) => [document.metadata.id, document]));
            const canUseBuiltInSemantic = providerResolution.active !== null && (vectorResolution.active !== null || extensionVectorAdapter !== null);
            if (extensionSearchProvider) {
                try {
                    const providerResponse = await Promise.resolve(extensionSearchProvider.query({
                        query,
                        mode: effectiveMode,
                        tokens,
                        options,
                        settings,
                        documents: filteredDocuments,
                    }));
                    hits = normalizeExtensionProviderHits(extensionSearchProvider.providerName, providerResponse, filteredById);
                }
                catch (error) {
                    if (!canUseBuiltInSemantic) {
                        throw new PmCliError(`Extension search provider "${extensionSearchProvider.providerName}" failed: ${error instanceof Error ? error.message : String(error)}`, EXIT_CODE.GENERIC_FAILURE);
                    }
                }
            }
            if (hits === keywordHits) {
                const implicitHybridMode = !modeWasExplicit && effectiveMode === "hybrid";
                const { provider, vectorStore } = requireSemanticDependencies(effectiveMode, providerResolution, vectorResolution, extensionVectorAdapter !== null);
                hits = await computeSemanticOrHybridHits({
                    requestedMode: effectiveMode,
                    query,
                    filteredDocuments,
                    keywordHits,
                    hybridSemanticWeight,
                    limit,
                    maxResults,
                    provider,
                    vectorStore,
                    extensionVectorAdapter,
                    settings,
                    ...(implicitHybridMode
                        ? {
                            embeddingTimeoutMs: IMPLICIT_HYBRID_EMBEDDING_TIMEOUT_MS,
                            vectorQueryTimeoutMs: IMPLICIT_HYBRID_VECTOR_TIMEOUT_MS,
                        }
                        : {}),
                });
            }
        }
        catch (error) {
            const canFallbackToKeyword = !modeWasExplicit && effectiveMode === "hybrid";
            if (!canFallbackToKeyword) {
                throw error;
            }
            effectiveMode = "keyword";
            hits = keywordHits;
            warnings.push(buildImplicitSemanticFallbackWarning(error));
        }
    }
    const thresholded = hits.filter((entry) => entry.score >= scoreThreshold);
    const sorted = sortHits(thresholded, statusRegistry);
    const resolvedLimit = effectiveMode === "keyword" ? limit : (limit ?? maxResults);
    const limited = resolvedLimit === undefined ? sorted : sorted.slice(0, resolvedLimit);
    const projectedItems = projectSearchHits(limited, projection);
    const projectionFields = projection.mode === "full" ? null : [...projection.fields];
    return {
        query: query.trim(),
        mode: effectiveMode,
        items: projectedItems,
        count: projectedItems.length,
        filters: {
            mode: effectiveMode,
            type: options.type ?? null,
            tag: options.tag ?? null,
            priority: options.priority ?? null,
            deadline_before: options.deadlineBefore ?? null,
            deadline_after: options.deadlineAfter ?? null,
            include_linked: includeLinked,
            title_exact: titleExact,
            phrase_exact: phraseExact,
            score_threshold: scoreThreshold,
            hybrid_semantic_weight: effectiveMode === "hybrid" ? hybridSemanticWeight : null,
            limit: options.limit ?? null,
            runtime_filters: runtimeFieldFilters,
            projection: projection.mode,
            fields: projectionFields,
        },
        projection: {
            mode: projection.mode,
            fields: projectionFields,
        },
        now: nowIso(),
        ...(warnings.length > 0 ? { warnings } : {}),
    };
}
