// @unbrained/pm-cli/dist/cli/commands/stats.js
import fs from "node:fs/promises";
import path from "node:path";
import { getActiveExtensionRegistrations, runActiveOnReadHooks } from "../../core/extensions/index.js";
import { pathExists } from "../../core/fs/fs-utils.js";
import { enforceHistoryStreamPolicyForItems } from "../../core/history/history-stream-policy.js";
import { resolveItemTypeRegistry } from "../../core/item/type-registry.js";
import { resolveRuntimeStatusRegistry } from "../../core/schema/runtime-schema.js";
import { EXIT_CODE } from "../../core/shared/constants.js";
import { PmCliError } from "../../core/shared/errors.js";
import { nowIso } from "../../core/shared/time.js";
import { listAllFrontMatter } from "../../core/store/item-store.js";
import { getSettingsPath, resolvePmRoot } from "../../core/store/paths.js";
import { readSettings } from "../../core/store/settings.js";
function zeroByType(itemTypes) {
    return itemTypes.reduce((acc, value) => {
        acc[value] = 0;
        return acc;
    }, {});
}
function zeroByStatus(statuses) {
    return statuses.reduce((acc, value) => {
        acc[value] = 0;
        return acc;
    }, {});
}
function countNonEmptyLines(raw) {
    if (raw.trim().length === 0) {
        return 0;
    }
    return raw
        .split(/\r?\n/u)
        .filter((line) => line.trim().length > 0)
        .length;
}
async function readHistorySummary(pmRoot) {
    const historyDir = path.join(pmRoot, "history");
    if (!(await pathExists(historyDir))) {
        return {
            history_streams: 0,
            history_entries: 0,
        };
    }
    await runActiveOnReadHooks({
        path: historyDir,
        scope: "project",
    });
    const historyFiles = (await fs.readdir(historyDir))
        .filter((entry) => entry.endsWith(".jsonl"))
        .sort((a, b) => a.localeCompare(b));
    let historyEntries = 0;
    for (const file of historyFiles) {
        const historyPath = path.join(historyDir, file);
        const raw = await fs.readFile(historyPath, "utf8");
        await runActiveOnReadHooks({
            path: historyPath,
            scope: "project",
        });
        historyEntries += countNonEmptyLines(raw);
    }
    return {
        history_streams: historyFiles.length,
        history_entries: historyEntries,
    };
}
export async function runStats(global) {
    const pmRoot = resolvePmRoot(process.cwd(), global.path);
    if (!(await pathExists(getSettingsPath(pmRoot)))) {
        throw new PmCliError(`Tracker is not initialized at ${pmRoot}. Run pm init first.`, EXIT_CODE.NOT_FOUND);
    }
    const settings = await readSettings(pmRoot);
    const typeRegistry = resolveItemTypeRegistry(settings, getActiveExtensionRegistrations());
    const statusRegistry = resolveRuntimeStatusRegistry(settings.schema);
    const items = await listAllFrontMatter(pmRoot, settings.item_format, typeRegistry.type_to_folder, undefined, settings.schema);
    await enforceHistoryStreamPolicyForItems({
        pmRoot,
        settings,
        itemIds: items.map((item) => item.id),
        commandLabel: "stats",
    });
    const byType = zeroByType(typeRegistry.types);
    const byStatus = zeroByStatus(statusRegistry.definitions.map((definition) => definition.id));
    for (const item of items) {
        if (byType[item.type] === undefined) {
            byType[item.type] = 0;
        }
        byType[item.type] += 1;
        if (byStatus[item.status] === undefined) {
            byStatus[item.status] = 0;
        }
        byStatus[item.status] += 1;
    }
    const historySummary = await readHistorySummary(pmRoot);
    return {
        totals: {
            items: items.length,
            history_streams: historySummary.history_streams,
            history_entries: historySummary.history_entries,
        },
        by_type: byType,
        by_status: byStatus,
        generated_at: nowIso(),
    };
}
