// @unbrained/pm-cli/dist/cli/commands/test-all.js
import { pathExists } from "../../core/fs/fs-utils.js";
import { getActiveExtensionRegistrations } from "../../core/extensions/index.js";
import { resolveItemTypeRegistry } from "../../core/item/type-registry.js";
import { parseOptionalNumber } from "../../core/item/parse.js";
import { normalizeStatusInput } from "../../core/item/status.js";
import { resolveRuntimeStatusRegistry } from "../../core/schema/runtime-schema.js";
import { EXIT_CODE } from "../../core/shared/constants.js";
import { PmCliError } from "../../core/shared/errors.js";
import { nowIso } from "../../core/shared/time.js";
import { listAllFrontMatter } from "../../core/store/item-store.js";
import { getSettingsPath, resolveGlobalPmRoot, resolvePmRoot } from "../../core/store/paths.js";
import { readSettings } from "../../core/store/settings.js";
import { appendTrackedTestRunSummary } from "../../core/test/item-test-run-tracking.js";
import { countFailureCategories, runLinkedTests, runTest, summarizeContextPreflight, } from "./test.js";
function parseStatus(raw, statusRegistry) {
    if (raw === undefined) {
        return undefined;
    }
    const normalized = normalizeStatusInput(raw, statusRegistry);
    if (!normalized) {
        const allowedStatuses = statusRegistry.definitions.map((definition) => definition.id);
        throw new PmCliError(`Invalid --status value "${raw}". Allowed: ${allowedStatuses.join(", ")}`, EXIT_CODE.USAGE);
    }
    return normalized;
}
function parseTimeout(raw) {
    if (raw === undefined) {
        return undefined;
    }
    return parseOptionalNumber(raw, "timeout");
}
function parseNonNegativeInteger(raw, flag) {
    if (raw === undefined) {
        return undefined;
    }
    const parsed = Number(raw);
    if (!Number.isInteger(parsed) || parsed < 0) {
        throw new PmCliError(`${flag} must be a non-negative integer`, EXIT_CODE.USAGE);
    }
    return parsed;
}
function resolveTrackingAuthor(fallback) {
    const candidate = process.env.PM_AUTHOR ?? fallback;
    const trimmed = candidate.trim();
    return trimmed.length > 0 ? trimmed : "unknown";
}
function resolveTrackedRunId() {
    const fromEnv = process.env.PM_BACKGROUND_TEST_RUN_ID?.trim();
    if (fromEnv && fromEnv.length > 0) {
        return fromEnv;
    }
    return `test-all-local-${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 8)}`;
}
function normalizeCommand(command) {
    return command.trim().replaceAll(/\s+/g, " ");
}
function normalizeEnvSetSignature(value) {
    if (!value || Object.keys(value).length === 0) {
        return "{}";
    }
    return JSON.stringify(Object.fromEntries(Object.entries(value).sort(([left], [right]) => left.localeCompare(right))));
}
function normalizeEnvClearSignature(value) {
    if (!value || value.length === 0) {
        return "[]";
    }
    return JSON.stringify([...value].sort((left, right) => left.localeCompare(right)));
}
function normalizePmContextModeSignature(value) {
    const normalized = value?.trim().toLowerCase();
    return normalized && normalized.length > 0 ? normalized : "none";
}
function normalizeAssertionSignature(test) {
    const normalized = {
        assert_stdout_contains: [...new Set(test.assert_stdout_contains ?? [])].sort((left, right) => left.localeCompare(right)),
        assert_stdout_regex: [...new Set(test.assert_stdout_regex ?? [])].sort((left, right) => left.localeCompare(right)),
        assert_stderr_contains: [...new Set(test.assert_stderr_contains ?? [])].sort((left, right) => left.localeCompare(right)),
        assert_stderr_regex: [...new Set(test.assert_stderr_regex ?? [])].sort((left, right) => left.localeCompare(right)),
        assert_stdout_min_lines: typeof test.assert_stdout_min_lines === "number" ? test.assert_stdout_min_lines : undefined,
        assert_json_field_equals: Object.fromEntries(Object.entries(test.assert_json_field_equals ?? {}).sort(([left], [right]) => left.localeCompare(right))),
        assert_json_field_gte: Object.fromEntries(Object.entries(test.assert_json_field_gte ?? {}).sort(([left], [right]) => left.localeCompare(right))),
    };
    return JSON.stringify(normalized);
}
function buildLinkedTestKey(test) {
    const command = test.command?.trim();
    if (command && command.length > 0) {
        const envSet = normalizeEnvSetSignature(test.env_set);
        const envClear = normalizeEnvClearSignature(test.env_clear);
        const pmContextMode = normalizePmContextModeSignature(test.pm_context_mode);
        const sharedHostSafe = test.shared_host_safe === true ? "true" : "false";
        const assertions = normalizeAssertionSignature(test);
        return `command:${test.scope}:${normalizeCommand(command)}:${envSet}:${envClear}:${pmContextMode}:${sharedHostSafe}:${assertions}`;
    }
    const linkedPath = test.path?.trim() ?? "";
    return `path:${test.scope}:${linkedPath}`;
}
function maxTimeoutSeconds(current, candidate) {
    if (candidate === undefined) {
        /* c8 ignore next - exercised implicitly by duplicate timeout normalization */
        return current;
    }
    if (current === undefined || candidate > current) {
        return candidate;
    }
    return current;
}
function countStatuses(runResults) {
    let passed = 0;
    let failed = 0;
    let skipped = 0;
    for (const result of runResults) {
        if (result.status === "passed") {
            passed += 1;
            continue;
        }
        if (result.status === "failed") {
            failed += 1;
            continue;
        }
        skipped += 1;
    }
    return { passed, failed, skipped };
}
function mergeFailureCategoryCounts(target, source) {
    for (const [key, value] of Object.entries(source)) {
        target[key] += value;
    }
}
export async function runTestAll(options, global) {
    const pmRoot = resolvePmRoot(process.cwd(), global.path);
    if (!(await pathExists(getSettingsPath(pmRoot)))) {
        throw new PmCliError(`Tracker is not initialized at ${pmRoot}. Run pm init first.`, EXIT_CODE.NOT_FOUND);
    }
    const settings = await readSettings(pmRoot);
    const statusRegistry = resolveRuntimeStatusRegistry(settings.schema);
    const typeRegistry = resolveItemTypeRegistry(settings, getActiveExtensionRegistrations());
    const statusFilter = parseStatus(options.status, statusRegistry);
    const limitFilter = parseNonNegativeInteger(options.limit, "--limit");
    const offsetFilter = parseNonNegativeInteger(options.offset, "--offset") ?? 0;
    const allItems = await listAllFrontMatter(pmRoot, settings.item_format, typeRegistry.type_to_folder, undefined, settings.schema);
    const statusFilteredItems = allItems
        .filter((item) => (statusFilter ? item.status === statusFilter : true))
        .sort((a, b) => a.id.localeCompare(b.id));
    const filteredItems = limitFilter === undefined
        ? statusFilteredItems.slice(offsetFilter)
        : statusFilteredItems.slice(offsetFilter, offsetFilter + limitFilter);
    const defaultTimeoutSeconds = parseTimeout(options.timeout);
    const sourceRoots = {
        projectPmRoot: pmRoot,
        globalPmRoot: resolveGlobalPmRoot(process.cwd()),
    };
    const runStartedAt = nowIso();
    const trackingEnabled = settings.testing.record_results_to_items === true;
    const trackingRunId = resolveTrackedRunId();
    const trackingAuthor = resolveTrackingAuthor(settings.author_default);
    const trackingAttemptRaw = process.env.PM_BACKGROUND_TEST_RUN_ATTEMPT?.trim();
    const trackingParsedAttempt = trackingAttemptRaw ? Number.parseInt(trackingAttemptRaw, 10) : Number.NaN;
    const trackingResumedFrom = process.env.PM_BACKGROUND_TEST_RUN_RESUMED_FROM?.trim();
    const trackingWarnings = [];
    const results = [];
    const seenTestKeys = new Set();
    let passed = 0;
    let failed = 0;
    let skipped = 0;
    let linkedTests = 0;
    const failureCategories = countFailureCategories([]);
    const itemTests = [];
    for (const item of filteredItems) {
        const readResult = await runTest(item.id, {
            run: false,
        }, {
            ...global,
            path: pmRoot,
        });
        linkedTests += readResult.tests.length;
        itemTests.push({ item, tests: readResult.tests });
    }
    const effectiveTimeoutByKey = new Map();
    for (const { tests } of itemTests) {
        for (const test of tests) {
            const key = buildLinkedTestKey(test);
            if (!effectiveTimeoutByKey.has(key)) {
                effectiveTimeoutByKey.set(key, test.timeout_seconds);
                continue;
            }
            effectiveTimeoutByKey.set(key, maxTimeoutSeconds(effectiveTimeoutByKey.get(key), test.timeout_seconds));
        }
    }
    for (const { item, tests } of itemTests) {
        const testsToRun = [];
        const keyedTests = tests.map((test) => {
            const key = buildLinkedTestKey(test);
            const duplicate = seenTestKeys.has(key);
            if (!duplicate) {
                seenTestKeys.add(key);
                const effectiveTimeoutSeconds = effectiveTimeoutByKey.get(key);
                testsToRun.push(effectiveTimeoutSeconds === undefined ? test : { ...test, timeout_seconds: effectiveTimeoutSeconds });
            }
            return { test, key, duplicate };
        });
        const executedResults = testsToRun.length > 0
            ? await runLinkedTests(testsToRun, defaultTimeoutSeconds, {
                progress: options.progress,
                sourceRoots,
                envSet: options.envSet,
                envClear: options.envClear,
                sharedHostSafe: options.sharedHostSafe,
                pmContext: options.pmContext,
                overrideLinkedPmContext: options.overrideLinkedPmContext,
                failOnContextMismatch: options.failOnContextMismatch,
                failOnEmptyTestRun: options.failOnEmptyTestRun,
                requireAssertionsForPm: options.requireAssertionsForPm,
                checkContext: options.checkContext,
                autoPmContext: options.autoPmContext,
            })
            : [];
        let executedIndex = 0;
        const runResults = keyedTests.map(({ test, key, duplicate }) => {
            if (!duplicate) {
                const executed = executedResults[executedIndex];
                executedIndex += 1;
                return executed;
            }
            return {
                command: test.command,
                path: test.path,
                status: "skipped",
                error: `Duplicate linked test skipped (key=${key}).`,
            };
        });
        const summary = countStatuses(runResults);
        const itemFailureCategories = countFailureCategories(runResults);
        mergeFailureCategoryCounts(failureCategories, itemFailureCategories);
        passed += summary.passed;
        failed += summary.failed;
        skipped += summary.skipped;
        if (trackingEnabled) {
            try {
                await appendTrackedTestRunSummary({
                    pmRoot,
                    settings,
                    itemId: item.id,
                    author: trackingAuthor,
                    message: `Track test-all run summary (${trackingRunId})`,
                    entry: {
                        run_id: trackingRunId,
                        kind: "test-all",
                        status: summary.failed > 0 || (options.failOnSkipped === true && summary.skipped > 0) ? "failed" : "passed",
                        started_at: runStartedAt,
                        finished_at: nowIso(),
                        recorded_at: nowIso(),
                        attempt: Number.isFinite(trackingParsedAttempt) && trackingParsedAttempt >= 1 ? trackingParsedAttempt : undefined,
                        resumed_from: trackingResumedFrom && trackingResumedFrom.length > 0 ? trackingResumedFrom : undefined,
                        passed: summary.passed,
                        failed: summary.failed,
                        skipped: summary.skipped,
                        items: 1,
                        linked_tests: tests.length,
                        fail_on_skipped_triggered: options.failOnSkipped === true && summary.skipped > 0 ? true : undefined,
                    },
                });
            }
            catch (error) {
                trackingWarnings.push(`test_result_tracking_failed:${item.id}:${error instanceof Error ? error.message : String(error)}`);
            }
        }
        results.push({
            ok: summary.failed === 0 && !(options.failOnSkipped === true && summary.skipped > 0),
            id: item.id,
            status: item.status,
            test_count: tests.length,
            passed: summary.passed,
            failed: summary.failed,
            skipped: summary.skipped,
            run_results: runResults,
            failure_categories: itemFailureCategories,
        });
    }
    const failOnSkippedTriggered = options.failOnSkipped === true && skipped > 0;
    if (options.checkContext === true) {
        const allRunResults = results.flatMap((entry) => entry.run_results);
        const preflight = summarizeContextPreflight(allRunResults);
        trackingWarnings.push(`context_preflight:checked_pm_commands=${preflight.checked_pm_commands};` +
            `tracker_read_commands=${preflight.tracker_read_commands};` +
            `mismatches=${preflight.mismatches};` +
            `auto_remediated=${preflight.auto_remediated}`);
    }
    return {
        ok: failed === 0 && failOnSkippedTriggered !== true,
        totals: {
            items: filteredItems.length,
            linked_tests: linkedTests,
            passed,
            failed,
            skipped,
            failure_categories: failureCategories,
        },
        failed,
        passed,
        skipped,
        fail_on_skipped_triggered: failOnSkippedTriggered ? true : undefined,
        warnings: trackingWarnings.length > 0 ? trackingWarnings : undefined,
        results,
    };
}
