// @unbrained/pm-cli/dist/cli/commands/test-runs.js
import os from "node:os";
import { pathExists } from "../../core/fs/fs-utils.js";
import { EXIT_CODE } from "../../core/shared/constants.js";
import { PmCliError } from "../../core/shared/errors.js";
import { getBackgroundTestRunStatus, listBackgroundTestRuns, readBackgroundTestRunLogs, resumeBackgroundTestRun, runBackgroundTestRunWorker, spawnBackgroundTestRunWorker, startBackgroundTestRun, stopBackgroundTestRun, } from "../../core/test/background-runs.js";
import { getSettingsPath, resolveGlobalPmRoot, resolvePmRoot } from "../../core/store/paths.js";
import { readSettings } from "../../core/store/settings.js";
import { parseLimit } from "../shared-parsers.js";
const BACKGROUND_STATUS_VALUES = ["queued", "running", "passed", "failed", "stopped", "canceled"];
const BACKGROUND_STREAM_VALUES = ["stdout", "stderr", "both"];
function normalizeStatus(value) {
    if (value === undefined) {
        return undefined;
    }
    const normalized = value.trim().toLowerCase();
    if (BACKGROUND_STATUS_VALUES.includes(normalized)) {
        return normalized;
    }
    throw new PmCliError(`Invalid --status value "${value}"`, EXIT_CODE.USAGE);
}
function normalizeStream(value) {
    if (value === undefined || value.trim().length === 0) {
        return "stderr";
    }
    const normalized = value.trim().toLowerCase();
    if (BACKGROUND_STREAM_VALUES.includes(normalized)) {
        return normalized;
    }
    throw new PmCliError(`Invalid --stream value "${value}"`, EXIT_CODE.USAGE);
}
function resolveWhoamiFallback() {
    try {
        const username = os.userInfo().username.trim();
        if (username.length > 0) {
            return username;
        }
    }
    catch {
        // Fall back to environment-derived attribution.
    }
    return undefined;
}
function resolveRequestedBy(author, fallback) {
    const candidates = [
        author,
        process.env.PM_AUTHOR,
        fallback,
        process.env.USER,
        process.env.LOGNAME,
        process.env.USERNAME,
        resolveWhoamiFallback(),
    ];
    for (const candidate of candidates) {
        if (typeof candidate !== "string") {
            continue;
        }
        const trimmed = candidate.trim();
        if (trimmed.length > 0) {
            return trimmed;
        }
    }
    return "unknown";
}
async function ensureInitialized(pmRoot) {
    if (!(await pathExists(getSettingsPath(pmRoot)))) {
        throw new PmCliError(`Tracker is not initialized at ${pmRoot}. Run pm init first.`, EXIT_CODE.NOT_FOUND);
    }
}
export async function runStartBackgroundRun(options, global) {
    const pmRoot = resolvePmRoot(process.cwd(), global.path);
    const globalPmRoot = resolveGlobalPmRoot(process.cwd());
    await ensureInitialized(pmRoot);
    const settings = await readSettings(pmRoot);
    const requestedBy = resolveRequestedBy(options.author, settings.author_default);
    const started = await startBackgroundTestRun({
        pmRoot,
        globalPmRoot,
        kind: options.kind,
        commandArgs: options.commandArgs,
        requestedBy,
        targetId: options.targetId,
        statusFilter: options.statusFilter,
    });
    if (!started.started) {
        return {
            started: false,
            duplicate_of: started.duplicate_of,
            run: started.run,
        };
    }
    const spawned = await spawnBackgroundTestRunWorker({
        pmRoot,
        runId: started.run.id,
        noExtensions: options.noExtensions === true,
    });
    return {
        started: true,
        run: spawned,
    };
}
export async function runTestRunsList(options, global) {
    const pmRoot = resolvePmRoot(process.cwd(), global.path);
    await ensureInitialized(pmRoot);
    const status = normalizeStatus(options.status);
    const limit = parseLimit(options.limit, "limit");
    const runs = await listBackgroundTestRuns(pmRoot, { status, limit });
    return {
        runs,
        count: runs.length,
        filters: {
            status,
            limit,
        },
    };
}
export async function runTestRunsStatus(runId, global) {
    const pmRoot = resolvePmRoot(process.cwd(), global.path);
    await ensureInitialized(pmRoot);
    const status = await getBackgroundTestRunStatus(pmRoot, runId);
    return {
        run: status.run,
        health: status.health,
    };
}
export async function runTestRunsLogs(runId, options, global) {
    const pmRoot = resolvePmRoot(process.cwd(), global.path);
    await ensureInitialized(pmRoot);
    const stream = normalizeStream(options.stream);
    const tail = parseLimit(options.tail, "tail");
    const logs = await readBackgroundTestRunLogs(pmRoot, runId, stream, tail);
    return {
        run: logs.run,
        stream: logs.stream,
        tail: logs.tail,
        stdout: logs.stdout,
        stderr: logs.stderr,
    };
}
export async function runTestRunsStop(runId, options, global) {
    const pmRoot = resolvePmRoot(process.cwd(), global.path);
    await ensureInitialized(pmRoot);
    const stopped = await stopBackgroundTestRun(pmRoot, runId, options.force === true);
    return {
        run: stopped.run,
        signal_sent: stopped.signal_sent,
    };
}
export async function runTestRunsResume(runId, options, global) {
    const pmRoot = resolvePmRoot(process.cwd(), global.path);
    await ensureInitialized(pmRoot);
    const settings = await readSettings(pmRoot);
    const requestedBy = resolveRequestedBy(options.author, settings.author_default);
    const resumed = await resumeBackgroundTestRun(pmRoot, runId, requestedBy, options.noExtensions === true);
    return {
        resumed_from: runId,
        run: resumed,
    };
}
export async function runTestRunsWorker(runId, global) {
    const pmRoot = resolvePmRoot(process.cwd(), global.path);
    await ensureInitialized(pmRoot);
    const finished = await runBackgroundTestRunWorker(pmRoot, runId, global.noExtensions === true);
    return {
        id: finished.id,
        status: finished.status,
        exit_code: finished.exit_code,
    };
}
