// @unbrained/pm-cli/dist/core/store/front-matter-cache.js
import fs from "node:fs/promises";
import path from "node:path";
import { createHash } from "node:crypto";
import { runActiveOnReadHooks } from "../extensions/index.js";
import { parseItemDocument } from "../item/item-format.js";
import { ITEM_FILE_EXTENSIONS, getItemFormatFromPath } from "./paths.js";
const CACHE_VERSION = 3;
const CACHE_FILENAME = "metadata-cache.json";
function computeContextFingerprint(preferredFormat, typeToFolder, schema) {
    const hash = createHash("sha256");
    hash.update(`format:${preferredFormat ?? "default"}`);
    const sortedTypes = Object.entries(typeToFolder)
        .sort(([a], [b]) => a.localeCompare(b))
        .map(([type, folder]) => `${type}=${folder}`)
        .join(",");
    hash.update(`|types:${sortedTypes}`);
    if (schema) {
        hash.update(`|schema:${JSON.stringify(schema)}`);
    }
    return hash.digest("hex").slice(0, 16);
}
function getCachePath(pmRoot) {
    return path.join(pmRoot, "runtime", CACHE_FILENAME);
}
async function loadCache(pmRoot) {
    try {
        const raw = await fs.readFile(getCachePath(pmRoot), "utf8");
        const parsed = JSON.parse(raw);
        if (parsed.version !== CACHE_VERSION || typeof parsed.entries !== "object") {
            return null;
        }
        return parsed;
    }
    catch {
        return null;
    }
}
async function persistCache(pmRoot, envelope) {
    const cachePath = getCachePath(pmRoot);
    await fs.mkdir(path.dirname(cachePath), { recursive: true });
    await fs.writeFile(cachePath, JSON.stringify(envelope), "utf8");
}
function appendWarning(warnings, warning) {
    if (warnings && !warnings.includes(warning)) {
        warnings.push(warning);
    }
}
/**
 * List all item documents using a persistent on-disk metadata cache.
 * Only parses files whose mtime/size have changed since the last cached run.
 */
export async function listAllDocumentsCached(pmRoot, preferredFormat, typeToFolder, warnings, schema) {
    const contextFingerprint = computeContextFingerprint(preferredFormat, typeToFolder, schema);
    const existingCache = await loadCache(pmRoot);
    const previousEntries = existingCache && existingCache.context_fingerprint === contextFingerprint
        ? existingCache.entries
        : {};
    const entries = Object.entries(typeToFolder);
    const dirResults = await Promise.all(entries.map(async ([, folder]) => {
        const dirPath = path.join(pmRoot, folder);
        try {
            const files = await fs.readdir(dirPath);
            return { folder, dirPath, files };
        }
        catch (error) {
            if (typeof error === "object" && error !== null && "code" in error && error.code !== "ENOENT") {
                appendWarning(warnings, `item_list_directory_read_failed:${folder}`);
            }
            return { folder, dirPath, files: [] };
        }
    }));
    const newEntries = {};
    const documentsById = new Map();
    const parseTasks = [];
    for (const { folder, dirPath, files } of dirResults) {
        for (const file of files) {
            if (!ITEM_FILE_EXTENSIONS.some((ext) => file.toLowerCase().endsWith(ext))) {
                continue;
            }
            const filePath = path.join(dirPath, file);
            const relativePath = path.relative(pmRoot, filePath);
            parseTasks.push((async () => {
                try {
                    const stat = await fs.stat(filePath);
                    const mtimeMs = stat.mtimeMs;
                    const ctimeMs = stat.ctimeMs;
                    const { size } = stat;
                    const cached = previousEntries[relativePath];
                    let metadata;
                    let bodyLength;
                    const itemFormat = getItemFormatFromPath(filePath);
                    // Preserve onRead hook semantics even when metadata is served from cache.
                    await runActiveOnReadHooks({ path: filePath, scope: "project" });
                    if (cached && cached.mtime_ms === mtimeMs && cached.ctime_ms === ctimeMs && cached.size === size) {
                        metadata = cached.metadata;
                        bodyLength = cached.body_length;
                    }
                    else {
                        const raw = await fs.readFile(filePath, "utf8");
                        const parsed = parseItemDocument(raw, {
                            format: itemFormat,
                            schema,
                            onWarning: (w) => appendWarning(warnings, w),
                        });
                        metadata = parsed.metadata;
                        bodyLength = parsed.body.length;
                    }
                    newEntries[relativePath] = {
                        mtime_ms: mtimeMs,
                        ctime_ms: ctimeMs,
                        size,
                        metadata,
                        body_length: bodyLength,
                    };
                    const existing = documentsById.get(metadata.id);
                    if (!existing) {
                        documentsById.set(metadata.id, {
                            document: { metadata, body: "" },
                            itemFormat,
                        });
                    }
                    else {
                        const shouldReplace = preferredFormat
                            ? itemFormat === preferredFormat && existing.itemFormat !== preferredFormat
                            : itemFormat === "toon" && existing.itemFormat !== "toon";
                        if (shouldReplace) {
                            documentsById.set(metadata.id, {
                                document: { metadata, body: "" },
                                itemFormat,
                            });
                        }
                    }
                }
                catch {
                    appendWarning(warnings, `item_list_item_read_failed:${folder}/${file}`);
                }
            })());
        }
    }
    await Promise.all(parseTasks);
    persistCache(pmRoot, {
        version: CACHE_VERSION,
        context_fingerprint: contextFingerprint,
        entries: newEntries,
    }).catch(() => { });
    return [...documentsById.values()]
        .sort((left, right) => left.document.metadata.id.localeCompare(right.document.metadata.id))
        .map((entry) => entry.document);
}
