// @unbrained/pm-cli/dist/core/store/item-format-migration.js
import fs from "node:fs/promises";
import path from "node:path";
import { runActiveOnWriteHooks } from "../extensions/index.js";
import { pathExists, readFileIfExists, removeFileIfExists, writeFileAtomic } from "../fs/fs-utils.js";
import { parseItemDocument, serializeItemDocument } from "../item/item-format.js";
import { TYPE_TO_FOLDER } from "../shared/constants.js";
import { getItemFormatFromPath, getItemPath } from "./paths.js";
function alternateItemFormat(_targetFormat) {
    return "json_markdown";
}
function normalizeRelativePath(pmRoot, absolutePath) {
    return path.relative(pmRoot, absolutePath).replaceAll("\\", "/");
}
function errorSummary(error) {
    return String(error).replaceAll(/[^a-zA-Z0-9._-]+/g, "_").slice(0, 120);
}
export async function migrateItemFilesToFormat(pmRoot, targetFormat, op = "item_format:migrate", typeToFolder = TYPE_TO_FOLDER, schema) {
    if (targetFormat !== "toon") {
        throw new Error("Only toon item-format migration targets are supported. Markdown item files are legacy read-only input.");
    }
    const migratedIds = new Set();
    const removedPaths = new Set();
    const warnings = [];
    let scanned = 0;
    const alternateFormat = alternateItemFormat(targetFormat);
    const typeEntries = Object.entries(typeToFolder);
    for (const [itemType, folder] of typeEntries) {
        const directoryPath = path.join(pmRoot, folder);
        let files = [];
        try {
            files = await fs.readdir(directoryPath);
        }
        catch {
            continue;
        }
        const variantsById = new Map();
        for (const file of files) {
            const absolutePath = path.join(directoryPath, file);
            const itemFormat = getItemFormatFromPath(absolutePath);
            if (!itemFormat) {
                continue;
            }
            const extension = path.extname(file);
            const itemId = file.slice(0, file.length - extension.length).trim();
            if (itemId.length === 0) {
                continue;
            }
            const entry = variantsById.get(itemId) ?? {};
            entry[itemFormat] = absolutePath;
            variantsById.set(itemId, entry);
        }
        const itemIds = [...variantsById.keys()].sort((left, right) => left.localeCompare(right));
        for (const itemId of itemIds) {
            const variants = variantsById.get(itemId);
            scanned += 1;
            const sourcePath = (variants[targetFormat] ?? variants[alternateFormat]);
            const sourceFormat = sourcePath === variants[targetFormat] ? targetFormat : alternateFormat;
            try {
                const sourceRaw = await fs.readFile(sourcePath, "utf8");
                const parsedDocument = parseItemDocument(sourceRaw, {
                    format: sourceFormat,
                    schema,
                    onWarning: (warning) => warnings.push(`item_format_migration_parse_warning:${itemId}:${warning}`),
                });
                const targetPath = getItemPath(pmRoot, itemType, itemId, targetFormat, typeToFolder);
                const serializedTarget = serializeItemDocument(parsedDocument, { format: targetFormat, schema });
                const existingTargetRaw = await readFileIfExists(targetPath);
                if (existingTargetRaw !== serializedTarget) {
                    await writeFileAtomic(targetPath, serializedTarget);
                    warnings.push(...(await runActiveOnWriteHooks({
                        path: targetPath,
                        scope: "project",
                        op,
                    })));
                    migratedIds.add(itemId);
                }
                const alternatePath = variants[alternateFormat];
                if (alternatePath && alternatePath !== targetPath && (await pathExists(alternatePath))) {
                    await removeFileIfExists(alternatePath);
                    removedPaths.add(normalizeRelativePath(pmRoot, alternatePath));
                    warnings.push(...(await runActiveOnWriteHooks({
                        path: alternatePath,
                        scope: "project",
                        op: `${op}:remove`,
                    })));
                    migratedIds.add(itemId);
                }
            }
            catch (error) {
                warnings.push(`item_format_migration_skipped:${normalizeRelativePath(pmRoot, sourcePath)}:${errorSummary(error)}`);
            }
        }
    }
    return {
        target_format: targetFormat,
        scanned,
        migrated: [...migratedIds].sort((left, right) => left.localeCompare(right)),
        removed: [...removedPaths].sort((left, right) => left.localeCompare(right)),
        warnings,
    };
}
