// @unbrained/pm-cli/dist/core/store/paths.js
import os from "node:os";
import path from "node:path";
import { PM_DIRNAME, SETTINGS_FILENAME, TYPE_TO_FOLDER } from "../shared/constants.js";
export const ITEM_FILE_EXTENSION_BY_FORMAT = {
    json_markdown: ".md",
    toon: ".toon",
};
const ITEM_FORMAT_BY_EXTENSION = {
    ".md": "json_markdown",
    ".toon": "toon",
};
export const ITEM_FILE_EXTENSIONS = [".md", ".toon"];
export function resolvePmRoot(cwd, cliPath) {
    const envPath = process.env.PM_PATH;
    const selected = cliPath?.trim() || envPath?.trim() || PM_DIRNAME;
    return path.resolve(cwd, selected);
}
export function resolveGlobalPmRoot(cwd) {
    const envPath = process.env.PM_GLOBAL_PATH?.trim();
    const selected = envPath && envPath.length > 0 ? envPath : path.join(os.homedir(), ".pm-cli");
    return path.resolve(cwd, selected);
}
export function getSettingsPath(pmRoot) {
    return path.join(pmRoot, SETTINGS_FILENAME);
}
function deriveDefaultTypeFolder(type) {
    const normalized = type
        .trim()
        .toLowerCase()
        .replace(/[^a-z0-9]+/g, "-")
        .replace(/^-+|-+$/g, "");
    if (normalized.length === 0) {
        return "items";
    }
    return normalized.endsWith("s") ? normalized : `${normalized}s`;
}
export function getTypeDirPath(pmRoot, type, typeToFolder = TYPE_TO_FOLDER) {
    const folder = typeToFolder[type] ?? deriveDefaultTypeFolder(type);
    return path.join(pmRoot, folder);
}
export function getItemPath(pmRoot, type, id, itemFormat = "toon", typeToFolder = TYPE_TO_FOLDER) {
    return path.join(getTypeDirPath(pmRoot, type, typeToFolder), `${id}${ITEM_FILE_EXTENSION_BY_FORMAT[itemFormat]}`);
}
export function getItemFormatFromPath(itemPath) {
    const extension = path.extname(itemPath).toLowerCase();
    return ITEM_FORMAT_BY_EXTENSION[extension] ?? null;
}
export function getHistoryPath(pmRoot, id) {
    return path.join(pmRoot, "history", `${id}.jsonl`);
}
export function getLockPath(pmRoot, id) {
    return path.join(pmRoot, "locks", `${id}.lock`);
}
export function getRuntimePath(pmRoot) {
    return path.join(pmRoot, "runtime");
}
export function getTestRunsPath(pmRoot) {
    return path.join(getRuntimePath(pmRoot), "test-runs");
}
export function getTestRunsRecordsPath(pmRoot) {
    return path.join(getTestRunsPath(pmRoot), "runs");
}
export function getTestRunRecordPath(pmRoot, runId) {
    return path.join(getTestRunsRecordsPath(pmRoot), `${runId}.json`);
}
export function getTestRunsStdoutPath(pmRoot) {
    return path.join(getTestRunsPath(pmRoot), "stdout");
}
export function getTestRunsStderrPath(pmRoot) {
    return path.join(getTestRunsPath(pmRoot), "stderr");
}
export function getTestRunStdoutPath(pmRoot, runId) {
    return path.join(getTestRunsStdoutPath(pmRoot), `${runId}.log`);
}
export function getTestRunStderrPath(pmRoot, runId) {
    return path.join(getTestRunsStderrPath(pmRoot), `${runId}.log`);
}
export function getTestRunsResultsPath(pmRoot) {
    return path.join(getTestRunsPath(pmRoot), "results");
}
export function getTestRunResultPath(pmRoot, runId) {
    return path.join(getTestRunsResultsPath(pmRoot), `${runId}.json`);
}
