// @unbrained/pm-cli/dist/core/store/settings.js
import { z } from "zod";
import { runActiveOnReadHooks, runActiveOnWriteHooks } from "../extensions/index.js";
import { GOVERNANCE_PRESET_DEFAULTS, SETTINGS_DEFAULTS } from "../shared/constants.js";
import { readFileIfExists, writeFileAtomic } from "../fs/fs-utils.js";
import { ensureRuntimeSchemaFileScaffold, loadRuntimeSchemaFromOptionalFiles, normalizeRuntimeSchemaSettings, } from "../schema/runtime-schema.js";
import { getSettingsPath } from "./paths.js";
import { orderObject } from "../shared/serialization.js";
const itemTypeOptionSchema = z.object({
    key: z.string(),
    values: z.array(z.string()),
    required: z.boolean().optional(),
    aliases: z.array(z.string()).optional(),
    description: z.string().optional(),
});
const itemTypeCommandOptionPolicySchema = z.object({
    command: z.union([z.literal("create"), z.literal("update")]),
    option: z.string(),
    required: z.boolean().optional(),
    visible: z.boolean().optional(),
    enabled: z.boolean().optional(),
});
const itemTypeDefinitionSchema = z.object({
    name: z.string(),
    folder: z.string().optional(),
    aliases: z.array(z.string()).optional(),
    required_create_fields: z.array(z.string()).optional(),
    required_create_repeatables: z.array(z.string()).optional(),
    options: z.array(itemTypeOptionSchema).optional(),
    command_option_policies: z.array(itemTypeCommandOptionPolicySchema).optional(),
});
const runtimeStatusDefinitionSchema = z.object({
    id: z.string(),
    aliases: z.array(z.string()).optional(),
    roles: z
        .array(z.union([
        z.literal("draft"),
        z.literal("active"),
        z.literal("blocked"),
        z.literal("terminal"),
        z.literal("terminal_done"),
        z.literal("terminal_canceled"),
        z.literal("default_open"),
        z.literal("default_close"),
        z.literal("default_cancel"),
    ]))
        .optional(),
    description: z.string().optional(),
    order: z.number().optional(),
});
const runtimeFieldDefinitionSchema = z.object({
    key: z.string(),
    metadata_key: z.string().optional(),
    front_matter_key: z.string().optional(),
    cli_flag: z.string().optional(),
    cli_aliases: z.array(z.string()).optional(),
    description: z.string().optional(),
    type: z.union([z.literal("string"), z.literal("number"), z.literal("boolean"), z.literal("string_array")]).optional(),
    commands: z
        .array(z.union([
        z.literal("create"),
        z.literal("update"),
        z.literal("update_many"),
        z.literal("list"),
        z.literal("search"),
        z.literal("calendar"),
        z.literal("context"),
    ]))
        .optional(),
    repeatable: z.boolean().optional(),
    required: z.boolean().optional(),
    required_on_create: z.boolean().optional(),
    required_types: z.array(z.string()).optional(),
    allow_unset: z.boolean().optional(),
});
const runtimeSchemaSettingsSchema = z
    .object({
    version: z.number().int().optional(),
    files: z
        .object({
        types: z.string().optional(),
        statuses: z.string().optional(),
        fields: z.string().optional(),
        workflows: z.string().optional(),
    })
        .optional(),
    statuses: z.array(runtimeStatusDefinitionSchema).optional(),
    fields: z.array(runtimeFieldDefinitionSchema).optional(),
    workflow: z
        .object({
        draft_status: z.string().optional(),
        open_status: z.string().optional(),
        in_progress_status: z.string().optional(),
        blocked_status: z.string().optional(),
        close_status: z.string().optional(),
        canceled_status: z.string().optional(),
    })
        .optional(),
    unknown_field_policy: z.union([z.literal("allow"), z.literal("warn"), z.literal("reject")]).optional(),
})
    .optional();
const governanceSettingsSchema = z
    .object({
    preset: z.union([z.literal("minimal"), z.literal("default"), z.literal("strict"), z.literal("custom")]).optional(),
    ownership_enforcement: z.union([z.literal("none"), z.literal("warn"), z.literal("strict")]).optional(),
    create_mode_default: z.union([z.literal("progressive"), z.literal("strict")]).optional(),
    close_validation_default: z.union([z.literal("off"), z.literal("warn"), z.literal("strict")]).optional(),
    parent_reference: z.union([z.literal("warn"), z.literal("strict_error")]).optional(),
    metadata_profile: z.union([z.literal("core"), z.literal("strict"), z.literal("custom")]).optional(),
    force_required_for_stale_lock: z.boolean().optional(),
})
    .optional();
const settingsSchema = z.object({
    version: z.number().int(),
    id_prefix: z.string(),
    author_default: z.string(),
    item_format: z.union([z.literal("toon"), z.literal("json_markdown")]).optional(),
    locks: z.object({
        ttl_seconds: z.number().int(),
    }),
    output: z.object({
        default_format: z.union([z.literal("toon"), z.literal("json")]),
    }),
    history: z
        .object({
        missing_stream: z.union([z.literal("auto_create"), z.literal("strict_error")]),
    })
        .optional(),
    validation: z
        .object({
        sprint_release_format: z.union([z.literal("warn"), z.literal("strict_error")]),
        parent_reference: z.union([z.literal("warn"), z.literal("strict_error")]).optional(),
        metadata_profile: z.union([z.literal("core"), z.literal("strict"), z.literal("custom")]).optional(),
        metadata_required_fields: z.array(z.string()).optional(),
        lifecycle_stale_blocker_reason_patterns: z.array(z.string()).optional(),
        lifecycle_closure_like_blocked_reason_patterns: z.array(z.string()).optional(),
        lifecycle_closure_like_resolution_patterns: z.array(z.string()).optional(),
        lifecycle_closure_like_actual_result_patterns: z.array(z.string()).optional(),
    })
        .optional(),
    governance: governanceSettingsSchema,
    workflow: z
        .object({
        definition_of_done: z.array(z.string()),
    })
        .optional(),
    testing: z
        .object({
        record_results_to_items: z.boolean(),
    })
        .optional(),
    telemetry: z
        .object({
        enabled: z.boolean(),
        first_run_prompt_completed: z.boolean().optional(),
        capture_level: z.union([z.literal("minimal"), z.literal("redacted"), z.literal("max")]).optional(),
        endpoint: z.string().optional(),
        installation_id: z.string().optional(),
        retention_days: z.number().int().positive().optional(),
    })
        .optional(),
    item_types: z
        .object({
        definitions: z.array(itemTypeDefinitionSchema),
    })
        .optional(),
    schema: runtimeSchemaSettingsSchema,
    context: z
        .object({
        default_depth: z.union([z.literal("brief"), z.literal("standard"), z.literal("deep")]).optional(),
        activity_limit: z.number().int().positive().optional(),
        stale_threshold_days: z.number().int().positive().optional(),
        sections: z
            .object({
            hierarchy: z.boolean().optional(),
            activity: z.boolean().optional(),
            progress: z.boolean().optional(),
            blockers: z.boolean().optional(),
            files: z.boolean().optional(),
            workload: z.boolean().optional(),
            staleness: z.boolean().optional(),
            tests: z.boolean().optional(),
        })
            .optional(),
    })
        .optional(),
    extensions: z.object({
        enabled: z.array(z.string()),
        disabled: z.array(z.string()),
    }),
    search: z.object({
        score_threshold: z.number(),
        hybrid_semantic_weight: z.number().optional(),
        max_results: z.number().int(),
        embedding_model: z.string(),
        embedding_batch_size: z.number().int(),
        scanner_max_batch_retries: z.number().int(),
        provider: z.string().optional(),
    }),
    providers: z.object({
        openai: z.object({
            base_url: z.string(),
            api_key: z.string(),
            model: z.string(),
        }),
        ollama: z.object({
            base_url: z.string(),
            model: z.string(),
        }),
    }),
    vector_store: z.object({
        adapter: z.string().optional(),
        qdrant: z.object({
            url: z.string(),
            api_key: z.string(),
        }),
        lancedb: z.object({
            path: z.string(),
        }),
    }),
});
const SETTINGS_WRITE_OP = "settings:write";
function resolveGovernanceKnobsFromPreset(preset) {
    return structuredClone(GOVERNANCE_PRESET_DEFAULTS[preset]);
}
function normalizeGovernancePreset(value) {
    if (value === "minimal" || value === "default" || value === "strict" || value === "custom") {
        return value;
    }
    return SETTINGS_DEFAULTS.governance.preset;
}
function normalizeGovernanceForPersist(governance) {
    if (governance.preset === "custom") {
        return {
            preset: "custom",
            ownership_enforcement: governance.ownership_enforcement,
            create_mode_default: governance.create_mode_default,
            close_validation_default: governance.close_validation_default,
            parent_reference: governance.parent_reference,
            metadata_profile: governance.metadata_profile,
            force_required_for_stale_lock: governance.force_required_for_stale_lock,
        };
    }
    return {
        preset: governance.preset,
    };
}
export function resolveGovernanceKnobs(settings) {
    const rawGovernance = settings.governance ?? {};
    const preset = normalizeGovernancePreset(rawGovernance.preset);
    if (preset === "custom") {
        const baseline = resolveGovernanceKnobsFromPreset("default");
        return {
            preset,
            ownership_enforcement: rawGovernance.ownership_enforcement ?? baseline.ownership_enforcement,
            create_mode_default: rawGovernance.create_mode_default ?? baseline.create_mode_default,
            close_validation_default: rawGovernance.close_validation_default ?? baseline.close_validation_default,
            parent_reference: rawGovernance.parent_reference ?? baseline.parent_reference,
            metadata_profile: rawGovernance.metadata_profile ?? baseline.metadata_profile,
            force_required_for_stale_lock: rawGovernance.force_required_for_stale_lock ?? baseline.force_required_for_stale_lock,
        };
    }
    return {
        preset,
        ...resolveGovernanceKnobsFromPreset(preset),
    };
}
function cloneDefaults() {
    return structuredClone(SETTINGS_DEFAULTS);
}
function buildFallbackSettingsReadResult(warning) {
    return {
        settings: cloneDefaults(),
        metadata: {
            has_explicit_item_format: false,
        },
        warnings: warning ? [warning] : [],
    };
}
function hasExplicitItemFormat(raw) {
    if (typeof raw !== "object" || raw === null || Array.isArray(raw)) {
        return false;
    }
    const itemFormat = raw.item_format;
    return itemFormat === "toon" || itemFormat === "json_markdown";
}
function normalizeStringList(values) {
    return [...new Set((values ?? []).map((value) => value.trim()).filter((value) => value.length > 0))].sort((left, right) => left.localeCompare(right));
}
function normalizeValidationMetadataRequiredFields(values) {
    const normalized = [...new Set((values ?? []).map((value) => value.trim().toLowerCase().replaceAll("-", "_")))]
        .filter((value) => value.length > 0)
        .sort((left, right) => left.localeCompare(right));
    return normalized.filter((value) => [
        "author",
        "acceptance_criteria",
        "estimated_minutes",
        "close_reason",
        "reviewer",
        "risk",
        "confidence",
        "sprint",
        "release",
    ].includes(value));
}
function normalizeValidationPatternList(values) {
    return [...new Set((values ?? []).map((value) => value.trim().toLowerCase()).filter((value) => value.length > 0))].sort((left, right) => left.localeCompare(right));
}
function normalizeItemTypeOptionDefinition(option) {
    const key = option.key.trim();
    if (key.length === 0) {
        return null;
    }
    const values = normalizeStringList(option.values);
    const aliases = normalizeStringList(option.aliases);
    const description = option.description?.trim();
    return {
        key,
        values,
        required: option.required === true ? true : undefined,
        aliases: aliases.length > 0 ? aliases : undefined,
        description: description && description.length > 0 ? description : undefined,
    };
}
function normalizeItemTypeCommandOptionPolicy(policy) {
    const option = policy.option.trim();
    if (option.length === 0) {
        return null;
    }
    return {
        command: policy.command,
        option,
        required: policy.required,
        visible: policy.visible,
        enabled: policy.enabled,
    };
}
function normalizeItemTypeDefinition(definition) {
    const name = definition.name.trim();
    if (name.length === 0) {
        return null;
    }
    const hasRequiredCreateFields = definition.required_create_fields !== undefined;
    const hasRequiredCreateRepeatables = definition.required_create_repeatables !== undefined;
    const hasOptions = definition.options !== undefined;
    const hasCommandOptionPolicies = definition.command_option_policies !== undefined;
    const folder = definition.folder?.trim();
    const aliases = normalizeStringList(definition.aliases);
    const requiredCreateFields = normalizeStringList(definition.required_create_fields);
    const requiredCreateRepeatables = normalizeStringList(definition.required_create_repeatables);
    const options = (definition.options ?? [])
        .map((option) => normalizeItemTypeOptionDefinition(option))
        .filter((option) => option !== null)
        .sort((left, right) => left.key.localeCompare(right.key));
    const commandOptionPolicies = (() => {
        const dedupedByKey = new Map();
        for (const policy of definition.command_option_policies ?? []) {
            const normalized = normalizeItemTypeCommandOptionPolicy(policy);
            if (!normalized) {
                continue;
            }
            dedupedByKey.set(`${normalized.command}:${normalized.option.toLowerCase()}`, normalized);
        }
        return [...dedupedByKey.values()].sort((left, right) => left.command === right.command
            ? left.option.localeCompare(right.option)
            : left.command.localeCompare(right.command));
    })();
    return {
        name,
        folder: folder && folder.length > 0 ? folder : undefined,
        aliases: aliases.length > 0 ? aliases : undefined,
        required_create_fields: hasRequiredCreateFields ? requiredCreateFields : undefined,
        required_create_repeatables: hasRequiredCreateRepeatables ? requiredCreateRepeatables : undefined,
        options: hasOptions ? options : undefined,
        command_option_policies: hasCommandOptionPolicies ? commandOptionPolicies : undefined,
    };
}
export function normalizeItemTypeDefinitions(definitions) {
    const normalized = (definitions ?? [])
        .map((definition) => normalizeItemTypeDefinition(definition))
        .filter((definition) => definition !== null);
    const dedupedByName = new Map();
    for (const definition of normalized) {
        dedupedByName.set(definition.name.toLowerCase(), definition);
    }
    return [...dedupedByName.values()].sort((left, right) => left.name.localeCompare(right.name));
}
function mergeSettings(raw) {
    const defaults = cloneDefaults();
    const parsed = settingsSchema.safeParse(raw);
    if (!parsed.success) {
        return defaults;
    }
    const settings = parsed.data;
    const governance = resolveGovernanceKnobs({
        governance: settings.governance ?? { preset: "default" },
    });
    return {
        ...defaults,
        ...settings,
        item_format: settings.item_format === "json_markdown" ? "toon" : (settings.item_format ?? defaults.item_format),
        locks: { ...defaults.locks, ...settings.locks },
        output: { ...defaults.output, ...settings.output },
        history: { ...defaults.history, ...(settings.history ?? {}) },
        validation: {
            ...defaults.validation,
            ...(settings.validation ?? {}),
            parent_reference: governance.parent_reference,
            metadata_profile: governance.metadata_profile,
            metadata_required_fields: normalizeValidationMetadataRequiredFields(settings.validation?.metadata_required_fields),
            lifecycle_stale_blocker_reason_patterns: normalizeValidationPatternList(settings.validation?.lifecycle_stale_blocker_reason_patterns ??
                defaults.validation.lifecycle_stale_blocker_reason_patterns),
            lifecycle_closure_like_blocked_reason_patterns: normalizeValidationPatternList(settings.validation?.lifecycle_closure_like_blocked_reason_patterns ??
                defaults.validation.lifecycle_closure_like_blocked_reason_patterns),
            lifecycle_closure_like_resolution_patterns: normalizeValidationPatternList(settings.validation?.lifecycle_closure_like_resolution_patterns ??
                defaults.validation.lifecycle_closure_like_resolution_patterns),
            lifecycle_closure_like_actual_result_patterns: normalizeValidationPatternList(settings.validation?.lifecycle_closure_like_actual_result_patterns ??
                defaults.validation.lifecycle_closure_like_actual_result_patterns),
        },
        governance,
        workflow: {
            definition_of_done: [...(settings.workflow?.definition_of_done ?? defaults.workflow.definition_of_done)],
        },
        testing: {
            record_results_to_items: settings.testing?.record_results_to_items ?? defaults.testing.record_results_to_items,
        },
        telemetry: {
            enabled: settings.telemetry?.enabled ?? defaults.telemetry.enabled,
            first_run_prompt_completed: settings.telemetry?.first_run_prompt_completed ?? defaults.telemetry.first_run_prompt_completed,
            capture_level: settings.telemetry?.capture_level ?? defaults.telemetry.capture_level,
            endpoint: settings.telemetry?.endpoint ?? defaults.telemetry.endpoint,
            installation_id: settings.telemetry?.installation_id ?? defaults.telemetry.installation_id,
            retention_days: settings.telemetry?.retention_days ?? defaults.telemetry.retention_days,
        },
        item_types: {
            definitions: normalizeItemTypeDefinitions(settings.item_types?.definitions),
        },
        schema: normalizeRuntimeSchemaSettings(settings.schema ?? defaults.schema),
        context: {
            default_depth: settings.context?.default_depth ?? defaults.context.default_depth,
            activity_limit: settings.context?.activity_limit ?? defaults.context.activity_limit,
            stale_threshold_days: settings.context?.stale_threshold_days ?? defaults.context.stale_threshold_days,
            sections: {
                ...defaults.context.sections,
                ...(settings.context?.sections ?? {}),
            },
        },
        extensions: {
            enabled: [...settings.extensions.enabled],
            disabled: [...settings.extensions.disabled],
        },
        search: { ...defaults.search, ...settings.search },
        providers: {
            openai: { ...defaults.providers.openai, ...settings.providers.openai },
            ollama: { ...defaults.providers.ollama, ...settings.providers.ollama },
        },
        vector_store: {
            adapter: settings.vector_store.adapter ?? defaults.vector_store.adapter,
            qdrant: { ...defaults.vector_store.qdrant, ...settings.vector_store.qdrant },
            lancedb: { ...defaults.vector_store.lancedb, ...settings.vector_store.lancedb },
        },
    };
}
export function serializeSettings(settings) {
    const governance = resolveGovernanceKnobs(settings);
    const normalizedSettings = {
        ...settings,
        item_format: "toon",
        validation: {
            ...settings.validation,
            parent_reference: governance.parent_reference,
            metadata_profile: governance.metadata_profile,
            metadata_required_fields: normalizeValidationMetadataRequiredFields(settings.validation?.metadata_required_fields),
            lifecycle_stale_blocker_reason_patterns: normalizeValidationPatternList(settings.validation?.lifecycle_stale_blocker_reason_patterns ??
                SETTINGS_DEFAULTS.validation.lifecycle_stale_blocker_reason_patterns),
            lifecycle_closure_like_blocked_reason_patterns: normalizeValidationPatternList(settings.validation?.lifecycle_closure_like_blocked_reason_patterns ??
                SETTINGS_DEFAULTS.validation.lifecycle_closure_like_blocked_reason_patterns),
            lifecycle_closure_like_resolution_patterns: normalizeValidationPatternList(settings.validation?.lifecycle_closure_like_resolution_patterns ??
                SETTINGS_DEFAULTS.validation.lifecycle_closure_like_resolution_patterns),
            lifecycle_closure_like_actual_result_patterns: normalizeValidationPatternList(settings.validation?.lifecycle_closure_like_actual_result_patterns ??
                SETTINGS_DEFAULTS.validation.lifecycle_closure_like_actual_result_patterns),
        },
        governance,
        item_types: {
            definitions: normalizeItemTypeDefinitions(settings.item_types?.definitions),
        },
        schema: normalizeRuntimeSchemaSettings(settings.schema),
    };
    const ordered = orderObject({
        ...normalizedSettings,
        governance: normalizeGovernanceForPersist(governance),
    }, [
        "version",
        "id_prefix",
        "author_default",
        "item_format",
        "locks",
        "output",
        "history",
        "validation",
        "governance",
        "workflow",
        "testing",
        "telemetry",
        "item_types",
        "schema",
        "context",
        "extensions",
        "search",
        "providers",
        "vector_store",
    ]);
    ordered.locks = orderObject(ordered.locks, ["ttl_seconds"]);
    ordered.output = orderObject(ordered.output, ["default_format"]);
    ordered.history = orderObject(ordered.history, ["missing_stream"]);
    ordered.validation = orderObject(ordered.validation, [
        "sprint_release_format",
        "parent_reference",
        "metadata_profile",
        "metadata_required_fields",
        "lifecycle_stale_blocker_reason_patterns",
        "lifecycle_closure_like_blocked_reason_patterns",
        "lifecycle_closure_like_resolution_patterns",
        "lifecycle_closure_like_actual_result_patterns",
    ]);
    ordered.governance = orderObject(ordered.governance, [
        "preset",
        "ownership_enforcement",
        "create_mode_default",
        "close_validation_default",
        "parent_reference",
        "metadata_profile",
        "force_required_for_stale_lock",
    ]);
    ordered.workflow = orderObject(ordered.workflow, ["definition_of_done"]);
    ordered.testing = orderObject(ordered.testing, ["record_results_to_items"]);
    ordered.telemetry = orderObject(ordered.telemetry, [
        "enabled",
        "first_run_prompt_completed",
        "capture_level",
        "endpoint",
        "installation_id",
        "retention_days",
    ]);
    ordered.item_types = orderObject(ordered.item_types, ["definitions"]);
    ordered.schema = orderObject(ordered.schema, [
        "version",
        "files",
        "statuses",
        "fields",
        "workflow",
        "unknown_field_policy",
    ]);
    ordered.schema.files = orderObject((ordered.schema.files ?? {}), ["types", "statuses", "fields", "workflows"]);
    ordered.schema.workflow = orderObject((ordered.schema.workflow ?? {}), ["draft_status", "open_status", "in_progress_status", "blocked_status", "close_status", "canceled_status"]);
    ordered.context = orderObject(ordered.context, [
        "default_depth",
        "activity_limit",
        "stale_threshold_days",
        "sections",
    ]);
    ordered.context.sections = orderObject((ordered.context.sections ?? {}), ["hierarchy", "activity", "progress", "blockers", "files", "workload", "staleness", "tests"]);
    ordered.extensions = orderObject(ordered.extensions, ["enabled", "disabled"]);
    ordered.search = orderObject(ordered.search, [
        "score_threshold",
        "hybrid_semantic_weight",
        "max_results",
        "embedding_model",
        "embedding_batch_size",
        "scanner_max_batch_retries",
        "provider",
    ]);
    ordered.providers = orderObject(ordered.providers, ["openai", "ollama"]);
    ordered.providers.openai = orderObject((ordered.providers.openai ?? {}), ["base_url", "api_key", "model"]);
    ordered.providers.ollama = orderObject((ordered.providers.ollama ?? {}), ["base_url", "model"]);
    ordered.vector_store = orderObject(ordered.vector_store, ["adapter", "qdrant", "lancedb"]);
    ordered.vector_store.qdrant = orderObject((ordered.vector_store.qdrant ?? {}), ["url", "api_key"]);
    ordered.vector_store.lancedb = orderObject((ordered.vector_store.lancedb ?? {}), ["path"]);
    return `${JSON.stringify(ordered, null, 2)}\n`;
}
export async function readSettingsWithMetadata(pmRoot) {
    const settingsPath = getSettingsPath(pmRoot);
    const raw = await readFileIfExists(settingsPath);
    if (raw === null) {
        return buildFallbackSettingsReadResult();
    }
    await runActiveOnReadHooks({
        path: settingsPath,
        scope: "project",
    });
    let parsed;
    try {
        parsed = JSON.parse(raw);
    }
    catch {
        return buildFallbackSettingsReadResult("settings_read_invalid_json");
    }
    const validated = settingsSchema.safeParse(parsed);
    if (!validated.success) {
        return buildFallbackSettingsReadResult("settings_read_invalid_schema");
    }
    try {
        const mergedSettings = mergeSettings(validated.data);
        const schemaScaffold = await ensureRuntimeSchemaFileScaffold(pmRoot, mergedSettings.schema);
        const loadedSchemaSections = await loadRuntimeSchemaFromOptionalFiles(pmRoot, mergedSettings.schema);
        const settings = {
            ...mergedSettings,
            item_types: {
                definitions: normalizeItemTypeDefinitions([
                    ...mergedSettings.item_types.definitions,
                    ...(loadedSchemaSections.type_definitions_from_file ?? []),
                ]),
            },
            schema: loadedSchemaSections.schema,
        };
        return {
            settings,
            metadata: {
                has_explicit_item_format: hasExplicitItemFormat(parsed),
            },
            warnings: [
                ...(validated.data.item_format === "json_markdown"
                    ? ["settings_item_format_legacy_json_markdown_coerced_to_toon"]
                    : []),
                ...schemaScaffold.created_paths.map((createdPath) => `runtime_schema_bootstrap_created:${createdPath}`),
                ...loadedSchemaSections.warnings,
            ],
        };
    }
    catch {
        return buildFallbackSettingsReadResult("settings_read_merge_failed");
    }
}
export async function readSettings(pmRoot) {
    return (await readSettingsWithMetadata(pmRoot)).settings;
}
export async function writeSettings(pmRoot, settings, op = SETTINGS_WRITE_OP) {
    const settingsPath = getSettingsPath(pmRoot);
    await writeFileAtomic(settingsPath, serializeSettings(settings));
    await runActiveOnWriteHooks({
        path: settingsPath,
        scope: "project",
        op,
    });
}
