// @unbrained/pm-cli/dist/core/search/cache.js
import path from "node:path";
import { getActiveExtensionRegistrations } from "../extensions/index.js";
import { pathExists, readFileIfExists, removeFileIfExists, writeFileAtomic } from "../fs/fs-utils.js";
import { resolveItemTypeRegistry } from "../item/type-registry.js";
import { toErrorMessage } from "../shared/primitives.js";
import { locateItem, readLocatedItem } from "../store/item-store.js";
import { getSettingsPath } from "../store/paths.js";
import { readSettings } from "../store/settings.js";
import { executeEmbeddingBatchesWithRetry } from "./embedding-batches.js";
import { buildSemanticCorpusInput } from "./corpus.js";
import { resolveEmbeddingProviders } from "./providers.js";
import { resolveSettingsWithSemanticRuntimeDefaults } from "./semantic-defaults.js";
import { executeVectorDelete, executeVectorUpsert, resolveVectorStores } from "./vector-stores.js";
import { nowIso } from "../shared/time.js";
export const SEARCH_CACHE_ARTIFACT_PATHS = ["index/manifest.json", "search/embeddings.jsonl"];
export const VECTORIZATION_STATUS_LEDGER_PATH = "search/vectorization-status.json";
function formatInvalidationWarning(relativePath, error) {
    return `search_cache_invalidation_failed:${relativePath}:${String(error)}`;
}
function toUniqueSorted(values) {
    return [...new Set([...values].map((value) => value.trim()).filter((value) => value.length > 0))]
        .sort((left, right) => left.localeCompare(right));
}
function isValidUpdatedAt(value) {
    return typeof value === "string" && value.trim().length > 0 && Number.isFinite(Date.parse(value));
}
function normalizeVectorizationLedgerEntries(entries) {
    const normalized = new Map();
    for (const [rawId, rawUpdatedAt] of Object.entries(entries)) {
        const id = rawId.trim();
        if (id.length === 0 || !isValidUpdatedAt(rawUpdatedAt)) {
            continue;
        }
        normalized.set(id, rawUpdatedAt);
    }
    return Object.fromEntries([...normalized.entries()].sort((left, right) => left[0].localeCompare(right[0])));
}
export async function readVectorizationStatusLedger(pmRoot) {
    const ledgerPath = path.join(pmRoot, VECTORIZATION_STATUS_LEDGER_PATH);
    const raw = await readFileIfExists(ledgerPath);
    if (raw === null || raw.trim().length === 0) {
        return {
            entries: {},
            warnings: [],
        };
    }
    let parsed;
    try {
        parsed = JSON.parse(raw);
    }
    catch {
        return {
            entries: {},
            warnings: ["search_vectorization_status_ledger_invalid"],
        };
    }
    if (typeof parsed !== "object" ||
        parsed === null ||
        parsed.version !== 1 ||
        !Array.isArray(parsed.items)) {
        return {
            entries: {},
            warnings: ["search_vectorization_status_ledger_invalid"],
        };
    }
    const mapped = new Map();
    for (const entry of parsed.items) {
        if (typeof entry !== "object" || entry === null) {
            return {
                entries: {},
                warnings: ["search_vectorization_status_ledger_invalid"],
            };
        }
        const id = entry.id;
        const updatedAt = entry.updated_at;
        if (typeof id !== "string" || id.trim().length === 0 || !isValidUpdatedAt(updatedAt)) {
            return {
                entries: {},
                warnings: ["search_vectorization_status_ledger_invalid"],
            };
        }
        mapped.set(id.trim(), updatedAt);
    }
    return {
        entries: Object.fromEntries([...mapped.entries()].sort((left, right) => left[0].localeCompare(right[0]))),
        warnings: [],
    };
}
export async function writeVectorizationStatusLedger(pmRoot, entries) {
    const normalizedEntries = normalizeVectorizationLedgerEntries(entries);
    const ledgerPath = path.join(pmRoot, VECTORIZATION_STATUS_LEDGER_PATH);
    const serialized = `${JSON.stringify({
        version: 1,
        generated_at: nowIso(),
        items: Object.entries(normalizedEntries).map(([id, updated_at]) => ({
            id,
            updated_at,
        })),
    }, null, 2)}\n`;
    await writeFileAtomic(ledgerPath, serialized);
}
function buildVectorPayload(item) {
    return {
        id: item.id,
        type: item.type,
        status: item.status,
        priority: item.priority,
        updated_at: item.updated_at,
    };
}
function buildSkippedSemanticRefreshResult(itemIds, warning) {
    return {
        refreshed: [],
        skipped: itemIds,
        warnings: [warning],
    };
}
async function resolveSemanticRefreshRuntimeContext(pmRoot, normalizedItemIds, options) {
    if (!options.settings && !(await pathExists(getSettingsPath(pmRoot)))) {
        return buildSkippedSemanticRefreshResult(normalizedItemIds, "search_semantic_refresh_skipped:settings_not_initialized");
    }
    let settings;
    try {
        settings = options.settings ?? (await readSettings(pmRoot));
    }
    catch (error) {
        return buildSkippedSemanticRefreshResult(normalizedItemIds, `search_semantic_refresh_skipped:settings_read_failed:${toErrorMessage(error)}`);
    }
    const effectiveSettings = options.apply_runtime_defaults
        ? resolveSettingsWithSemanticRuntimeDefaults(settings).settings
        : settings;
    const provider = resolveEmbeddingProviders(effectiveSettings).active;
    if (!provider) {
        return buildSkippedSemanticRefreshResult(normalizedItemIds, "search_semantic_refresh_skipped:provider_unconfigured");
    }
    const vectorStore = resolveVectorStores(effectiveSettings).active;
    if (!vectorStore) {
        return buildSkippedSemanticRefreshResult(normalizedItemIds, "search_semantic_refresh_skipped:vector_store_unconfigured");
    }
    return {
        settings: effectiveSettings,
        provider,
        vectorStore,
    };
}
async function collectSemanticRefreshWorkload(pmRoot, idPrefix, preferredFormat, normalizedItemIds, typeToFolder, schema) {
    const warnings = [];
    const skipped = new Set();
    const missing = new Set();
    const documents = [];
    for (const itemId of normalizedItemIds) {
        const located = await locateItem(pmRoot, itemId, idPrefix, preferredFormat, typeToFolder);
        if (!located) {
            missing.add(itemId);
            continue;
        }
        try {
            const loaded = await readLocatedItem(located, { schema });
            documents.push({
                id: loaded.document.metadata.id,
                document: loaded.document,
            });
        }
        catch (error) {
            skipped.add(located.id);
            warnings.push(`search_semantic_refresh_item_read_failed:${located.id}:${toErrorMessage(error)}`);
        }
    }
    documents.sort((left, right) => left.id.localeCompare(right.id));
    return {
        documents,
        missingIds: toUniqueSorted(missing),
        skippedIds: toUniqueSorted(skipped),
        warnings,
    };
}
async function refreshLocatedSemanticVectors(settings, provider, vectorStore, documents) {
    if (documents.length === 0) {
        return { refreshed: [], skipped: [], warnings: [] };
    }
    try {
        const corpusInputs = documents.map((entry) => buildSemanticCorpusInput(entry.document, {
            providerName: provider.name,
        }));
        const embeddingResult = await executeEmbeddingBatchesWithRetry(provider, settings, corpusInputs);
        await executeVectorUpsert(vectorStore, documents.map((entry, index) => ({
            id: entry.id,
            vector: embeddingResult.vectors[index],
            payload: buildVectorPayload(entry.document.metadata),
        })));
        return {
            refreshed: documents.map((entry) => entry.id),
            skipped: [],
            warnings: embeddingResult.warnings,
        };
    }
    catch (error) {
        return {
            refreshed: [],
            skipped: documents.map((entry) => entry.id),
            warnings: [`search_semantic_refresh_failed:${toErrorMessage(error)}`],
        };
    }
}
async function pruneMissingSemanticVectors(vectorStore, missingIds) {
    if (missingIds.length === 0) {
        return { refreshed: [], skipped: [], warnings: [] };
    }
    try {
        await executeVectorDelete(vectorStore, missingIds);
        return {
            refreshed: [],
            skipped: [],
            warnings: [],
        };
    }
    catch (error) {
        return {
            refreshed: [],
            skipped: missingIds,
            warnings: [`search_semantic_refresh_delete_failed:${toErrorMessage(error)}`],
        };
    }
}
export async function invalidateSearchCacheArtifacts(pmRoot) {
    const invalidated = [];
    const warnings = [];
    for (const relativePath of SEARCH_CACHE_ARTIFACT_PATHS) {
        const artifactPath = path.join(pmRoot, relativePath);
        try {
            if (!(await pathExists(artifactPath))) {
                continue;
            }
            await removeFileIfExists(artifactPath);
            invalidated.push(relativePath);
        }
        catch (error) {
            warnings.push(formatInvalidationWarning(relativePath, error));
        }
    }
    return {
        invalidated,
        warnings,
    };
}
export async function refreshSemanticEmbeddingsForMutatedItems(pmRoot, itemIds, options = {}) {
    const normalizedItemIds = toUniqueSorted(itemIds);
    if (normalizedItemIds.length === 0) {
        return {
            refreshed: [],
            skipped: [],
            warnings: [],
        };
    }
    const runtimeContext = await resolveSemanticRefreshRuntimeContext(pmRoot, normalizedItemIds, options);
    if (!("settings" in runtimeContext)) {
        return runtimeContext;
    }
    const typeRegistry = resolveItemTypeRegistry(runtimeContext.settings, getActiveExtensionRegistrations());
    const workload = await collectSemanticRefreshWorkload(pmRoot, runtimeContext.settings.id_prefix, runtimeContext.settings.item_format, normalizedItemIds, typeRegistry.type_to_folder, runtimeContext.settings.schema);
    const refreshedResult = await refreshLocatedSemanticVectors(runtimeContext.settings, runtimeContext.provider, runtimeContext.vectorStore, workload.documents);
    const pruneResult = await pruneMissingSemanticVectors(runtimeContext.vectorStore, workload.missingIds);
    const refreshedIdSet = new Set(refreshedResult.refreshed);
    const refreshedEntries = Object.fromEntries(workload.documents
        .filter((entry) => refreshedIdSet.has(entry.id))
        .map((entry) => [entry.id, entry.document.metadata.updated_at]));
    const ledgerWarnings = [];
    if (Object.keys(refreshedEntries).length > 0 || workload.missingIds.length > 0) {
        const ledgerRead = await readVectorizationStatusLedger(pmRoot);
        const nextEntries = {
            ...ledgerRead.entries,
            ...normalizeVectorizationLedgerEntries(refreshedEntries),
        };
        for (const missingId of workload.missingIds) {
            delete nextEntries[missingId];
        }
        try {
            await writeVectorizationStatusLedger(pmRoot, nextEntries);
        }
        catch (error) {
            ledgerWarnings.push(`search_vectorization_status_ledger_write_failed:${toErrorMessage(error)}`);
        }
        ledgerWarnings.push(...ledgerRead.warnings);
    }
    return {
        refreshed: toUniqueSorted([...refreshedResult.refreshed, ...pruneResult.refreshed]),
        skipped: toUniqueSorted([...workload.skippedIds, ...refreshedResult.skipped, ...pruneResult.skipped]),
        warnings: [...workload.warnings, ...refreshedResult.warnings, ...pruneResult.warnings, ...ledgerWarnings],
    };
}
export async function refreshSearchArtifactsForMutation(pmRoot, itemIds) {
    const invalidation = await invalidateSearchCacheArtifacts(pmRoot);
    const semanticRefresh = await refreshSemanticEmbeddingsForMutatedItems(pmRoot, itemIds);
    return {
        invalidated: invalidation.invalidated,
        refreshed: semanticRefresh.refreshed,
        skipped: semanticRefresh.skipped,
        warnings: [...invalidation.warnings, ...semanticRefresh.warnings],
    };
}
