// @unbrained/pm-cli/dist/core/search/http-client.js
import { toErrorMessage } from "../shared/primitives.js";
export const DEFAULT_SEARCH_HTTP_TIMEOUT_MS = 30_000;
export function normalizeSearchHttpTimeoutMs(timeoutMs, requestLabel, defaultTimeoutMs = DEFAULT_SEARCH_HTTP_TIMEOUT_MS) {
    const resolved = timeoutMs ?? defaultTimeoutMs;
    if (!Number.isFinite(resolved) || resolved <= 0) {
        throw new Error(`${requestLabel} timeout must be a positive finite number`);
    }
    return Math.floor(resolved);
}
export function resolveSearchHttpFetcher(fetcher, requestLabel) {
    if (fetcher) {
        return fetcher;
    }
    if (typeof globalThis.fetch === "function") {
        return globalThis.fetch.bind(globalThis);
    }
    throw new Error(`${requestLabel} execution requires a fetch implementation`);
}
export async function readFailedSearchHttpResponseBody(response) {
    try {
        return (await response.text()).replaceAll(/\s+/g, " ").trim();
    }
    catch (error) {
        return `(failed to read response body: ${toErrorMessage(error)})`;
    }
}
export async function executeSearchJsonRequest(options) {
    const controller = new AbortController();
    const timeoutHandle = setTimeout(() => {
        controller.abort();
    }, options.timeoutMs);
    try {
        let response;
        try {
            response = await options.fetcher(options.endpoint, {
                method: options.method,
                headers: options.headers,
                body: JSON.stringify(options.body),
                signal: controller.signal,
            });
        }
        catch (error) {
            if (error instanceof Error && error.name === "AbortError") {
                throw new Error(`${options.requestLabel} timed out after ${options.timeoutMs}ms`);
            }
            throw new Error(`${options.requestLabel} execution failed: ${toErrorMessage(error)}`);
        }
        if (!response.ok) {
            const responseBody = await readFailedSearchHttpResponseBody(response);
            const detail = responseBody.length > 0 ? `: ${responseBody}` : "";
            throw new Error(`${options.requestLabel} failed with status ${response.status} ${response.statusText}${detail}`);
        }
        try {
            return await response.json();
        }
        catch (error) {
            throw new Error(`${options.responseLabel} JSON parse failed: ${toErrorMessage(error)}`);
        }
    }
    finally {
        clearTimeout(timeoutHandle);
    }
}
