// @unbrained/pm-cli/dist/core/search/providers.js
import { executeSearchJsonRequest, normalizeSearchHttpTimeoutMs, resolveSearchHttpFetcher, } from "./http-client.js";
import { isFiniteNumberArray, toNonEmptyString, trimTrailingSlashes, } from "../shared/primitives.js";
function normalizeOpenAiEmbeddingsEndpoint(baseUrl) {
    const normalizedBaseUrl = trimTrailingSlashes(baseUrl);
    const normalizedLower = normalizedBaseUrl.toLowerCase();
    if (normalizedLower.endsWith("/embeddings")) {
        return normalizedBaseUrl;
    }
    if (normalizedLower.endsWith("/v1")) {
        return `${normalizedBaseUrl}/embeddings`;
    }
    return `${normalizedBaseUrl}/v1/embeddings`;
}
function resolveOpenAiProvider(settings) {
    const baseUrl = toNonEmptyString(settings.providers?.openai?.base_url);
    const model = toNonEmptyString(settings.providers?.openai?.model);
    if (!baseUrl || !model) {
        return null;
    }
    const apiKey = toNonEmptyString(settings.providers?.openai?.api_key);
    return {
        name: "openai",
        base_url: baseUrl,
        model,
        ...(apiKey ? { api_key: apiKey } : {}),
    };
}
function resolveOllamaProvider(settings) {
    const baseUrl = toNonEmptyString(settings.providers?.ollama?.base_url);
    const model = toNonEmptyString(settings.providers?.ollama?.model);
    if (!baseUrl || !model) {
        return null;
    }
    return {
        name: "ollama",
        base_url: baseUrl,
        model,
    };
}
function normalizeEmbeddingInputs(input) {
    const inputs = (Array.isArray(input) ? input : [input])
        .map((entry) => entry.trim())
        .filter((entry) => entry.length > 0);
    if (inputs.length === 0) {
        throw new Error("Embedding input must include at least one non-empty string");
    }
    return inputs;
}
function dedupeEmbeddingInputs(inputs) {
    const uniqueInputs = [];
    const originalToUniqueIndex = [];
    const uniqueIndexByInput = new Map();
    for (const entry of inputs) {
        const existingUniqueIndex = uniqueIndexByInput.get(entry);
        if (existingUniqueIndex === undefined) {
            const nextUniqueIndex = uniqueInputs.length;
            uniqueInputs.push(entry);
            uniqueIndexByInput.set(entry, nextUniqueIndex);
            originalToUniqueIndex.push(nextUniqueIndex);
            continue;
        }
        originalToUniqueIndex.push(existingUniqueIndex);
    }
    return {
        uniqueInputs,
        originalToUniqueIndex,
    };
}
function buildOrderedOpenAiEntries(data) {
    const openAiEntries = data.map((entry, position) => ({
        embedding: entry.embedding,
        index: entry.index,
        position,
    }));
    const hasExplicitIndex = openAiEntries.some((entry) => entry.index !== undefined);
    if (hasExplicitIndex) {
        for (const entry of openAiEntries) {
            if (!Number.isInteger(entry.index)) {
                throw new TypeError(`OpenAI embedding response entry at position ${entry.position} is missing a valid integer index`);
            }
        }
        return [...openAiEntries].sort((a, b) => {
            const byIndex = a.index - b.index;
            if (byIndex !== 0) {
                return byIndex;
            }
            return a.position - b.position;
        });
    }
    return openAiEntries;
}
export function resolveEmbeddingProviders(settings) {
    const openAi = resolveOpenAiProvider(settings);
    const ollama = resolveOllamaProvider(settings);
    const available = [openAi, ollama].filter((entry) => entry !== null);
    return {
        active: available[0] ?? null,
        available,
    };
}
export function resolveEmbeddingRequestTarget(provider) {
    const baseUrl = trimTrailingSlashes(provider.base_url);
    if (provider.name === "openai") {
        return {
            provider: "openai",
            endpoint: normalizeOpenAiEmbeddingsEndpoint(baseUrl),
            model: provider.model,
        };
    }
    return {
        provider: "ollama",
        endpoint: `${baseUrl}/api/embed`,
        model: provider.model,
    };
}
export function buildEmbeddingRequestPlan(provider, input) {
    const normalizedInputs = normalizeEmbeddingInputs(input);
    if (provider.name === "openai") {
        return {
            target: resolveEmbeddingRequestTarget(provider),
            method: "POST",
            headers: {
                "content-type": "application/json",
                ...(provider.api_key ? { authorization: `Bearer ${provider.api_key}` } : {}),
            },
            body: {
                model: provider.model,
                input: normalizedInputs.length === 1 ? normalizedInputs[0] : normalizedInputs,
            },
        };
    }
    return {
        target: resolveEmbeddingRequestTarget(provider),
        method: "POST",
        headers: {
            "content-type": "application/json",
        },
        body: {
            model: provider.model,
            input: normalizedInputs,
        },
    };
}
export function normalizeEmbeddingResponse(provider, response) {
    if (provider.name === "openai") {
        const data = response.data;
        if (!Array.isArray(data) || data.length === 0) {
            throw new Error("OpenAI embedding response must include a non-empty data array");
        }
        const orderedEntries = buildOrderedOpenAiEntries(data);
        return orderedEntries.map((entry, index) => {
            const embedding = entry.embedding;
            if (!isFiniteNumberArray(embedding)) {
                throw new Error(`OpenAI embedding response entry at index ${index} is missing a numeric embedding vector`);
            }
            return [...embedding];
        });
    }
    const singleEmbedding = response.embedding;
    if (isFiniteNumberArray(singleEmbedding)) {
        return [[...singleEmbedding]];
    }
    const multipleEmbeddings = response.embeddings;
    if (Array.isArray(multipleEmbeddings) && multipleEmbeddings.length > 0) {
        return multipleEmbeddings.map((entry, index) => {
            if (!isFiniteNumberArray(entry)) {
                throw new Error(`Ollama embedding response entry at index ${index} is missing a numeric embedding vector`);
            }
            return [...entry];
        });
    }
    throw new Error("Ollama embedding response must include embedding or embeddings vectors");
}
export async function executeEmbeddingRequest(provider, input, options = {}) {
    const timeoutMs = normalizeSearchHttpTimeoutMs(options.timeout_ms, "Embedding request");
    const fetcher = resolveSearchHttpFetcher(options.fetcher, "Embedding request");
    const normalizedInputs = normalizeEmbeddingInputs(input);
    const dedupedInputs = dedupeEmbeddingInputs(normalizedInputs);
    const requestPlan = buildEmbeddingRequestPlan(provider, dedupedInputs.uniqueInputs);
    const payload = await executeSearchJsonRequest({
        endpoint: requestPlan.target.endpoint,
        method: requestPlan.method,
        headers: requestPlan.headers,
        body: requestPlan.body,
        timeoutMs,
        fetcher,
        requestLabel: "Embedding request",
        responseLabel: "Embedding response",
    });
    const vectors = normalizeEmbeddingResponse(provider, payload);
    if (vectors.length !== dedupedInputs.uniqueInputs.length) {
        throw new Error(`Embedding response cardinality mismatch: expected ${dedupedInputs.uniqueInputs.length} vector(s), received ${vectors.length}`);
    }
    return dedupedInputs.originalToUniqueIndex.map((uniqueIndex) => [...vectors[uniqueIndex]]);
}
