// @unbrained/pm-cli/dist/core/search/vector-stores.js
import { mkdir, readFile, rename, unlink, writeFile } from "node:fs/promises";
import { basename, dirname, join, resolve } from "node:path";
import { executeSearchJsonRequest, normalizeSearchHttpTimeoutMs, resolveSearchHttpFetcher, } from "./http-client.js";
import { isFiniteNumberArray, toErrorMessage, toNonEmptyString, trimTrailingSlashes, } from "../shared/primitives.js";
const DEFAULT_COLLECTION = "pm_items";
const LANCE_DB_LOCAL_SNAPSHOT_DIR = ".pm-cli-local-vectors";
const LANCE_DB_LOCAL_SNAPSHOT_VERSION = 1;
const lanceDbLocalTables = new Map();
function normalizeVector(value) {
    if (!isFiniteNumberArray(value) || value.length === 0) {
        throw new Error("Vector values must be a non-empty numeric array");
    }
    return [...value];
}
function normalizeLimit(value) {
    if (!Number.isFinite(value) || value <= 0) {
        throw new Error("Vector query limit must be a positive number");
    }
    return Math.floor(value);
}
function normalizeQdrantQueryResponse(payload) {
    const result = payload.result;
    if (!Array.isArray(result)) {
        throw new TypeError("Qdrant query response must include a result array");
    }
    const hits = result.map((entry, index) => {
        const idValue = entry.id;
        const idCandidate = typeof idValue === "number" ? String(idValue) : idValue;
        const id = toNonEmptyString(idCandidate);
        if (!id) {
            throw new Error(`Qdrant query response entry at index ${index} is missing a non-empty id`);
        }
        const score = entry.score;
        if (typeof score !== "number" || !Number.isFinite(score)) {
            throw new TypeError(`Qdrant query response entry at index ${index} is missing a finite numeric score`);
        }
        const payloadValue = entry.payload;
        if (payloadValue !== undefined &&
            (typeof payloadValue !== "object" || payloadValue === null || Array.isArray(payloadValue))) {
            throw new Error(`Qdrant query response entry at index ${index} must provide payload as an object when set`);
        }
        return {
            id,
            score,
            ...(payloadValue ? { payload: payloadValue } : {}),
        };
    });
    hits.sort((left, right) => {
        if (left.score !== right.score) {
            return right.score - left.score;
        }
        return left.id.localeCompare(right.id);
    });
    return hits;
}
function normalizeQdrantUpsertResponse(payload) {
    const nestedStatus = toNonEmptyString(payload.result?.status);
    if (nestedStatus) {
        return { status: nestedStatus };
    }
    const topLevelStatus = toNonEmptyString(payload.status);
    if (topLevelStatus) {
        return { status: topLevelStatus };
    }
    throw new Error("Qdrant upsert response must include status metadata");
}
async function executeRemoteVectorPlan(endpoint, plan, timeoutMs, fetcher, requestKind) {
    return await executeSearchJsonRequest({
        endpoint,
        method: plan.method,
        headers: plan.headers,
        body: plan.body,
        timeoutMs,
        fetcher,
        requestLabel: `Vector ${requestKind} request`,
        responseLabel: `Vector ${requestKind} response`,
    });
}
function resolveQdrantStore(settings) {
    const url = toNonEmptyString(settings.vector_store?.qdrant?.url);
    if (!url) {
        return null;
    }
    const apiKey = toNonEmptyString(settings.vector_store?.qdrant?.api_key);
    return {
        name: "qdrant",
        url,
        ...(apiKey ? { api_key: apiKey } : {}),
    };
}
function resolveLanceDbStore(settings) {
    const lancedbPath = toNonEmptyString(settings.vector_store?.lancedb?.path);
    if (!lancedbPath) {
        return null;
    }
    return {
        name: "lancedb",
        path: lancedbPath,
    };
}
function normalizeVectorRecords(records) {
    if (!Array.isArray(records) || records.length === 0) {
        throw new Error("Vector upsert records must include at least one entry");
    }
    return records.map((record, index) => {
        const id = toNonEmptyString(record.id);
        if (!id) {
            throw new Error(`Vector upsert record at index ${index} is missing a non-empty id`);
        }
        const vector = normalizeVector(record.vector);
        const payload = record.payload;
        if (payload !== undefined && (typeof payload !== "object" || payload === null || Array.isArray(payload))) {
            throw new Error(`Vector upsert record at index ${index} must provide payload as an object when set`);
        }
        return {
            id,
            vector,
            ...(payload ? { payload } : {}),
        };
    });
}
function normalizeVectorDeleteIds(ids) {
    if (!Array.isArray(ids) || ids.length === 0) {
        throw new Error("Vector delete ids must include at least one entry");
    }
    const normalizedIds = ids
        .map((id, index) => {
        const normalized = toNonEmptyString(id);
        if (!normalized) {
            throw new Error(`Vector delete id at index ${index} is missing a non-empty value`);
        }
        return normalized;
    })
        .sort((left, right) => left.localeCompare(right));
    const uniqueIds = [];
    for (const id of normalizedIds) {
        if (uniqueIds.at(-1) !== id) {
            uniqueIds.push(id);
        }
    }
    return uniqueIds;
}
function resolveQdrantDeleteTarget(upsertTarget) {
    return upsertTarget.replace(/\/points\?wait=true$/, "/points/delete?wait=true");
}
function getLanceDbLocalTableKey(storePath, table) {
    return getLanceDbSnapshotPath(storePath, table);
}
function getLanceDbSnapshotPath(storePath, table) {
    return join(resolve(storePath), LANCE_DB_LOCAL_SNAPSHOT_DIR, `${table}.json`);
}
function isNodeErrorWithCode(error, code) {
    return typeof error === "object" && error !== null && error.code === code;
}
function normalizeSnapshotRecord(entry, index, snapshotPath) {
    const id = toNonEmptyString(entry.id);
    if (!id) {
        throw new Error(`LanceDB local snapshot '${snapshotPath}' record at index ${index} is missing a non-empty id`);
    }
    const vector = normalizeVector(entry.vector);
    const payload = entry.payload;
    if (payload !== undefined && (typeof payload !== "object" || payload === null || Array.isArray(payload))) {
        throw new Error(`LanceDB local snapshot '${snapshotPath}' record '${id}' must provide payload as an object when set`);
    }
    return {
        id,
        vector,
        ...(payload ? { payload: payload } : {}),
    };
}
function parseLanceDbSnapshot(snapshotPath, expectedTable, raw) {
    let parsed;
    try {
        parsed = JSON.parse(raw);
    }
    catch (error) {
        throw new Error(`LanceDB local snapshot at '${snapshotPath}' is not valid JSON: ${toErrorMessage(error)}`);
    }
    if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) {
        throw new Error(`LanceDB local snapshot at '${snapshotPath}' must be a JSON object`);
    }
    const version = parsed.version;
    if (version !== LANCE_DB_LOCAL_SNAPSHOT_VERSION) {
        throw new Error(`LanceDB local snapshot at '${snapshotPath}' must include version=${LANCE_DB_LOCAL_SNAPSHOT_VERSION}`);
    }
    const table = toNonEmptyString(parsed.table);
    if (!table) {
        throw new Error(`LanceDB local snapshot at '${snapshotPath}' must include a non-empty table value`);
    }
    if (table !== expectedTable) {
        throw new Error(`LanceDB local snapshot at '${snapshotPath}' table mismatch: expected '${expectedTable}', received '${table}'`);
    }
    const recordsValue = parsed.records;
    if (!Array.isArray(recordsValue)) {
        throw new TypeError(`LanceDB local snapshot at '${snapshotPath}' must include a records array`);
    }
    const tableRecords = new Map();
    for (let index = 0; index < recordsValue.length; index += 1) {
        const record = normalizeSnapshotRecord(recordsValue[index], index, snapshotPath);
        tableRecords.set(record.id, record);
    }
    return tableRecords;
}
async function loadLanceDbLocalTable(storePath, table) {
    const key = getLanceDbLocalTableKey(storePath, table);
    const cached = lanceDbLocalTables.get(key);
    if (cached) {
        return cached;
    }
    const snapshotPath = getLanceDbSnapshotPath(storePath, table);
    let loaded = new Map();
    try {
        const raw = await readFile(snapshotPath, "utf8");
        loaded = parseLanceDbSnapshot(snapshotPath, table, raw);
    }
    catch (error) {
        if (!isNodeErrorWithCode(error, "ENOENT")) {
            throw error;
        }
    }
    lanceDbLocalTables.set(key, loaded);
    return loaded;
}
function buildSnapshotRecords(table) {
    const records = [...table.values()];
    records.sort((left, right) => left.id.localeCompare(right.id));
    return records.map((record) => ({
        id: record.id,
        vector: [...record.vector],
        ...(record.payload ? { payload: record.payload } : {}),
    }));
}
async function removeSnapshotFile(snapshotPath) {
    try {
        await unlink(snapshotPath);
    }
    catch (error) {
        if (!isNodeErrorWithCode(error, "ENOENT")) {
            throw new Error(`LanceDB local snapshot delete failed at '${snapshotPath}': ${toErrorMessage(error)}`);
        }
    }
}
async function persistLanceDbLocalTable(storePath, tableName, table) {
    const snapshotPath = getLanceDbSnapshotPath(storePath, tableName);
    if (table.size === 0) {
        await removeSnapshotFile(snapshotPath);
        return;
    }
    const snapshotDir = dirname(snapshotPath);
    try {
        await mkdir(snapshotDir, { recursive: true });
    }
    catch (error) {
        throw new Error(`LanceDB local snapshot directory create failed at '${snapshotDir}': ${toErrorMessage(error)}`);
    }
    const tempPath = join(snapshotDir, `${basename(snapshotPath)}.tmp-${process.pid}-${Date.now()}-${Math.random().toString(16).slice(2)}`);
    const serialized = `${JSON.stringify({
        version: LANCE_DB_LOCAL_SNAPSHOT_VERSION,
        table: tableName,
        records: buildSnapshotRecords(table),
    }, null, 2)}\n`;
    try {
        await writeFile(tempPath, serialized, "utf8");
        await rename(tempPath, snapshotPath);
    }
    catch (error) {
        await unlink(tempPath).catch(() => { });
        throw new Error(`LanceDB local snapshot write failed at '${snapshotPath}': ${toErrorMessage(error)}`);
    }
}
function l2Norm(vector) {
    let sumSq = 0;
    for (let index = 0; index < vector.length; index += 1) {
        sumSq += vector[index] * vector[index];
    }
    return Math.sqrt(sumSq);
}
function cosineSimilarity(left, right) {
    let dotProd = 0;
    for (let index = 0; index < left.length; index += 1) {
        dotProd += left[index] * right[index];
    }
    const leftNorm = l2Norm(left);
    const rightNorm = l2Norm(right);
    if (leftNorm === 0 || rightNorm === 0) {
        return 0;
    }
    return dotProd / (leftNorm * rightNorm);
}
export function resolveVectorStores(settings) {
    const qdrant = resolveQdrantStore(settings);
    const lancedb = resolveLanceDbStore(settings);
    const available = [qdrant, lancedb].filter((entry) => entry !== null);
    return {
        active: available[0] ?? null,
        available,
    };
}
export function resolveVectorStoreRequestTarget(store) {
    if (store.name === "qdrant") {
        const baseUrl = trimTrailingSlashes(store.url);
        return {
            store: "qdrant",
            query_target: `${baseUrl}/collections/${DEFAULT_COLLECTION}/points/search`,
            upsert_target: `${baseUrl}/collections/${DEFAULT_COLLECTION}/points?wait=true`,
        };
    }
    const encodedPath = encodeURIComponent(store.path);
    return {
        store: "lancedb",
        query_target: `lancedb://${encodedPath}#${DEFAULT_COLLECTION}`,
        upsert_target: `lancedb://${encodedPath}#${DEFAULT_COLLECTION}`,
    };
}
export function buildVectorQueryPlan(store, vector, limit) {
    const target = resolveVectorStoreRequestTarget(store);
    const normalizedVector = normalizeVector(vector);
    const normalizedLimit = normalizeLimit(limit);
    if (store.name === "qdrant") {
        return {
            target,
            method: "POST",
            headers: {
                "content-type": "application/json",
                ...(store.api_key ? { "api-key": store.api_key } : {}),
            },
            body: {
                vector: normalizedVector,
                limit: normalizedLimit,
                with_payload: true,
            },
        };
    }
    return {
        target,
        method: "LOCAL",
        headers: {},
        body: {
            table: DEFAULT_COLLECTION,
            vector: normalizedVector,
            limit: normalizedLimit,
        },
    };
}
export function buildVectorUpsertPlan(store, records) {
    const target = resolveVectorStoreRequestTarget(store);
    const normalizedRecords = normalizeVectorRecords(records);
    if (store.name === "qdrant") {
        return {
            target,
            method: "POST",
            headers: {
                "content-type": "application/json",
                ...(store.api_key ? { "api-key": store.api_key } : {}),
            },
            body: {
                points: normalizedRecords,
            },
        };
    }
    return {
        target,
        method: "LOCAL",
        headers: {},
        body: {
            table: DEFAULT_COLLECTION,
            records: normalizedRecords,
        },
    };
}
export function buildVectorDeletePlan(store, ids) {
    const target = resolveVectorStoreRequestTarget(store);
    const normalizedIds = normalizeVectorDeleteIds(ids);
    if (store.name === "qdrant") {
        return {
            target,
            method: "POST",
            headers: {
                "content-type": "application/json",
                ...(store.api_key ? { "api-key": store.api_key } : {}),
            },
            body: {
                points: normalizedIds,
            },
        };
    }
    return {
        target,
        method: "LOCAL",
        headers: {},
        body: {
            table: DEFAULT_COLLECTION,
            ids: normalizedIds,
        },
    };
}
export async function executeVectorQuery(store, vector, limit, options = {}) {
    const plan = buildVectorQueryPlan(store, vector, limit);
    if (plan.method === "LOCAL") {
        const lanceDbStore = store;
        const queryBody = plan.body;
        const table = await loadLanceDbLocalTable(lanceDbStore.path, queryBody.table);
        if (table.size === 0) {
            return [];
        }
        const queryVector = normalizeVector(queryBody.vector);
        const queryLimit = normalizeLimit(queryBody.limit);
        const hits = [];
        let dimensionMismatchCount = 0;
        for (const record of table.values()) {
            if (record.vector.length !== queryVector.length) {
                dimensionMismatchCount++;
                continue;
            }
            hits.push({
                id: record.id,
                score: cosineSimilarity(queryVector, record.vector),
                ...(record.payload ? { payload: record.payload } : {}),
            });
        }
        if (dimensionMismatchCount > 0 && options.warnings) {
            options.warnings.push(`vector_dimension_mismatch:${dimensionMismatchCount} records skipped (expected ${queryVector.length} dimensions)`);
        }
        hits.sort((left, right) => {
            if (left.score !== right.score) {
                return right.score - left.score;
            }
            return left.id.localeCompare(right.id);
        });
        return hits.slice(0, queryLimit);
    }
    const timeoutMs = normalizeSearchHttpTimeoutMs(options.timeout_ms, "Vector request");
    const fetcher = resolveSearchHttpFetcher(options.fetcher, "Vector request");
    const payload = await executeRemoteVectorPlan(plan.target.query_target, {
        method: "POST",
        headers: plan.headers,
        body: plan.body,
    }, timeoutMs, fetcher, "query");
    return normalizeQdrantQueryResponse(payload);
}
export async function executeVectorUpsert(store, records, options = {}) {
    const plan = buildVectorUpsertPlan(store, records);
    if (plan.method === "LOCAL") {
        const lanceDbStore = store;
        const upsertBody = plan.body;
        const key = getLanceDbLocalTableKey(lanceDbStore.path, upsertBody.table);
        const table = await loadLanceDbLocalTable(lanceDbStore.path, upsertBody.table);
        for (const record of upsertBody.records) {
            table.set(record.id, record);
        }
        await persistLanceDbLocalTable(lanceDbStore.path, upsertBody.table, table);
        lanceDbLocalTables.set(key, table);
        return { status: "ok" };
    }
    const timeoutMs = normalizeSearchHttpTimeoutMs(options.timeout_ms, "Vector request");
    const fetcher = resolveSearchHttpFetcher(options.fetcher, "Vector request");
    const payload = await executeRemoteVectorPlan(plan.target.upsert_target, {
        method: "POST",
        headers: plan.headers,
        body: plan.body,
    }, timeoutMs, fetcher, "upsert");
    return normalizeQdrantUpsertResponse(payload);
}
export async function executeVectorDelete(store, ids, options = {}) {
    const plan = buildVectorDeletePlan(store, ids);
    if (plan.method === "LOCAL") {
        const lanceDbStore = store;
        const deleteBody = plan.body;
        const key = getLanceDbLocalTableKey(lanceDbStore.path, deleteBody.table);
        const table = await loadLanceDbLocalTable(lanceDbStore.path, deleteBody.table);
        if (table.size === 0) {
            return { status: "ok" };
        }
        for (const id of deleteBody.ids) {
            table.delete(id);
        }
        await persistLanceDbLocalTable(lanceDbStore.path, deleteBody.table, table);
        if (table.size === 0) {
            lanceDbLocalTables.delete(key);
        }
        else {
            lanceDbLocalTables.set(key, table);
        }
        return { status: "ok" };
    }
    const timeoutMs = normalizeSearchHttpTimeoutMs(options.timeout_ms, "Vector request");
    const fetcher = resolveSearchHttpFetcher(options.fetcher, "Vector request");
    const payload = await executeRemoteVectorPlan(resolveQdrantDeleteTarget(plan.target.upsert_target), {
        method: "POST",
        headers: plan.headers,
        body: plan.body,
    }, timeoutMs, fetcher, "delete");
    return normalizeQdrantUpsertResponse(payload);
}
