// @unbrained/pm-cli/dist/core/telemetry/observability.js
import { EXIT_CODE } from "../shared/constants.js";
const SETUP_ROOT_COMMANDS = new Set(["init", "config", "completion", "completion-tags"]);
const QUERY_ROOT_COMMANDS = new Set([
    "list",
    "list-all",
    "list-open",
    "list-in-progress",
    "list-blocked",
    "list-closed",
    "list-canceled",
    "list-draft",
    "search",
    "get",
    "context",
    "calendar",
    "history",
    "activity",
    "aggregate",
    "deps",
    "contracts",
    "stats",
    "dedupe-audit",
]);
const MUTATION_ROOT_COMMANDS = new Set([
    "create",
    "update",
    "update-many",
    "append",
    "close",
    "delete",
    "restore",
    "claim",
    "release",
    "comments",
    "notes",
    "learnings",
    "files",
    "docs",
    "start-task",
    "pause-task",
    "close-task",
]);
const TESTING_ROOT_COMMANDS = new Set(["test", "test-all", "test-runs", "test-verify", "trace-test", "test-ping"]);
const DIAGNOSTICS_ROOT_COMMANDS = new Set(["health", "validate", "normalize", "reindex", "gc", "extension-doctor"]);
function normalizeCommandPath(commandPath) {
    return commandPath
        .trim()
        .replaceAll(/\s+/g, " ")
        .toLowerCase();
}
function normalizeErrorCode(errorCode) {
    const normalized = errorCode?.trim().toLowerCase();
    return normalized && normalized.length > 0 ? normalized : undefined;
}
export function deriveTelemetryCommandTaxonomy(commandPath) {
    const normalizedPath = normalizeCommandPath(commandPath);
    const tokens = normalizedPath.length > 0 ? normalizedPath.split(" ") : ["<unknown>"];
    const root = tokens[0] ?? "<unknown>";
    const leaf = tokens[tokens.length - 1] ?? root;
    let family = "other";
    if (SETUP_ROOT_COMMANDS.has(root)) {
        family = "setup";
    }
    else if (QUERY_ROOT_COMMANDS.has(root)) {
        family = "query";
    }
    else if (MUTATION_ROOT_COMMANDS.has(root)) {
        family = "mutation";
    }
    else if (TESTING_ROOT_COMMANDS.has(root)) {
        family = "testing";
    }
    else if (root === "extension") {
        family = "extension";
    }
    else if (DIAGNOSTICS_ROOT_COMMANDS.has(root)) {
        family = "diagnostics";
    }
    return {
        command_path: normalizedPath.length > 0 ? normalizedPath : "<unknown>",
        command_root: root,
        command_leaf: leaf,
        command_depth: tokens.length,
        command_family: family,
    };
}
export function inferTelemetryErrorCode(params) {
    if (params.ok) {
        return undefined;
    }
    const explicit = normalizeErrorCode(params.errorCode);
    if (explicit) {
        return explicit;
    }
    const message = (params.errorMessage ?? "").trim().toLowerCase();
    if (message.includes("unknown command")) {
        return "unknown_command";
    }
    if (message.includes("unknown option")) {
        return "unknown_option";
    }
    if (message.includes("missing required options") || message.includes("missing required option")) {
        return "missing_required_option";
    }
    if (message.includes("missing required argument")) {
        return "missing_required_argument";
    }
    if (message.includes("no update flags provided")) {
        return "no_update_fields";
    }
    if (message.includes("is assigned to") && message.includes("use --force")) {
        return "ownership_conflict";
    }
    if (message.includes("is locked")) {
        return "lock_conflict";
    }
    if (message.includes("already terminal") && message.includes("use --force")) {
        return "terminal_state_conflict";
    }
    if (message.includes("tracker is not initialized")) {
        return "tracker_not_initialized";
    }
    if (message.includes(" not found")) {
        return "item_not_found";
    }
    if (message.includes("use \"pm close <id> <text>\" to close an item") ||
        (message.includes("invalid --status value") && message.includes("\"closed\""))) {
        return "close_through_update";
    }
    if (message.startsWith("invalid ") || message.includes(" must be ")) {
        return "invalid_argument_value";
    }
    if (message.includes("strict create mode requires concrete values for --")) {
        return "missing_required_option";
    }
    if (message.includes("either as positional") && message.includes("not both")) {
        return "invalid_command_usage";
    }
    const exitCode = Number.isFinite(params.exitCode) ? Math.max(0, Math.trunc(params.exitCode ?? 0)) : undefined;
    if (exitCode === EXIT_CODE.USAGE) {
        return "invalid_command_usage";
    }
    if (exitCode === EXIT_CODE.NOT_FOUND) {
        return "item_not_found";
    }
    if (exitCode === EXIT_CODE.CONFLICT) {
        return "lock_conflict";
    }
    if (exitCode === EXIT_CODE.DEPENDENCY_FAILED) {
        return "dependency_failed";
    }
    return "command_failed";
}
export function deriveTelemetryCommandResolution(params) {
    if (params.ok) {
        return "success";
    }
    const normalizedCode = normalizeErrorCode(params.errorCode);
    if (normalizedCode === "unknown_command") {
        return "nonexistent_command";
    }
    if (normalizedCode === "unknown_option") {
        return "invalid_option";
    }
    if (normalizedCode === "missing_required_option") {
        return "missing_required_option";
    }
    if (normalizedCode === "missing_required_argument") {
        return "missing_required_argument";
    }
    if (params.errorCategory === "usage") {
        return "invalid_usage";
    }
    if (params.errorCategory === "validation") {
        return "validation_failed";
    }
    if (params.errorCategory === "conflict") {
        return "conflict";
    }
    if (params.errorCategory === "runtime") {
        return "runtime_failed";
    }
    return "unknown_failed";
}
