// @unbrained/pm-cli/dist/core/shared/time.js
import { PmCliError } from "./errors.js";
import { EXIT_CODE } from "./constants.js";
const RELATIVE_DEADLINE = /^([+-]?)(\d+)([hdwm])$/i;
const COMPOUND_RELATIVE_DEADLINE = /^[+-]?\d+[hdwm](?:[+-]\d+[hdwm])+$/i;
const COMPACT_DATE = /^(\d{4})(\d{2})(\d{2})$/;
const COMPACT_DATETIME = /^(\d{4})(\d{2})(\d{2})(?:[T\s]?)(\d{2})(\d{2})(\d{2})?([.,]\d{1,3})?(Z|[+-]\d{2}:?\d{2})?$/i;
const HYPHEN_TIME = /^(\d{4}-\d{2}-\d{2})[T\s](\d{2})-(\d{2})(?:-(\d{2}))?([.,]\d{1,3})?(Z|[+-]\d{2}:?\d{2})?$/i;
const COMPACT_TIME = /^(\d{4}-\d{2}-\d{2})[T\s](\d{2})(\d{2})(\d{2})?([.,]\d{1,3})?(Z|[+-]\d{2}:?\d{2})?$/i;
export function nowIso() {
    return new Date().toISOString();
}
export function isTimestampLiteral(input) {
    return Number.isFinite(Date.parse(input));
}
export function compareTimestampStrings(left, right) {
    const leftMs = Date.parse(left);
    const rightMs = Date.parse(right);
    if (Number.isFinite(leftMs) && Number.isFinite(rightMs) && leftMs !== rightMs) {
        return leftMs - rightMs;
    }
    return left.localeCompare(right);
}
function normalizeFraction(raw) {
    if (!raw)
        return "";
    const digits = raw.slice(1);
    return `.${digits.padEnd(3, "0").slice(0, 3)}`;
}
function normalizeOffset(raw) {
    if (!raw)
        return "";
    if (raw.toUpperCase() === "Z")
        return "Z";
    const compact = /^([+-]\d{2})(\d{2})$/.exec(raw);
    if (compact)
        return `${compact[1]}:${compact[2]}`;
    return raw;
}
function normalizeTimestampCandidates(input) {
    const candidates = [];
    const push = (value) => {
        if (!value || value === input || candidates.includes(value))
            return;
        candidates.push(value);
    };
    const compactDate = COMPACT_DATE.exec(input);
    if (compactDate) {
        const [, year, month, day] = compactDate;
        push(`${year}-${month}-${day}`);
    }
    const compactDateTime = COMPACT_DATETIME.exec(input);
    if (compactDateTime) {
        const [, year, month, day, hour, minute, secondRaw, fractionRaw, offsetRaw] = compactDateTime;
        const second = secondRaw ? `:${secondRaw}` : "";
        push(`${year}-${month}-${day}T${hour}:${minute}${second}${normalizeFraction(fractionRaw)}${normalizeOffset(offsetRaw)}`);
    }
    const hyphenTime = HYPHEN_TIME.exec(input);
    if (hyphenTime) {
        const [, datePart, hour, minute, secondRaw, fractionRaw, offsetRaw] = hyphenTime;
        const second = secondRaw ? `:${secondRaw}` : "";
        push(`${datePart}T${hour}:${minute}${second}${normalizeFraction(fractionRaw)}${normalizeOffset(offsetRaw)}`);
    }
    const compactTime = COMPACT_TIME.exec(input);
    if (compactTime) {
        const [, datePart, hour, minute, secondRaw, fractionRaw, offsetRaw] = compactTime;
        const second = secondRaw ? `:${secondRaw}` : "";
        push(`${datePart}T${hour}:${minute}${second}${normalizeFraction(fractionRaw)}${normalizeOffset(offsetRaw)}`);
    }
    const spaceDateTime = /^(\d{4}-\d{2}-\d{2})\s+(.+)$/.exec(input);
    if (spaceDateTime) {
        const [, datePart, timePart] = spaceDateTime;
        push(`${datePart}T${timePart}`);
    }
    return candidates;
}
function parseTimestampWithFallbacks(input) {
    const direct = Date.parse(input);
    if (Number.isFinite(direct))
        return direct;
    for (const candidate of normalizeTimestampCandidates(input)) {
        const parsed = Date.parse(candidate);
        if (Number.isFinite(parsed))
            return parsed;
    }
    return Number.NaN;
}
function daysInUtcMonth(year, monthIndex) {
    return new Date(Date.UTC(year, monthIndex + 1, 0)).getUTCDate();
}
function addUtcMonths(now, amount) {
    const result = new Date(now.getTime());
    const startDay = result.getUTCDate();
    const targetMonthIndex = result.getUTCMonth() + amount;
    const targetYear = result.getUTCFullYear() + Math.floor(targetMonthIndex / 12);
    const targetMonth = ((targetMonthIndex % 12) + 12) % 12;
    const day = Math.min(startDay, daysInUtcMonth(targetYear, targetMonth));
    result.setUTCFullYear(targetYear, targetMonth, day);
    return result;
}
export function resolveIsoOrRelative(input, now = new Date(), fieldLabel = "deadline") {
    const trimmed = input.trim();
    if (trimmed.toLowerCase() === "now") {
        return now.toISOString();
    }
    const relative = RELATIVE_DEADLINE.exec(trimmed);
    if (relative) {
        const sign = relative[1] === "-" ? -1 : 1;
        const amount = Number.parseInt(relative[2], 10) * sign;
        const unit = relative[3].toLowerCase();
        if (unit === "m") {
            return addUtcMonths(now, amount).toISOString();
        }
        const msPerUnit = unit === "h" ? 60 * 60 * 1000 : unit === "d" ? 24 * 60 * 60 * 1000 : 7 * 24 * 60 * 60 * 1000;
        return new Date(now.getTime() + amount * msPerUnit).toISOString();
    }
    const timestamp = parseTimestampWithFallbacks(trimmed);
    if (!Number.isFinite(timestamp)) {
        const normalizedLabel = fieldLabel.trim().length > 0 ? fieldLabel.trim() : "deadline";
        const guidance = COMPOUND_RELATIVE_DEADLINE.test(trimmed)
            ? "Compound relative expressions like +3d+1h are not supported; use a single relative token (for example +3d) or an ISO/date string."
            : 'Use ISO/date string input, "now", or relative +6h/-6h/+1d/-1d/+2w/-2w/+6m/-6m.';
        throw new PmCliError(`Invalid ${normalizedLabel} value "${input}". ${guidance}`, EXIT_CODE.USAGE);
    }
    return new Date(timestamp).toISOString();
}
