// @unbrained/pm-cli/dist/core/sentry/helpers.js
import { getSentry } from "./instrument.js";
import { PmCliError } from "../shared/errors.js";
import { deriveTelemetryCommandResolution, deriveTelemetryCommandTaxonomy, } from "../telemetry/observability.js";
let activeCommandSpan;
function setSpanAttribute(span, key, value) {
    if (value === undefined) {
        return;
    }
    const attributeTarget = span;
    if (typeof attributeTarget.setAttribute === "function") {
        attributeTarget.setAttribute(key, value);
    }
}
export function sentrySetCommandContext(command, args, options, metadata) {
    const Sentry = getSentry();
    if (!Sentry)
        return;
    const taxonomy = deriveTelemetryCommandTaxonomy(command);
    Sentry.setTag("pm.command", command);
    Sentry.setTag("pm.command_root", taxonomy.command_root);
    Sentry.setTag("pm.command_family", taxonomy.command_family);
    if (typeof metadata?.source_context === "string" && metadata.source_context.trim().length > 0) {
        Sentry.setTag("pm.source_context", metadata.source_context);
    }
    const safeArgs = args.map((arg) => arg.startsWith("--") ? arg.split("=")[0] : arg);
    Sentry.setContext("pm.command", {
        name: command,
        root: taxonomy.command_root,
        leaf: taxonomy.command_leaf,
        family: taxonomy.command_family,
        args: safeArgs,
        option_keys: Object.keys(options).sort(),
        source_context: metadata?.source_context,
        source_context_source: metadata?.source_context_source,
    });
    Sentry.addBreadcrumb({
        category: "pm.command",
        message: `pm ${command}`,
        level: "info",
        data: {
            args: safeArgs,
            command_root: taxonomy.command_root,
            command_family: taxonomy.command_family,
            source_context: metadata?.source_context,
        },
    });
}
export function sentryStartCommandSpan(command) {
    const Sentry = getSentry();
    if (!Sentry)
        return;
    activeCommandSpan = Sentry.startInactiveSpan({
        op: "pm.command",
        name: `pm ${command}`,
        forceTransaction: true,
    });
}
export function sentryFinishCommandSpan(ok, error, metadata) {
    if (!activeCommandSpan)
        return;
    const normalizedOk = ok ? "true" : "false";
    const normalizedExitCode = typeof metadata?.exit_code === "number" ? String(metadata.exit_code) : undefined;
    const Sentry = getSentry();
    if (Sentry) {
        Sentry.setTag("pm.ok", normalizedOk);
        Sentry.setTag("pm.command_ok", normalizedOk);
        if (typeof metadata?.exit_code === "number") {
            Sentry.setTag("pm.exit_code", String(metadata.exit_code));
            Sentry.setTag("pm.command_exit_code", String(metadata.exit_code));
        }
        if (typeof metadata?.error_code === "string" && metadata.error_code.trim().length > 0) {
            Sentry.setTag("pm.error_code", metadata.error_code);
        }
        if (typeof metadata?.error_category === "string" && metadata.error_category.trim().length > 0) {
            Sentry.setTag("pm.error_category", metadata.error_category);
        }
        if (typeof metadata?.command_resolution === "string" && metadata.command_resolution.trim().length > 0) {
            Sentry.setTag("pm.command_resolution", metadata.command_resolution);
        }
        if (typeof metadata?.resolution_stage === "string" && metadata.resolution_stage.trim().length > 0) {
            Sentry.setTag("pm.resolution_stage", metadata.resolution_stage);
        }
    }
    setSpanAttribute(activeCommandSpan, "pm.ok", normalizedOk);
    setSpanAttribute(activeCommandSpan, "pm.command_ok", normalizedOk);
    setSpanAttribute(activeCommandSpan, "pm.exit_code", normalizedExitCode);
    setSpanAttribute(activeCommandSpan, "pm.command_exit_code", normalizedExitCode);
    setSpanAttribute(activeCommandSpan, "pm.error_code", metadata?.error_code);
    setSpanAttribute(activeCommandSpan, "pm.error_category", metadata?.error_category);
    setSpanAttribute(activeCommandSpan, "pm.command_resolution", metadata?.command_resolution);
    setSpanAttribute(activeCommandSpan, "pm.resolution_stage", metadata?.resolution_stage);
    activeCommandSpan.setStatus(ok ? { code: 1 } : { code: 2, message: error ?? "command_failed" });
    activeCommandSpan.end();
    activeCommandSpan = undefined;
}
export function sentryCaptureCliError(error) {
    if (!shouldCaptureCliError(error))
        return;
    const Sentry = getSentry();
    if (!Sentry)
        return;
    if (error instanceof Error) {
        const extras = {};
        if ("exitCode" in error && typeof error.exitCode === "number") {
            extras.exit_code = error.exitCode;
        }
        if ("context" in error && typeof error.context === "object") {
            extras.error_context = error.context;
        }
        Sentry.captureException(error, { extra: extras });
    }
    else {
        Sentry.captureException(new Error(String(error)));
    }
}
export function sentryLogCliUsageError(params) {
    const Sentry = getSentry();
    if (!Sentry)
        return;
    const resolvedCommandResolution = params.command_resolution ??
        deriveTelemetryCommandResolution({
            ok: false,
            errorCode: params.error_code,
            errorCategory: params.error_category,
        });
    const resolvedResolutionStage = params.resolution_stage ?? "unknown";
    const payload = {
        "pm.command": params.command,
        "pm.error_code": params.error_code,
        "pm.error_category": params.error_category,
        "pm.exit_code": params.exit_code,
        "pm.error_message": params.error_message,
        "pm.command_resolution": resolvedCommandResolution,
        "pm.resolution_stage": resolvedResolutionStage,
        "pm.source_context": params.source_context ?? "",
    };
    const loggerCandidate = Sentry
        .logger;
    if (loggerCandidate && typeof loggerCandidate.warn === "function") {
        loggerCandidate.warn("pm_cli_usage_error", payload);
        return;
    }
    Sentry.captureMessage(`pm_cli_usage_error:${params.error_code}`, {
        level: "warning",
        tags: {
            "pm.command": params.command,
            "pm.error_code": params.error_code,
            "pm.error_category": params.error_category,
            "pm.exit_code": String(params.exit_code),
            "pm.command_resolution": resolvedCommandResolution,
            "pm.resolution_stage": resolvedResolutionStage,
            "pm.source_context": params.source_context ?? "unknown",
        },
        extra: payload,
    });
}
export function shouldCaptureCliError(error) {
    return !(error instanceof PmCliError);
}
export async function sentryFlush() {
    const Sentry = getSentry();
    if (!Sentry)
        return;
    try {
        await Sentry.flush(3000);
    }
    catch {
        // Sentry flush must never block CLI exit.
    }
}
