// @unbrained/pm-cli/dist/types.js
export const BUILTIN_ITEM_TYPE_VALUES = [
    "Epic",
    "Feature",
    "Task",
    "Chore",
    "Issue",
    "Decision",
    "Event",
    "Reminder",
    "Milestone",
    "Meeting",
];
export const ITEM_TYPE_VALUES = BUILTIN_ITEM_TYPE_VALUES;
export const STATUS_VALUES = [
    "draft",
    "open",
    "in_progress",
    "blocked",
    "closed",
    "canceled",
];
export const RUNTIME_STATUS_ROLE_VALUES = [
    "draft",
    "active",
    "blocked",
    "terminal",
    "terminal_done",
    "terminal_canceled",
    "default_open",
    "default_close",
    "default_cancel",
];
export const RUNTIME_FIELD_TYPE_VALUES = ["string", "number", "boolean", "string_array"];
export const RUNTIME_FIELD_COMMAND_VALUES = ["create", "update", "update_many", "list", "search", "calendar", "context"];
export const RUNTIME_UNKNOWN_FIELD_POLICY_VALUES = ["allow", "warn", "reject"];
export const DEPENDENCY_KIND_VALUES = [
    "blocks",
    "parent",
    "child",
    "parent_child",
    "child_of",
    "related",
    "related_to",
    "discovered_from",
    "blocked_by",
    "incident_from",
    "epic",
    "supersedes",
    "task",
];
export const SCOPE_VALUES = ["project", "global"];
export const RISK_VALUES = ["low", "medium", "high", "critical"];
export const ISSUE_SEVERITY_VALUES = ["low", "medium", "high", "critical"];
export const CONFIDENCE_TEXT_VALUES = ["low", "medium", "high"];
export const ITEM_FORMAT_VALUES = ["toon", "json_markdown"];
export const SPRINT_RELEASE_FORMAT_POLICY_VALUES = ["warn", "strict_error"];
export const PARENT_REFERENCE_POLICY_VALUES = ["warn", "strict_error"];
export const VALIDATE_METADATA_PROFILE_VALUES = ["core", "strict", "custom"];
export const GOVERNANCE_PRESET_VALUES = ["minimal", "default", "strict", "custom"];
export const GOVERNANCE_OWNERSHIP_ENFORCEMENT_VALUES = ["none", "warn", "strict"];
export const GOVERNANCE_CREATE_MODE_DEFAULT_VALUES = ["progressive", "strict"];
export const GOVERNANCE_CLOSE_VALIDATION_DEFAULT_VALUES = ["off", "warn", "strict"];
export const VALIDATE_METADATA_REQUIRED_FIELD_VALUES = [
    "author",
    "acceptance_criteria",
    "estimated_minutes",
    "close_reason",
    "reviewer",
    "risk",
    "confidence",
    "sprint",
    "release",
];
export const RECURRENCE_FREQUENCY_VALUES = ["daily", "weekly", "monthly", "yearly"];
export const RECURRENCE_WEEKDAY_VALUES = ["mon", "tue", "wed", "thu", "fri", "sat", "sun"];
export const CONTEXT_DEPTH_VALUES = ["brief", "standard", "deep"];
export const CONTEXT_SECTION_VALUES = [
    "hierarchy",
    "activity",
    "progress",
    "blockers",
    "files",
    "workload",
    "staleness",
    "tests",
];
