// @unbrained/pm-cli/dist/core/lock/lock.js
import fs from "node:fs/promises";
import { runActiveOnReadHooks, runActiveOnWriteHooks, runActiveServiceOverride } from "../extensions/index.js";
import { EXIT_CODE } from "../shared/constants.js";
import { PmCliError } from "../shared/errors.js";
import { toErrorMessage } from "../shared/primitives.js";
import { getLockPath } from "../store/paths.js";
import { nowIso } from "../shared/time.js";
function parseLockInfo(raw) {
    let parsed;
    try {
        parsed = JSON.parse(raw);
    }
    catch {
        return {
            info: null,
            warnings: ["lock_info_invalid_json"],
        };
    }
    if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) {
        return {
            info: null,
            warnings: ["lock_info_invalid_shape"],
        };
    }
    const candidate = parsed;
    const id = candidate.id;
    const pid = candidate.pid;
    const owner = candidate.owner;
    const createdAt = candidate.created_at;
    const ttlSeconds = candidate.ttl_seconds;
    if (typeof id !== "string" ||
        typeof pid !== "number" ||
        !Number.isFinite(pid) ||
        typeof owner !== "string" ||
        typeof createdAt !== "string" ||
        typeof ttlSeconds !== "number" ||
        !Number.isFinite(ttlSeconds)) {
        return {
            info: null,
            warnings: ["lock_info_invalid_shape"],
        };
    }
    return {
        info: {
            id,
            pid,
            owner,
            created_at: createdAt,
            ttl_seconds: ttlSeconds,
        },
        warnings: [],
    };
}
async function readLockInfo(lockPath) {
    try {
        const raw = await fs.readFile(lockPath, "utf8");
        await runActiveOnReadHooks({
            path: lockPath,
            scope: "project",
        });
        return parseLockInfo(raw);
    }
    catch (error) {
        if (isErrno(error, "ENOENT")) {
            return {
                info: null,
                warnings: [],
            };
        }
        return {
            info: null,
            warnings: ["lock_info_read_failed"],
        };
    }
}
function isErrno(error, code) {
    return typeof error === "object" && error !== null && "code" in error && error.code === code;
}
async function emitLockWriteHook(lockPath, op) {
    await runActiveOnWriteHooks({
        path: lockPath,
        scope: "project",
        op,
    });
}
function buildLockPayload(id, owner, ttlSeconds) {
    return {
        id,
        pid: process.pid,
        owner,
        created_at: nowIso(),
        ttl_seconds: ttlSeconds,
    };
}
async function createLockFile(lockPath, id, owner, ttlSeconds) {
    const handle = await fs.open(lockPath, "wx");
    try {
        await handle.writeFile(`${JSON.stringify(buildLockPayload(id, owner, ttlSeconds), null, 2)}\n`, "utf8");
    }
    finally {
        await handle.close();
    }
    await emitLockWriteHook(lockPath, "lock:create");
}
async function unlinkLockWithHook(lockPath, op) {
    try {
        await fs.unlink(lockPath);
        await emitLockWriteHook(lockPath, op);
    }
    catch {
        // Lock cleanup is best-effort.
    }
}
function isStaleLock(info, ttlSeconds) {
    const createdAtMs = info?.created_at ? Date.parse(info.created_at) : Number.NaN;
    const ageMs = Number.isFinite(createdAtMs) ? Date.now() - createdAtMs : Number.POSITIVE_INFINITY;
    return ageMs > ttlSeconds * 1000;
}
function lockOwnerSuffix(info) {
    return info?.owner ? ` (owner ${info.owner})` : "";
}
async function handleExistingLock(lockPath, id, ttlSeconds, force, forceRequiredForStaleLock) {
    const lockInfo = await readLockInfo(lockPath);
    if (!isStaleLock(lockInfo.info, ttlSeconds)) {
        throw new PmCliError(`Item ${id} is locked${lockOwnerSuffix(lockInfo.info)}`, EXIT_CODE.CONFLICT);
    }
    if (!force && forceRequiredForStaleLock) {
        const warningSuffix = lockInfo.warnings.length > 0 ? ` (${lockInfo.warnings.join(",")})` : "";
        throw new PmCliError(`Item ${id} lock is stale${warningSuffix}; rerun with --force when supported for this command`, EXIT_CODE.CONFLICT);
    }
    await unlinkLockWithHook(lockPath, "lock:stale_remove");
}
export async function acquireLock(pmRoot, id, ttlSeconds, owner, force = false, forceRequiredForStaleLock = true) {
    const lockOverride = await runActiveServiceOverride("lock_acquire", {
        pm_root: pmRoot,
        id,
        ttl_seconds: ttlSeconds,
        owner,
        force,
        force_required_for_stale_lock: forceRequiredForStaleLock,
    });
    if (lockOverride.handled) {
        const releaseFromFunction = typeof lockOverride.result === "function" ? lockOverride.result : null;
        const releaseFromObject = typeof lockOverride.result === "object" &&
            lockOverride.result !== null &&
            "release" in lockOverride.result &&
            typeof lockOverride.result.release === "function"
            ? lockOverride.result.release
            : null;
        const release = releaseFromFunction ?? releaseFromObject;
        if (release) {
            return async () => {
                await Promise.resolve(release());
                await runActiveServiceOverride("lock_release", {
                    pm_root: pmRoot,
                    id,
                    owner,
                });
            };
        }
    }
    const lockPath = getLockPath(pmRoot, id);
    for (let attempt = 0; attempt < 3; attempt += 1) {
        try {
            await createLockFile(lockPath, id, owner, ttlSeconds);
            return async () => {
                await unlinkLockWithHook(lockPath, "lock:release");
                await runActiveServiceOverride("lock_release", {
                    pm_root: pmRoot,
                    id,
                    owner,
                });
            };
        }
        catch (error) {
            if (!isErrno(error, "EEXIST")) {
                throw new PmCliError(`Failed to acquire lock for ${id}: ${toErrorMessage(error)}`, EXIT_CODE.GENERIC_FAILURE);
            }
            await handleExistingLock(lockPath, id, ttlSeconds, force, forceRequiredForStaleLock);
        }
    }
    throw new PmCliError(`Failed to acquire lock for ${id}`, EXIT_CODE.CONFLICT);
}
